## Introduction {.page_break_before}

<!-- "---" indicates a number or result that is to be filled out -->

Key sentence: To identify cell types of interest in host organisms, single-cell RNA-seq is a powerful tool, but it is not readily accessible to organisms without well-annotated reference genomes.

Zoonotic diseases such as SARS-CoV-1 and SARS-CoV-2 are often harbored in wild animals such as bats, that do not have well annotated genomes, but the asymptomatic nature of infection in those organisms and how it could be applied to humans is of immediate concern, due to the recent SARS, MERS, and Ebola outbreaks, and the current COVID-19 pandemic.
Identifying cell types in these host organisms enables studying the broad array of cell types available to living creatures, and can help understand human disease by seeing cell types that are "normal" in one animal, but resemble human disease cell types.
To identify cell types of interest in host organisms, single-cell RNA-seq is a powerful tool, but it is not readily accessible to organisms without well-annotated reference genomes.
Current methods for identifying cell types use single-cell RNA-sequencing, however, require three things: (1) high-quality genome assembly, (2) high-quality predicted gene annotations and (3) high-quality orthologous gene mappings across species.
However, the task of identifying even known cell types in species with poorly annotated genomes is nontrivial, as 99.999% of the approximate 10 million animal species predicted to exist on Earth have no submitted genome assembly [@doi:10.1371/journal.pbio.1001127; @url:https://www.ncbi.nlm.nih.gov/genome/browse#!/overview/] and even fewer have gene predictions and orthologous gene mappings to human, which remains an open problem [@doi:10.1038/nrg.2016.127; @doi:10.1146/annurev-cellbio-100616-060818].
But, we can nonetheless generate single-cell RNA-seq data from these species with poorly annotated genomes, and if we can leverage the annotations from existing organisms, we can better understand why host animals are able to tolerate infection by a pathogen that is deadly in humans.

Comparative transcriptomics traditionally relies on orthologous gene mapping, which requires a genome.
A group that is working on this is the Quest for Orthologs Consortium, which has established benchmarking  [@doi:10.1038/nmeth.3830; @doi:10.1093/nar/gkaa308; @doi:10.1093/bioinformatics/btu492; @doi:10.3389/fgene.2017.00165] services.
Another option besides using 1:1 orthologs is using orthogroups [@doi:10.1186/s13059-015-0721-2], aligning genes/reads to a common genome (if the sepecies are closely related, e.g. within 10 million years as for great apes)
There are many strategies for comparative transcriptomics for single cells [@doi:10.3389/fcell.2019.00175].
One of the first fundamental step is creatinga  common gene epxression matrix to compare the cell types.
The most commonly used strategy is subsetting the matrix down to 1:1 orthologous genes, which already have been computed by a database such as ENSEMBL [@doi:10.1093/nar/gkz966; @doi:10.1093/nar/gkz966].
Another strategy is aligning reads to the native species' genome, and using a tool such as LiftOver to remap coordinates into a well annotated genome such as human, and use those genes.
A method of quickly identifying potential orthlogs is using reduced amino acid alphabets [@doi:10.1093/gigascience/giz118].
All of these tools, the problem with them is that they require well annotated genomes, or genome assemblies.
What we can do with k-mers is use a peripherally related genome to align to, and then extract the protein-coding reads and k-mers that are present, then directly compare them.
These tools are not available to the 99.9% of vertebrates that do not have a reference genome, but transcriptome work, single-cell RNA-seq data is now so easy to obtain, that people already want to do analyses with their single cell data even if their organism doesn' thave a great reference genome.
And that's what we are enabling.

K-mers from sequencing data for single cells have been used to show that they can identify single cell types as well as gene counts [@url:https://www.biorxiv.org/content/10.1101/723833v1].
Recently, there has been literature to use subsampled k-mers to identify metagenomes within metagenomic data, but this uses subsampling of the data [@doi:10.1186/s13059-016-0997-x; @doi:10.21105/joss.00027; @doi:10.12688/f1000research.19675.1]. It turns out that k-mers are quite redundant and it is not necessary to use all of them for genome identification.
Specifically, it is possible to use a fractionals ubsampling of the original data to represent the data faithfully as if all k-mers were stored, as implemented in the Sourmash software [@doi:10.21105/joss.00027; @doi:10.12688/f1000research.19675.1].
k-mers or k-long usbsstrings hav ea long history in bioinformatics [@isbn:9780990374633] and have been widely used for genome and transcriptome assembly problems. They were recently applied to metagenomics for taxonomic classification.

We show that we use a subsampled set of th ek-mers to still identify cell types across and within species.

HOwever, for applying to cell types across species, we add the laywer of using reduced amino acid alphabet k-mres.

We combine the two concepts of k-mer subsampling and homology identification via reduced amino acid alphabets into what we term "k-mer homology."

Thus, there is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for orthologous gene mapping, gene annotations, or a reference genome.
Short, $k$-long sequence substrings, or $k$-mers, have been proposed for clustering single cells [@url:https://doi.org/10.1101/723833] and here we implemented $k$-mers from putatitvely translated RNA-seq reads with reduced amino acid alphabets [@doi:10.1093/bioinformatics/btp164 @doi:10.1093/gigascience/giz118; @doi:10.1093/bioinformatics/10.4.453; @10.1186/1471-2105-12-159; @doi:10.1093/protein/13.3.149; @doi:10.1093/bioinformatics/btp164; @doi:10.1093/nar/gkh180], to find shared cell types across species, and further identify *de novo* orthologous genes by querying the predicted protein sequences to a reference database.
This method relies solely on divergence time between species, which we show can be estimated from RNA-seq nucleotide $k$-mers (Supplemental Figure [@sfig:sfig1]).
As the direct assignment of protein-coding sequence skips both traditional alignment and gene orthology assignment it can a) be applied to transcriptomes from organisms with no or poorly annotated genomes, and b) identify putative functions of protein sequences contributing to shared cell types.

We benchmark the prediction of protein-coding sequence using simulated human RNA-seq data from single-copy genes present across all clades of mammals [@doi:10.1093/bioinformatics/btv351]. We show that one can extract putative protein-coding reads from XX mammalian species in BUSCO, and present the best parameters for different divergence times.

<!-- We first apply this method on a bulk comparative transcriptomic dataset consisting of nine amniote species and six tissues [@doi:10.1038/nature10532], showing that we achieve similar clustering results as using only reads mapping to 1:1 orthologs or Hierarchical Orthologous Groups (HOGs) [@doi:10.1371/journal.pone.0053786; @doi:10.1093/bib/bbr034; @doi:10.1093/bioinformatics/btu492] of protein-coding genes, but are able to resolve ... which can only be seen by using the $k$-mer method. -->
We first demonstrate the utility of this method by comparing lung tissue transcriptomes from organisms diverged by approximately 100 million years [@url:http://timetree.org/]: mouse from *Tabula Muris Senis* [@doi:10.1038/s41586-020-2496-1],
Across this evolutionary distance, only XX 1:1 orthologous genes exist as found by ...
<!-- and XX HOGs via orthologous matrix (OMA) [@doi:10.7717/peerj.6231; @doi:10.1093/bioinformatics/btx229] -->
We then add single-cell lung data from *Rhinolophus sinicus*, Chinese greater horseshoe bat, whose single-cell transcriptome was recently published [@url:https://www.biorxiv.org/content/10.1101/2020.06.30.175778v1]..
Using the k-mer homology method, we are able to distinguish the subtypes of T cells in the *R. sinicus* which were hidden using read counts alone, due to k-mers from genes not present in the genome assembly.
We then show the ability to find cell type enriched sequences that are absent from the reference genome, in particular interferon pathway genes and natural killer receptor genes.

Thus, we have shown the reference-free method using the $k$-mers from single cells is a novel, annotation-agnostic method for comparing cells across species that is capable of identifying cell states unique to a particular organism, helping to build the cell type evolutionary tree of life.

Similar to $k$-mer based approaches for transcript quantification [@doi:10.1038/nbt.2862; @doi:10.1038/nbt.3519; @doi:10.1038/nmeth.4197], we implemented $k$-mer based gene expression analyses across species, but instead of using DNA $k$-mers, our critical innovation was using translated protein $k$-mers.
We utilized sequence bloom trees (SBTs) [@doi:10.1038/nbt.3442] using a bottom-up approach to build them similar to previous work [@doi:10.1089/cmb.2017.0265; @doi:10.1089/cmb.2017.0258] to ensure localization of new datasets, meaning, if two leaves share a parent, they are guaranteed to be more similar than two leaves that do not share a parent.

For widespread accessibility and usage, we implemented `sencha` into two distinct Nextflow pipelines following software best practices such as testing and continuous integration: (1) `nf-core/kmermaid` to compare translated transcriptomes across divergent species, and (2) `czbiohub/nf-predictorthologs` to infer functions of translated sequences.

## Results

### Reliable identification of putative protein-coding sequence with `leatea translate` across ~100 million years of evolution

To extract putative protein codings equences from RNA-seq reads, we first established whether we could predict protein-coding sequences from simulated human RNA-seq data. We reliably identified putative protein-coding reads using our tool, `sencha`.

`sencha` works by performing six frame translation of RNAs, discarding reading frames with stop codons, and decomposing each frame into amino acid k-mers, and checking for the k-mer present in the reference proteome (Supplemental Fig. {@fig:sfig1}A). It is not a completely reference-free method as it does require a reference proteome, but we wanted to know how well the translation worked across species. To test this, we simulated human RNA-seq data and used reference proteomes from other mammals as the test dataset. We found that we were able to achieve a classification power of above 90% using a k-mer size of 8, which was interestingly optimized across all possible reference proteomes (Supplemental Fig. {@fig:sfig1}C). We found that a protein alphabet was most prooductive in translating the sequences across various species.
Using the reference proteomes from different animals represents extracting protein-coding reads from an organisms whose closely related species in the reference database is 10, 50, etc million years diverged.

Using the reference protomes, we ffound that a small fraction of overlapping k-mers in the putative protein-coding read were required to identify the read as putatively protein-coding. Our sensitivity to remove non-coding reads is quite high. We chose a minimum containment threshold of 0.05 as this maximized the number of putative protein coding reads while succcessfully discarding non-coding reads.

We then simulated




Putative protein-coding reads from simulated Human RNA-seq data can be identified across the mammalian lineage

![
Figure 1.
**A.** Overview of transcriptomic datasets from three species: *Homo sapiens* (human), *Rhinolophus sinicus* (Chinese horseshoe bat), and the early chordate *Botryllus schlosseri*. For *B. schlosseri*, we have FACS-purified bulk RNA-seq populations, including circulating cell types similar to mammalian hematopoietic stem cells (HCS), lymphoid, and myeloid populations. For *H. sapiens* and *R. sinicus*, we have droplet-based single-cell RNA-seq data for bone marrow, lung, intestine, spleen, thymus, and whole blood tissues.
**B.** Overview of computational method. First, in the `kmermaid` pipeline, RNA-seq reads are translated to protein via the Sencha tool, then converted to a reduced amino acid alphabet, decomposed into $k$-mers, and then those $k$-mers are subsampled. Next, the `nf-predictorthologs` pipeline performs differential $k$-mer expression, much like differential gene expression, to find $k$-mers that are enriched in a population, and queries for those differential $k$-mers in databases to identify them.
](images/SVG/figure1.svg){#fig:fig1 width="100%"}




### Degenerate amino acid alphabet k-mers recapitulate known biology and finds genes absent from even well-annotated genomes

Given our protein-coding reads, we could now use redudced amino acid alphbatest as a lightweight substitute for orthology across species. First, we shows that we can identify cell type across individual mice and human.

We were able to idnetify cell types across mice by using DNA k-mers of size --- and protein k-mers of size --- and Dayhoff k-mers of size ---.

Across human-tomouse, we found that DNA k-mers which vary quite greatly across species, were too variable to use to identify cell types. Instead, we found that even protein k-mers cannot reliably identify cell types across species. We turned to reduced amino acid alphabets such as dayhoff (Fig. {@fig:fig2}A).

We found we needed a minimum k-mer size of --- to have an adjusted rand index at least as high as --- to enable reliable classification of cell types witin species. Across species, we neded to use a reduced amino acid alphabet and a k-mer size of --- (Fig. {@fig:fig2}B-C).

Once we identified common cell types across species, we then dug into genes that are enriched in the cell types that may not be present even in the annotated genome. Or present in the orthologous gene mapping (Fig. {@fig:fig2}D).

Indeed, we idenitified --- genes that are not present in the genomes of human and mouse.

Additionally, we used k-mers from orthogroups, orthologous groups of genes, defined at the mammalian clade, of genes summarized to common groups (Fig. {@fig:fig2}E). [cite orthogroups paper]
This enables finding expression of genes that may not be defined in the gneome, and certainly genes that do not appear, are excluded by, subsetting to 1:1 orthologs, which is a common strategy for cross-species transcripome comparisons.

- Mouse:mouse comparison
  - identifies minimum parameters for within-species celltype identification
    - DNA, protein, dayhoff
  - finding genes not in the genome
- Human:mouse comparison
  - identifies minimum parameters for cross-species analyses
    - DNA, protein, dayhoff
      - Dayhoff, dna ksize: 42 (aa ksize: 14)
  - identifies gene expression for genes that don't have 1:1 orthologs
  - Identify genes not present in one or more genomes

Comparison to other methods??

To compare the performance of k-mer homology to traditonal cross-species methods, we used --- tools to compare label propagation across species using 1:1 orthologous gene datasets. We found that using 1:1 orthologs was not good and resulted in a classification adjusted rand index score of --- which was not good.


To recapitulate known biology in the mammalian lung, we compared lung cell atlases from species related within the mammalian suborder Euarchontoglires, whose most recent common ancestor is between 85 to 95 million years ago, we applied these methods to a mouse single-cell RNA-seq atlas, *Tabula Muris Senis* [@doi:10.1038/s41586-020-2496-1], to a human
lung cell atlas [@url:https://www.biorxiv.org/content/10.1101/742320v2]. We find that similar to the gene expression data, the epithelial and endothelial cell types are quite separate

![
Figure 2.
**A.** Left, UMAP of human and and mouse lung transcriptomes using $k$-mer similarity in reduced amino acid space, BBKNN-corrected for species. Right, UMAP of human and mouse transcriptomes using counts, BBKNN-corrected for species.
**B.** Silhouette scores on species and cell type groups (narrow_group).
**C.** Comparison of gene lists from differential $k$-mer expression, as compared to differential gene expression.
**D.** Barplot showing whether differentially expressed k-mers were found in the genome assembly, and whether they were in a gene annotated as an ortholog, or not.
](images/SVG/figure2.svg){#fig:fig2 width="100%"}


### Reduced amino acid k-mers from transcriptomes enabled identification of unannotated immune genes in Chinese horseshoe bat suggesting immune adaptations

We refined the lung cell type annotation from the Chinese horseshoe bat using our k-mer based cell type label propagation method. Using thes refined lables, we found that --- in the bat cell types were more similar to --- classical monocytes --- in mouse and human than they tended to be to their original cell types (Fig. {@fig:fig3}A).
This led us to then perform differential exprssion to find the genes that may be enriched in the bat cell types ovedrall. Using both logicistic regression on the k-mer abundances, and expression of orthogroups containment in th ecells, we found --- genes were widespreadly expressed in the bat, as compared to mouse and human (Fig. {@fig:fig3}B).
We then dug into the genes that are expressed, but not found inthe aligned genome or the gene anotation, and found that --- gene was a key contributor to --- natural killer cell --- type identity.

- Mouse:bat
  - celltype mapping
- mouse:bat:human
  - Identify gene expression for genes missing in genome


To identify molecular cell types unique to the Chinese horseshoe bat within mammals, we added a multi-organ, whole organism single-cell RNA-seq atlas of *R. sinicus*. We found expression of interferon genes, which have not been identified in the bat genome, but we were able to detect them in the transcriptome. In the bat scrnaseq paper, they performed qPCR to show expression of interferon genes, even though they weren't identified in the genome, to show those genes do truly exist.

To apply the method to a novel biological problem and create insights into previously unknown biology, we applied the k-mer homology method to a recently published cell atlas of the Chinese horseshoe bat, *R. sinicus* [@url:https://www.biorxiv.org/content/10.1101/2020.06.30.175778v1].

![
Figure 3.
**A.** Left, UMAP of human, mouse, and *R. sinicus* (Chinese horseshoe bat) transcriptomes using $k$-mers similarity in reduced amino acid space, BBKNN-corrected for species. Colored by cell compartment.
**B.** Barplot showing whether differentially expressed k-mers were found in the genome assembly, and whether they were in a gene annotated as an ortholog, or not.
**C.** UMAP of human, mouse and bat immune transcriptomes, colored by interferon-like gene expression.
**D.** Alignment of assembled Bat interferon protein sequences with known human sequences.
](images/SVG/figure3.svg){#fig:fig3 width="100%"}






## Discussion

We implemented the concept of lightweight orthology assignment using k-mers to the problem of cross-species RNA-seq analyses and achieved insight into unannotated genes and genes absent from the genome assembly. By removing the orthology inference step, `kmermaid` opens up the possibility of finding shared and divergent tissue and cell types across a broad range of species, paving the way for evolutionary analyses of cell types across species. `kmermaid` can be used in *de novo* setting for non-model organisms, finding similar cell types within an organism, or finding similar cell types relative to a reference organism, without the need for a reference genome or transcriptome. The memory usage of `kmermaid` is quite low, using only XXMB for extracting coding sequences and 50MB for assigning protein k-mer signatures. As the number of RNA-seq datasets, especially single-cell RNA-seq datasets continues to grow, we expect `kmermaid` to be widely used for identifying cell types in non-model organisms.

In summary, we developed a method to identify both known cell types in a non-model organism using a reference atlas from another organism, without the need for a genome or gene annotation from the non-model organism.
This method can be used to combine single-cell cell atlases from well-annotated,  model organisms, with sequencing data from poorly annotated non-model organisms, to directly find homologous cell types and orthologous genes.
By eliminating read alignment and orthologous gene mapping, `kmermaid` enables comparison of transcriptomes of the remaining 99.999% Eukaryotic species on Earth without submitted genome assemblies, with the cell atlases of a handful of model organisms to identify shared and novel cell types, and *de novo* identify orthologous genes.
By identifying homologous cell types across a broad variety of species, we come closer to an understanding of the evolution of genes, cells, and thus life itself.

`kmermaid` is free and open-source software and is available as Supplementary Data and at http://github.com/czbiohub/kmermaid and as a scalable Nextflow workflow at http://github.com/nf-core/nf-kmermaid.


<!--

## Outline

- Kmers can approximate orthologies
  - Jaccard similarity of orthologues is higher than non-orthologues
  - Benchmarking using https://orthology.benchmarkservice.org/cgi-bin/gateway.pl
  - Finding orthologues
    - Gold standard
      - ENSEMBL COMPARA
      - Quest for Orthologs consortium, Altenhoff, A. M., Boeckmann, B., Capella-Gutierrez, S., Dalquen, D. A., DeLuca, T., et al. (2016). Standardized benchmarking in the quest for orthologs. Nature Methods, 13(5), 425–430. http://doi.org/10.1038/nmeth.3830 [@doi:10.1038/nmeth.3830]
    - Orthologous groups/Conserved Domain Database [@url:https://www.ebi.ac.uk/miriam/main/collections/MIR:00000119]


<!-- ## Figure 2 -- $k$-mers from lossily-encoded putative protein-coding reads faithfully pull out reads from protein-coding genes within amniotes -->


- Overview of kmermaid pipeline
  - Comparison of tissue across species
    - Partition reads to coding/noncoding bins
    - MinHash the Dayhoff-encoded coding sequences
    - Jaccard similarity on the MinHashes
- Which reads are found to have coding features but didn’t map to the genome?
- Do these features map to novel genes or gene fusions?
- Kmers can find correct reading from of RNA-seq reads
  - Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data
- Comparison to other methods: RNASamba [@doi:10.1101/620880]

<!-- ### Figure 3 -- $k$-mers can pull out only reads from transcription factors and Amniotes can be compared on the MinHashes of their protein-coding sequences -->

<!-- ![Figure 3.](images/figure3.svg){#fig:fig3 width="100%"} -->

- Kmers can find only transcription factor reads of TFs from RNA-seq reads
  - Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data

### Some potential references

Gene expression evolution through duplications

- Farre, D., & Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 [@doi:10.1093/molbev/msp242]
- Thornton, J. W., & DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 [@doi:10.1146/annurev.genom.1.1.41]
- Farre, D., & Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 [@doi:10.1093/molbev/msp242]

Taxa-restricted genes

- Human-specific genes in fetal neocortex
Florio, M., Heide, M., Pinson, A., Brandl, H., Albert, M., Winkler, S., et al. (2018). Evolution and cell-type specificity of human-specific genes preferentially expressed in progenitors of fetal neocortex. eLife, 7, D635. http://doi.org/10.7554/eLife.32332 [@doi:10.7554/eLife.32332]
- Insects -- Santos, M. E., Le Bouquin, A., Crumière, A. J. J., & Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 [@doi:10.1126/science.aan2748]


Correlated evolution of celltypes?

- Liang, C., Musser, J. M., Cloutier, A., Prum, R. O., & Wagner, G. P. (2018). Pervasive Correlated Evolution in Gene Expression Shapes Cell and Tissue Type Transcriptomes. Genome Biology and Evolution, 10(2), 538–552. http://doi.org/10.1093/gbe/evy016 [@doi:10.1093/gbe/evy016]

Cell type homology

- Thornton, J. W., & DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 [@doi:10.1146/annurev.genom.1.1.41]
- Tschopp, P., & Tabin, C. J. (2017). Deep homology in the age of next-generation sequencing. Philosophical Transactions of the Royal Society B: Biological Sciences, 372(1713), 20150475–8. http://doi.org/10.1098/rstb.2015.0475 [@doi:10.1098/rstb.2015.0475]
- Hejnol, A., & Lowe, C. J. (2015). Embracing the comparative approach: how robust phylogenies and broader developmental sampling impacts the understanding of nervous system evolution. Philosophical Transactions of the Royal Society B: Biological Sciences, 370(1684), 20150045–16. http://doi.org/10.1098/rstb.2015.0045 [@doi:10.1098/rstb.2015.0045]
- Santos, M. E., Le Bouquin, A., Crumière, A. J. J., & Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 [@doi:10.1126/science.aan2748]
- Mammalian decidual cell

Cell type evolution

- Erkenbrack, E. M., Maziarz, J. D., Griffith, O. W., Liang, C., Chavan, A. R., Nnamani, M. C., & Wagner, G. P. (2018). The mammalian decidual cell evolved from a cellular stress response. PLOS Biology, 16(8), e2005594–27. http://doi.org/10.1371/journal.pbio.2005594 [@doi:10.1371/journal.pbio.2005594]

<!-- # Discussion --> -->
