## Introduction {.page_break_before}

<!-- "---" indicates a number or result that is to be filled out -->

### Key sentence: To identify cell types of interest in host organisms, single-cell RNA-seq is a powerful tool, but it is not readily accessible to organisms without well-annotated reference genomes.

The current COVID-19 pandemic has exposed inabilities of understanding zoonotic diseases in wild animals such as bats, due to poorly annotated genomes.
However, the asymptomatic nature of infection in zoonotic host organisms and how it could be applied to humans is of immediate concern.
<!-- non-human host organisms such as the bat *Rhinolophus sinicus*. -->
Zoonoses such as SARS-CoV-1 and SARS-CoV-2 are often harbored in wild animals such as bats, that do not have well annotated genomes, but the asymptomatic nature of infection in those organisms and how it could be applied to humans is of immediate concern, due to the recent SARS, MERS, and Ebola outbreaks, and the current COVID-19 pandemic.
Identifying cell types in these host organisms enables studying the broad array of cell types available to living creatures, and can help understand human disease by seeing cell types that are "normal" in one animal, but resemble human disease cell types.
To identify cell types of interest in host organisms, single-cell RNA-seq is a powerful tool, but it is not readily accessible to organisms without well-annotated reference genomes, as 99.999% of the approximate 10 million animal species predicted to exist on Earth have no submitted genome assembly [@doi:10.1371/journal.pbio.1001127; @url:https://www.ncbi.nlm.nih.gov/genome/browse#!/overview/].
Current methods for identifying cell types use single-cell RNA-sequencing, however, require three things: (1) high-quality genome assembly, (2) high-quality predicted gene annotations and (3) high-quality orthologous gene mappings across species.
But, we can nonetheless generate single-cell RNA-seq data from these species with poorly annotated genomes, and if we can leverage the annotations from existing organisms, we can better understand why host animals are able to tolerate infection by a pathogen that is deadly in humans.

### key: The most common strategies for cross-species transcriptomics lose data due to unaligned sequences, unannotated genes, and unmatched genes across species.


There are many strategies for comparative transcriptomics for single cells [@doi:10.3389/fcell.2019.00175].
The most commonly used strategies for cross-species gene expression analyses lose data due to unaligned sequences, unannotated genes, and unmatched genes across species.

Unaligned sequences are largely due to errors in genome assembly, or
While high-quality end-to-end genome assemblies are now within reach with the recently published teleomere-to-telomere human genome assembly [@doi:10.1038/s41586-020-2547-7], and high-quality assemblies, including gene annotations, will soon be a possibility thanks to thhe efforts of the Vertebrate Genome Project [@https://doi.org/10.1101/2020.05.22.110833; @https://doi.org/10.1101/2020.06.30.177956].
While the genome assembly provides substantial informationa bout the genomic structure of a species, it does not yet have the matching genes across species.

### key: Matching genes across species is an open and unsolved problem.

One of the first fundamental step is creating a common gene expression matrix to compare the cell types.
Matching genes across species is perfolrmed via a process called orthologous gene mapping.
A group that is working on this is the Quest for Orthologs Consortium, which has established benchmarking  [@doi:10.1038/nmeth.3830; @doi:10.1093/nar/gkaa308; @doi:10.1093/bioinformatics/btu492; @doi:10.3389/fgene.2017.00165] services.
Another option besides using 1:1 orthologs is using orthogroups [@doi:10.1186/s13059-015-0721-2], aligning genes/reads to a common genome if the species are closely related, such as for primates [@doi:10.1038/s41586-019-1654-9].

and even fewer have gene predictions and orthologous gene mappings to human, which remains an open problem [@doi:10.1038/nrg.2016.127; @doi:10.1146/annurev-cellbio-100616-060818].
The most commonly used strategy is subsetting the matrix down to 1:1 orthologous genes, which already have been computed by a database such as ENSEMBL [@doi:10.1093/nar/gkz966; @doi:10.1093/nar/gkz966].
Comparative transcriptomics traditionally relies on orthologous gene mapping, which requires a genome.



Another strategy for more distantly related species is to only use transcription factor genes that are part of "core regulatory complex" (CoRC), which again subsets the data to only certain genes.
Not just at alignment, but at aligning to annotated genes, at subsetting to 1:1 orthologs -- these are all data lossy steps.
While the first two steps are dependent on the genome assembly and gene annotation quailty, the subsetting to 1:1 orthologs cannot be solved with improved assembly and gene annotation tools.
It can be solved with using other strategies for finding orthologous genes.
Another strategy is aligning reads to the native species' genome, and using a tool such as LiftOver to remap coordinates into a well annotated genome such as human, and use those genes.

For all of these tools, the problem with them is that they lose data at every step.
These tools are not available to the 99.9% of vertebrates that do not have a reference genome, but transcriptome work, single-cell RNA-seq data is now so easy to obtain, that people already want to do analyses with their single cell data even if their organism doesn' thave a great reference genome.
And that's what we are enabling.

Once the genomes of the species are obtained, the work is not over.
For each pariwise species comparison, the orthologous genes must be found.
There are many tools for this, such as OrthoFinder [@doi:10.1186/s13059-015-0721-2], eggNOG [@doi:10.1093/nar/gkv1248; @doi:10.1093/molbev/msx148], and highly curated databases such as ENSEMBL [citation].
The time to run these tools is a long time.
But that's not really the point... what we really care about is using the RNA-seq data directly and not losing data due to mismatched orthology, unannotated genes, or unaligned seuqences.




### key: A method of not losing data due to unaligned sequences or unannotated genes is using all the data in the form of subsampled $k$-mers.


$k$-mers from sequencing data for single cells have been used to show that they can identify single cell types as well as gene counts [@https://doi.org/10.1101/723833].
Recently, there has been literature to use subsampled $k$-mers to identify metagenomes within metagenomic data, but this uses subsampling of the data [@doi:10.1186/s13059-016-0997-x; @doi:10.21105/joss.00027; @doi:10.12688/f1000research.19675.1].
It turns out that $k$-mers are quite redundant and it is not necessary to use all of them for genome identification.
Specifically, it is possible to use a fractional subsampling of the original data to represent the data faithfully as if all $k$-mers were stored, as implemented in the Sourmash software [@doi:10.21105/joss.00027; @doi:10.12688/f1000research.19675.1].
$k$-mers or k-long substrings have a long history in bioinformatics [@isbn:9780990374633] and have been widely used for genome and transcriptome assembly problems. They were recently applied to metagenomics for taxonomic classification.

We show that we use a subsampled set of the $k$-mers to still identify cell types across and within species.

However, for applying to cell types across species, we add the layer of using reduced amino acid alphabet $k$-mers.

However, for applying to cell types across species, we add the layer of using protein $k$-mers.

### key: Protein sequence can be inferred from RNA-seq due to nullomers.

Nullomers are sequences that are not present in the genome annotation or protein sequences [@doi:10.1142/9789812772435_0034; @doi:10.1371/journal.pone.0164540; @https://doi.org/10.1101/2020.06.25.170431; @https://doi.org/10.1101/2020.03.02.972422] of a species.
While this seems like there are a contradiction between wanting to use all the data, and then only using the data that is present in the database annotation, we show we are able to predict protein coding reads using simulated data.
The signal of a non-protein coding sequence is very obvious -- the true coding frames have k-mers in common with the known database, and the non-true coding frames do not have k-mers in common.
There are only so many k-mers that are compatible with eukaryotic, or mammalian life.
There are amino acid combinations that are not seen in the natural world.
Whether nullomers are not present due to natural selection or due to CpG hypermutability [@doi:10.1371/journal.pone.0001022] remains under debate, there are this phenomena is real and can be exploited to identify putative protein-coding sequences.
By using the "negative space" of unobserved protein sequences, we can identify reading frames of RNA to protein-coding sequences that are unlikely.
Non-productive Reading frames of RNA-seq can be found in the nullomers.
Protein sequences that have been observed are inferred in the k-mers.

### key: Gene annotations can be inferred from RNA-sequencing data, which is easily obtained with commodity tools

What we can do with $k$-mers is use a peripherally related genome to align to, and then extract the protein-coding reads and $k$-mers that are present, then directly compare them.

Transcribed seqeunces in the form of RNA-seq data can be inferred to be transcribed genes for now.


### key: Reduced amino acid alphabets provide a lightweight method of identifying matching genes across species.

Not only do we use protein k-mers, but we use reduced amino acid alphabet k-mers.

Amino acids can be categorized into their biochemical properites.
Many times,a  protein can retain its same phsyical structure and function even with amino acid changes.
These amino acid changes may be retaining the phyiochemical properites of the amino acid.
The space of amino acid substituion was explored back in the 90s to find similarity searches for proteins.
The BLOSUM62, PAM substituion matrices were a method of assigning higher penalties for amino acids that changed their biochemical type.
Reduced amino acid alphabets are a method of reducing blame in changing of amino acid sequences, which is especially compatible with k-mers as if a single letter in a k-mer changes, then the whole k-mer no longer matches.

A method of quickly identifying potential orthologs is using reduced amino acid alphabets [@doi:10.1093/gigascience/giz118].
Reduced amino acid alphabets have been used for database searches of related protein sequences [@doi:10.1186/1471-2105-12-159] and for protein structure prediction
[@doi:10.1093/bioinformatics/btp164 @doi:10.1093/gigascience/giz118; @doi:10.1093/bioinformatics/10.4.453; @doi:10.1186/1471-2105-12-159; @doi:10.1093/protein/13.3.149; @doi:10.1093/bioinformatics/btp164; @doi:10.1093/nar/gkh180]


We combine the two concepts of $k$-mer subsampling and homology identification via reduced amino acid alphabets into what we term "$k$-mer homology."


### key: There is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for orthologous gene mapping, gene annotations, or a genome assembly.

We translate our protein sequences to protein.
We need to predict protein seuqnece, and turns out most reading frames of RNA don't need to be predicted -- they are already near zero.
They have stop codons.

Thus, there is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for orthologous gene mapping, gene annotations, or a reference genome.
Short, $k$-long sequence substrings, or $k$-mers, have been proposed for clustering single cells [@url:https://doi.org/10.1101/723833] and here we implemented $k$-mers from putatitvely translated RNA-seq reads with reduced amino acid alphabets [@doi:10.1093/bioinformatics/btp164 @doi:10.1093/gigascience/giz118; @doi:10.1093/bioinformatics/10.4.453; @doi:10.1186/1471-2105-12-159; @doi:10.1093/protein/13.3.149; @doi:10.1093/bioinformatics/btp164; @doi:10.1093/nar/gkh180], to find shared cell types across species, and further identify *de novo* orthologous genes by querying the predicted protein sequences to a reference database.
This method relies solely on divergence time between species, which we show can be estimated from RNA-seq nucleotide $k$-mers (Supplemental Figure [@sfig:sfig1]).
As the direct assignment of protein-coding sequence skips both traditional alignment and gene orthology assignment it can a) be applied to transcriptomes from organisms with no or poorly annotated genomes, and b) identify putative functions of protein sequences contributing to shared cell types.

We benchmark the prediction of protein-coding sequence using simulated human RNA-seq data from single-copy genes present across all clades of mammals [@doi:10.1093/bioinformatics/btv351].
We show that one can extract putative protein-coding reads from --- mammalian species in BUSCO, and present the best parameters for different divergence times.

<!-- We first apply this method on a bulk comparative transcriptomic dataset consisting of nine amniote species and six tissues [@doi:10.1038/nature10532], showing that we achieve similar clustering results as using only reads mapping to 1:1 orthologs or Hierarchical Orthologous Groups (HOGs) [@doi:10.1371/journal.pone.0053786; @doi:10.1093/bib/bbr034; @doi:10.1093/bioinformatics/btu492] of protein-coding genes, but are able to resolve ... which can only be seen by using the $k$-mer method. -->
We first demonstrate the utility of this method by comparing lung tissue transcriptomes from organisms diverged by approximately 100 million years [@url:http://timetree.org/]: mouse from *Tabula Muris Senis* [@doi:10.1038/s41586-020-2496-1],
Across this evolutionary distance, only --- 1:1 orthologous genes exist as found by ...
<!-- and XX HOGs via orthologous matrix (OMA) [@doi:10.7717/peerj.6231; @doi:10.1093/bioinformatics/btx229] -->
We then add single-cell lung data from *Rhinolophus sinicus*, Chinese greater horseshoe bat, whose single-cell transcriptome was recently published [@https://doi.org/10.1101/2020.06.30.175778]..
Using the $k$-mer homology method, we are able to distinguish the subtypes of T cells in the *R. sinicus* which were hidden using read counts alone, due to $k$-mers from genes not present in the genome assembly.
We then show the ability to find cell type enriched sequences that are absent from the reference genome, in particular interferon pathway genes and natural killer receptor genes.


### key: $k$-mer homology is a transcriptome-first method for cross-species analyses implementing via $k$-mers and reduced amino acid alphabets

Thus, we have shown the reference-free method using the $k$-mers from single cells is a novel, annotation-agnostic method for comparing cells across species that is capable of identifying cell states unique to a particular organism, helping to build the cell type evolutionary tree of life.

Similar to $k$-mer based approaches for transcript quantification [@doi:10.1038/nbt.2862; @doi:10.1038/nbt.3519; @doi:10.1038/nmeth.4197], we implemented $k$-mer based gene expression analyses across species, but instead of using DNA $k$-mers, our critical innovation was using translated protein $k$-mers.
We utilized sequence bloom trees (SBTs) [@doi:10.1038/nbt.3442] using a bottom-up approach to build them similar to previous work [@doi:10.1089/cmb.2017.0265; @doi:10.1089/cmb.2017.0258] to ensure localization of new datasets, meaning, if two leaves share a parent, they are guaranteed to be more similar than two leaves that do not share a parent.

For widespread accessibility and usage, we implemented `leaftea` into two distinct Nextflow pipelines following software best practices such as testing and continuous integration: (1) `nf-core/kmermaid` to compare translated transcriptomes across divergent species, and (2) `czbiohub/nf-predictorthologs` to infer functions of translated sequences.

![
Figure 1.
**A.** Overview of transcriptomic datasets from three species: *Homo sapiens* (human), *Rhinolophus sinicus* (Chinese horseshoe bat), and the early chordate *Botryllus schlosseri*. For *B. schlosseri*, we have FACS-purified bulk RNA-seq populations, including circulating cell types similar to mammalian hematopoietic stem cells (HCS), lymphoid, and myeloid populations. For *H. sapiens* and *R. sinicus*, we have droplet-based single-cell RNA-seq data for bone marrow, lung, intestine, spleen, thymus, and whole blood tissues.
**B.** Overview of computational method. First, in the `kmermaid` pipeline, RNA-seq reads are translated to protein via the leaftea tool, then converted to a reduced amino acid alphabet, decomposed into $k$-mers, and then those $k$-mers are subsampled. Next, the `nf-predictorthologs` pipeline performs differential $k$-mer expression, much like differential gene expression, to find $k$-mers that are enriched in a population, and queries for those differential $k$-mers in databases to identify them.
](images/SVG/figure1.svg){#fig:fig1 width="100%"}

## Results



### Reliable identification of putative protein-coding sequence with `leatea translate` across ~150 million years of evolution

| Amino acid              | Property              | Dayhoff |
|:------------------------|:----------------------|:--------|
| `C`                     | Sulfur polymerization | `a`     |
| `A`, `G`, `P`, `S`, `T` | Small                 | `b`     |
| `D`, `E`, `N`, `Q`      | Acid and amide        | `c`     |
| `H`, `K`, `R`           | Basic                 | `d`     |
| `I`, `L`, `M`, `V`      | Hydrophobic           | `e`     |
| `F`, `W`, `Y`           | Aromatic              | `f`     |
Table: Dayhoff encoding of amino acids by categories of biochemical properties, into a six-letter alphabet. Adapted from [@doi:10.1186/1471-2105-9-367]. For example, the protein sequence `LIVING` would become `eeeecb`. {#tbl:dayhoff-alphabet}

![
Figure 2. Application of k-mer homology to mammalian lung cell type identification recapitulates known biology and identifies enriched genes not present in reference genome assembly.
**A.** Overview of `leafte translate` method. First, each read is translated into all six possible protein-coding translation frames. Next, reading frames with stop codons are eliminated. Each protein-coding frame is $k$-merized, then the fraction of $k$-mers which appear in the known protein-coding database is computed. Reading frames containing 5% or more $k$-mers matching the reference proteome are inferred to be putatively protein-coding.
**B.** Experimental set up to identify best k-mer sizes for translating protein-coding sequence, given a reference proteome of divergent organism. First, human RNA-seq reads were simulated from genes present in all extant mammals. Next, we used highly curated reference proteomes from BUSCO mammalia, representing almost 200 million years of evolution, across a wide range of mammalian species. Using `leaftea`, we partition the reads into protein-coding and noncoding using a variety of amino acid alphabets and k-mer sizes, and then compute classification metrics.
**C.** Predictive power of detecting protein-coding sequence across ~200 million years of evolution, using simulated human RNA-seq reads and mammalian reference proteomes. x-axis, k-mer size used to translate the RNA sequence; y-axis, receiver operating characteristic area under the curve (ROC AUC) of protein-coding prediction at that k-mer size. Dotted vertical lines indicate k-mer size producing maximum ROC AUC for each divergence time.
**D.** Classification metrics of protein-coding prediction with both protein and Dayhoff alphabets, using best-performing k-mer sizes for each alphabet from (**C**). Left, precision; middle, recall; right, $F_{1/2}$ score. x-axis, Estimated divergence time of reference proteome relative to human, from timetree.org; y-axis, value of classification metric.
](images/SVG/figure2.svg){#fig:fig2 width="100%"}

To extract putative protein coding sequences from RNA-seq reads, we first established whether we could predict protein-coding sequences from simulated human RNA-seq data.
We reliably identified putative protein-coding reads using our tool, `leaftea`.
`leaftea` works by performing six frame translation of RNAs, discarding reading frames with stop codons, and decomposing each frame into amino acid $k$-mers, and checking for the $k$-mer present in the reference proteome (Fig. {@fig:fig1}A).
In detecting putative protein coding sequence, there many be small amino acid changes that do not change the overall function of the protein, and to accomodate this, we offer the option to translate using either the original 20-letter amino acid alphabet, or a six-letter reduced amino acid alphabet, the Dayhoff encoding [@tbl:dayhoff-alphabet].
As many organisms of interest are not annotated, but a different organism with a related ancestor is, we wanted to understand the limits of protein-coding detection when the best-annoated organism in the database is divergent with up to approximately 200 million years.
To test this, we simulated human RNA-seq data from genes present in all extant mammals as determined by BUSCO [@doi:10.1007/978-1-4939-9173-0_14], and used reference proteomes from eighteen mammals, ranging in divergence from 0 to almost 200 million years [@url:timetree.org] (Fig. {@fig:fig1}B).
We found that we were able maximize classification power, measured via receiver operating characteristic area under the curve (ROC AUC), for most divergence times for a protein alphabet with a $k$-mer size of 8, and 17 for a Dayhoff encoding (Fig. {@fig:fig1}C, left and middle).
We found that a protein alphabet with a $k$-mer size of 8 was most productive in translating the sequences for closely related species, within Boroeutheria, a mangnorder of mammals related within approximately 100 million years of one another [@url:timetree.org], containing the superorders Euarchontoglires (includes rodents and primates) and Laurasiatheria (includes bat, hedgehog, hooved animals, cats, and dogs), with a mean ROC AUC of 0.873898 among well-annotated reference proteomes.
For more distantly related species such as platypus, a the Dayhoff alphabet had a ROC AUC of 0.547203, however, the precision of rejecting non-coding reads remained quite high across distantly related species (Fig. {@fig:fig1}D, left).
As this metric performs better in precision, we used an $F_\Beta$ score, where $\Beta = 1/2$, to indicate a 2x preference for precision over recall (Fig. {@fig:fig1}D, right).
<!-- Interested in understanding whether `leaftea`'s strength lies in detecting true protein-coding reads, or eliminating non-coding reads, we computed the precision and recall of protein-coding classification (Fig. {@fig:fig1D}).
We found that
Using the reference proteomes from different animals represents extracting protein-coding reads from an organisms whose closely related species in the reference database is 10, 50, etc million years diverged.
Using the reference proteomes, we found that a small fraction of overlapping $k$-mers in the putative protein-coding read were required to identify the read as putatively protein-coding. Our sensitivity to remove non-coding reads is quite high.
We chose a minimum containment threshold of 0.05 as this maximized the number of putative protein coding reads while successfully discarding non-coding reads. -->
This shows that while it is possible to detect putative protein-coding sequence even when the best annotated organism in the database almost 200 million years diverged from the organism in question.


<!-- Putative protein-coding reads from simulated Human RNA-seq data can be identified across the mammalian lineage -->

### Degenerate amino acid alphabet $k$-mers identify common cell types across species and finds genes absent from even well-annotated genomes




#### key: Protein k-mers can be used to predict cell types within species and across individuals

![
Figure 3.
**A.** Experimental set up to validate effectiveness of building a cell type database from one species, to label cells from another species. First, we build a cell type database using sequence bloom trees (SBTs) from mouse cells, then search using a human cell for the best matching mouse cell.
**B.** Mean of classification metrics across mouse:human cell type lookup experiments, highlighting the k-mer size that maximizes the mean for each metric and alphabet. Vertical lines, standard deviation of the mean. x-axis, k-mer size in DNA space, i.e. a k-mer size of 21 corresponds to a an amino acid length of 7. y-axis, value of classification metric. Left, adjusted rand score; right, F1 score.
**C.** Molecules and k-mer sizes maximizing classification metrics from (**E**).
**D.** Confusion matrices of classification using Dayhoff encoding and k-mer size of ---, the best performing metrics from (**E**). x-axis, predicted cell label; y-axis, ground truth cell label.
](images/SVG/figure3.svg){#fig:fig3 width="100%"}




Previous work has used k-mers from protein-coding genes of single-cell RNA-seq data within one individual of a species to distinguish cell types [@https://doi.org/10.1101/723833], and we are interested in first comparing k-mer effectiveness across individuals of different species.
Similar to their results, we find that using DNA from coding nucleotides successfully distinguishes cell types within individuals of the same species.
However, we find that even within species, the self-lookup of cell types is hindered by technical artifact genes.
We removed k-mers from technical artifact genes, such as ribosomal and mitochondrial genes, and genes associated with single-cell dissociation [@doi:10.1038/s41467-017-02772-x; @doi:10.1016/j.celrep.2017.10.037; @doi:10.1038/nmeth.4437; @doi:10.1186/s13059-020-02032-0; @https://doi.org/10.1101/2020.06.30.178673].
We find that a DNA k-mer size of --- works well to distinguish cell types across species.
However, across species, then there is an obvious "batch effect" present as a result of individual DNA differences in the genome.
We find that using protein k-mers with a k-mer size of --- works well to "soften the rough edges" intriduced by individual genomic variation for cell types.
Additionally, we see that we can use a reduced amino acid alphabet within mouse, across mice, to identify similar cell types.
Using the Tabula Muris Senis data, wE built a cell type database from one mouse, and queried its own cells. We found that a k-mer size of --- with DNA had the best [F1 score / Adjusted Rand Index (ARI)] of --- cell type classification within one male mouse.
Interested in how the genomic diffferences across individuals affect DNA-based cell type classification, we built a cell type database from one male mouse, and computed cell similarities across mice.
We found that across mice, the DNA was still able to identify cell types, but the protein k-mers had a higher F1 score of ---.
Finally, we also tested a male-female comparison, where the databse was built from male cells, and the query was female cells.
We found that a DNA k-mer size of --- resulted in an F1 score of ---.



#### key: Human cell types can be predicted from mouse cells using k-mers from reduced amino acid alphabets

Given putative protein-coding sequences from RNA-seq reads, we now use reduced amino acid alphabets as a lightweight substitute for homology across species.
First, we show that we can identify cell type across individual mice and human.
<!-- Maybe this is for the supplemental text? -->
We were able to identify cell types across mice by using DNA $k$-mers of size --- and protein $k$-mers of size --- and Dayhoff $k$-mers of size ---.
Across human-to-mouse, we found that DNA $k$-mers which vary quite greatly across species, were too variable to use to identify cell types.
Instead, we found that even protein $k$-mers cannot reliably identify cell types across species.
We turned to reduced amino acid alphabets such as Dayhoff (Fig. {@fig:fig2}A).
We found we needed a minimum $k$-mer size of --- to have an adjusted rand index at least as high as --- to enable reliable classification of cell types within species.
Across species, we needed to use a reduced amino acid alphabet and a $k$-mer size of --- (Fig. {@fig:fig2}B-C).
Once we identified common cell types across species, we then dug into genes that are enriched in the cell types that may not be present even in the annotated genome.
Or present in the orthologous gene mapping (Fig. {@fig:fig2}D).
Indeed, we identified --- genes that are not present in the genomes of human and mouse.
Additionally, we used $k$-mers from orthogroups, orthologous groups of genes, defined at the mammalian clade, of genes summarized to common groups (Fig. {@fig:fig2}E). [cite orthogroups paper]
This enables finding expression of genes that may not be defined in the genome, and certainly genes that do not appear, are excluded by, subsetting to 1:1 orthologs, which is a common strategy for cross-species transcriptome comparisons.

<!-- - Mouse:mouse comparison
  - identifies minimum parameters for within-species celltype identification
    - DNA, protein, dayhoff
  - finding genes not in the genome
- Human:mouse comparison
  - identifies minimum parameters for cross-species analyses
    - DNA, protein, dayhoff
      - Dayhoff, dna ksize: 42 (aa ksize: 14)
  - identifies gene expression for genes that don't have 1:1 orthologs
  - Identify genes not present in one or more genomes -->


#### key: Reduced amino acid alphabet k-mers enable identifying cell type enriched sequences that may not be present in gene annotation or genome assembly, or orthologous gene mappings

To recapitulate known biology in the mammalian lung, we compared lung cell atlases from species related within the mammalian suborder Euarchontoglires, whose most recent common ancestor is between 85 to 95 million years ago, we applied these methods to a mouse single-cell RNA-seq atlas, *Tabula Muris Senis* [@doi:10.1038/s41586-020-2496-1], to a human
lung cell atlas [@doi:10.1038/s41586-020-2922-4].
We used a reduced amino acid alphabet of Dayhoff to categorize the amino acids into biochemical properties (Fig. {@fig:fig2}A).

To reduce effets of batches, we removed risobomal sequences from any k-mers resembling sequences from mammalian ribosomes.
As a control, we first built SBTs on one mouse, and queried with cells from a different mouse.
We found that --- k-size and --- alphabet was appropriate to identify cell types across individuals with a very similar genomic background.
We were able to identify the correct cell type ---% of the time.
Next, we built SBTs across humans, querying with a different individual and built an SBT on a different individual.
Similarly, we found  that --- k-size and --- alphabets were successful at identifying cell types across individuals.
We were able to identify the correct cell type ---% of the time.



We built a sequence bloom tree database of cells from *Tabula Muris Senis* and applied the cell type annotations to the human data.
We found that by searching for the number of k-mers overlapping the query cell, using a containment search [@url:https://github.com/dib-lab/2020-paper-sourmash-gather], we were able to correctly identify the cell type ---% of the time (Fig. {@fig:fig2}B).
There were minimal misclassifications and most misclassifications occur as a result of ---.
Curious whether these cell types had any enriched sequences that weren't present in the annotated genomes, we then performed differential k-mer expression to identify enriched sequences in the cell types.
We found that compared to gene expression, we had a ---% overlap with the genes that are known to be enriched by gene expression of 1:1 orthologs, and we found --- genes that were enriched in the cell types but were not 1:1 orthologs, and thus would have been missed by traditional methods.

<!-- Comparison to other methods?? -->

To compare the performance of $k$-mer homology to traditional cross-species methods, we used --- tools to compare label propagation across species using 1:1 orthologous gene datasets.
We found that using 1:1 orthologs was not good and resulted in a classification adjusted rand index score of --- which was not good.


#### key: Cell type enriched sequences may include sequences that are not present in reference genome assemblies

To find expression of genes that many not be in the genome, we took two approaches.
First, we used differential k-mer expression to enrich for particular sequences that were highly expressed in a cell type.
Then, we computed signatures for the last 100aa of each amino acid sequence of orthogroups to get a relative gene expression.
We found that the overlap between differential gene expression 1:1 orthologs, and the differential k-mer expression and orthogroup expression, we found --- % overlap between al three, which had a --- p-val7ue of dignificance using a chi-squared or hypergeoemetric test of enrichment.
However, with the k-mers, we were able to find genes such as ---, --- and --- which are enriched in the cell type ---, but are not a 1:1 ortholog, and are instead a --- ortholog.
Some of these were even not present in the genome assembly, for example --- is present in the human genome assembly but not in the mouse assembly.
Thus, using k-mers, we are able to identify genes that are enriched in cell types but are not present in the genome assembly, in unannotated genes, or not 1:1 orthologs.

<!-- We find that similar to the gene expression data, the epithelial and endothelial cell types are quite separate -->




### Reduced amino acid $k$-mers from transcriptomes enabled identification of unannotated immune genes in Chinese horseshoe bat suggesting immune adaptations

We refined the lung cell type annotation from the Chinese horseshoe bat using our $k$-mer based cell type label propagation method. Using these refined labels, we found that --- in the bat cell types were more similar to --- classical monocytes --- in mouse and human than they tended to be to their original cell types (Fig. {@fig:fig3}A).
This led us to then perform differential expression to find the genes that may be enriched in the bat cell types overall.
Using both logistic regression on the $k$-mer abundances, and expression of orthogroups containment in the cells, we found --- genes were widespreadly expressed in the bat, as compared to mouse and human (Fig. {@fig:fig3}B).
We then dug into the genes that are expressed, but not found in the aligned genome or the gene annotation, and found that --- gene was a key contributor to --- natural killer cell --- type identity.

- Mouse:bat
  - celltype mapping
- mouse:bat:human
  - Identify gene expression for genes missing in genome


To identify molecular cell types unique to the Chinese horseshoe bat within mammals, we added a multi-organ, whole organism single-cell RNA-seq atlas of *R. sinicus*. We found expression of interferon genes, which have not been identified in the bat genome, but we were able to detect them in the transcriptome. In the bat scrnaseq paper, they performed qPCR to show expression of interferon genes, even though they weren't identified in the genome, to show those genes do truly exist.

To apply the method to a novel biological problem and create insights into previously unknown biology, we applied the $k$-mer homology method to a recently published cell atlas of the Chinese horseshoe bat, *R. sinicus* [@https://doi.org/10.1101/2020.06.30.175778].








## Discussion

### key: Putative protein-coding sequence can be detected across ~150 million years of evolution

With the increasing accessibility of single-cell RNA-seq, it is more and more challenging to organize existing data.
 but the analysis of said data is more and more difficult if the organism is not well anotated in the databses.
However, we are able to recover the putative protein coding sequences, even when the most recent ancestor of the closest related organism in the database is within ~150 million years, meaning we can study any possible mammal.
[What iss the average distance between unknown/unannotated organisms and annotated organisms? Can this be quantified?]
We used RNA-seq data directly due to the fact that we can extract putative protein-coding sequences from the RNA-seq data.
Our precision is high, meaning we are able to discard true non-coding frames very easily.
Even for species that are distantly related, ~150 million years between human and the distantly related --- species, we were able to recover ---% of the known protein-coding frames from the simulated human data.
The most distantly related organism, platypus, with an estimated divergence time from human ranging from 167-192 million years (@url:timetree.org), was able to identify true protein-coding sequence in human ---% of the time.
This indicates that even if the organism is --- million years diverged from the most closely related organism in the database, it is still possible to recover putative protein-coding.
Thus, this opens up the possibility of studying nearly ---% of organisms on the planet, as most are related within 150 million years of annotated organisms. [Olga: Ask Erich Schwarz for help on how to compute this]

### key: Cell type databases built on protein k-mers can assign cell types across individuals within species

### key: Reduced amino acid alphabet k-mers enable lighweight orthology assignment

### key: Cell type databases built on reduced amino acid alphabet protein k-mers can assign cell types across species


### key: Lightweight orthology assignment using $k$-mers enabled reference-independent cross-species RNA-seq analyses and achieved insight into unannotated genes and genes absent from the genome assembly.

We implemented the concept of lightweight orthology assignment using $k$-mers to the problem of cross-species RNA-seq analyses and achieved insight into unannotated genes and genes absent from the genome assembly.
By removing the orthology inference step, `kmermaid` opens up the possibility of finding shared and divergent tissue and cell types across a broad range of species, paving the way for evolutionary analyses of cell types across species.
`kmermaid` can be used in *de novo* setting for non-model organisms, finding similar cell types within an organism, or finding similar cell types relative to a reference organism, without the need for a reference genome or transcriptome.
The memory usage of `kmermaid` is quite low, using only ---MB for extracting coding sequences and 50MB for assigning protein $k$-mer signatures.
As the number of RNA-seq datasets, especially single-cell RNA-seq datasets continues to grow, we expect `kmermaid` to be widely used for identifying cell types in non-model organisms.


### key: Reduced amino acid alphabet k-mer enrichment allowed for inserting /identifying genes that were not present in the genome assembly or gene annotation, or gene orthology

We were able to identify putative homologs across species by finding genes that were enriched in common cell types

### key: We built an orthologous gene tree and were able to identify genes in cells across all species
<!-- Is this a separate paper? -->
Using orthogroups, we buitl a signature tree database of orthologous genes.
We then were able to quantify gene expresion
We were able to detect genes that were not present in the genome, but were present in the orthologous genees databases.
We were thus able ot directly compare cell types across species with a common set of orthologous genes.
We used sourmash gather to identify genes that were expressed across mujltiple cell types

### In summary, we developed a method to identify cell types in a non-model organism using a reference atlas from another organism, without the need for a genome or gene annotation from the non-model organism.

In summary, we developed a method to identify both known cell types in a non-model organism using a reference atlas from another organism, without the need for a genome or gene annotation from the non-model organism.
This method can be used to combine single-cell cell atlases from well-annotated,  model organisms, with sequencing data from poorly annotated non-model organisms, to directly find homologous cell types and orthologous genes.
By eliminating read alignment and orthologous gene mapping, `kmermaid` enables comparison of transcriptomes of the remaining 99.999% Eukaryotic species on Earth without submitted genome assemblies, with the cell atlases of a handful of model organisms to identify shared and novel cell types, and *de novo* identify orthologous genes.
By identifying homologous cell types across a broad variety of species, we come closer to an understanding of the evolution of genes, cells, and thus life itself.

`kmermaid` is free and open-source software and is available as Supplementary Data and at http://github.com/czbiohub/kmermaid and as a scalable Nextflow workflow at http://github.com/nf-core/nf-kmermaid.


<!--

## Outline

- Kmers can approximate orthologies
  - Jaccard similarity of orthologues is higher than non-orthologues
  - Benchmarking using https://orthology.benchmarkservice.org/cgi-bin/gateway.pl
  - Finding orthologues
    - Gold standard
      - ENSEMBL COMPARA
      - Quest for Orthologs consortium, Altenhoff, A. M., Boeckmann, B., Capella-Gutierrez, S., Dalquen, D. A., DeLuca, T., et al. (2016). Standardized benchmarking in the quest for orthologs. Nature Methods, 13(5), 425–430. http://doi.org/10.1038/nmeth.3830 [@doi:10.1038/nmeth.3830]
    - Orthologous groups/Conserved Domain Database [@url:https://www.ebi.ac.uk/miriam/main/collections/MIR:00000119]


<!-- ## Figure 2 -- $k$-mers from lossily-encoded putative protein-coding reads faithfully pull out reads from protein-coding genes within amniotes -->


- Overview of kmermaid pipeline
  - Comparison of tissue across species
    - Partition reads to coding/noncoding bins
    - MinHash the Dayhoff-encoded coding sequences
    - Jaccard similarity on the MinHashes
- Which reads are found to have coding features but didn’t map to the genome?
- Do these features map to novel genes or gene fusions?
- Kmers can find correct reading from of RNA-seq reads
  - Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data
- Comparison to other methods: RNASamba [@doi:10.1101/620880]

<!-- ### Figure 3 -- $k$-mers can pull out only reads from transcription factors and Amniotes can be compared on the MinHashes of their protein-coding sequences -->

<!-- ![Figure 3.](images/figure3.svg){#fig:fig3 width="100%"} -->


<!-- ### Some potential references

Gene expression evolution through duplications

- Farre, D., & Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 [@doi:10.1093/molbev/msp242]
- Thornton, J. W., & DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 [@doi:10.1146/annurev.genom.1.1.41]
- Farre, D., & Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 [@doi:10.1093/molbev/msp242]

Taxa-restricted genes

- Human-specific genes in fetal neocortex
Florio, M., Heide, M., Pinson, A., Brandl, H., Albert, M., Winkler, S., et al. (2018). Evolution and cell-type specificity of human-specific genes preferentially expressed in progenitors of fetal neocortex. eLife, 7, D635. http://doi.org/10.7554/eLife.32332 [@doi:10.7554/eLife.32332]
- Insects -- Santos, M. E., Le Bouquin, A., Crumière, A. J. J., & Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 [@doi:10.1126/science.aan2748]


Correlated evolution of celltypes?

- Liang, C., Musser, J. M., Cloutier, A., Prum, R. O., & Wagner, G. P. (2018). Pervasive Correlated Evolution in Gene Expression Shapes Cell and Tissue Type Transcriptomes. Genome Biology and Evolution, 10(2), 538–552. http://doi.org/10.1093/gbe/evy016 [@doi:10.1093/gbe/evy016]

Cell type homology

- Thornton, J. W., & DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 [@doi:10.1146/annurev.genom.1.1.41]
- Tschopp, P., & Tabin, C. J. (2017). Deep homology in the age of next-generation sequencing. Philosophical Transactions of the Royal Society B: Biological Sciences, 372(1713), 20150475–8. http://doi.org/10.1098/rstb.2015.0475 [@doi:10.1098/rstb.2015.0475]
- Hejnol, A., & Lowe, C. J. (2015). Embracing the comparative approach: how robust phylogenies and broader developmental sampling impacts the understanding of nervous system evolution. Philosophical Transactions of the Royal Society B: Biological Sciences, 370(1684), 20150045–16. http://doi.org/10.1098/rstb.2015.0045 [@doi:10.1098/rstb.2015.0045]
- Santos, M. E., Le Bouquin, A., Crumière, A. J. J., & Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 [@doi:10.1126/science.aan2748]
- Mammalian decidual cell

Cell type evolution

- Erkenbrack, E. M., Maziarz, J. D., Griffith, O. W., Liang, C., Chavan, A. R., Nnamani, M. C., & Wagner, G. P. (2018). The mammalian decidual cell evolved from a cellular stress response. PLOS Biology, 16(8), e2005594–27. http://doi.org/10.1371/journal.pbio.2005594 [@doi:10.1371/journal.pbio.2005594] -->

<!-- # Discussion --> -->
