<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Olga Borisovna Botvinnik" />
  <meta name="author" content="Venkata Naga Pranathi Vemuri" />
  <meta name="author" content="N. Tessa Pierce" />
  <meta name="author" content="Phoenix Aja Logan" />
  <meta name="author" content="Saba Nafees" />
  <meta name="author" content="C. Titus Brown" />
  <meta name="author" content="Jim Karkanias" />
  <meta name="dcterms.date" content="2020-08-19" />
  <meta name="keywords" content="comparative transcriptomics, single-cell RNA-seq, k-mer, hashing, MinHash" />
  <title>Sencha directly translates RNA-seq reads to enable functional prediction of transcriptomic ‘dark matter’ across species</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Sencha directly translates RNA-seq reads to enable functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta name="citation_title" content="Sencha directly translates RNA-seq reads to enable functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta property="og:title" content="Sencha directly translates RNA-seq reads to enable functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta property="twitter:title" content="Sencha directly translates RNA-seq reads to enable functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta name="dc.date" content="2020-08-19" />
  <meta name="citation_publication_date" content="2020-08-19" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Olga Borisovna Botvinnik" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0003-4412-7970" />
  <meta name="twitter:creator" content="@olgabot" />
  <meta name="citation_author" content="Venkata Naga Pranathi Vemuri" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0002-5748-9594" />
  <meta name="twitter:creator" content="@pranuvemuri" />
  <meta name="citation_author" content="N. Tessa Pierce" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0002-2942-5331" />
  <meta name="twitter:creator" content="@saltyscientist" />
  <meta name="citation_author" content="Phoenix Aja Logan" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0003-4581-0552" />
  <meta name="twitter:creator" content="@phoenixlogan" />
  <meta name="citation_author" content="Saba Nafees" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_institution" content="Department of Biological Sciences, Texas Tech University" />
  <meta name="citation_author_institution" content="Department of Mathematics &amp; Statistics, Texas Tech University" />
  <meta name="citation_author_orcid" content="0000-0002-3292-7703" />
  <meta name="twitter:creator" content="@sabanafeesTTU" />
  <meta name="citation_author" content="C. Titus Brown" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0001-6001-2677" />
  <meta name="twitter:creator" content="@ctitusbrown" />
  <meta name="citation_author" content="Jim Karkanias" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0002-8057-6055" />
  <meta name="twitter:creator" content="@jkarkanias" />
  <link rel="canonical" href="https://czbiohub.github.io/de-novo-orthology-paper/" />
  <meta property="og:url" content="https://czbiohub.github.io/de-novo-orthology-paper/" />
  <meta property="twitter:url" content="https://czbiohub.github.io/de-novo-orthology-paper/" />
  <meta name="citation_fulltext_html_url" content="https://czbiohub.github.io/de-novo-orthology-paper/" />
  <meta name="citation_pdf_url" content="https://czbiohub.github.io/de-novo-orthology-paper/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://czbiohub.github.io/de-novo-orthology-paper/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://czbiohub.github.io/de-novo-orthology-paper/v/de03029d26c1274ef8d61874c114abd0dc7b19f0/" />
  <meta name="manubot_html_url_versioned" content="https://czbiohub.github.io/de-novo-orthology-paper/v/de03029d26c1274ef8d61874c114abd0dc7b19f0/" />
  <meta name="manubot_pdf_url_versioned" content="https://czbiohub.github.io/de-novo-orthology-paper/v/de03029d26c1274ef8d61874c114abd0dc7b19f0/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
  
  <!-- pandoc-eqnos: equation style -->
  <style>
    .eqnos { display: inline-block; position: relative; width: 100%; }
    .eqnos br { display: none; }
    .eqnos-number { position: absolute; right: 0em; top: 50%; line-height: 0; }
  </style>
</head>
<body>
<header id="title-block-header">
<h1 class="title">Sencha directly translates RNA-seq reads to enable functional prediction of transcriptomic ‘dark matter’ across species</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://czbiohub.github.io/de-novo-orthology-paper/v/de03029d26c1274ef8d61874c114abd0dc7b19f0/">permalink</a>)
was automatically generated
from <a href="https://github.com/czbiohub/de-novo-orthology-paper/tree/de03029d26c1274ef8d61874c114abd0dc7b19f0">czbiohub/de-novo-orthology-paper@de03029</a>
on August 19, 2020.
</em></small></p>
<p><span class="banner lightred"><span class="fas fa-info-circle fa-lg"></span> <strong>This is an in progress manuscript.</strong></span></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Olga Borisovna Botvinnik</strong> <sup><a href="#correspondence">✉</a></sup><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4412-7970">0000-0003-4412-7970</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/olgabot">olgabot</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/olgabot">olgabot</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
<li><p><strong>Venkata Naga Pranathi Vemuri</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-5748-9594">0000-0002-5748-9594</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/pranathivemuri">pranathivemuri</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/pranuvemuri">pranuvemuri</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
<li><p><strong>N. Tessa Pierce</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2942-5331">0000-0002-2942-5331</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/bluegenes">bluegenes</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/saltyscientist">saltyscientist</a><br>
<small>
Department of Population Health and Reproduction, University of California, Davis
· Funded by NSF 1711984
</small></p></li>
<li><p><strong>Phoenix Aja Logan</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4581-0552">0000-0003-4581-0552</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/phoenixAja">phoenixAja</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/phoenixlogan">phoenixlogan</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
<li><p><strong>Saba Nafees</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-3292-7703">0000-0002-3292-7703</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/snafees">snafees</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/sabanafeesTTU">sabanafeesTTU</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub; Department of Biological Sciences, Texas Tech University; Department of Mathematics &amp; Statistics, Texas Tech University
</small></p></li>
<li><p><strong>C. Titus Brown</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6001-2677">0000-0001-6001-2677</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/ctb">ctb</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/ctitusbrown">ctitusbrown</a><br>
<small>
Department of Population Health and Reproduction, University of California, Davis
· Funded by Moore Foundation GBMF4551
</small></p></li>
<li><p><strong>Jim Karkanias</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8057-6055">0000-0002-8057-6055</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/jkensai">jkensai</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/jkarkanias">jkarkanias</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
</ul>
<h2 id="abstract">Abstract</h2>
<p>We introduce <code>sencha</code>, a novel computational method for extracting high-confidence protein-coding sequences from RNA-seq data into non-coding reads and protein-coding reads which can then be manipulated to reduced amino acid alphabets which are directly comparable across large evolutionary distances.
As the direct assignment of protein-coding sequence skips both traditional alignment and gene orthology assignment it can a) be applied to transcriptomes from organisms with no or poorly annotated genomes, and b) identify putative functions of protein sequences contributing to shared cell types.
<!-- Thus, we can identify orthologous cell types while discovering *de novo* orthologous genes across species. -->
Thus, we first leverage short 15-20-long amino acid words in reduced amino acid alphabets to identify homologous cell types across organisms without the need for a reference genome.
Second, we use the homologous cell types from before to find <em>de novo</em> homologous genes across all species, within similar cell types.
<!-- this is actually really cool but a fairly complex concept. maybe split it up?

1 very cool idea is that you can, in an automated fashion, leverage the idea of a reduced AA code in K-mers to ID similar cell types across organisms w/o reference genomes

a second cool feature, resulting from the first, is that you can do this discovery piece.

i feel like impact of each gets lost by having such an efficient sentence.. --></p>
<p>We applied these methods to the problem of understanding the unique ability of bats to harbor zoonotic diseases, especially coronaviruses such as SARS, MERS, SARS-CoV-2, an issue of fundamental importance to human health.
However, important immunological genes such as Interferons have not been adequately identified in the bat genome and thus that aspect of bat biology could not be identified.
With our methods, we were able to detect widespread expression of interferon genes in immune tissues of the SARS reservoir species, and the purported SARS-COV-2 species <em>Rhinolophus sinicus</em> (Chinese horseshoe bat) single-cell RNA-seq transcriptomes, as compared to human tissues.
The <em>R. sinicus</em> interferon genes were slightly different from human interferon genes […]
Additionally, we were able to identify a lymphoid-like cell population in the early chordate <em>Botryllus schlosseri</em> bulk RNA-seq, an early chordate (divergence ~684 millions of years ago), and validate the expression of the lymphoid genes using [qRT-PCR].
By enabling analyses across divergent species’ transcriptomes in a genome-, gene annotation-, and orthology-agnostic manner, <code>sencha</code> illustrates the potential of non-model organisms in building the cell type evolutionary tree of life.</p>
<h2 id="introduction">Introduction</h2>
<!-- Maybe what I'm looking for here is like (1) Why do we care about ID'ing different cell populations (2) how is it normally done? (3) what are the limitations (4) what are new enabling technologies that are untapped to fix the limitations (5) what is your hypothesis (that you can use K-mers to translate scRNAseq from non-ref organisms (6) a brief outline of what you did.

the rest of this could then be blended into the results section to better explain (closer to the data) why you did what you did (then show the results right after the "intro" like text)
-->
<!-- so you're like (1) ID'ing cell types is an important thing (2) scRNA seq is super helpful for this (3) BUT #2 is only true really if you have a reference genome (4) there is a ton of shit we'd want to have cell type IDs for buy dont have ref. genomes. BUT we do have scRNA from them (5) therefore, there is an outstanding need to develop strategies to learn cell types in species w/o first generating genomes (6) #leaftea -->
<p>Identifying cell types in multicellular animals (metazoans) enables studying the broad array of cell types available to living creatures, and can help understand human disease by seeing cell types that are adjacent to human “normal” cell types, but resemble human disease cell types.
Current methods for identifying cell types use single-cell RNA-sequencing, however, these methods require three things: (1) high-quality genome assembly, (2) high-quality predicted gene annotations and (3) high-quality orthologous gene mappings across species.
However, the task of identifying even known cell types in species with poorly annotated genomes is nontrivial, as 99.999% of the predicted 8.7 million Eukaryotic species on Earth have no submitted genome assembly <span class="citation" data-cites="OkFljXKC xJN8WB5G">[<a href="#ref-OkFljXKC" role="doc-biblioref">1</a>,<a href="#ref-xJN8WB5G" role="doc-biblioref">2</a>]</span> and even fewer have gene predictions and orthologous gene mappings to human, which remains an open problem <span class="citation" data-cites="ogAGO9KH wdzgXUGy">[<a href="#ref-ogAGO9KH" role="doc-biblioref">3</a>,<a href="#ref-wdzgXUGy" role="doc-biblioref">4</a>]</span>.
But, we can nontheless generate single-cell RNA-seq data from these species with poorly annotated genomes, and if we can leverage the annotations from existing organisms, we can
Thus, there is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for orthologous gene mapping, gene annotations, or a reference genome.
Short, <span class="math inline">\(k\)</span>-long sequence substrings, or <span class="math inline">\(k\)</span>-mers, have been proposed for clustering single cells <span class="citation" data-cites="wtk13QGK">[<a href="#ref-wtk13QGK" role="doc-biblioref">5</a>]</span> and here we implemented <span class="math inline">\(k\)</span>-mers from putatitvely translated RNA-seq reads with reduced amino acid alphabets <span class="citation" data-cites="Aq7zYbOg 10.1186/1471-2105-12-159 f5u3dDtc Aq7zYbOg QMYjnr6l">[<span class="citeproc-not-found" data-reference-id="10.1186/1471-2105-12-159"><strong>???</strong></span>,<a href="#ref-Aq7zYbOg" role="doc-biblioref">6</a>,<a href="#ref-Aq7zYbOg" role="doc-biblioref">6</a>,<a href="#ref-f5u3dDtc" role="doc-biblioref">7</a>,<a href="#ref-QMYjnr6l" role="doc-biblioref">8</a>]</span>, to find shared cell types across species, and further identify <em>de novo</em> orthologous genes by querying the predicted protein sequences to a reference database.
This method relies solely on divergence time between species, which we show can be estimated from RNA-seq nucleotide <span class="math inline">\(k\)</span>-mers (Supplemental Figure <span class="citation" data-cites="sfig:sfig1">[<span class="citeproc-not-found" data-reference-id="sfig:sfig1"><strong>???</strong></span>]</span>).
As the direct assignment of protein-coding sequence skips both traditional alignment and gene orthology assignment it can a) be applied to transcriptomes from organisms with no or poorly annotated genomes, and b) identify putative functions of protein sequences contributing to shared cell types.</p>
<p>As the direct assignment of protein-coding sequence skips both traditional alignment and gene orthology assignment it can a) be applied to transcriptomes from organisms with no or poorly annotated genomes, and b) identify putative functions of protein sequences contributing to shared cell types.</p>
<p>We benchmark the genome-agnostic method on the Quest for Orthologs Opisthokonta dataset, showing that <span class="math inline">\(k\)</span>-mers from reduced amino acid alphabets are sufficient to estimate orthology.
Using human amino acid sequences, we show that one can extract putative protein-coding reads from 35 Opisthokonta species in Quest for Orthologs, and present the best <span class="math inline">\(k\)</span>-mer size and alphabet for different divergence times.</p>
<p>We first apply this method on a bulk comparative transcriptomic dataset consisting of nine amniote species and six tissues <span class="citation" data-cites="fBCZ87Cf">[<a href="#ref-fBCZ87Cf" role="doc-biblioref">11</a>]</span>, showing that we achieve similar clustering results as using only reads mapping to 1:1 orthologs or Hierarchical Orthologous Groups (HOGs) <span class="citation" data-cites="137npblqI 5RECVcFl 19AeqN3SK">[<a href="#ref-137npblqI" role="doc-biblioref">12</a>,<a href="#ref-5RECVcFl" role="doc-biblioref">13</a>,<a href="#ref-19AeqN3SK" role="doc-biblioref">14</a>]</span> of protein-coding genes, but are able to resolve … which can only be seen by using the <span class="math inline">\(k\)</span>-mer method.
We further demonstrate the utility of this method by comparing transcriptomes from organisms diverged by approximately 676 million years <span class="citation" data-cites="1A6CqUMhm">[<a href="#ref-1A6CqUMhm" role="doc-biblioref">15</a>]</span>: a single-cell atlas of a model organism, mouse from <em>Tabula Muris Senis</em> <span class="citation" data-cites="DXegEVg5">[<a href="#ref-DXegEVg5" role="doc-biblioref">16</a>]</span>, and bulk RNA-seq from <em>Botryllus schlosseri</em> <span class="citation" data-cites="YFejwX4f">[<a href="#ref-YFejwX4f" role="doc-biblioref">17</a>]</span>, a colonial tunicate which exhibits cell populations similar to the myeloid immune lineage.
Across this evolutionary distance, only XX 1:1 orthologous genes exist as found by … and XX HOGs via orthologous matrix (OMA) <span class="citation" data-cites="AlRzKu3c xwCPkKHO">[<a href="#ref-AlRzKu3c" role="doc-biblioref">18</a>,<a href="#ref-xwCPkKHO" role="doc-biblioref">19</a>]</span>
We show that the myeloid-like cells from <em>B. schlosseri</em> not only cluster with the myeloid immune cells from <em>Tabula Muris Senis</em> <span class="citation" data-cites="rSyxV4O8">[<a href="#ref-rSyxV4O8" role="doc-biblioref">20</a>]</span>, we also find <em>de novo</em> orthologous genes, such as …
We find that using <span class="math inline">\(k\)</span>-mers has the advantage of resolving … in comparison to using read counts from 1:1 gene orthologs.
Using <span class="math inline">\(k\)</span>-mers, we were able to resolve cell types … , which was hidden using read counts alone.
Thus, we have shown the reference-free method using the <span class="math inline">\(k\)</span>-mers from single cells is a novel, annotation-agnostic method for comparing cells across species that is capable of identifying cell states unique to a particular organism, helping to build the cell type evolutionary tree of life.</p>
<p>Similar to <span class="math inline">\(k\)</span>-mer based approaches for transcript quantification <span class="citation" data-cites="RIPzCufe 12KZMHMQl vrqQcFyx">[<a href="#ref-RIPzCufe" role="doc-biblioref">21</a>,<a href="#ref-12KZMHMQl" role="doc-biblioref">22</a>,<a href="#ref-vrqQcFyx" role="doc-biblioref">23</a>]</span>, we implemented <span class="math inline">\(k\)</span>-mer based gene expression analyses across species, but instead of using DNA <span class="math inline">\(k\)</span>-mers, our critical innovation was using translated protein <span class="math inline">\(k\)</span>-mers.
We utilized sequence bloom trees (SBTs) <span class="citation" data-cites="DfZvKOpG">[<a href="#ref-DfZvKOpG" role="doc-biblioref">24</a>]</span> using a bottom-up approach to build them similar to previous work <span class="citation" data-cites="1HpdDzfQX AfFgGSHX">[<a href="#ref-1HpdDzfQX" role="doc-biblioref">25</a>,<a href="#ref-AfFgGSHX" role="doc-biblioref">26</a>]</span> to ensure localization of new datasets, meaning, if two leaves share a parent, they are guaranteed to be more similar than two leaves that do not share a parent.</p>
<p>For widespread accessibility and usage, we implemented <code>sencha</code> into two distinct Nextflow pipelines following software best practices such as testing and continuous integration: (1) <code>nf-core/kmermaid</code> to compare translated transcriptomes across divergent species, and (2) <code>czbiohub/nf-predictorthologs</code> to infer functions of translated sequences.</p>
<h2 id="results">Results</h2>
<div id="fig:fig1" class="fignos">
<figure>
<img src="images/SVG/figure1.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 1:</span> Figure 1.
<strong>A.</strong> Overview of transcriptomic datasets from three species: <em>Homo sapiens</em> (human), <em>Rhinolophus sinicus</em> (Chinese horseshoe bat), and the early chordate <em>Botryllus schlosseri</em>. For <em>B. schlosseri</em>, we have FACS-purified bulk RNA-seq populations, including circulating cell types similar to mammalian hematopoietic stem cells (HCS), lymphoid, and myeloid populations. For <em>H. sapiens</em> and <em>R. sinicus</em>, we have droplet-based single-cell RNA-seq data for bone marrow, lung, intestine, spleen, thymus, and whole blood tissues.
<strong>B.</strong> Overview of computational method. First, in the <code>kmermaid</code> pipeline, RNA-seq reads are translated to protein via the Sencha tool, then converted to a reduced amino acid alphabet, decomposed into <span class="math inline">\(k\)</span>-mers, and then those <span class="math inline">\(k\)</span>-mers are subsampled. Next, the <code>nf-predictorthologs</code> pipeline performs differential <span class="math inline">\(k\)</span>-mer expression, much like differential gene expression, to find <span class="math inline">\(k\)</span>-mers that are enriched in a population, and queries for those differential <span class="math inline">\(k\)</span>-mers in databases to identify them.
<strong>C.</strong> Reduced amino acid alphabets allow for flexibility in sequence <span class="math inline">\(k\)</span>-mers. The Dayhoff 6-letter and hydrophobic-polar 2-letter encodings are reduced alphabets that retain biochemical properties of amino acids that have been conserved over many years. The single amino acid change of isoleucine (<code>I</code>) to leucine (<code>L</code>) in the protein sequence <code>PRTEINSEQ</code> to <code>PRTELNSEQ</code>, but has no effect in the Dayhoff or hydrophobic-polar encodings.
<strong>D.</strong> Given a <span class="math inline">\(k\)</span>-mer of interest, we can categorize the <span class="math inline">\(k\)</span>-mer into several bins: (1) <span class="math inline">\(k\)</span>-mer not present in genome assembly; (2) <span class="math inline">\(k\)</span>-mer in genome assembly, but in a repetitive or multi-mapped sequence which are often ignored in downstream analyses; (3) <span class="math inline">\(k\)</span>-mer uniquely mapped, but not in a gene; (4) <span class="math inline">\(k\)</span>-mer in a gene, but not in a known ortholog; and (5) <span class="math inline">\(k\)</span>-mer in a known ortholog.</figcaption>
</figure>
</div>
<p>To determine whether short segments of sequences could detect gene orthologues, we <span class="math inline">\(k\)</span>-merized orthologous genes derived from the ENSEMBL version 97 <span class="citation" data-cites="lBlqrqAO">[<a href="#ref-lBlqrqAO" role="doc-biblioref">27</a>]</span> COMPARA database <span class="citation" data-cites="j0y6s0kf">[<a href="#ref-j0y6s0kf" role="doc-biblioref">28</a>]</span> (Figure [<a href="#fig:fig1">1</a>]).
We compared human protein sequences to orthologous chimpanzee, mouse, (orangutan, bonobo, gorilla, macaque, opossum, platypus, chicken) protein sequences, as these are species used in <span class="citation" data-cites="fBCZ87Cf">[<a href="#ref-fBCZ87Cf" role="doc-biblioref">11</a>]</span>.
In addition to <span class="math inline">\(k\)</span>-merizing the protein-coding sequence, we also re-encoded the protein-coding sequence into a six-letter Dayhoff alphabet <span class="citation" data-cites="66KDHtpG">[<a href="#ref-66KDHtpG" role="doc-biblioref">29</a>]</span>, a nine-letter encoding <span class="citation" data-cites="Rqlso7kk">[<a href="#ref-Rqlso7kk" role="doc-biblioref">9</a>]</span>, and a two-letter hydrophobic-polar encodings <span class="citation" data-cites="1GiMcbPLs E2Kp7kA0">[<a href="#ref-1GiMcbPLs" role="doc-biblioref">30</a>,<a href="#ref-E2Kp7kA0" role="doc-biblioref">31</a>]</span>, show in Table [<a href="#tbl:sequence-encodings">1</a>].</p>
<!-- We found that, consistent with previous knowledge, that 1:1 orthologues had higher $k$-mer similarities as determined by the Jaccard Index. This approach is similar to SwiftOrtho [@doi:10.1093/gigascience/giz118], a k-mer based orthology relationship finder.

Additionally, more recently diverged genes had higher $k$-mer similarity as well.

Across tissues of the same time from the Brawand 2011 [@doi:10.1038/nature10532] dataset, we extracted protein-coding sequences, generated dayhoff signatures of k-mer size length 12, extracted hashes and thus k-mers shared by samples from the same tissue, went back to the original protein sequence, and searched NCBI RefSeq NR for potential proteins.
For each sample, we observed that shared k-mers appeared in 1:1 orthologous genes XX% of the time 1:many orthologs YY% of the time, many:many orthologs ZZ% of the time, in genes not known to be orthologs AA% of the time, in unannotated regions AA% of the time, in multimapped reads BB% of the time, and in unmapped reads CC% of the time.
Overall, we observed XX de novo orthologs in each tissue.
We removed genes that were already known to be orthologous. -->
<p>To recapitulate known biology demonstrated by multi-tissue, whole-organism cell atlases related within the mammalian suborder Euarchontoglires, whose most recent common ancestor is between 85 to 95 million years ago, we applied these methods to a mouse single-cell RNA-seq atlas, <em>Tabula Muris Senis</em> <span class="citation" data-cites="DXegEVg5">[<a href="#ref-DXegEVg5" role="doc-biblioref">16</a>]</span>, a mouse lemur single-cell RNA-seq atlas, <em>Tabula Microcebus</em> <span class="citation" data-cites="https://www.nature.com/articles/d41586-019-01789-0">[<span class="citeproc-not-found" data-reference-id="https://www.nature.com/articles/d41586-019-01789-0"><strong>???</strong></span>]</span>, and a multi-tissue human cell atlas.</p>
<div id="fig:fig2" class="fignos">
<figure>
<img src="images/SVG/figure2.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 2:</span> Figure 2.
<strong>A.</strong> Left, UMAP of human, mouse lemur, and mouse transcriptomes using <span class="math inline">\(k\)</span>-mers similarity in reduced amino acid space, BBKNN-corrected for species. Right, UMAP of human, mouse lemur, and mouse transcriptomes using counts, BBKNN-corrected for species.
<strong>B.</strong> Silhouette plot of Leiden clustering on <span class="math inline">\(k\)</span>-mers, relative to known cell ontology classes.
<strong>C.</strong> Comparison of gene lists from differential <span class="math inline">\(k\)</span>-mer expression, as compared to differential gene expression.
<strong>D.</strong> Barplot showing whether differentially expressed k-mers were found in the genome assembly, and whether they were in a gene annotated as an ortholog, or not.</figcaption>
</figure>
</div>
<!-- To identify novel orthologs among annotated species related by ~300 million years, we first applied these methods on a seminal comparative transcriptomics dataset of human,  chimpanzee, mouse, orangutan, bonobo, gorilla, macaque, opossum, platypus, and chicken [@doi:10.1038/nature10532]. We found .. -->
<p>To identify molecular cell types unique to the Chinese horseshoe bat within mammals, we added a multi-organ, whole organism single-cell RNA-seq atlas of <em>Rhinolophus sinicus</em>. We found expression of interferon genes, which have not been identified in the bat genome, but we were able to detect them in the transcriptome. In the bat scrnaseq paper, they performed qPCR to show expression of interferon genes, even though they weren’t identified in the genome, to show those genes do truly exist.</p>
<!-- To identify common developmental trajectories across >1000 million years of divergence, we merged transcriptomes from *Hydra* [@doi:10.1126/science.aav9314], zebrafish [@doi:10.1126/science.aar4362], and mouse [@doi:10.1038/s41586-019-0969-x]. We demonstrate that the transcriptional signal from ectoderm, mesoderm, and endoderm are conserved when using a protein k-mer size of XX and an XX amino acid encoding. -->
<div id="fig:fig3" class="fignos">
<figure>
<img src="images/SVG/figure3.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 3:</span> Figure 3.
<strong>A.</strong> Left, UMAP of human, mouse lemur, mouse, and <em>Rhinolophus sinicus</em> (Chinese horseshoe bat) transcriptomes using <span class="math inline">\(k\)</span>-mers similarity in reduced amino acid space, BBKNN-corrected for species. Colored by cell compartment.
<strong>B.</strong> Barplot showing whether differentially expressed k-mers were found in the genome assembly, and whether they were in a gene annotated as an ortholog, or not.
<strong>C.</strong> UMAP of human, mouse lemur, mouse and bat immune transcriptomes, colored by interferon-like gene expression.
<strong>D.</strong> Alignment of assembled Bat interferon protein sequences with known human sequences.</figcaption>
</figure>
</div>
<h2 id="discussion">Discussion</h2>
<p><code>kmermaid</code> implements the concept of lightweight orthology assignment using k-mers to the problem of cross-species RNA-seq analyses and achieves unprecedented speed of analysis. By removing the orthology inference step, <code>kmermaid</code> opens up the possibilty of finding shared and divergent tissue and cell types across a broad range of species, paving the way for evolutionary analyses of cell types across species. <code>kmermaid</code> can be used in <em>de novo</em> setting for non-model organisms, finding similar cell types within an organism, or finding similar cell types relative to a reference organism, without the need for a reference genome or transcriptome. The memory usage of <code>kmermaid</code> is quite low, using only 50MB for extracting coding sequences and 50MB for assigning protein k-mer signatures. As the number of RNA-seq datasets, especially single-cell RNA-seq datasets continues to grow, we expect <code>kmermaid</code> to be widely used for identifying cell types in non-model organisms.</p>
<p>In summary, we developed a method to identify both known cell types in a non-model organism using a reference atlas from another organism, without the need for a genome or gene annotation from the non-model organism.
This method can be used to combine single-cell cell atlases from well-annotated, model organisms, with sequencing data from poorly annotated non-model organisms, to directly find homologous cell types and orthologous genes.
By eliminating read alignment and orthologous gene mapping, <code>kmermaid</code> enables comparison of transcriptomes of the remaining 99.999% Eukaryotic species on Earth without submitted genome assemblies, with the cell atlases of a handful of model organisms to identify shared and novel cell types, and <em>de novo</em> identify orthologous genes.
By identifying homologous cell types across a broad variety of species, we come closer to an understanding of the evolution of genes, cells, and thus life itself.</p>
<p><code>kmermaid</code> is free and open-source software and is available as Supplementary Data and at http://github.com/czbiohub/kmermaid and as a scalable Nextflow workflow at http://github.com/nf-core/nf-kmermaid.</p>
<h2 id="outline">Outline</h2>
<ul>
<li>Kmers can approximate orthologies
<ul>
<li>Jaccard similarity of orthologues is higher than non-orthologues</li>
<li>Benchmarking using https://orthology.benchmarkservice.org/cgi-bin/gateway.pl</li>
<li>Finding orthologues
<ul>
<li>Gold standard
<ul>
<li>ENSEMBL COMPARA</li>
<li>Quest for Orthologs consortium, Altenhoff, A. M., Boeckmann, B., Capella-Gutierrez, S., Dalquen, D. A., DeLuca, T., et al. (2016). Standardized benchmarking in the quest for orthologs. Nature Methods, 13(5), 425–430. http://doi.org/10.1038/nmeth.3830 <span class="citation" data-cites="LkrS4xXB">[<a href="#ref-LkrS4xXB" role="doc-biblioref">32</a>]</span></li>
</ul></li>
<li>Orthologous groups/Conserved Domain Database <span class="citation" data-cites="bkF0801R">[<a href="#ref-bkF0801R" role="doc-biblioref">33</a>]</span></li>
</ul></li>
</ul></li>
</ul>
<!-- ## Figure 2 -- $k$-mers from lossily-encoded putative protein-coding reads faithfully pull out reads from protein-coding genes within amniotes -->
<ul>
<li>Overview of kmermaid pipeline
<ul>
<li>Comparison of tissue across species
<ul>
<li>Partition reads to coding/noncoding bins</li>
<li>MinHash the Dayhoff-encoded coding sequences</li>
<li>Jaccard similarity on the MinHashes</li>
</ul></li>
</ul></li>
<li>Which reads are found to have coding features but didn’t map to the genome?</li>
<li>Do these features map to novel genes or gene fusions?</li>
<li>Kmers can find correct reading from of RNA-seq reads
<ul>
<li>Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data</li>
</ul></li>
<li>Comparison to other methods: RNASamba <span class="citation" data-cites="AtPPptbH">[<a href="#ref-AtPPptbH" role="doc-biblioref">34</a>]</span></li>
</ul>
<!-- ### Figure 3 -- $k$-mers can pull out only reads from transcription factors and Amniotes can be compared on the MinHashes of their protein-coding sequences -->
<!-- ![Figure 3.](images/figure3.svg){#fig:fig3 width="100%"} -->
<ul>
<li>Kmers can find only transcription factor reads of TFs from RNA-seq reads
<ul>
<li>Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data</li>
</ul></li>
</ul>
<h3 id="some-potential-references">Some potential references</h3>
<p>Gene expression evolution through duplications</p>
<ul>
<li>Farre, D., &amp; Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 <span class="citation" data-cites="1DcTxE8Z3">[<a href="#ref-1DcTxE8Z3" role="doc-biblioref">35</a>]</span></li>
<li>Thornton, J. W., &amp; DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 <span class="citation" data-cites="uQLS9NYU">[<a href="#ref-uQLS9NYU" role="doc-biblioref">36</a>]</span></li>
<li>Farre, D., &amp; Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 <span class="citation" data-cites="1DcTxE8Z3">[<a href="#ref-1DcTxE8Z3" role="doc-biblioref">35</a>]</span></li>
</ul>
<p>Taxa-restricted genes</p>
<ul>
<li>Human-specific genes in fetal neocortex
Florio, M., Heide, M., Pinson, A., Brandl, H., Albert, M., Winkler, S., et al. (2018). Evolution and cell-type specificity of human-specific genes preferentially expressed in progenitors of fetal neocortex. eLife, 7, D635. http://doi.org/10.7554/eLife.32332 <span class="citation" data-cites="nbBjh1sL">[<a href="#ref-nbBjh1sL" role="doc-biblioref">37</a>]</span></li>
<li>Insects – Santos, M. E., Le Bouquin, A., Crumière, A. J. J., &amp; Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 <span class="citation" data-cites="lJHQuvIH">[<a href="#ref-lJHQuvIH" role="doc-biblioref">38</a>]</span></li>
</ul>
<p>Correlated evolution of celltypes?</p>
<ul>
<li>Liang, C., Musser, J. M., Cloutier, A., Prum, R. O., &amp; Wagner, G. P. (2018). Pervasive Correlated Evolution in Gene Expression Shapes Cell and Tissue Type Transcriptomes. Genome Biology and Evolution, 10(2), 538–552. http://doi.org/10.1093/gbe/evy016 <span class="citation" data-cites="FTv2KYrZ">[<a href="#ref-FTv2KYrZ" role="doc-biblioref">39</a>]</span></li>
</ul>
<p>Cell type homology</p>
<ul>
<li>Thornton, J. W., &amp; DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 <span class="citation" data-cites="uQLS9NYU">[<a href="#ref-uQLS9NYU" role="doc-biblioref">36</a>]</span></li>
<li>Tschopp, P., &amp; Tabin, C. J. (2017). Deep homology in the age of next-generation sequencing. Philosophical Transactions of the Royal Society B: Biological Sciences, 372(1713), 20150475–8. http://doi.org/10.1098/rstb.2015.0475 <span class="citation" data-cites="hdRo8a7z">[<a href="#ref-hdRo8a7z" role="doc-biblioref">40</a>]</span></li>
<li>Hejnol, A., &amp; Lowe, C. J. (2015). Embracing the comparative approach: how robust phylogenies and broader developmental sampling impacts the understanding of nervous system evolution. Philosophical Transactions of the Royal Society B: Biological Sciences, 370(1684), 20150045–16. http://doi.org/10.1098/rstb.2015.0045 <span class="citation" data-cites="1GRtpoh49">[<a href="#ref-1GRtpoh49" role="doc-biblioref">41</a>]</span></li>
<li>Santos, M. E., Le Bouquin, A., Crumière, A. J. J., &amp; Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 <span class="citation" data-cites="lJHQuvIH">[<a href="#ref-lJHQuvIH" role="doc-biblioref">38</a>]</span></li>
<li>Mammalian decidual cell</li>
</ul>
<p>Cell type evolution</p>
<ul>
<li>Erkenbrack, E. M., Maziarz, J. D., Griffith, O. W., Liang, C., Chavan, A. R., Nnamani, M. C., &amp; Wagner, G. P. (2018). The mammalian decidual cell evolved from a cellular stress response. PLOS Biology, 16(8), e2005594–27. http://doi.org/10.1371/journal.pbio.2005594 <span class="citation" data-cites="8cGr6TE">[<a href="#ref-8cGr6TE" role="doc-biblioref">42</a>]</span></li>
</ul>
<!-- # Discussion -->
<h2 id="online-methods">Online Methods</h2>
<!-- This is just a placeholder to separate the supplemental methods from the rest of the paper.
All remaining supplemental sections should fall under this, meaning they need to be formatted
with heading 3 with three hashes: `###` -->
<h3 id="implementation">Implementation</h3>
<h4 id="reduced-alphabets">Reduced alphabets</h4>
<p>At the core of <code>sencha</code> is the ability to cheaply compare sequences using <span class="math inline">\(k\)</span>-mers.
As <span class="math inline">\(k\)</span>-mers are very brittle to substitutions and thus to compare across species, one must allow for minor base substitutions that still maintain similar chemical or structural properties.
A reduced alphabet can encode useful information into a smaller alphabet space, and enable sequence comparisons across a broader variety of species than the original alphabet alone.</p>
<h5 id="reduced-amino-acid-alphabets">Reduced amino acid alphabets</h5>
<p>Reduced amino acid alphabets have been useful for over 50 years <span class="citation" data-cites="66KDHtpG">[<a href="#ref-66KDHtpG" role="doc-biblioref">29</a>]</span> in finding related protein sequences <span class="citation" data-cites="Aq7zYbOg 16Kv6RIe3 sZoryWIB f5u3dDtc QMYjnr6l">[<a href="#ref-Aq7zYbOg" role="doc-biblioref">6</a>,<a href="#ref-f5u3dDtc" role="doc-biblioref">7</a>,<a href="#ref-QMYjnr6l" role="doc-biblioref">8</a>,<a href="#ref-16Kv6RIe3" role="doc-biblioref">10</a>,<a href="#ref-sZoryWIB" role="doc-biblioref">43</a>]</span>.
Recently, a reduced amino acid alphabet (specifically, <code>aa9</code> below) combined with <span class="math inline">\(k\)</span>-mers were used to find homologous protein-coding sequences <span class="citation" data-cites="Rqlso7kk">[<a href="#ref-Rqlso7kk" role="doc-biblioref">9</a>]</span>.
We build on this concept by enabling prediction of protein-coding sequences from RNA-seq reads, and by enabling users to perform a parameter sweep in an all-by-all comparison to identify putative homologs using a variety of alphabet metrics.</p>
<h6 id="dayhoff-and-hp-alphabets">Dayhoff and <code>HP</code> alphabets</h6>
<div id="tbl:sequence-encodings" class="tablenos">
<table>
<caption><span>Table 1:</span> Dayhoff and hydrophobic-polar encodings are a reduced amino acid
alphabet allowing for permissive cross-species sequence comparisons. For
example, the amino acid sequence <code>SASHAFIERCE</code> would be Dayhoff-encoded
to <code>bbbdbfecdac</code>, and HP-encoded to <code>phpphhhpppp</code>, as below. </caption>
<thead>
<tr class="header">
<th style="text-align: left;">Amino acid</th>
<th style="text-align: left;">Property</th>
<th style="text-align: left;">Dayhoff</th>
<th style="text-align: left;">Hydrophobic-polar (HP)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>C</code></td>
<td style="text-align: left;">Sulfur polymerization</td>
<td style="text-align: left;"><code>a</code></td>
<td style="text-align: left;"><code>p</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>A</code>, <code>G</code>, <code>P</code>, <code>S</code>, <code>T</code></td>
<td style="text-align: left;">Small</td>
<td style="text-align: left;"><code>b</code></td>
<td style="text-align: left;"><code>AGP</code>: <code>h</code><br><code>ST</code>: <code>p</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>D</code>, <code>E</code>, <code>N</code>, <code>Q</code></td>
<td style="text-align: left;">Acid and amide</td>
<td style="text-align: left;"><code>c</code></td>
<td style="text-align: left;"><code>p</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>H</code>, <code>K</code>, <code>R</code></td>
<td style="text-align: left;">Basic</td>
<td style="text-align: left;"><code>d</code></td>
<td style="text-align: left;"><code>p</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>I</code>, <code>L</code>, <code>M</code>, <code>V</code></td>
<td style="text-align: left;">Hydrophobic</td>
<td style="text-align: left;"><code>e</code></td>
<td style="text-align: left;"><code>h</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>F</code>, <code>W</code>, <code>Y</code></td>
<td style="text-align: left;">Aromatic</td>
<td style="text-align: left;"><code>f</code></td>
<td style="text-align: left;"><code>h</code></td>
</tr>
</tbody>
</table>
</div>
<pre><code>protein20: SASHAFIERCE
dayhoff6:  bbbdbfecdac
hp2:       phpphhhpppp</code></pre>
<h6 id="all-implemented-alphabets-with-citations-not-as-nicely-organized">All implemented alphabets (with citations, not as nicely organized)</h6>
<!-- Copied this google spreadsheet https://docs.google.com/spreadsheets/d/1RDuQD0aRyv-FnQJbjRosHEJV85_9BNrzmmvgFRQSgN8/edit#gid=0 into https://thisdavej.com/copy-table-in-excel-and-paste-as-a-markdown-table/, reformatted with https://atom.io/packages/markdown-table-formatter -->
<p>[NOTE: maybe this should go into the supplementary? The main alphabets that have been successful for me are dayhoff and HP]</p>
<table>
<colgroup>
<col style="width: 32%" />
<col style="width: 12%" />
<col style="width: 54%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Citation</th>
<th style="text-align: left;">Alphabet</th>
<th style="text-align: left;">Amino acid groups</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Phillips, R., <em>et al</em>. (2012). <span class="citation" data-cites="1GiMcbPLs">[<a href="#ref-1GiMcbPLs" role="doc-biblioref">30</a>]</span></td>
<td style="text-align: left;">hp2</td>
<td style="text-align: left;"><code>AFGILMPVWY</code> <code>CDEHKNQRST</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Peterson, E. L., <em>et al</em>. (2009) <span class="citation" data-cites="Aq7zYbOg">[<a href="#ref-Aq7zYbOg" role="doc-biblioref">6</a>]</span></td>
<td style="text-align: left;">gbmr4</td>
<td style="text-align: left;"><code>G</code> <code>ADKERNTSQ</code> <code>YFLIVMCWH</code> <code>P</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Dayhoff, M. O., &amp; Eck, R. V. (1968). <span class="citation" data-cites="66KDHtpG">[<a href="#ref-66KDHtpG" role="doc-biblioref">29</a>]</span></td>
<td style="text-align: left;">dayhoff6</td>
<td style="text-align: left;"><code>AGPST</code> <code>HRK</code> <code>DENQ</code> <code>FWY</code> <code>ILMV</code> <code>C</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">This paper</td>
<td style="text-align: left;">botvinnik8</td>
<td style="text-align: left;"><code>AG</code> <code>DE</code> <code>RK</code> <code>NQ</code> <code>ST</code> <code>FY</code> <code>LIV</code> <code>CMWHP</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Hu, X., &amp; Friedberg, I. (2019). <span class="citation" data-cites="Rqlso7kk">[<a href="#ref-Rqlso7kk" role="doc-biblioref">9</a>]</span></td>
<td style="text-align: left;">aa9</td>
<td style="text-align: left;"><code>G</code> <code>AST</code> <code>KR</code> <code>EQ</code> <code>DN</code> <code>CFILMVY</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Peterson, E. L., <em>et al</em>. (2009) <span class="citation" data-cites="Aq7zYbOg">[<a href="#ref-Aq7zYbOg" role="doc-biblioref">6</a>]</span></td>
<td style="text-align: left;">sdm12</td>
<td style="text-align: left;"><code>G</code> <code>A</code> <code>D</code> <code>KER</code> <code>N</code> <code>TSQ</code> <code>YF</code> <code>LIVM</code> <code>C</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Peterson, E. L., <em>et al</em>. (2009) <span class="citation" data-cites="Aq7zYbOg">[<a href="#ref-Aq7zYbOg" role="doc-biblioref">6</a>]</span></td>
<td style="text-align: left;">hsdm17</td>
<td style="text-align: left;"><code>G</code> <code>A</code> <code>D</code> <code>KE</code> <code>R</code> <code>N</code> <code>T</code> <code>S</code> <code>Q</code> <code>Y</code> <code>F</code> <code>LIV</code> <code>M</code> <code>C</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Dayhoff, M. O., &amp; Eck, R. V. (1968). <span class="citation" data-cites="66KDHtpG">[<a href="#ref-66KDHtpG" role="doc-biblioref">29</a>]</span></td>
<td style="text-align: left;">protein20</td>
<td style="text-align: left;"><code>G</code> <code>A</code> <code>D</code> <code>E</code> <code>K</code> <code>R</code> <code>N</code> <code>T</code> <code>S</code> <code>Q</code> <code>Y</code> <code>F</code> <code>L</code> <code>I</code> <code>V</code> <code>M</code> <code>C</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
</tbody>
</table>
<h5 id="reduced-nucleotide-alphabets">Reduced nucleotide alphabets</h5>
<p>The IUPAC degenerate nucleotide code <span class="citation" data-cites="11LSoLYLW">[<a href="#ref-11LSoLYLW" role="doc-biblioref">44</a>]</span> includes several two-letter codes for the original 4-letter nucleobase alphabet.
The first, Weak/Strong, indicates the strength of the hydrogen bond across the double strand.
The bond of adenine to thymine has two hydrogen bonds, making it weak; and the bond of guanine to cytosine has three hydrogen bonds, making it 50% stronger.
The second, Purine/Pyrimidine, encodes the ring size of the nucleobase, where Adenine and Guanine both have larger Purine double rings, while Cytosine and Thymine/Uracil have smaller Pyrimidine rings.
The third, Amino/Keto, designates the main functional group of the ring, where Adenine and Cytosine both have an Amino group, while Guanine and Thymine/Uracil both have a Keto group.</p>
<!-- This spreadsheet:https://docs.google.com/spreadsheets/d/1RDuQD0aRyv-FnQJbjRosHEJV85_9BNrzmmvgFRQSgN8/edit#gid=104831627 was copied into this formatter: https://thisdavej.com/copy-table-in-excel-and-paste-as-a-markdown-table/ -->
<table>
<colgroup>
<col style="width: 10%" />
<col style="width: 15%" />
<col style="width: 13%" />
<col style="width: 19%" />
<col style="width: 41%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Nucleotide</th>
<th style="text-align: left;">Hydrogen Bonding</th>
<th style="text-align: left;">Ring type</th>
<th style="text-align: left;">Ring functional group</th>
<th style="text-align: left;">Nucleobase chemical structure</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">A</td>
<td style="text-align: left;">Weak (W)</td>
<td style="text-align: left;">Purine (R)</td>
<td style="text-align: left;">Amino (M)</td>
<td style="text-align: left;"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/Adenine.svg/120px-Adenine.svg.png" title="fig:" alt="Adenine chemical structure" /></td>
</tr>
<tr class="even">
<td style="text-align: left;">C</td>
<td style="text-align: left;">Strong (S)</td>
<td style="text-align: left;">Pyrimidine (Y)</td>
<td style="text-align: left;">Amino (M)</td>
<td style="text-align: left;"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/1/10/Cytosine_chemical_structure.png/120px-Cytosine_chemical_structure.png" title="fig:" alt="Cytosine chemical structure" /></td>
</tr>
<tr class="odd">
<td style="text-align: left;">G</td>
<td style="text-align: left;">Strong (S)</td>
<td style="text-align: left;">Purine (R)</td>
<td style="text-align: left;">Keto (K)</td>
<td style="text-align: left;"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/3f/Guanin.svg/150px-Guanin.svg.png" title="fig:" alt="Guanine chemical structure" /></td>
</tr>
<tr class="even">
<td style="text-align: left;">T</td>
<td style="text-align: left;">Weak (W)</td>
<td style="text-align: left;">Pyrimidine (Y)</td>
<td style="text-align: left;">Keto (K)</td>
<td style="text-align: left;"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/ab/Thymine_skeletal.svg/120px-Thymine_skeletal.svg.png" title="fig:" alt="Thymine chemical structure" /></td>
</tr>
<tr class="odd">
<td style="text-align: left;">U</td>
<td style="text-align: left;">Weak (W)</td>
<td style="text-align: left;">Pyrimidine (Y)</td>
<td style="text-align: left;">Keto (K)</td>
<td style="text-align: left;"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b4/Uracil.svg/120px-Uracil.svg.png" title="fig:" alt="Uracil chemical structure" /></td>
</tr>
</tbody>
</table>
<p>Thus, the nucleotide string <code>GATTACA</code> would be re-encoded into the following:</p>
<pre><code>Nucleotide:        GATTACA
Weak/Strong:       SWWWWSW
Purine/Pyrimidine: RRYYRYR
Functional group:  KMKKMMM</code></pre>
<h4 id="sencha-translate"><code>sencha translate</code></h4>
<figure>
<img src="images/SVG/supplemental_figure1.svg" id="sfig1:supplemental_figure1" data-tag="supplemental_figure1" style="width:100.0%" alt="" /><figcaption>Supplemental Figure 1: Overview of <code>sencha translate</code> <strong>B.</strong> First, each read is translated into all six possible protein-coding translation frames. Next, reading frames with stop codons are eliminated. Each protein-coding frame is <span class="math inline">\(k\)</span>-merized, then the fraction of <span class="math inline">\(k\)</span>-mers which appear in the known protein-coding database is computed. Frames which contain a fraction of coding frames exceeding the threshold are inferred to be putatively protein-coding. <strong>B.</strong> Worked example of an RNA-seq read with a single putatitive reading frame. <strong>C.</strong> Worked example of an RNA-seq read with multiple reading frames, and a <code>UCSC</code> genome browser shot of the read showing that both reading frames are present in the annotation.</figcaption>
</figure>
<h5 id="set-jaccard-threshold-of-translate-by-controlling-false-positive-rate-of-protein-coding-prediction">Set Jaccard threshold of <code>translate</code> by controlling false positive rate of protein-coding prediction</h5>
<p>To set a threshold of the minimum Jaccard overlap between a translated read’s frame and the reference proteome, the most statistically principled way is to control the false positive rate of predicing a protein-coding read.</p>
<h6 id="probability-of-random-k-mers-from-a-read">Probability of random <span class="math inline">\(k\)</span>-mers from a read</h6>
<p>If <span class="math inline">\(k\)</span>-mers from reads were independent, identically distributed (i.i.d.) variables, then a translated read of length <span class="math inline">\(L_{\mathrm{translated}}\)</span> drawing letters from the alphabet <span class="math inline">\(\Sigma\)</span>, whose size is <span class="math inline">\(|\Sigma|\)</span>, would contain</p>
<p><span id="eq:n_kmers_per_read_iid" class="eqnos"><span class="math display">\[\left( \frac{1}{\left| \Sigma \right|^k} \right)^{L_{\mathrm{translated}} - k + 1}\]</span><span class="eqnos-number">(1)</span></span> </p>
<p>However, <span class="math inline">\(k\)</span>-mers drawn from reads are not i.i.d.
Let’s take a simple example.
If we have a two-letter alphabet, <span class="math inline">\(\Sigma = \\{a, b\\},\)</span>, thus <span class="math inline">\(|\Sigma| = 2\)</span>.
Let us use an example sequence <span class="math inline">\(S = abbabba\)</span>.
If <span class="math inline">\(k = 4\)</span>, then the first <span class="math inline">\(k\)</span>-mer is <span class="math inline">\(abba\)</span>.
The second <span class="math inline">\(k\)</span>-mer is thus either <span class="math inline">\(bbaa\)</span> or <span class="math inline">\(bbab\)</span>, with equal probability.
We can generalize this: Given the first <span class="math inline">\(k\)</span>-mer, the first <span class="math inline">\(k-1\)</span> letters from the second <span class="math inline">\(k\)</span>-mer are known, and thus the probability of guessing the next <span class="math inline">\(k\)</span>-mer is <span class="math inline">\(\frac{1}{\left|\Sigma\right|}\)</span>.</p>
<figure>
<img src="images/abba_sequence_prob_future_kmers.png" id="sfig:prob_kmers_in_read" data-tag="prob_kmers_in_read" style="width:50.0%" alt="" /><figcaption>Probability of future <span class="math inline">\(k\)</span>-mers is influenced by the existence of previous <span class="math inline">\(k\)</span>-mers.</figcaption>
</figure>
<p>Thus, the probability of a random <span class="math inline">\(k\)</span>-mer from a sequencing read is completely dependent on the alphabet size <span class="math inline">\(|\Sigma|\)</span> and its translated sequence length, <span class="math inline">\(L_{\mathrm{translated}}\)</span>:</p>
<p><span id="eq:prob_kmers_in_read" class="eqnos"><span class="math display">\[
\begin{align}
\mathrm{Pr}(\mathrm{FPR}) &amp;= \left(\frac{1}{|\Sigma|}\right)^k \times \left( \frac{1}{|\Sigma|} \right)^{L_{\mathrm{translated}} - k }\\
&amp;=\frac{1}{| \Sigma |^k  \times |\Sigma|^{L_{\mathrm{translated} - k} }}\\
&amp;= \frac{1}{| \Sigma|^{L_{\mathrm{translated}}}}
\end{align}
\]</span><span class="eqnos-number">(2)</span></span> </p>
<h6 id="bloom-filter-collision-probability">Bloom filter collision probability</h6>
<p>The probability of error of the <code>khmer</code> bloom filter implementation <span class="citation" data-cites="1DpHWPZC9">[<a href="#ref-1DpHWPZC9" role="doc-biblioref">45</a>]</span> used in <code>sencha</code>, given <span class="math inline">\(N\)</span> distinct <span class="math inline">\(k\)</span>-mers counted, a hash table size of <span class="math inline">\(H\)</span>, and <span class="math inline">\(Z\)</span> total number of hash tables, is</p>
<p><span id="eq:bloom_filter_fpr" class="eqnos"><span class="math display">\[\mathrm{Pr}(\mathrm{FPR}_{\mathrm{bloom}}) = \left( 1 - \exp^{N/H}\right)^Z.\]</span><span class="eqnos-number">(3)</span></span> </p>
<p>Theoretically, the total number of <span class="math inline">\(k\)</span>-mers is limited by the alphabet size and choice of <span class="math inline">\(k\)</span>.
Empirically, the number of possible <span class="math inline">\(k\)</span>-mers is limited by the <span class="math inline">\(k\)</span>-mers which are compatible with life, and by <span class="math inline">\(k=5\)</span>, the number of theoretical protein <span class="math inline">\(k\)</span>-mers exceeds the number of observed protein <span class="math inline">\(k\)</span>-mers.
Additionally, the mass of all possible <span class="math inline">\(k\)</span>-mers of a certain size, exceeds the mass of the planet earth by <span class="math inline">\(k = X\)</span> [get the data for this].
The UniProtKB Opisthokonta manually reviewed dataset contains <span class="math inline">\(4.8 \times 10^7\)</span> <span class="math inline">\(7\)</span>-mers in the protein alphabet.
Thus, we can give an upper bound to the number of theoretical <span class="math inline">\(k\)</span>-mers to be <span class="math inline">\(10^8\)</span>.
Therefore, the total number of <span class="math inline">\(k\)</span>-mers in the bloom filter is,</p>
<p><span id="eq:n_proteome_kmers" class="eqnos"><span class="math display">\[N = \min\left( 10^8, |\Sigma|^k \right).\]</span><span class="eqnos-number">(4)</span></span> </p>
<figure>
<img src="images/ksize_vs_n_kmers_theoretical_observed.png" alt="" /><figcaption>Number of theoretical <span class="math inline">\(k\)</span>-mers given alphabet size, compared to observed <span class="math inline">\(k\)</span>-mers in ENSEMBL human translated proteome. The number of observed <span class="math inline">\(k\)</span>-mers is distinct from the number of theoretical <span class="math inline">\(k\)</span>-mers, as the total number of observed <span class="math inline">\(k\)</span>-mers is limted by <span class="math inline">\(k\)</span>-mers compatible with life. Rerun this with uniprot uniref data.</figcaption>
</figure>
<h6 id="false-positive-rate-of-protein-coding-prediction">False positive rate of protein-coding prediction</h6>
<p>Combining Equations <a href="#eq:prob_kmers_in_read">2</a>, <a href="#eq:n_proteome_kmers">4</a>, and <a href="#eq:bloom_filter_fpr">3</a>, for an RNA-seq read of length <span class="math inline">\(L\)</span> where its translated length <span class="math inline">\(L_{\mathrm{translated}} = \lfloor \frac{L}{3} \rfloor\)</span>, containing a possible six frames of translation, then the false-positive rate (FPR) protein-coding read is,</p>
<p><span id="eq:fpr_protein_coding" class="eqnos"><span class="math display">\[
\begin{align}
\mathrm{Pr}\left(\mathrm{FPR}\right) &amp;=
6 \times \left( 1 - \exp^{\min\left( 10^8, |\Sigma|^k \right)} \right)^Z \times \frac{1}{\left| \Sigma \right|^{L_{\mathrm{translated}}}}.
\end{align}
\]</span><span class="eqnos-number">(5)</span></span> </p>
<p>Probability of a random <span class="math inline">\(k\)</span>-mer being present in a perfect protein index
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\frac{N}{\left| \Sigma \right| ^k}
\]</span><span class="eqnos-number">(6)</span></span> </p>
<p>Probability of <em>all</em> <span class="math inline">\(L - k + 1\)</span> <span class="math inline">\(k\)</span>-mers of a read being present in a perfect protein index
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\left( \frac{N}{\left| \Sigma \right| ^k} \right) ^ { L - k + 1 }
\]</span><span class="eqnos-number">(7)</span></span> </p>
<p>Probability of a random <span class="math inline">\(k\)</span>-mer being present in an <em>imperfect</em> protein bloom filter with false positive rate or “collision rate” <span class="math inline">\(p_{\mathrm{collision}}\)</span>
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\frac{N}{p_{\mathrm{collision}} * \left| \Sigma \right| ^k}
\]</span><span class="eqnos-number">(8)</span></span> </p>
<p>Probability of <em>all</em> <span class="math inline">\(L - k + 1\)</span> <span class="math inline">\(k\)</span>-mers of a read being present in an <em>imperfect</em> protein bloom filter with false positive rate or “collision rate” <span class="math inline">\(p_{\mathrm{collision}}\)</span>
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\left( \frac{N}{p_{\mathrm{collision}} * \left| \Sigma \right| ^k} \right) ^ { L - k + 1 }
\]</span><span class="eqnos-number">(9)</span></span> </p>
<h5 id="similarity-thresholds-for-percentage-of-matching-k-mers">Similarity thresholds for percentage of matching <span class="math inline">\(k\)</span>-mers</h5>
<p>A single SNP in a read affects <span class="math inline">\(k\)</span> <span class="math inline">\(k\)</span>-mers.</p>
<h4 id="sencha-compare-kmer-content-performs-all-by-all-or-pairwise-k-mer-similarity-of-protein-or-nucleotide-sequences-using-reduced-alphabets"><code>sencha compare-kmer-content</code> performs all-by-all or pairwise k-mer similarity of protein or nucleotide sequences using reduced alphabets</h4>
<figure>
<img src="images/SVG/figure3.svg" id="sfig:figure3" data-tag="figure3" style="width:100.0%" alt="" /><figcaption>Overview of <code>sencha compare-kmer-content</code> <strong>A.</strong> Protein sequences are <span class="math inline">\(k\)</span>-merized by converting into a bag of words using a sliding window of size <span class="math inline">\(k\)</span>, potentially re-encoded to a lossy alphabet, and then their fraction of overlapping <span class="math inline">\(k\)</span>-mers is computed into a Jaccard similarity. <strong>B.</strong> One option for <code>sencha compare-kmer-content</code> is to specify a pair of sequence files, and compute a background of <span class="math inline">\(k\)</span>-mer similarty using randomly shuffled pairs. <strong>C.</strong> Another option for <code>sencha compare-kmer-content</code> is to do an all-by-all <span class="math inline">\(k\)</span>-mer similarity comparison.</figcaption>
</figure>
<h3 id="benchmarking">Benchmarking</h3>
<p>Methods go here.</p>
<h4 id="computational">Computational</h4>
<figure>
<img src="images/SVG/supplemental_figure2.svg" id="sfig:sfig2" data-tag="sfig2" style="width:100.0%" alt="" /><figcaption>Supplemental Figure 2. <strong>D.</strong> Example of predicting protein-coding sequence using Brawand2011 RNA-seq data, and human proteome as the reference. x-axis, percentage of reads falling into that category, y-axis, the species which the reads are from.</figcaption>
</figure>
<h5 id="k-mer-comparison-of-orthologous-genes"><span class="math inline">\(k\)</span>-mer comparison of orthologous genes</h5>
<p>We used ENSEMBL version 97.
We did things.
One sentence per line.
Prefer DOI for references, but for Biorxiv use the URL.
DOI example: <span class="citation" data-cites="LkrS4xXB">[<a href="#ref-LkrS4xXB" role="doc-biblioref">32</a>]</span>.
Biorxiv example: <span class="citation" data-cites="46R7aU4N">[<a href="#ref-46R7aU4N" role="doc-biblioref">46</a>]</span>.
Multiple citations per line example: <span class="citation" data-cites="LkrS4xXB 46R7aU4N">[<a href="#ref-LkrS4xXB" role="doc-biblioref">32</a>,<a href="#ref-46R7aU4N" role="doc-biblioref">46</a>]</span>.</p>
<h5 id="extraction-of-putative-coding-reads-from-rna-seq">Extraction of putative coding reads from RNA-seq</h5>
<p>We did things.
One sentence per line.
Prefer DOI for references, but for Biorxiv use the URL.
DOI example: <span class="citation" data-cites="LkrS4xXB">[<a href="#ref-LkrS4xXB" role="doc-biblioref">32</a>]</span>.
Biorxiv example: <span class="citation" data-cites="46R7aU4N">[<a href="#ref-46R7aU4N" role="doc-biblioref">46</a>]</span>.
Multiple citations per line example: <span class="citation" data-cites="LkrS4xXB 46R7aU4N">[<a href="#ref-LkrS4xXB" role="doc-biblioref">32</a>,<a href="#ref-46R7aU4N" role="doc-biblioref">46</a>]</span>.</p>
<h4 id="prediction-of-protein-coding-sequences-across-a-variety-of-species">Prediction of protein-coding sequences across a variety of species</h4>
<p>We used <code>sencha translate</code> to obtain putative protein-coding sequences from a comaparative transcriptomic dataset spanning nine species and six tissues <span class="citation" data-cites="fBCZ87Cf">[<a href="#ref-fBCZ87Cf" role="doc-biblioref">11</a>]</span>.</p>
<h4 id="read-preprocessing">Read preprocessing</h4>
<p>As the protein-coding score is assessed on the entire read, we recommend RNA-seq reads be removed of library artifacts to the best of the user’s ability.
This means, the adapters should be trimmed, and if there was a negative insert size such that the R1 and R2 reads overlap, then the read pairs should be merged.</p>
<h3 id="pipelines">Pipelines</h3>
<!-- ## Figure 3 -- Overview of `kmermaid` and `predictorthologs` pipelines -->
<div id="fig:sfig3" class="fignos">
<figure>
<img src="images/SVG/supplementary_figure3.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 4:</span> <strong>A.</strong> Overview of the <code>kmermaid</code> pipeline. (<strong>a</strong>, <strong>b</strong>, <strong>c</strong>) <code>kmermaid</code> consists of a protein-coding prediction phase (<strong>a</strong>) that is invoked by the command <code>sencha translate</code>, a k-mer sketch computation phase (<strong>b</strong>) invoked by the command <code>sourmash sketch</code>, a signature similarity comparison phase (<strong>c</strong>) invoked by the command <code>sourmash compare</code>, and an optional database-creation phase (<strong>d</strong>) invoked by the command <code>sourmash index</code>. The coding prediction phase has three components: (1) six-frame translation, removal of stop-codon frames, and subsequent <span class="math inline">\(k\)</span>-merization of RNA-sequencing reads; (2) a degenerate protein alphabet which allows for protein-coding detection from a wide variety of species; (3) a bloom filter containing known protein-coding sequences from a well annotated organism; and (4) computation of the Jaccard index of translated RNA-seq reading frames. The sketch computation phase involves randomly subsetting the degenerate peptide <span class="math inline">\(k\)</span>-mers using a MinHash algorithm. The sketch comparison phase consists of computing the Jaccard intersection of MinHashed degenerate peptide <span class="math inline">\(k\)</span>-mers between all pairs of samples.</figcaption>
</figure>
</div>
<div id="fig:kmermaid" class="fignos">
<figure>
<img src="images/SVG/kmermaid_workflow.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 5:</span> Supplemental Figure X.
<strong>A.</strong> Overview of <code>nf-core/kmermaid</code> pipeline to compare DNA/RNA/protein sequences on k-mer content.
1. If input is bam, extract per-cell sequences using <code>bam2fasta percell</code>.
2. Predict amino acid sequence of each RNA-seq read using <code>sencha translate</code>.
3. Randomly subsample amino acid k-mers via MinHash using <code>sourmash sketch</code>.
4. Compare all k-mer sketches to one another using <code>sourmash compare</code> to compute cell-cell Jaccard similarities.
5. Build sequence bloom tree using <code>sourmash index</code>.
6. Build k-nearest neighbor graph using sequence bloom tree.
7. Build UMAP off of KNN.</figcaption>
</figure>
</div>
<div id="fig:predictorthologs" class="fignos">
<figure>
<img src="images/SVG/predictorthologs_workflow.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 6:</span> Supplemental Figure X.
<strong>B.</strong> Overview of <code>czbiohub/nf-predictorthologs</code> pipeline to query putative function of protein sequences.
1. If input is bam, must also have a convert bam reads to raw fastq files using the <code>samtools fastq</code> subcommand (samtools version 1.9). If input is fastqs, go directly to second step.
2. Trim adapters, poly-A, polyG using the <code>fastp</code> tool.
3. Predict protein-coding sequence using khtools extract_coding, using conservative UniProt/SwissProt manually curated database as examples of known protein-coding sequences, for most stringent definition of protein-coding.
4. Query predicted protein in permissive NCBI RefSeq non-redundant protein database for most complete search query.</figcaption>
</figure>
</div>
<h3 id="installation">Installation</h3>
<p><code>sencha</code> can be installed with the Anaconda package manager, <code>conda</code> (preferred),</p>
<pre><code># Note: not actually on bioconda yet ... this is aspirational
conda install --channel bioconda sencha</code></pre>
<p>or from the Python Package Index (PyPI) with the Python package manager, <code>pip</code>,</p>
<pre><code># Note: not actually on PyPI yet ... this is aspirational
pip install sencha</code></pre>
<h3 id="usage">Usage</h3>
<h4 id="creation-of-amino-acid-k-mer-database-with-sencha-index">Creation of amino acid <span class="math inline">\(k\)</span>-mer database with <code>sencha index</code></h4>
<p>Before predicting protein-coding sequences, <code>sencha</code> must create a database of known amino acid <span class="math inline">\(k\)</span>-mers, which is stored in the form of a probabilistic set membership data structure known as a bloom filter.
<code>sencha</code> uses the bloom filter implementation in <code>khmer</code>/<code>oxli</code> <span class="citation" data-cites="G2CzfHHF 1DpHWPZC9">[<a href="#ref-1DpHWPZC9" role="doc-biblioref">45</a>,<a href="#ref-G2CzfHHF" role="doc-biblioref">47</a>]</span>, called a NodeGraph.
We created a dataset of known amino acid <span class="math inline">\(k\)</span>-mers from the manually annotated UniProtKB/Swiss-Prot databases <span class="citation" data-cites="1ZE22clL 1BBNQZrBs">[<a href="#ref-1ZE22clL" role="doc-biblioref">48</a>,<a href="#ref-1BBNQZrBs" role="doc-biblioref">49</a>]</span>.
We used only protein sequences observed in <em>Opisthokont</em> species <span class="citation" data-cites="Ty6517L8">[<a href="#ref-Ty6517L8" role="doc-biblioref">50</a>]</span>, previously known as a “Fungi/Metazoa” group that encompasseses “Fungus-like” <em>Holomycota</em> and “Animal-like” <em>Holozoa</em>. [NOTE: Does this need a figure/phylogenetic timetree?]</p>
<pre><code>sencha index \
  --tablesize 100000000 \
  --molecule protein \
  --peptide-ksize 7 \
  --save-as uniprot-reviewed_yes+taxonomy_2759__molecule-protein_ksize-7.bloomfilter \
  uniprot-reviewed_yes+taxonomy_2759.fasta.gz</code></pre>
<h4 id="prediction-of-protein-coding-sequences-with-sencha-translate">Prediction of protein-coding sequences with <code>sencha translate</code></h4>
<p>We then predicted protein coding reads using the created bloom filter using <code>sencha translate</code>.</p>
<pre><code>sencha translate \
  --molecule protein \
  --coding-nucleotide-fasta SRR306800_GSM752653_ggo_br_F_1__coding_reads_nucleotides.fasta \
  --csv SRR306800_GSM752653_ggo_br_F_1__coding_scores.csv \
  --json-summary SRR306800_GSM752653_ggo_br_F_1__coding_summary.json \
  --jaccard-threshold 0.8 \
  --peptides-are-index \
  uniprot-reviewed_yes+taxonomy_2759__molecule-protein_ksize-7.bloomfilter \
  SRR306800_GSM752653_ggo_br_F_1_trimmed.fq.gz &gt; SRR306800_GSM752653_ggo_br_F_1__coding_reads_peptides.fasta</code></pre>
<figure>
<img src="images/SVG/supplemental_figure2.svg" id="sfig2:sfig2" data-tag="supplemental_figure2" style="width:100.0%" alt="" /><figcaption>Applications of <code>sencha translate</code>. <strong>A.</strong> We simulated RNA-seq data using Opisthokonta species from the Quest for Orthologs dataset for true positive protein-coding RNAs, reads completely contained within intergenic, intronic, and UTR sequences as true positive noncoding RNAs, and reads partially overlapping a coding and noncoding region as an adversarial test set. We then predicted protein-coding sequences and computed false positive and false negative rates. False Positive coding reads were found to be … False negative noncoding reads were found to be … <strong>B.</strong> Number of putative protein-coding sequences per read. <strong>C.</strong> This method could also be used to extract only reads whose putative protein-coding sequences are transcription factors. <strong>D.</strong> We ran <code>sencha translate</code> on the five tissues and nine species from the Brawand 2011 dataset.</figcaption>
</figure>
<!-- ### `sencha compare-kmer-content` is a simple method to identify homologs

![Applications of `sencha compare-kmer-content`. **A.** We used `sencha compare-kmer-content` on pairs of orthologous protein sequences between humans and the remaining Opisthokonta species in the Quest for Orthologs dataset. x-axis, $k$-mer size, y-axis, mean difference. **B.** False positive calls by `sencha compare-kmer-content` are either paralogs or read-through protein products. **C.** We applied `sencha compare-kmer-content` to ... to find putative orthologs. We found ... the accuracy was ...](images/SVG/figure4.svg){#sfig:figure4 tag="figure4" width="100%"} -->
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references hanging-indent" role="doc-bibliography">
<div id="ref-OkFljXKC">
<p>1. <strong>How Many Species Are There on Earth and in the Ocean?</strong><br />
Camilo Mora, Derek P. Tittensor, Sina Adl, Alastair G. B. Simpson, Boris Worm<br />
<em>PLoS Biology</em> (2011-08-23) <a href="https://doi.org/fpr4z8">https://doi.org/fpr4z8</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pbio.1001127">10.1371/journal.pbio.1001127</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21886479">21886479</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3160336">PMC3160336</a></p>
</div>
<div id="ref-xJN8WB5G">
<p>2. <strong>Genome List - Genome - NCBI</strong><a href="https://www.ncbi.nlm.nih.gov/genome/browse">https://www.ncbi.nlm.nih.gov/genome/browse</a></p>
</div>
<div id="ref-ogAGO9KH">
<p>3. <strong>The origin and evolution of cell types</strong><br />
Detlev Arendt, Jacob M. Musser, Clare V. H. Baker, Aviv Bergman, Connie Cepko, Douglas H. Erwin, Mihaela Pavlicev, Gerhard Schlosser, Stefanie Widder, Manfred D. Laubichler, Günter P. Wagner<br />
<em>Nature Reviews Genetics</em> (2016-11-07) <a href="https://doi.org/f9b62x">https://doi.org/f9b62x</a><br />
DOI: <a href="https://doi.org/10.1038/nrg.2016.127">10.1038/nrg.2016.127</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27818507">27818507</a></p>
</div>
<div id="ref-wdzgXUGy">
<p>4. <strong>How Single-Cell Genomics Is Changing Evolutionary and Developmental Biology</strong><br />
John C. Marioni, Detlev Arendt<br />
<em>Annual Review of Cell and Developmental Biology</em> (2017-10-06) <a href="https://doi.org/ggb632">https://doi.org/ggb632</a><br />
DOI: <a href="https://doi.org/10.1146/annurev-cellbio-100616-060818">10.1146/annurev-cellbio-100616-060818</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28813177">28813177</a></p>
</div>
<div id="ref-wtk13QGK">
<p>5. <strong>K-mer counting with low memory consumption enables fast clustering of single-cell sequencing data without read alignment</strong><br />
Christina Huan Shi, Kevin Y. Yip<br />
<em>bioRxiv</em> (2019-08-02) <a href="https://www.biorxiv.org/content/10.1101/723833v1">https://www.biorxiv.org/content/10.1101/723833v1</a><br />
DOI: <a href="https://doi.org/10.1101/723833">10.1101/723833</a></p>
</div>
<div id="ref-Aq7zYbOg">
<p>6. <strong>Reduced amino acid alphabets exhibit an improved sensitivity and selectivity in fold assignment</strong><br />
Eric L. Peterson, Jané Kondev, Julie A. Theriot, Rob Phillips<br />
<em>Bioinformatics</em> (2009-04-07) <a href="https://doi.org/btqmnp">https://doi.org/btqmnp</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btp164">10.1093/bioinformatics/btp164</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19351620">19351620</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2732308">PMC2732308</a></p>
</div>
<div id="ref-f5u3dDtc">
<p>7. <strong>Simplified amino acid alphabets for protein fold recognition and implications for folding</strong><br />
Lynne Reed Murphy, Anders Wallqvist, Ronald M. Levy<br />
<em>Protein Engineering, Design and Selection</em> (2000-03) <a href="https://doi.org/bdtngh">https://doi.org/bdtngh</a><br />
DOI: <a href="https://doi.org/10.1093/protein/13.3.149">10.1093/protein/13.3.149</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/10775656">10775656</a></p>
</div>
<div id="ref-QMYjnr6l">
<p>8. <strong>Local homology recognition and distance measures in linear time using compressed amino acid alphabets</strong><br />
R. C. Edgar<br />
<em>Nucleic Acids Research</em> (2004-01-02) <a href="https://doi.org/ckg5d4">https://doi.org/ckg5d4</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkh180">10.1093/nar/gkh180</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/14729922">14729922</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC373290">PMC373290</a></p>
</div>
<div id="ref-Rqlso7kk">
<p>9. <strong>SwiftOrtho: A fast, memory-efficient, multiple genome orthology classifier</strong><br />
Xiao Hu, Iddo Friedberg<br />
<em>GigaScience</em> (2019-10-01) <a href="https://doi.org/ggcr5x">https://doi.org/ggcr5x</a><br />
DOI: <a href="https://doi.org/10.1093/gigascience/giz118">10.1093/gigascience/giz118</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31648300">31648300</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6812468">PMC6812468</a></p>
</div>
<div id="ref-16Kv6RIe3">
<p>10. <strong>Fast databank searching with a reduced amino-acid alphabet</strong><br />
Claudine Landès, Jean-Loup Risler<br />
<em>Bioinformatics</em> (1994) <a href="https://doi.org/cvrjmw">https://doi.org/cvrjmw</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/10.4.453">10.1093/bioinformatics/10.4.453</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/7804879">7804879</a></p>
</div>
<div id="ref-fBCZ87Cf">
<p>11. <strong>The evolution of gene expression levels in mammalian organs</strong><br />
David Brawand, Magali Soumillon, Anamaria Necsulea, Philippe Julien, Gábor Csárdi, Patrick Harrigan, Manuela Weier, Angélica Liechti, Ayinuer Aximu-Petri, Martin Kircher, … Henrik Kaessmann<br />
<em>Nature</em> (2011-10) <a href="https://doi.org/fcvk54">https://doi.org/fcvk54</a><br />
DOI: <a href="https://doi.org/10.1038/nature10532">10.1038/nature10532</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22012392">22012392</a></p>
</div>
<div id="ref-137npblqI">
<p>12. <strong>Inferring Hierarchical Orthologous Groups from Orthologous Gene Pairs</strong><br />
Adrian M. Altenhoff, Manuel Gil, Gaston H. Gonnet, Christophe Dessimoz<br />
<em>PLoS ONE</em> (2013-01-14) <a href="https://doi.org/ggkv2j">https://doi.org/ggkv2j</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0053786">10.1371/journal.pone.0053786</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23342000">23342000</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3544860">PMC3544860</a></p>
</div>
<div id="ref-5RECVcFl">
<p>13. <strong>Conceptual framework and pilot study to benchmark phylogenomic databases based on reference gene trees</strong><br />
B. Boeckmann, M. Robinson-Rechavi, I. Xenarios, C. Dessimoz<br />
<em>Briefings in Bioinformatics</em> (2011-07-07) <a href="https://doi.org/c78rwm">https://doi.org/c78rwm</a><br />
DOI: <a href="https://doi.org/10.1093/bib/bbr034">10.1093/bib/bbr034</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21737420">21737420</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3178055">PMC3178055</a></p>
</div>
<div id="ref-19AeqN3SK">
<p>14. <strong>Big data and other challenges in the quest for orthologs</strong><br />
E. L. L. Sonnhammer, T. Gabaldon, A. W. Sousa da Silva, M. Martin, M. Robinson-Rechavi, B. Boeckmann, P. D. Thomas, C. Dessimoz, <br />
<em>Bioinformatics</em> (2014-07-26) <a href="https://doi.org/f6ntvb">https://doi.org/f6ntvb</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btu492">10.1093/bioinformatics/btu492</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25064571">25064571</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4201156">PMC4201156</a></p>
</div>
<div id="ref-1A6CqUMhm">
<p>15. <strong>TimeTree :: The Timescale of Life</strong><a href="http://timetree.org/">http://timetree.org/</a></p>
</div>
<div id="ref-DXegEVg5">
<p>16. <strong>A Single Cell Transcriptomic Atlas Characterizes Aging Tissues in the Mouse</strong><br />
The Tabula Muris Consortium, Angela Oliveira Pisco, Aaron McGeever, Nicholas Schaum, Jim Karkanias, Norma F. Neff, Spyros Darmanis, Tony Wyss-Coray, Stephen R. Quake<br />
<em>bioRxiv</em> (2019-11-18) <a href="https://www.biorxiv.org/content/10.1101/661728v2">https://www.biorxiv.org/content/10.1101/661728v2</a><br />
DOI: <a href="https://doi.org/10.1101/661728">10.1101/661728</a></p>
</div>
<div id="ref-YFejwX4f">
<p>17. <strong>Complex mammalian-like haematopoietic system found in a colonial chordate</strong><br />
Benyamin Rosental, Mark Kowarsky, Jun Seita, Daniel M. Corey, Katherine J. Ishizuka, Karla J. Palmeri, Shih-Yu Chen, Rahul Sinha, Jennifer Okamoto, Gary Mantalas, … Ayelet Voskoboynik<br />
<em>Nature</em> (2018-12) <a href="https://doi.org/gfkzvm">https://doi.org/gfkzvm</a><br />
DOI: <a href="https://doi.org/10.1038/s41586-018-0783-x">10.1038/s41586-018-0783-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30518860">30518860</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6347970">PMC6347970</a></p>
</div>
<div id="ref-AlRzKu3c">
<p>18. <strong>Assigning confidence scores to homoeologs using fuzzy logic</strong><br />
Natasha M. Glover, Adrian Altenhoff, Christophe Dessimoz<br />
<em>PeerJ</em> (2019-01-11) <a href="https://doi.org/ggkv2k">https://doi.org/ggkv2k</a><br />
DOI: <a href="https://doi.org/10.7717/peerj.6231">10.7717/peerj.6231</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30648004">30648004</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6330999">PMC6330999</a></p>
</div>
<div id="ref-xwCPkKHO">
<p>19. <strong>Orthologous Matrix (OMA) algorithm 2.0: more robust to asymmetric evolutionary rates and more scalable hierarchical orthologous group inference</strong><br />
Clément-Marie Train, Natasha M Glover, Gaston H Gonnet, Adrian M Altenhoff, Christophe Dessimoz<br />
<em>Bioinformatics</em> (2017-07-12) <a href="https://doi.org/ggkv2h">https://doi.org/ggkv2h</a><br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btx229">10.1093/bioinformatics/btx229</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28881964">28881964</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5870696">PMC5870696</a></p>
</div>
<div id="ref-rSyxV4O8">
<p>20. <strong>A single-cell transcriptomic atlas characterizes ageing tissues in the mouse</strong><em>Nature</em> (2020-07) <a href="https://doi.org/gg426c">https://doi.org/gg426c</a><br />
DOI: <a href="https://doi.org/10.1038/s41586-020-2496-1">10.1038/s41586-020-2496-1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32669714">32669714</a></p>
</div>
<div id="ref-RIPzCufe">
<p>21. <strong>Sailfish enables alignment-free isoform quantification from RNA-seq reads using lightweight algorithms</strong><br />
Rob Patro, Stephen M Mount, Carl Kingsford<br />
<em>Nature Biotechnology</em> (2014-04-20) <a href="https://doi.org/gfghc2">https://doi.org/gfghc2</a><br />
DOI: <a href="https://doi.org/10.1038/nbt.2862">10.1038/nbt.2862</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24752080">24752080</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4077321">PMC4077321</a></p>
</div>
<div id="ref-12KZMHMQl">
<p>22. <strong>Near-optimal probabilistic RNA-seq quantification</strong><br />
Nicolas L Bray, Harold Pimentel, Páll Melsted, Lior Pachter<br />
<em>Nature Biotechnology</em> (2016-04-04) <a href="https://doi.org/f8nvsp">https://doi.org/f8nvsp</a><br />
DOI: <a href="https://doi.org/10.1038/nbt.3519">10.1038/nbt.3519</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27043002">27043002</a></p>
</div>
<div id="ref-vrqQcFyx">
<p>23. <strong>Salmon provides fast and bias-aware quantification of transcript expression</strong><br />
Rob Patro, Geet Duggal, Michael I Love, Rafael A Irizarry, Carl Kingsford<br />
<em>Nature Methods</em> (2017-03-06) <a href="https://doi.org/gcw9f5">https://doi.org/gcw9f5</a><br />
DOI: <a href="https://doi.org/10.1038/nmeth.4197">10.1038/nmeth.4197</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28263959">28263959</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5600148">PMC5600148</a></p>
</div>
<div id="ref-DfZvKOpG">
<p>24. <strong>Fast search of thousands of short-read sequencing experiments</strong><br />
Brad Solomon, Carl Kingsford<br />
<em>Nature Biotechnology</em> (2016-02-08) <a href="https://doi.org/f8ddk3">https://doi.org/f8ddk3</a><br />
DOI: <a href="https://doi.org/10.1038/nbt.3442">10.1038/nbt.3442</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26854477">26854477</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4804353">PMC4804353</a></p>
</div>
<div id="ref-1HpdDzfQX">
<p>25. <strong>Improved Search of Large Transcriptomic Sequencing Databases Using Split Sequence Bloom Trees</strong><br />
Brad Solomon, Carl Kingsford<br />
<em>Journal of Computational Biology</em> (2018-07) <a href="https://doi.org/gdxhgz">https://doi.org/gdxhgz</a><br />
DOI: <a href="https://doi.org/10.1089/cmb.2017.0265">10.1089/cmb.2017.0265</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29641248">29641248</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6067102">PMC6067102</a></p>
</div>
<div id="ref-AfFgGSHX">
<p>26. <strong>AllSome Sequence Bloom Trees</strong><br />
Chen Sun, Robert S. Harris, Rayan Chikhi, Paul Medvedev<br />
<em>Journal of Computational Biology</em> (2018-05) <a href="https://doi.org/gdkn3v">https://doi.org/gdkn3v</a><br />
DOI: <a href="https://doi.org/10.1089/cmb.2017.0258">10.1089/cmb.2017.0258</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29620920">29620920</a></p>
</div>
<div id="ref-lBlqrqAO">
<p>27. <strong>Ensembl 2018</strong><br />
Daniel R Zerbino, Premanand Achuthan, Wasiu Akanni, M Ridwan Amode, Daniel Barrell, Jyothish Bhai, Konstantinos Billis, Carla Cummins, Astrid Gall, Carlos García Girón, … Paul Flicek<br />
<em>Nucleic Acids Research</em> (2017-11-16) <a href="https://doi.org/gcwg6r">https://doi.org/gcwg6r</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gkx1098">10.1093/nar/gkx1098</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29155950">29155950</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5753206">PMC5753206</a></p>
</div>
<div id="ref-j0y6s0kf">
<p>28. <strong>Ensembl comparative genomics resources</strong><br />
Javier Herrero, Matthieu Muffato, Kathryn Beal, Stephen Fitzgerald, Leo Gordon, Miguel Pignatelli, Albert J. Vilella, Stephen M. J. Searle, Ridwan Amode, Simon Brent, … Paul Flicek<br />
<em>Database</em> (2016) <a href="https://doi.org/ggb9tv">https://doi.org/ggb9tv</a><br />
DOI: <a href="https://doi.org/10.1093/database/bav096">10.1093/database/bav096</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26896847">26896847</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4761110">PMC4761110</a></p>
</div>
<div id="ref-66KDHtpG">
<p>29. <strong>Atlas of protein sequence and structure</strong><br />
Margaret O Dayhoff<br />
<em>National Biomedical Research Foundation.</em> (1969)</p>
</div>
<div id="ref-1GiMcbPLs">
<p>30. <strong>Physical biology of the cell</strong><br />
Rob Phillips, Julie Theriot, Jane Kondev, Hernan Garcia<br />
<em>Garland Science</em> (2012)</p>
</div>
<div id="ref-E2Kp7kA0">
<p>31. <strong>Theory for the folding and stability of globular proteins</strong><br />
Ken A. Dill<br />
<em>Biochemistry</em> (1985-03-12) <a href="https://doi.org/fnj5k7">https://doi.org/fnj5k7</a><br />
DOI: <a href="https://doi.org/10.1021/bi00327a032">10.1021/bi00327a032</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/3986190">3986190</a></p>
</div>
<div id="ref-LkrS4xXB">
<p>32. <strong>Standardized benchmarking in the quest for orthologs</strong><br />
Adrian M AltenhoffBrigitte Boeckmann, Salvador Capella-Gutierrez, Daniel A Dalquen, Todd DeLuca, Kristoffer Forslund, Jaime Huerta-Cepas, Benjamin Linard, Cécile Pereira, … Christophe Dessimoz<br />
<em>Nature Methods</em> (2016-04-04) <a href="https://doi.org/f3rpzx">https://doi.org/f3rpzx</a><br />
DOI: <a href="https://doi.org/10.1038/nmeth.3830">10.1038/nmeth.3830</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27043882">27043882</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4827703">PMC4827703</a></p>
</div>
<div id="ref-bkF0801R">
<p>33. <strong>Identifiers.org</strong><a href="https://identifiers.org/">https://identifiers.org/</a></p>
</div>
<div id="ref-AtPPptbH">
<p>34. <strong>RNAsamba: coding potential assessment using ORF and whole transcript sequence information</strong><br />
Antonio P. Camargo, Vsevolod Sourkov, Marcelo F. Carazzolle<br />
<em>Cold Spring Harbor Laboratory</em> (2019-04-28) <a href="https://doi.org/ggdtxk">https://doi.org/ggdtxk</a><br />
DOI: <a href="https://doi.org/10.1101/620880">10.1101/620880</a></p>
</div>
<div id="ref-1DcTxE8Z3">
<p>35. <strong>Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates</strong><br />
D. Farre, M. M. Alba<br />
<em>Molecular Biology and Evolution</em> (2009-10-12) <a href="https://doi.org/dxrtmd">https://doi.org/dxrtmd</a><br />
DOI: <a href="https://doi.org/10.1093/molbev/msp242">10.1093/molbev/msp242</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19822635">19822635</a></p>
</div>
<div id="ref-uQLS9NYU">
<p>36. <strong>GENEFAMILYEVOLUTION ANDHOMOLOGY: Genomics Meets Phylogenetics</strong><br />
Joseph W. Thornton, Rob DeSalle<br />
<em>Annual Review of Genomics and Human Genetics</em> (2000-09) <a href="https://doi.org/bjp5pm">https://doi.org/bjp5pm</a><br />
DOI: <a href="https://doi.org/10.1146/annurev.genom.1.1.41">10.1146/annurev.genom.1.1.41</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/11701624">11701624</a></p>
</div>
<div id="ref-nbBjh1sL">
<p>37. <strong>Evolution and cell-type specificity of human-specific genes preferentially expressed in progenitors of fetal neocortex</strong><br />
Marta Florio, Michael Heide, Anneline Pinson, Holger Brandl, Mareike Albert, Sylke Winkler, Pauline Wimberger, Wieland B Huttner, Michael Hiller<br />
<em>eLife</em> (2018-03-21) <a href="https://doi.org/gc678k">https://doi.org/gc678k</a><br />
DOI: <a href="https://doi.org/10.7554/elife.32332">10.7554/elife.32332</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29561261">29561261</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5898914">PMC5898914</a></p>
</div>
<div id="ref-lJHQuvIH">
<p>38. <strong>Taxon-restricted genes at the origin of a novel trait allowing access to a new environment</strong><br />
M. Emília Santos, Augustin Le Bouquin, Antonin J. J. Crumière, Abderrahman Khila<br />
<em>Science</em> (2017-10-19) <a href="https://doi.org/gcgjbs">https://doi.org/gcgjbs</a><br />
DOI: <a href="https://doi.org/10.1126/science.aan2748">10.1126/science.aan2748</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29051384">29051384</a></p>
</div>
<div id="ref-FTv2KYrZ">
<p>39. <strong>Pervasive Correlated Evolution in Gene Expression Shapes Cell and Tissue Type Transcriptomes</strong><br />
Cong Liang, Jacob M Musser, Alison Cloutier, Richard O Prum, Günter P Wagner<br />
<em>Genome Biology and Evolution</em> (2018-01-23) <a href="https://doi.org/gc69v9">https://doi.org/gc69v9</a><br />
DOI: <a href="https://doi.org/10.1093/gbe/evy016">10.1093/gbe/evy016</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29373668">29373668</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5800078">PMC5800078</a></p>
</div>
<div id="ref-hdRo8a7z">
<p>40. <strong>Deep homology in the age of next-generation sequencing</strong><br />
Patrick Tschopp, Clifford J. Tabin<br />
<em>Philosophical Transactions of the Royal Society B: Biological Sciences</em> (2017-02-05) <a href="https://doi.org/gfzpbg">https://doi.org/gfzpbg</a><br />
DOI: <a href="https://doi.org/10.1098/rstb.2015.0475">10.1098/rstb.2015.0475</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27994118">27994118</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5182409">PMC5182409</a></p>
</div>
<div id="ref-1GRtpoh49">
<p>41. <strong>Embracing the comparative approach: how robust phylogenies and broader developmental sampling impacts the understanding of nervous system evolution</strong><br />
Andreas Hejnol, Christopher J. Lowe<br />
<em>Philosophical Transactions of the Royal Society B: Biological Sciences</em> (2015-12-19) <a href="https://doi.org/ggcd2m">https://doi.org/ggcd2m</a><br />
DOI: <a href="https://doi.org/10.1098/rstb.2015.0045">10.1098/rstb.2015.0045</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26554039">26554039</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4650123">PMC4650123</a></p>
</div>
<div id="ref-8cGr6TE">
<p>42. <strong>The mammalian decidual cell evolved from a cellular stress response</strong><br />
Eric M. Erkenbrack, Jamie D. Maziarz, Oliver W. Griffith, Cong Liang, Arun R. Chavan, Mauris C. Nnamani, Günter P. Wagner<br />
<em>PLOS Biology</em> (2018-08-24) <a href="https://doi.org/gd5b9s">https://doi.org/gd5b9s</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pbio.2005594">10.1371/journal.pbio.2005594</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30142145">30142145</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6108454">PMC6108454</a></p>
</div>
<div id="ref-sZoryWIB">
<p>43. <strong>RAPSearch: a fast protein similarity search tool for short reads</strong><br />
Yuzhen Ye, Jeong-Hyeon Choi, Haixu Tang<br />
<em>BMC Bioinformatics</em> (2011-05-15) <a href="https://doi.org/dgt5rt">https://doi.org/dgt5rt</a><br />
DOI: <a href="https://doi.org/10.1186/1471-2105-12-159">10.1186/1471-2105-12-159</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21575167">21575167</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3113943">PMC3113943</a></p>
</div>
<div id="ref-11LSoLYLW">
<p>44. <strong>Incomplete nucleic acid sequences</strong><a href="https://www.qmul.ac.uk/sbcs/iubmb/misc/naseq.html">https://www.qmul.ac.uk/sbcs/iubmb/misc/naseq.html</a></p>
</div>
<div id="ref-1DpHWPZC9">
<p>45. <strong>These Are Not the K-mers You Are Looking For: Efficient Online K-mer Counting Using a Probabilistic Data Structure</strong><br />
Qingpeng Zhang, Jason Pell, Rosangela Canino-Koning, Adina Chuang Howe, C. Titus Brown<br />
<em>PLoS ONE</em> (2014-07-25) <a href="https://doi.org/f6kb9b">https://doi.org/f6kb9b</a><br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0101271">10.1371/journal.pone.0101271</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25062443">25062443</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4111482">PMC4111482</a></p>
</div>
<div id="ref-46R7aU4N">
<p>46. <strong>OrthoFinder: phylogenetic orthology inference for comparative genomics</strong><br />
David M. Emms, Steven Kelly<br />
<em>bioRxiv</em> (2019-04-24) <a href="https://www.biorxiv.org/content/10.1101/466201v2">https://www.biorxiv.org/content/10.1101/466201v2</a><br />
DOI: <a href="https://doi.org/10.1101/466201">10.1101/466201</a></p>
</div>
<div id="ref-G2CzfHHF">
<p>47. <strong>The khmer software package: enabling efficient nucleotide sequence analysis</strong><br />
Michael R. Crusoe, Hussien F. Alameldin, Sherine Awad, Elmar Boucher, Adam Caldwell, Reed Cartwright, Amanda Charbonneau, Bede Constantinides, Greg Edvenson, Scott Fay, … C. Titus Brown<br />
<em>F1000Research</em> (2015-09-25) <a href="https://doi.org/9qp">https://doi.org/9qp</a><br />
DOI: <a href="https://doi.org/10.12688/f1000research.6924.1">10.12688/f1000research.6924.1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26535114">26535114</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4608353">PMC4608353</a></p>
</div>
<div id="ref-1ZE22clL">
<p>48. <strong>UniProt: a worldwide hub of protein knowledge</strong><em>Nucleic Acids Research</em> (2018-11-05) <a href="https://doi.org/gfwqck">https://doi.org/gfwqck</a><br />
DOI: <a href="https://doi.org/10.1093/nar/gky1049">10.1093/nar/gky1049</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30395287">30395287</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323992">PMC6323992</a></p>
</div>
<div id="ref-1BBNQZrBs">
<p>49. <strong>UniProtKB/Swiss-Prot, the Manually Annotated Section of the UniProt KnowledgeBase: How to Use the Entry View</strong><br />
Emmanuel Boutet, Damien Lieberherr, Michael Tognolli, Michel Schneider, Parit Bansal, Alan J. Bridge, Sylvain Poux, Lydie Bougueleret, Ioannis Xenarios<br />
<em>Plant Bioinformatics</em> (2016) <a href="https://doi.org/f79kbb">https://doi.org/f79kbb</a><br />
DOI: <a href="https://doi.org/10.1007/978-1-4939-3167-5_2">10.1007/978-1-4939-3167-5_2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26519399">26519399</a></p>
</div>
<div id="ref-Ty6517L8">
<p>50. <strong>Opisthokont</strong><br />
Wikipedia<br />
(2020-05-28) <a href="https://en.wikipedia.org/w/index.php?title=Opisthokont&amp;oldid=959328343">https://en.wikipedia.org/w/index.php?title=Opisthokont&amp;oldid=959328343</a></p>
</div>
</div>
<!-- default theme -->

<style>
    /* import google fonts */
    @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
    @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

    /* -------------------------------------------------- */
    /* global */
    /* -------------------------------------------------- */

    /* all elements */
    * {
        /* force sans-serif font unless specified otherwise */
        font-family: "Open Sans", "Helvetica", sans-serif;

        /* prevent text inflation on some mobile browsers */
        -webkit-text-size-adjust: none !important;
        -moz-text-size-adjust: none !important;
        -o-text-size-adjust: none !important;
        text-size-adjust: none !important;
    }

    @media only screen {
        /* "page" element */
        body {
            position: relative;
            box-sizing: border-box;
            font-size: 12pt;
            line-height: 1.5;
            max-width: 8.5in;
            margin: 20px auto;
            padding: 40px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* "page" element */
        body {
            padding: 20px;
            margin: 0;
            border-radius: 0;
            border: none;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
            background: none;
        }
    }

    /* -------------------------------------------------- */
    /* headings */
    /* -------------------------------------------------- */

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
        margin: 20px 0;
        padding: 0;
        font-weight: bold;
    }

    /* biggest heading */
    h1 {
        margin: 40px 0;
        text-align: center;
    }

    /* second biggest heading */
    h2 {
        margin-top: 30px;
        padding-bottom: 5px;
        border-bottom: solid 1px #bdbdbd;
    }

    /* -------------------------------------------------- */
    /* manuscript header */
    /* -------------------------------------------------- */

    /* manuscript title */
    header > h1 {
        margin: 0;
    }

    /* manuscript title caption text (ie "automatically generated on") */
    header + p {
        text-align: center;
        margin-top: 10px;
    }

    /* -------------------------------------------------- */
    /* text elements */
    /* -------------------------------------------------- */

    /* links */
    a {
        color: #2196f3;
        overflow-wrap: break-word;
    }

    /* normal links (not empty, not button link, not syntax highlighting link) */
    a:not(:empty):not(.button):not(.sourceLine) {
        padding-left: 1px;
        padding-right: 1px;
    }

    /* superscripts and subscripts */
    sub,
    sup {
        /* prevent from affecting line height */
        line-height: 0;
    }

    /* unordered and ordered lists*/
    ul,
    ol {
        padding-left: 20px;
    }

    /* class for styling text semibold */
    .semibold {
        font-weight: 600;
    }

    /* class for styling elements horizontally left aligned */
    .left {
        display: block;
        text-align: left;
        margin-left: auto;
        margin-right: 0;
        justify-content: left;
    }

    /* class for styling elements horizontally centered */
    .center {
        display: block;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
        justify-content: center;
    }

    /* class for styling elements horizontally right aligned */
    .right {
        display: block;
        text-align: right;
        margin-left: 0;
        margin-right: auto;
        justify-content: right;
    }

    /* -------------------------------------------------- */
    /* section elements */
    /* -------------------------------------------------- */

    /* horizontal divider line */
    hr {
        border: none;
        height: 1px;
        background: #bdbdbd;
    }

    /* paragraphs, horizontal dividers, figures, tables, code */
    p,
    hr,
    figure,
    table,
    pre {
        /* treat all as "paragraphs", with consistent vertical margins */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* figures */
    /* -------------------------------------------------- */

    /* figure */
    figure {
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure caption */
    figcaption {
        padding: 0;
        padding-top: 10px;
    }

    /* figure image element */
    figure img {
        max-width: 100%;
        display: block;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure auto-number */
    img + figcaption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* tables */
    /* -------------------------------------------------- */

    /* table */
    table {
        border-collapse: collapse;
        border-spacing: 0;
        width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* table cells */
    th,
    td {
        border: solid 1px #bdbdbd;
        padding: 10px;
        /* squash table if too wide for page by forcing line breaks */
        overflow-wrap: break-word;
        word-break: break-word;
    }

    /* header row and even rows */
    th,
    tr:nth-child(2n) {
        background-color: #fafafa;
    }

    /* odd rows */
    tr:nth-child(2n + 1) {
        background-color: #ffffff;
    }

    /* table caption */
    caption {
        text-align: left;
        padding: 0;
        padding-bottom: 10px;
    }

    /* table auto-number */
    table > caption > span:first-of-type,
    div.table_wrapper > table > caption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* code */
    /* -------------------------------------------------- */

    /* multi-line code block */
    pre {
        padding: 10px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
        break-inside: avoid;
        text-align: left;
    }

    /* inline code, ie code within normal text */
    :not(pre) > code {
        padding: 0 4px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
    }

    /* code text */
    /* apply all children, to reach syntax highlighting sub-elements */
    code,
    code * {
        /* force monospace font */
        font-family: "Source Code Pro", "Courier New", monospace;
    }

    /* -------------------------------------------------- */
    /* quotes */
    /* -------------------------------------------------- */

    /* quoted text */
    blockquote {
        margin: 0;
        padding: 0;
        border-left: 4px solid #bdbdbd;
        padding-left: 16px;
        break-inside: avoid;
    }

    /* -------------------------------------------------- */
    /* banners */
    /* -------------------------------------------------- */

    /* info banners */
    .banner {
        box-sizing: border-box;
        display: block;
        position: relative;
        width: 100%;
        margin-top: 20px;
        margin-bottom: 20px;
        padding: 20px;
        text-align: center;
    }

    /* paragraph in banner */
    .banner > p {
        margin: 0;
    }

    /* -------------------------------------------------- */
    /* highlight colors */
    /* -------------------------------------------------- */

    .white {
        background: #ffffff;
    }
    .lightgrey {
        background: #eeeeee;
    }
    .grey {
        background: #757575;
    }
    .darkgrey {
        background: #424242;
    }
    .black {
        background: #000000;
    }
    .lightred {
        background: #ffcdd2;
    }
    .lightyellow {
        background: #ffecb3;
    }
    .lightgreen {
        background: #dcedc8;
    }
    .lightblue {
        background: #e3f2fd;
    }
    .lightpurple {
        background: #f3e5f5;
    }
    .red {
        background: #f44336;
    }
    .orange {
        background: #ff9800;
    }
    .yellow {
        background: #ffeb3b;
    }
    .green {
        background: #4caf50;
    }
    .blue {
        background: #2196f3;
    }
    .purple {
        background: #9c27b0;
    }
    .white,
    .lightgrey,
    .lightred,
    .lightyellow,
    .lightgreen,
    .lightblue,
    .lightpurple,
    .orange,
    .yellow,
    .white a,
    .lightgrey a,
    .lightred a,
    .lightyellow a,
    .lightgreen a,
    .lightblue a,
    .lightpurple a,
    .orange a,
    .yellow a {
        color: #000000;
    }
    .grey,
    .darkgrey,
    .black,
    .red,
    .green,
    .blue,
    .purple,
    .grey a,
    .darkgrey a,
    .black a,
    .red a,
    .green a,
    .blue a,
    .purple a {
        color: #ffffff;
    }

    /* -------------------------------------------------- */
    /* buttons */
    /* -------------------------------------------------- */

    /* class for styling links like buttons */
    .button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        margin: 5px;
        padding: 10px 20px;
        font-size: 0.75em;
        font-weight: 600;
        text-transform: uppercase;
        text-decoration: none;
        letter-spacing: 1px;
        background: none;
        color: #2196f3;
        border: solid 1px #bdbdbd;
        border-radius: 5px;
    }

    /* buttons when hovered */
    .button:hover:not([disabled]),
    .icon_button:hover:not([disabled]) {
        cursor: pointer;
        background: #f5f5f5;
    }

    /* buttons when disabled */
    .button[disabled],
    .icon_button[disabled] {
        opacity: 0.35;
        pointer-events: none;
    }

    /* class for styling buttons containg only single icon */
    .icon_button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        margin: 0;
        padding: 0;
        background: none;
        border-radius: 5px;
        border: none;
        width: 20px;
        height: 20px;
        min-width: 20px;
        min-height: 20px;
    }

    /* icon button inner svg image */
    .icon_button > svg {
        height: 16px;
    }

    /* -------------------------------------------------- */
    /* icons */
    /* -------------------------------------------------- */

    /* class for styling icons inline with text */
    .inline_icon {
        height: 1em;
        position: relative;
        top: 0.125em;
    }

    /* -------------------------------------------------- */
    /* print control */
    /* -------------------------------------------------- */

    @media print {
        @page {
            /* suggested printing margin */
            margin: 0.5in;
        }

        /* document and "page" elements */
        html, body {
            margin: 0;
            padding: 0;
            width: 100%;
            height: 100%;
        }

        /* "page" element */
        body {
            font-size: 11pt !important;
            line-height: 1.35;
        }

        /* all headings */
        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
            margin: 15px 0;
        }

        /* heading 1 */
        h1 {
            font-size: 1.75em;
        }

        /* heading 2 */
        h2 {
            font-size: 1.25em;
            margin-top: 0;
        }

        /* heading 3 */
        h3 {
            font-size: 1.10em;
        }

        /* figures and tables */
        figure, table {
            font-size: 0.85em;
        }

        /* table cells */
        th,
        td {
            padding: 5px;
        }

        /* shrink font awesome icons */
        i.fas,
        i.fab,
        i.far,
        i.fal {
            transform: scale(0.85);
        }

        /* decrease banner margins */
        .banner {
            margin-top: 15px;
            margin-bottom: 15px;
            padding: 15px;
        }

        /* class for centering an element vertically on its own page */
        .page_center {
            margin: auto;
            width: 100%;
            height: 100%;
            display: flex;
            align-items: center;
            vertical-align: middle;
            break-before: page;
            break-after: page;
        }

        /* always insert a page break before the element */
        .page_break_before {
            break-before: page;
        }

        /* always insert a page break after the element */
        .page_break_after {
            break-after: page;
        }

        /* avoid page break before the element */
        .page_break_before_avoid {
            break-before: avoid;
        }

        /* avoid page break after the element */
        .page_break_after_avoid {
            break-after: avoid;
        }

        /* avoid page break inside the element */
        .page_break_inside_avoid {
            break-inside: avoid;
        }
    }

    /* -------------------------------------------------- */
    /* override pandoc css quirks */
    /* -------------------------------------------------- */

    .sourceCode {
        /* prevent unsightly overflow in wide code blocks */
        overflow: auto !important;
    }

    div.sourceCode {
        /* prevent background fill on top-most code block  container */
        background: none !important;
    }

    .sourceCode * {
        /* force consistent line spacing */
        line-height: 1.5 !important;
    }

    div.sourceCode {
        /* style code block margins same as <pre> element */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* tablenos */
    /* -------------------------------------------------- */

    /* tablenos wrapper */
    .tablenos {
        /* show scrollbar on tables if necessary to prevent overflow */
        width: 100%;
        margin: 20px 0;
    }

    .tablenos > table {
        /* move margins from table to table_wrapper to allow margin collapsing */
        margin: 0;
    }

    @media only screen {
        /* tablenos wrapper */
        .tablenos {
            /* show scrollbar on tables if necessary to prevent overflow */
            overflow-x: auto !important;
        }

        .tablenos th,
        .tablenos td {
            overflow-wrap: unset !important;
            word-break: unset !important;
        }

        /* table in wrapper */
        .tablenos table,
        .tablenos table * {
            /* don't break table words */
            overflow-wrap: normal !important;
        }
    }

    /* -------------------------------------------------- */
    /* mathjax */
    /* -------------------------------------------------- */

    /* mathjax containers */
    .math.display > span:not(.MathJax_Preview) {
        /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
        display: flex !important;
        overflow-x: auto !important;
        overflow-y: hidden !important;
        justify-content: center;
        align-items: center;
        margin: 0 !important;
    }

    /* right click menu */
    .MathJax_Menu {
        border-radius: 5px !important;
        border: solid 1px #bdbdbd !important;
        box-shadow: none !important;
    }

    /* equation auto-number */
    span[id^="eq:"] > span.math.display + span {
        font-weight: 600;
    }

    /* equation */
    span[id^="eq:"] > span.math.display > span {
        /* nudge to make room for equation auto-number and anchor */
        margin-right: 60px !important;
    }

    /* -------------------------------------------------- */
    /* anchors plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anchor button */
        .anchor {
            opacity: 0;
            margin-left: 5px;
        }

        /* anchor buttons within <h2>'s */
        h2 .anchor {
            margin-left: 10px;
        }

        /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
        *:hover > .anchor,
        .anchor:hover,
        .anchor:focus {
            opacity: 1;
        }

        /* anchor button when hovered */
        .anchor:hover {
            cursor: pointer;
        }
    }

    /* always show anchor button on devices with no mouse/hover ability */
    @media (hover: none) {
        .anchor {
            opacity: 1;
        }
    }

    /* always hide anchor button on print */
    @media only print {
        .anchor {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* accordion plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* accordion arrow button */
        .accordion_arrow {
            margin-right: 10px;
        }

        /* arrow icon when <h2> data-collapsed attribute true */
        h2[data-collapsed="true"] > .accordion_arrow > svg {
            transform: rotate(-90deg);
        }

        /* all elements (except <h2>'s) when data-collapsed attribute true */
        *:not(h2)[data-collapsed="true"] {
            display: none;
        }

        /* accordion arrow button when hovered and <h2>'s when hovered */
        .accordion_arrow:hover,
        h2[data-collapsed="true"]:hover,
        h2[data-collapsed="false"]:hover {
            cursor: pointer;
        }
    }

    /* always hide accordion arrow button on print */
    @media only print {
        .accordion_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* tooltips plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* tooltip container */
        #tooltip {
            position: absolute;
            width: 50%;
            min-width: 240px;
            max-width: 75%;
            z-index: 1;
        }

        /* tooltip content */
        #tooltip_content {
            margin-bottom: 5px;
            padding: 20px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
            overflow-wrap: break-word;
        }

        /* tooltip copy of paragraphs and figures */
        #tooltip_content > p,
        #tooltip_content > figure {
            margin: 0;
            max-height: 320px;
            overflow-y: auto;
        }

        /* tooltip copy of <img> */
        #tooltip_content > figure > img {
            max-height: 260px;
        }

        /* navigation bar */
        #tooltip_nav_bar {
            margin-top: 10px;
            text-align: center;
        }

        /* navigation bar previous/next buton */
        #tooltip_nav_bar > .icon_button {
            position: relative;
            top: 3px;
        }

        /* navigation bar previous button */
        #tooltip_nav_bar > .icon_button:first-of-type {
            margin-right: 5px;
        }

        /* navigation bar next button */
        #tooltip_nav_bar > .icon_button:last-of-type {
            margin-left: 5px;
        }
    }

    /* always hide tooltip on print */
    @media only print {
        #tooltip {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* jump to first plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* jump button */
        .jump_arrow {
            position: relative;
            top: 0.125em;
            margin-right: 5px;
        }
    }

    /* always hide jump button on print */
    @media only print {
        .jump_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* link highlight plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anything with data-highlighted attribute true */
        [data-highlighted="true"] {
            background: #ffeb3b;
        }

        /* anything with data-selected attribute true */
        [data-selected="true"] {
            background: #ff8a65 !important;
        }

        /* animation definition for glow */
        @keyframes highlight_glow {
            0% {
                background: none;
            }
            10% {
                background: #bbdefb;
            }
            100% {
                background: none;
            }
        }

        /* anything with data-glow attribute true */
        [data-glow="true"] {
            animation: highlight_glow 2s;
        }
    }

    /* -------------------------------------------------- */
    /* table of contents plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* toc panel */
        #toc_panel {
            box-sizing: border-box;
            position: fixed;
            top: 0;
            left: 0;
            background: #ffffff;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            z-index: 2;
        }

        /* toc panel when closed */
        #toc_panel[data-open="false"] {
            min-width: 60px;
            width: 60px;
            height: 60px;
            border-right: solid 1px #bdbdbd;
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc panel when open */
        #toc_panel[data-open="true"] {
            min-width: 260px;
            max-width: 480px;
            /* keep panel edge consistent distance away from "page" edge */
            width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
            bottom: 0;
            border-right: solid 1px #bdbdbd;
        }

        /* toc panel header */
        #toc_header {
            box-sizing: border-box;
            display: flex;
            flex-direction: row;
            align-items: center;
            height: 60px;
            margin: 0;
            padding: 20px;
        }

        /* toc panel header when hovered */
        #toc_header:hover {
            cursor: pointer;
        }

        /* toc panel header when panel open */
        #toc_panel[data-open="true"] > #toc_header {
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc open/close header button */
        #toc_button {
            margin-right: 20px;
        }

        /* hide toc list and header text when closed */
        #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
        #toc_panel[data-open="false"] > #toc_list {
            display: none;
        }

        /* toc list of entries */
        #toc_list {
            box-sizing: border-box;
            width: 100%;
            padding: 20px;
            position: absolute;
            top: calc(60px + 1px);
            bottom: 0;
            overflow: auto;
        }

        /* toc entry, link to section in document */
        .toc_link {
            display: block;
            padding: 5px;
            position: relative;
            font-weight: 600;
            text-decoration: none;
        }

        /* toc entry when hovered or when "viewed" */
        .toc_link:hover,
        .toc_link[data-viewing="true"] {
            background: #f5f5f5;
        }

        /* toc entry, level 1 indentation */
        .toc_link[data-level="1"] {
            margin-left: 0;
        }

        /* toc entry, level 2 indentation */
        .toc_link[data-level="2"] {
            margin-left: 20px;
        }

        /* toc entry, level 3 indentation */
        .toc_link[data-level="3"] {
            margin-left: 40px;
        }

        /* toc entry, level 4 indentation */
        .toc_link[data-level="4"] {
            margin-left: 60px;
        }

        /* toc entry bullets */
        #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
            position: absolute;
            left: -15px;
            top: -1px;
            font-size: 1.5em;
        }

        /* toc entry, level 2 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
            content: "\2022";
        }

        /* toc entry, level 3 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
            content: "\25AB";
        }

        /* toc entry, level 4 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
            content: "-";
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* push <body> ("page") element down to make room for toc icon */
        .toc_body_nudge {
            padding-top: 60px;
        }

        /* toc icon when panel closed and not hovered */
        #toc_panel[data-open="false"]:not(:hover) {
            background: rgba(255, 255, 255, 0.75);
        }
    }

    /* always hide toc panel on print */
    @media only print {
        #toc_panel {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* lightbox plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* regular <img> in document when hovered */
        .lightbox_document_img:hover {
            cursor: pointer;
        }

        .body_no_scroll {
            overflow: hidden !important;
        }

        /* screen overlay */
        #lightbox_overlay {
            display: flex;
            flex-direction: column;
            position: fixed;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            background: rgba(0, 0, 0, 0.75);
            z-index: 3;
        }

        /* middle area containing lightbox image */
        #lightbox_image_container {
            flex-grow: 1;
            display: flex;
            justify-content: center;
            align-items: center;
            overflow: hidden;
            position: relative;
            padding: 20px;
        }

        /* bottom area containing caption */
        #lightbox_bottom_container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100px;
            min-height: 100px;
            max-height: 100px;
            background: rgba(0, 0, 0, 0.5);
        }

        /* image number info text box */
        #lightbox_number_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            left: 2px;
            top: 0;
            z-index: 4;
        }

        /* zoom info text box */
        #lightbox_zoom_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            right: 2px;
            top: 0;
            z-index: 4;
        }

        /* copy of image caption */
        #lightbox_caption {
            box-sizing: border-box;
            display: inline-block;
            width: 100%;
            max-height: 100%;
            padding: 10px 0;
            text-align: center;
            overflow-y: auto;
            color: #ffffff;
        }

        /* navigation previous/next button */
        .lightbox_button {
            width: 100px;
            height: 100%;
            min-width: 100px;
            min-height: 100%;
            color: #ffffff;
        }

        /* navigation previous/next button when hovered */
        .lightbox_button:hover {
            background: none !important;
        }

        /* navigation button icon */
        .lightbox_button > svg {
            height: 25px;
        }

        /* figure auto-number */
        #lightbox_caption > span:first-of-type {
            font-weight: bold;
            margin-right: 5px;
        }

        /* lightbox image when hovered */
        #lightbox_img:hover {
            cursor: grab;
        }

        /* lightbox image when grabbed */
        #lightbox_img:active {
            cursor: grabbing;
        }
    }

    /* when on screen < 480px wide */
    @media only screen and (max-width: 480px) {
        /* make navigation buttons skinnier on small screens to make more room for caption text */
        .lightbox_button {
            width: 50px;
            min-width: 50px;
        }
    }

    /* always hide lightbox on print */
    @media only print {
        #lightbox_overlay {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* hypothesis (annotations) plugin */
    /* -------------------------------------------------- */

    /* side panel */
    .annotator-frame {
        width: 280px !important;
        z-index: 0 !important;
    }

    /* match highlight color to rest of theme */
    .annotator-highlights-always-on .annotator-hl {
        background-color: #ffeb3b !important;
    }

    /* match focused color to rest of theme */
    .annotator-hl.annotator-hl-focused {
        background-color: #ff8a65 !important;
    }

    /* match bucket bar color to rest of theme */
    .annotator-bucket-bar {
        background: #f5f5f5 !important;
    }

    /* always hide toolbar and tooltip on print */
    @media only print {
        .annotator-frame {
            display: none !important;
        }

        hypothesis-adder {
            display: none !important;
        }
    }
</style>
<!-- anchors plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds an anchor next to each of a certain type
        // of element that provides a human-readable url to that specific
        // item/position in the document (eg "manuscript.html#abstract"). It
        // also makes it such that scrolling out of view of a target removes
        // its identifier from the url.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'anchors';

        // default plugin options
        const options = {
            // which types of elements to add anchors next to, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3, [id^="fig:"], [id^="tbl:"], [id^="eq:"]',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // add anchor to each element of specified types
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements)
                addAnchor(element);

            // attach scroll listener to window
            window.addEventListener('scroll', onScroll);
        }

        // when window is scrolled
        function onScroll() {
            // if url has hash and user has scrolled out of view of hash
            // target, remove hash from url
            const tolerance = 100;
            const target = getHashTarget();
            if (target) {
                if (
                    target.getBoundingClientRect().top >
                        window.innerHeight + tolerance ||
                    target.getBoundingClientRect().bottom < 0 - tolerance
                )
                    history.pushState(null, null, ' ');
            }
        }

        // add anchor to element
        function addAnchor(element) {
            let addTo; // element to add anchor button to

            // if figure or table, modify withId and addTo to get expected
            // elements
            if (element.id.indexOf('fig:') === 0) {
                addTo = element.querySelector('figcaption');
            } else if (element.id.indexOf('tbl:') === 0) {
                addTo = element.querySelector('caption');
            } else if (element.id.indexOf('eq:') === 0) {
                addTo = element.querySelector('.eqnos-number');
            }

            addTo = addTo || element;
            const id = element.id || null;

            // do not add anchor if element doesn't have assigned id.
            // id is generated by pandoc and is assumed to be unique and
            // human-readable
            if (!id)
                return;

            // create anchor button
            const anchor = document.createElement('a');
            anchor.innerHTML = document.querySelector('.icon_link').innerHTML;
            anchor.title = 'Link to this part of the document';
            anchor.classList.add('icon_button', 'anchor');
            anchor.dataset.ignore = 'true';
            anchor.href = '#' + id;
            addTo.appendChild(anchor);
        }

        // get element that is target of link or url hash
        function getHashTarget() {
            const hash = window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- link icon -->

<template class="icon_link">
    <!-- modified from: https://fontawesome.com/icons/link -->
    <svg width="16" height="16" viewBox="0 0 512 512">
        <path
            fill="currentColor"
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        ></path>
    </svg>
</template>
<!-- accordion plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows sections of content under <h2> headings
        // to be collapsible.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'accordion';

        // default plugin options
        const options = {
            // whether to always start expanded ('false'), always start
            // collapsed ('true'), or start collapsed when screen small ('auto')
            startCollapsed: 'auto',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <h2> heading
            const headings = document.querySelectorAll('h2');
            for (const heading of headings) {
                addArrow(heading);

                // start expanded/collapsed based on option
                if (
                    options.startCollapsed === 'true' ||
                    (options.startCollapsed === 'auto' && isSmallScreen())
                )
                    collapseHeading(heading);
                else
                    expandHeading(heading);
            }

            // attach hash change listener to window
            window.addEventListener('hashchange', onHashChange);
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                goToElement(target);
        }

        // add arrow to heading
        function addArrow(heading) {
            // add arrow button
            const arrow = document.createElement('button');
            arrow.innerHTML = document.querySelector(
                '.icon_angle_down'
            ).innerHTML;
            arrow.classList.add('icon_button', 'accordion_arrow');
            heading.insertBefore(arrow, heading.firstChild);

            // attach click listener to heading and button
            heading.addEventListener('click', onHeadingClick);
            arrow.addEventListener('click', onArrowClick);
        }

        // determine if on mobile-like device with small screen
        function isSmallScreen() {
            return Math.min(window.innerWidth, window.innerHeight) < 480;
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get element that is target of hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // when <h2> heading is clicked
        function onHeadingClick(event) {
            // only collapse if <h2> itself is target of click (eg, user did
            // not click on anchor within <h2>)
            if (event.target === this)
                toggleCollapse(this);
        }

        // when arrow button is clicked
        function onArrowClick() {
            toggleCollapse(this.parentNode);
        }

        // collapse section if expanded, expand if collapsed
        function toggleCollapse(heading) {
            if (heading.dataset.collapsed === 'false')
                collapseHeading(heading);
            else
                expandHeading(heading);
        }

        // elements to exclude from collapse, such as table of contents panel,
        // hypothesis panel, etc
        const exclude = '#toc_panel, div.annotator-frame, #lightbox_overlay';

        // collapse section
        function collapseHeading(heading) {
            heading.setAttribute('data-collapsed', 'true');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'true');
        }

        // expand section
        function expandHeading(heading) {
            heading.setAttribute('data-collapsed', 'false');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'false');
        }

        // get list of elements between this <h2> and next <h2> or <h1>
        // ("children" of the <h2> section)
        function getChildren(heading) {
            return nextUntil(heading, 'h2, h1', exclude);
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get list of elements after a start element up to element matching
        // query
        function nextUntil(element, query, exclude) {
            const elements = [];
            while (element = element.nextElementSibling, element) {
                if (element.matches(query))
                    break;
                if (!element.matches(exclude))
                    elements.push(element);
            }
            return elements;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
    <!-- modified from: https://fontawesome.com/icons/angle-down -->
    <svg width="16" height="16" viewBox="0 0 448 512">
        <path
            fill="currentColor"
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        ></path>
    </svg>
</template>
<!-- tooltips plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when the user hovers or
        // focuses a link to a citation or figure, a tooltip appears with a
        // preview of the reference content, along with arrows to navigate
        // between instances of the same reference in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tooltips';

        // default plugin options
        const options = {
            // whether user must click off to close tooltip instead of just
            // un-hovering
            clickClose: 'false',
            // delay (in ms) between opening and closing tooltip
            delay: '100',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach hover and focus listeners to link
                link.addEventListener('mouseover', onLinkHover);
                link.addEventListener('mouseleave', onLinkUnhover);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('touchend', onLinkTouch);
            }

            // attach mouse, key, and resize listeners to window
            window.addEventListener('mousedown', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('keyup', onKeyUp);
            window.addEventListener('resize', onResize);
        }

        // when link is hovered
        function onLinkHover() {
            // function to open tooltip
            const delayOpenTooltip = function() {
                openTooltip(this);
            }.bind(this);

            // run open function after delay
            this.openTooltipTimer = window.setTimeout(
                delayOpenTooltip,
                options.delay
            );
        }

        // when mouse leaves link
        function onLinkUnhover() {
            // cancel opening tooltip
            window.clearTimeout(this.openTooltipTimer);

            // don't close on unhover if option specifies
            if (options.clickClose === 'true')
                return;

            // function to close tooltip
            const delayCloseTooltip = function() {
                // if tooltip open and if mouse isn't over tooltip, close
                const tooltip = document.getElementById('tooltip');
                if (tooltip && !tooltip.matches(':hover'))
                    closeTooltip();
            };

            // run close function after delay
            this.closeTooltipTimer = window.setTimeout(
                delayCloseTooltip,
                options.delay
            );
        }

        // when link is focused (tabbed to)
        function onLinkFocus(event) {
            openTooltip(this);
        }

        // when link is touched on touch screen
        function onLinkTouch(event) {
            // attempt to force hover state on first tap always, and trigger
            // regular link click (and navigation) on second tap
            if (event.target === document.activeElement)
                event.target.click();
            else {
                document.activeElement.blur();
                event.target.focus();
            }
            if (event.cancelable)
                event.preventDefault();
            event.stopPropagation();
            return false;
        }

        // when mouse is clicked anywhere in window
        function onClick(event) {
            closeTooltip();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'tooltip_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'tooltip_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeTooltip();
                    break;
            }
        }

        // when window is resized or zoomed
        function onResize() {
            closeTooltip();
        }

        // get all links of types we wish to handle
        function getLinks() {
            const queries = [];
            // exclude buttons, anchor links, toc links, etc
            const exclude =
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            queries.push('a[href^="#ref-"]' + exclude); // citation links
            queries.push('a[href^="#fig:"]' + exclude); // figure links
            const query = queries.join(', ');
            return document.querySelectorAll(query);
        }

        // get links with same target, get index of link in set, get total
        // same links
        function getSameLinks(link) {
            const sameLinks = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    sameLinks.push(otherLink);
            }

            return {
                elements: sameLinks,
                index: sameLinks.indexOf(link),
                total: sameLinks.length
            };
        }

        // open tooltip
        function openTooltip(link) {
            // delete tooltip if it exists, start fresh
            closeTooltip();

            // make tooltip element
            const tooltip = makeTooltip(link);

            // if source couldn't be found and tooltip not made, exit
            if (!tooltip)
                return;

            // make navbar elements
            const navBar = makeNavBar(link);
            if (navBar)
                tooltip.firstElementChild.appendChild(navBar);

            // attach tooltip to page
            document.body.appendChild(tooltip);

            // position tooltip
            const position = function() {
                positionTooltip(link);
            };
            position();

            // if tooltip contains images, position again after they've loaded
            const imgs = tooltip.querySelectorAll('img');
            for (const img of imgs)
                img.addEventListener('load', position);
        }

        // close (delete) tooltip
        function closeTooltip() {
            const tooltip = document.getElementById('tooltip');
            if (tooltip)
                tooltip.remove();
        }

        // make tooltip
        function makeTooltip(link) {
            // get target element that link points to
            const source = getSource(link);

            // if source can't be found, exit
            if (!source)
                return;

            // create new tooltip
            const tooltip = document.createElement('div');
            tooltip.id = 'tooltip';
            const tooltipContent = document.createElement('div');
            tooltipContent.id = 'tooltip_content';
            tooltip.appendChild(tooltipContent);

            // make copy of source node and put in tooltip
            const sourceCopy = makeCopy(source);
            tooltipContent.appendChild(sourceCopy);

            // attach mouse event listeners
            tooltip.addEventListener('click', onTooltipClick);
            tooltip.addEventListener('mousedown', onTooltipClick);
            tooltip.addEventListener('touchstart', onTooltipClick);
            tooltip.addEventListener('mouseleave', onTooltipUnhover);

            // (for interaction with lightbox plugin)
            // transfer click on tooltip copied img to original img
            const sourceImg = source.querySelector('img');
            const sourceCopyImg = sourceCopy.querySelector('img');
            if (sourceImg && sourceCopyImg) {
                const clickImg = function() {
                    sourceImg.click();
                    closeTooltip();
                };
                sourceCopyImg.addEventListener('click', clickImg);
            }

            return tooltip;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // when tooltip is clicked
        function onTooltipClick(event) {
            // when user clicks on tooltip, stop click from transferring
            // outside of tooltip (eg, click off to close tooltip, or eg click
            // off to unhighlight same refs)
            event.stopPropagation();
        }

        // when tooltip is unhovered
        function onTooltipUnhover(event) {
            if (options.clickClose === 'true')
                return;

            // make sure new mouse/touch/focus no longer over tooltip or any
            // element within it
            const tooltip = document.getElementById('tooltip');
            if (!tooltip)
                return;
            if (this.contains(event.relatedTarget))
                return;

            closeTooltip();
        }

        // make nav bar to go betwen prev/next instances of same reference
        function makeNavBar(link) {
            // find other links to the same source
            const sameLinks = getSameLinks(link);

            // don't show nav bar when singular reference
            if (sameLinks.total <= 1)
                return;

            // find prev/next links with same target
            const prevLink = getPrevLink(link, sameLinks);
            const nextLink = getNextLink(link, sameLinks);

            // create nav bar
            const navBar = document.createElement('div');
            navBar.id = 'tooltip_nav_bar';
            const text = sameLinks.index + 1 + ' of ' + sameLinks.total;

            // create nav bar prev/next buttons
            const prevButton = document.createElement('button');
            const nextButton = document.createElement('button');
            prevButton.id = 'tooltip_prev_button';
            nextButton.id = 'tooltip_next_button';
            prevButton.title =
                'Jump to the previous occurence of this item in the document [←]';
            nextButton.title =
                'Jump to the next occurence of this item in the document [→]';
            prevButton.classList.add('icon_button');
            nextButton.classList.add('icon_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;
            navBar.appendChild(prevButton);
            navBar.appendChild(document.createTextNode(text));
            navBar.appendChild(nextButton);

            // attach click listeners to buttons
            prevButton.addEventListener('click', function() {
                onPrevNextClick(link, prevLink);
            });
            nextButton.addEventListener('click', function() {
                onPrevNextClick(link, nextLink);
            });

            return navBar;
        }

        // get previous link with same target
        function getPrevLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if < 1
            let index;
            if (sameLinks.index - 1 >= 0)
                index = sameLinks.index - 1;
            else
                index = sameLinks.total - 1;
            return sameLinks.elements[index];
        }

        // get next link with same target
        function getNextLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if > total
            let index;
            if (sameLinks.index + 1 <= sameLinks.total - 1)
                index = sameLinks.index + 1;
            else
                index = 0;
            return sameLinks.elements[index];
        }

        // get element that is target of link or url hash
        function getSource(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if ref or figure, modify target to get expected element
            if (id.indexOf('ref-') === 0)
                target = target.querySelector('p');
            else if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');

            return target;
        }

        // when prev/next arrow button is clicked
        function onPrevNextClick(link, prevNextLink) {
            if (link && prevNextLink)
                goToElement(prevNextLink, window.innerHeight * 0.5);
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // determine position to place tooltip based on link position in
        // viewport and tooltip size
        function positionTooltip(link, left, top) {
            const tooltipElement = document.getElementById('tooltip');
            if (!tooltipElement)
                return;

            // get convenient vars for position/dimensions of
            // link/tooltip/page/view
            link = getRectInPage(link);
            const tooltip = getRectInPage(tooltipElement);
            const view = getRectInPage();

            // horizontal positioning
            if (left)
                // use explicit value
                left = left;
            else if (link.left + tooltip.width < view.right)
                // fit tooltip to right of link
                left = link.left;
            else if (link.right - tooltip.width > view.left)
                // fit tooltip to left of link
                left = link.right - tooltip.width;
            // center tooltip in view
            else
                left = (view.right - view.left) / 2 - tooltip.width / 2;

            // vertical positioning
            if (top)
                // use explicit value
                top = top;
            else if (link.top - tooltip.height > view.top)
                // fit tooltip above link
                top = link.top - tooltip.height;
            else if (link.bottom + tooltip.height < view.bottom)
                // fit tooltip below link
                top = link.bottom;
            else {
                // center tooltip in view
                top = view.top + view.height / 2 - tooltip.height / 2;
                // nudge off of link to left/right if possible
                if (link.right + tooltip.width < view.right)
                    left = link.right;
                else if (link.left - tooltip.width > view.left)
                    left = link.left - tooltip.width;
            }

            tooltipElement.style.left = left + 'px';
            tooltipElement.style.top = top + 'px';
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get position of element relative to page
        function getRectInPage(element) {
            const rect = getRectInView(element);
            const body = getRectInView(document.body);

            const newRect = {};
            newRect.left = rect.left - body.left;
            newRect.top = rect.top - body.top;
            newRect.right = rect.right - body.left;
            newRect.bottom = rect.bottom - body.top;
            newRect.width = rect.width;
            newRect.height = rect.height;

            return newRect;
        }

        // (for interaction with accordion plugin)
        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // (for interaction with accordion plugin)
        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- jump to first plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds a button next to each reference entry,
        // figure, and table that jumps the page to the first occurrence of a
        // link to that item in the manuscript.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'jumpToFirst';

        // default plugin options
        const options = {
            // whether to add buttons next to reference entries
            references: 'true',
            // whether to add buttons next to figures
            figures: 'true',
            // whether to add buttons next to tables
            tables: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            if (options.references !== 'false')
                makeReferenceButtons();
            if (options.figures !== 'false')
                makeFigureButtons();
            if (options.tables !== 'false')
                makeTableButtons();
        }

        // when jump button clicked
        function onButtonClick() {
            const first = getFirstOccurrence(this.dataset.id);
            if (!first)
                return;

            // update url hash so navigating "back" in history will return
            // user to jump button
            window.location.hash = this.dataset.id;
            // scroll to link
            window.setTimeout(function() {
                goToElement(first, window.innerHeight * 0.5);
            }, 0);
        }

        // get first occurence of link to item in document
        function getFirstOccurrence(id) {
            let query = 'a';
            query += '[href="#' + id + '"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelector(query);
        }

        // add button next to each reference entry
        function makeReferenceButtons() {
            const references = document.querySelectorAll('div[id^="ref-"]');
            for (const reference of references) {
                // get reference id and element to add button to
                const id = reference.id;
                const container = reference.firstElementChild;
                const first = getFirstOccurrence(id);

                // if can't find link to reference, ignore
                if (!first)
                    continue;

                // make jump button
                let button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this reference in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.innerHTML = button.outerHTML + container.innerHTML;
                button = container.firstElementChild;
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeFigureButtons() {
            const figures = document.querySelectorAll('[id^="fig:"]');
            for (const figure of figures) {
                // get figure id and element to add button to
                const id = figure.id;
                const container = figure.querySelector('figcaption') || figure;
                const first = getFirstOccurrence(id);

                // if can't find link to figure, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this figure in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeTableButtons() {
            const tables = document.querySelectorAll('[id^="tbl:"]');
            for (const table of tables) {
                // get ref id and element to add button to
                const id = table.id;
                const container = table.querySelector('caption') || table;
                const first = getFirstOccurrence(id);

                // if can't find link to table, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this table in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
    <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
    <svg width="16" height="16" viewBox="0 0 320 512">
        <path
            fill="currentColor"
            d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
        ></path>
    </svg>
</template>
<!-- link highlight plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user hovers or
        // focuses a link, other links that have the same target will be
        // highlighted. It also makes it such that when clicking a link, the
        // target of the link (eg reference, figure, table) is briefly
        // highlighted.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'linkHighlight';

        // default plugin options
        const options = {
            // whether to also highlight links that go to external urls
            externalLinks: 'false',
            // whether user must click off to unhighlight instead of just
            // un-hovering
            clickUnhighlight: 'false',
            // whether to also highlight links that are unique
            highlightUnique: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach mouse and focus listeners to link
                link.addEventListener('mouseenter', onLinkFocus);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('mouseleave', onLinkUnhover);
            }

            // attach click and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('hashchange', onHashChange);

            // run hash change on window load in case user has navigated
            // directly to hash
            onHashChange();
        }

        // when link is focused (tabbed to) or hovered
        function onLinkFocus() {
            highlight(this);
        }

        // when link is unhovered
        function onLinkUnhover() {
            if (options.clickUnhighlight !== 'true')
                unhighlightAll();
        }

        // when the mouse is clicked anywhere in window
        function onClick(event) {
            unhighlightAll();
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                glowElement(target);
        }

        // get element that is target of link or url hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            return target;
        }

        // start glow sequence on an element
        function glowElement(element) {
            const startGlow = function() {
                onGlowEnd();
                element.dataset.glow = 'true';
                element.addEventListener('animationend', onGlowEnd);
            };
            const onGlowEnd = function() {
                element.removeAttribute('data-glow');
                element.removeEventListener('animationend', onGlowEnd);
            };
            startGlow();
        }

        // highlight link and all others with same target
        function highlight(link) {
            // force unhighlight all to start fresh
            unhighlightAll();

            // get links with same target
            if (!link)
                return;
            const sameLinks = getSameLinks(link);

            // if link unique and option is off, exit and don't highlight
            if (sameLinks.length <= 1 && options.highlightUnique !== 'true')
                return;

            // highlight all same links, and "select" (special highlight) this
            // one
            for (const sameLink of sameLinks) {
                if (sameLink === link)
                    sameLink.setAttribute('data-selected', 'true');
                else
                    sameLink.setAttribute('data-highlighted', 'true');
            }
        }

        // unhighlight all links
        function unhighlightAll() {
            const links = getLinks();
            for (const link of links) {
                link.setAttribute('data-selected', 'false');
                link.setAttribute('data-highlighted', 'false');
            }
        }

        // get links with same target
        function getSameLinks(link) {
            const results = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    results.push(otherLink);
            }
            return results;
        }

        // get all links of types we wish to handle
        function getLinks() {
            let query = 'a';
            if (options.externalLinks !== 'true')
                query += '[href^="#"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelectorAll(query);
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- table of contents plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin provides a "table of contents" (toc) panel on
        // the side of the document that allows the user to conveniently
        // navigate between sections of the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableOfContents';

        // default plugin options
        const options = {
            // which types of elements to add links for, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3',
            // whether default behavior is to be closed ('false'), open
            // ('true'), or only open when screen wide enough to fit panel
            // ('auto'). note: still always starts closed when page loads.
            open: 'auto',
            // if list item is more than this many characters, text will be
            // truncated
            charLimit: '50',
            // whether or not to show bullets next to each toc item
            bullets: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // make toc panel and populate with entries (links to document
            // sections)
            const panel = makePanel();
            if (!panel)
                return;
            makeEntries(panel);
            document.body.insertBefore(panel, document.body.firstChild);

            closePanel();

            // attach click, scroll, and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('scroll', onScroll);
            window.addEventListener('hashchange', onScroll);
            window.addEventListener('keyup', onKeyUp);
            onScroll();

            // add class to push document body down out of way of toc button
            document.body.classList.add('toc_body_nudge');
        }

        // determine if screen wide enough to fit toc panel
        function isSmallScreen() {
            // in default theme:
            // 816px = 8.5in = width of "page" (<body>) element
            // 260px = min width of toc panel (*2 for both sides of <body>)
            return window.innerWidth < 816 + 260 * 2;
        }

        // open/close panel based on option and screen size
        function openOrClosePanel() {
            if (
                options.open === 'true' ||
                (options.open === 'auto' && !isSmallScreen())
            )
                openPanel();
            else
                closePanel();
        }

        // when mouse is clicked anywhere in window
        function onClick() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                openOrClosePanel();
        }

        // when window is scrolled or hash changed
        function onScroll() {
            highlightViewed();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            // close on esc
            if (event.key === 'Escape')
                closePanel();
        }

        // find entry of currently viewed document section in toc and highlight
        function highlightViewed() {
            const firstId = getFirstInView(options.typesQuery);

            // get toc entries (links), unhighlight all, then highlight viewed
            const list = document.getElementById('toc_list');
            if (!firstId || !list)
                return;
            const links = list.querySelectorAll('a');
            for (const link of links)
                link.dataset.viewing = 'false';
            const link = list.querySelector('a[href="#' + firstId + '"]');
            if (!link)
                return;
            link.dataset.viewing = 'true';
        }

        // get first or previous toc listed element in top half of view
        function getFirstInView(query) {
            // get all elements matching query and with id
            const elements = document.querySelectorAll(query);
            const elementsWithIds = [];
            for (const element of elements) {
                if (element.id)
                    elementsWithIds.push(element);
            }


            // get first or previous element in top half of view
            for (let i = 0; i < elementsWithIds.length; i++) {
                const element = elementsWithIds[i];
                const prevElement = elementsWithIds[Math.max(0, i - 1)];
                if (element.getBoundingClientRect().top >= 0) {
                    if (
                        element.getBoundingClientRect().top <
                        window.innerHeight / 2
                    )
                        return element.id;
                    else
                        return prevElement.id;
                }
            }
        }

        // make panel
        function makePanel() {
            // create panel
            const panel = document.createElement('div');
            panel.id = 'toc_panel';
            if (options.bullets === 'true')
                panel.dataset.bullets = 'true';

            // create header
            const header = document.createElement('div');
            header.id = 'toc_header';

            // create toc button
            const button = document.createElement('button');
            button.id = 'toc_button';
            button.innerHTML = document.querySelector(
                '.icon_th_list'
            ).innerHTML;
            button.classList.add('icon_button');

            // create header text
            const text = document.createElement('h3');
            text.innerHTML = 'Table of Contents';

            // create container for toc list
            const list = document.createElement('div');
            list.id = 'toc_list';

            // attach click listeners
            panel.addEventListener('click', onPanelClick);
            header.addEventListener('click', onHeaderClick);
            button.addEventListener('click', onButtonClick);

            // attach elements
            header.appendChild(button);
            header.appendChild(text);
            panel.appendChild(header);
            panel.appendChild(list);

            return panel;
        }

        // create toc entries (links) to each element of the specified types
        function makeEntries(panel) {
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements) {
                // do not add link if element doesn't have assigned id
                if (!element.id)
                    continue;

                // create link/list item
                const link = document.createElement('a');
                link.classList.add('toc_link');
                switch (element.tagName.toLowerCase()) {
                    case 'h1':
                        link.dataset.level = '1';
                        break;
                    case 'h2':
                        link.dataset.level = '2';
                        break;
                    case 'h3':
                        link.dataset.level = '3';
                        break;
                    case 'h4':
                        link.dataset.level = '4';
                        break;
                }
                link.title = element.innerText;
                let text = element.innerText;
                if (text.length > options.charLimit)
                    text = text.slice(0, options.charLimit) + '...';
                link.innerHTML = text;
                link.href = '#' + element.id;
                link.addEventListener('click', onLinkClick);

                // attach link
                panel.querySelector('#toc_list').appendChild(link);
            }
        }

        // when panel is clicked
        function onPanelClick(event) {
            // stop click from propagating to window/document and closing panel
            event.stopPropagation();
        }

        // when header itself is clicked
        function onHeaderClick(event) {
            togglePanel();
        }

        // when button is clicked
        function onButtonClick(event) {
            togglePanel();
            // stop header underneath button from also being clicked
            event.stopPropagation();
        }

        // when link is clicked
        function onLinkClick() {
            openOrClosePanel();
        }

        // open panel if closed, close if opened
        function togglePanel() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                closePanel();
            else
                openPanel();
        }

        // open panel
        function openPanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'true';
        }

        // close panel
        function closePanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'false';
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- th list icon -->

<template class="icon_th_list">
    <!-- modified from: https://fontawesome.com/icons/th-list -->
    <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
        <path
            fill="currentColor"
            d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- lightbox plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user clicks on an
        // image, the image fills the screen and the user can pan/drag/zoom
        // the image and navigate between other images in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'lightbox';

        // default plugin options
        const options = {
            // list of possible zoom/scale factors
            zoomSteps:
                '0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1,' +
                '1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8',
            // whether to fit image to view ('fit'), display at 100% and shrink
            // if necessary ('shrink'), or always display at 100% ('100')
            defaultZoom: 'fit',
            // whether to zoom in/out toward center of view ('true') or mouse
            // ('false')
            centerZoom: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <img> element
            const imgs = document.querySelectorAll('figure > img');
            let count = 1;
            for (const img of imgs) {
                img.classList.add('lightbox_document_img');
                img.dataset.number = count;
                img.dataset.total = imgs.length;
                img.addEventListener('click', openLightbox);
                count++;
            }

            // attach mouse and key listeners to window
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('keyup', onKeyUp);
        }

        // when mouse is moved anywhere in window
        function onWindowMouseMove(event) {
            window.mouseX = event.clientX;
            window.mouseY = event.clientY;
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'lightbox_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'lightbox_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeLightbox();
                    break;
            }
        }

        // open lightbox
        function openLightbox() {
            const lightbox = makeLightbox(this);
            if (!lightbox)
                return;

            blurBody(lightbox);
            document.body.appendChild(lightbox);
        }

        // make lightbox
        function makeLightbox(img) {
            // delete lightbox if it exists, start fresh
            closeLightbox();

            // create screen overlay containing lightbox
            const overlay = document.createElement('div');
            overlay.id = 'lightbox_overlay';

            // create image info boxes
            const numberInfo = document.createElement('div');
            const zoomInfo = document.createElement('div');
            numberInfo.id = 'lightbox_number_info';
            zoomInfo.id = 'lightbox_zoom_info';

            // create container for image
            const imageContainer = document.createElement('div');
            imageContainer.id = 'lightbox_image_container';
            const lightboxImg = makeLightboxImg(
                img,
                imageContainer,
                numberInfo,
                zoomInfo
            );
            imageContainer.appendChild(lightboxImg);

            // create bottom container for caption and navigation buttons
            const bottomContainer = document.createElement('div');
            bottomContainer.id = 'lightbox_bottom_container';
            const caption = makeCaption(img);
            const prevButton = makePrevButton(img);
            const nextButton = makeNextButton(img);
            bottomContainer.appendChild(prevButton);
            bottomContainer.appendChild(caption);
            bottomContainer.appendChild(nextButton);

            // attach top middle and bottom to overlay
            overlay.appendChild(numberInfo);
            overlay.appendChild(zoomInfo);
            overlay.appendChild(imageContainer);
            overlay.appendChild(bottomContainer);

            return overlay;
        }

        // make <img> object that is intuitively draggable and zoomable
        function makeLightboxImg(
            sourceImg,
            container,
            numberInfoBox,
            zoomInfoBox
        ) {
            // create copy of source <img>
            const img = sourceImg.cloneNode(true);
            img.classList.remove('lightbox_document_img');
            img.removeAttribute('id');
            img.removeAttribute('width');
            img.removeAttribute('height');
            img.style.position = 'unset';
            img.style.margin = '0';
            img.style.padding = '0';
            img.style.width = '';
            img.style.height = '';
            img.style.minWidth = '';
            img.style.minHeight = '';
            img.style.maxWidth = '';
            img.style.maxHeight = '';
            img.id = 'lightbox_img';

            // build sorted list of unique zoomSteps, always including a 100%
            let zoomSteps = [];
            const optionsZooms = options.zoomSteps.split(/[^0-9.]/);
            for (const optionZoom of optionsZooms) {
                const newZoom = parseFloat(optionZoom);
                if (newZoom && !zoomSteps.includes(newZoom))
                    zoomSteps.push(newZoom);
            }
            if (!zoomSteps.includes(1))
                zoomSteps.push(1);
            zoomSteps = zoomSteps.sort(function sortNumber(a, b) {
                return a - b;
            });

            // <img> object property variables
            let zoom = 1;
            let translateX = 0;
            let translateY = 0;
            let clickMouseX = undefined;
            let clickMouseY = undefined;
            let clickTranslateX = undefined;
            let clickTranslateY = undefined;

            updateNumberInfo();

            // update image numbers displayed in info box
            function updateNumberInfo() {
                numberInfoBox.innerHTML =
                    sourceImg.dataset.number + ' of ' + sourceImg.dataset.total;
            }

            // update zoom displayed in info box
            function updateZoomInfo() {
                let zoomInfo = zoom * 100;
                if (!Number.isInteger(zoomInfo))
                    zoomInfo = zoomInfo.toFixed(2);
                zoomInfoBox.innerHTML = zoomInfo + '%';
            }

            // move to closest zoom step above current zoom
            const zoomIn = function() {
                for (const zoomStep of zoomSteps) {
                    if (zoomStep > zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                updateTransform();
            };

            // move to closest zoom step above current zoom
            const zoomOut = function() {
                zoomSteps.reverse();
                for (const zoomStep of zoomSteps) {
                    if (zoomStep < zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                zoomSteps.reverse();

                updateTransform();
            };

            // update display of <img> based on scale/translate properties
            const updateTransform = function() {
                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                // get new width/height after scale
                const rect = img.getBoundingClientRect();
                // limit translate
                translateX = Math.max(translateX, -rect.width / 2);
                translateX = Math.min(translateX, rect.width / 2);
                translateY = Math.max(translateY, -rect.height / 2);
                translateY = Math.min(translateY, rect.height / 2);

                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                updateZoomInfo();
            };

            // fit <img> to container
            const fit = function() {
                // no x/y offset, 100% zoom by default
                translateX = 0;
                translateY = 0;
                zoom = 1;

                // widths of <img> and container
                const imgWidth = img.naturalWidth;
                const imgHeight = img.naturalHeight;
                const containerWidth = parseFloat(
                    window.getComputedStyle(container).width
                );
                const containerHeight = parseFloat(
                    window.getComputedStyle(container).height
                );

                // how much zooming is needed to fit <img> to container
                const xRatio = imgWidth / containerWidth;
                const yRatio = imgHeight / containerHeight;
                const maxRatio = Math.max(xRatio, yRatio);
                const newZoom = 1 / maxRatio;

                // fit <img> to container according to option
                if (options.defaultZoom === 'shrink') {
                    if (maxRatio > 1)
                        zoom = newZoom;
                } else if (options.defaultZoom === 'fit')
                    zoom = newZoom;

                updateTransform();
            };

            // when mouse wheel is rolled anywhere in container
            const onContainerWheel = function(event) {
                if (!event)
                    return;

                // let ctrl + mouse wheel to zoom behave as normal
                if (event.ctrlKey)
                    return;

                // prevent normal scroll behavior
                event.preventDefault();
                event.stopPropagation();

                // point around which to scale img
                const viewRect = container.getBoundingClientRect();
                const viewX = (viewRect.left + viewRect.right) / 2;
                const viewY = (viewRect.top + viewRect.bottom) / 2;
                const originX = options.centerZoom === 'true' ? viewX : mouseX;
                const originY = options.centerZoom === 'true' ? viewY : mouseY;

                // get point on image under origin
                const oldRect = img.getBoundingClientRect();
                const oldPercentX = (originX - oldRect.left) / oldRect.width;
                const oldPercentY = (originY - oldRect.top) / oldRect.height;

                // increment/decrement zoom
                if (event.deltaY < 0)
                    zoomIn();
                if (event.deltaY > 0)
                    zoomOut();

                // get offset between previous image point and origin
                const newRect = img.getBoundingClientRect();
                const offsetX =
                    originX - (newRect.left + newRect.width * oldPercentX);
                const offsetY =
                    originY - (newRect.top + newRect.height * oldPercentY);

                // translate image to keep image point under origin
                translateX += offsetX;
                translateY += offsetY;

                // perform translate
                updateTransform();
            };

            // when container is clicked
            function onContainerClick(event) {
                // if container itself is target of click, and not other
                // element above it
                if (event.target === this)
                    closeLightbox();
            }

            // when mouse button is pressed on image
            const onImageMouseDown = function(event) {
                // store original mouse position relative to image
                clickMouseX = window.mouseX;
                clickMouseY = window.mouseY;
                clickTranslateX = translateX;
                clickTranslateY = translateY;
                event.stopPropagation();
                event.preventDefault();
            };

            // when mouse button is released anywhere in window
            const onWindowMouseUp = function(event) {
                // reset original mouse position
                clickMouseX = undefined;
                clickMouseY = undefined;
                clickTranslateX = undefined;
                clickTranslateY = undefined;

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mouseup', onWindowMouseUp);
            };

            // when mouse is moved anywhere in window
            const onWindowMouseMove = function(event) {
                if (
                    clickMouseX === undefined ||
                    clickMouseY === undefined ||
                    clickTranslateX === undefined ||
                    clickTranslateY === undefined
                )
                    return;

                // offset image based on original and current mouse position
                translateX = clickTranslateX + window.mouseX - clickMouseX;
                translateY = clickTranslateY + window.mouseY - clickMouseY;
                updateTransform();
                event.preventDefault();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mousemove', onWindowMouseMove);
            };

            // when window is resized
            const onWindowResize = function(event) {
                fit();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('resize', onWindowResize);
            };

            // attach the necessary event listeners
            img.addEventListener('dblclick', fit);
            img.addEventListener('mousedown', onImageMouseDown);
            container.addEventListener('wheel', onContainerWheel);
            container.addEventListener('mousedown', onContainerClick);
            container.addEventListener('touchstart', onContainerClick);
            window.addEventListener('mouseup', onWindowMouseUp);
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('resize', onWindowResize);

            // run fit() after lightbox atttached to document and <img> Loaded
            // so needed container and img dimensions available
            img.addEventListener('load', fit);

            return img;
        }

        // make caption
        function makeCaption(img) {
            const caption = document.createElement('div');
            caption.id = 'lightbox_caption';
            const captionSource = img.nextElementSibling;
            if (captionSource.tagName.toLowerCase() === 'figcaption') {
                const captionCopy = makeCopy(captionSource);
                caption.innerHTML = captionCopy.innerHTML;
            }

            caption.addEventListener('touchstart', function(event) {
                event.stopPropagation();
            });

            return caption;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // make button to jump to previous image in document
        function makePrevButton(img) {
            const prevButton = document.createElement('button');
            prevButton.id = 'lightbox_prev_button';
            prevButton.title = 'Jump to the previous image in the document [←]';
            prevButton.classList.add('icon_button', 'lightbox_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;

            // attach click listeners to button
            prevButton.addEventListener('click', function() {
                getPrevImg(img).click();
            });

            return prevButton;
        }

        // make button to jump to next image in document
        function makeNextButton(img) {
            const nextButton = document.createElement('button');
            nextButton.id = 'lightbox_next_button';
            nextButton.title = 'Jump to the next image in the document [→]';
            nextButton.classList.add('icon_button', 'lightbox_button');
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;

            // attach click listeners to button
            nextButton.addEventListener('click', function() {
                getNextImg(img).click();
            });

            return nextButton;
        }

        // get previous image in document
        function getPrevImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if < 1
            if (index - 1 >= 0)
                index--;
            else
                index = imgs.length - 1;
            return imgs[index];
        }

        // get next image in document
        function getNextImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if > total
            if (index + 1 <= imgs.length - 1)
                index++;
            else
                index = 0;
            return imgs[index];
        }

        // close lightbox
        function closeLightbox() {
            focusBody();

            const lightbox = document.getElementById('lightbox_overlay');
            if (lightbox)
                lightbox.remove();
        }

        // make all elements behind lightbox non-focusable
        function blurBody(overlay) {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.tabIndex = -1;
            document.body.classList.add('body_no_scroll');
        }

        // make all elements focusable again
        function focusBody() {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.removeAttribute('tabIndex');
            document.body.classList.remove('body_no_scroll');
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- attributes plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows arbitrary HTML attributes to be attached
        // to (almost) any element. Place an HTML comment inside or next to the
        // desired element in the format <!-- $attribute="value" -->

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'attributes';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // get list of comments in document
            const comments = findComments();

            for(const comment of comments)
                if (comment.parentElement)
                    addAttributes(
                        comment.parentElement,
                        comment.nodeValue.trim()
                    );
        }

        // add html attributes to specified element based on string of 
        // html attributes and values
        function addAttributes(element, text) {
            // regex's for finding attribute/value pairs in the format of
            // attribute="value" or attribute='value
            const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
            const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

            // loop through attribute/value pairs
            let match;
            while(match = text.match(regex2) || text.match(regex1)) {
                // get attribute and value from regex capture groups
                let attribute = match[1];
                let value = match[2];

                // remove from string
                text = text.substring(match.index + match[0].length);

                if (!attribute || !value)
                    break;

                // set attribute of parent element
                try {
                    element.setAttribute(attribute, value);
                } catch(error) {
                    console.log(error);
                }

                // special case for colspan
                if (attribute === 'colspan')
                    removeTableCells(element, value);
            }
        }

        // get list of comment elements in document
        function findComments() {
            const comments = [];

            // iterate over comment nodes in document
            function acceptNode(node) {
                return NodeFilter.FILTER_ACCEPT;
            }
            const iterator = document.createNodeIterator(
                document.body,
                NodeFilter.SHOW_COMMENT,
                acceptNode
            );
            let node;
            while(node = iterator.nextNode())
                comments.push(node);

            return comments;
        }

        // remove certain number of cells after specified cell
        function removeTableCells(cell, number) {
            number = parseInt(number);
            if (!number)
                return;

            // remove elements
            for(; number > 1; number--) {
                if (cell.nextElementSibling)
                    cell.nextElementSibling.remove();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- math plugin configuration -->

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        "CommonHTML": { linebreaks: { automatic: true } },
        "HTML-CSS": { linebreaks: { automatic: true } },
        "SVG": { linebreaks: { automatic: true } },
        "fast-preview": { disabled: true }
    });
</script>

<!-- math plugin -->

<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'MathJax' allows the proper rendering of
    // math/equations written in LaTeX.

    // https://www.mathjax.org/
</script>
<!-- annotations plugin configuration -->

<script>
    window.hypothesisConfig = function() {
        return {
            branding: {
                accentColor: '#2196f3',
                appBackgroundColor: '#f8f8f8',
                ctaBackgroundColor: '#f8f8f8',
                ctaTextColor: '#000000',
                selectionFontFamily: 'Open Sans, Helvetica, sans serif',
                annotationFontFamily: 'Open Sans, Helvetica, sans serif'
            }
        };
    };
</script>

<!-- annotations plugin -->

<script src='https://hypothes.is/embed.js'>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'Hypothesis' allows public annotation of the
    // manuscript.

    // https://web.hypothes.is/
</script>
<!-- analytics plugin -->

<!-- copy and paste code from Google Analytics or similar service here -->
</body>
</html>
