<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Olga Borisovna Botvinnik" />
  <meta name="author" content="Venkata Naga Pranathi Vemuri" />
  <meta name="author" content="N. Tessa Pierce" />
  <meta name="author" content="Phoenix Aja Logan" />
  <meta name="author" content="Saba Nafees" />
  <meta name="author" content="C. Titus Brown" />
  <meta name="author" content="Jim Karkanias" />
  <meta name="dcterms.date" content="2020-12-07" />
  <meta name="keywords" content="comparative transcriptomics, single-cell RNA-seq, k-mer, hashing, MinHash, gene orthology, orthologous genes, reduced amino acid alphabet, degenerate amino acid alphabet" />
  <title>k-mer homology is a transcriptome-first method enabling functional prediction of transcriptomic ‘dark matter’ across species</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="$k$-mer homology is a transcriptome-first method enabling functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta name="citation_title" content="$k$-mer homology is a transcriptome-first method enabling functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta property="og:title" content="$k$-mer homology is a transcriptome-first method enabling functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta property="twitter:title" content="$k$-mer homology is a transcriptome-first method enabling functional prediction of transcriptomic &#39;dark matter&#39; across species" />
  <meta name="dc.date" content="2020-12-07" />
  <meta name="citation_publication_date" content="2020-12-07" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Olga Borisovna Botvinnik" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0003-4412-7970" />
  <meta name="twitter:creator" content="@olgabot" />
  <meta name="citation_author" content="Venkata Naga Pranathi Vemuri" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0002-5748-9594" />
  <meta name="twitter:creator" content="@pranuvemuri" />
  <meta name="citation_author" content="N. Tessa Pierce" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0002-2942-5331" />
  <meta name="twitter:creator" content="@saltyscientist" />
  <meta name="citation_author" content="Phoenix Aja Logan" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0003-4581-0552" />
  <meta name="twitter:creator" content="@phoenixlogan" />
  <meta name="citation_author" content="Saba Nafees" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_institution" content="Department of Biological Sciences, Texas Tech University" />
  <meta name="citation_author_institution" content="Department of Mathematics &amp; Statistics, Texas Tech University" />
  <meta name="citation_author_orcid" content="0000-0002-3292-7703" />
  <meta name="twitter:creator" content="@sabanafeesTTU" />
  <meta name="citation_author" content="C. Titus Brown" />
  <meta name="citation_author_institution" content="Department of Population Health and Reproduction, University of California, Davis" />
  <meta name="citation_author_orcid" content="0000-0001-6001-2677" />
  <meta name="twitter:creator" content="@ctitusbrown" />
  <meta name="citation_author" content="Jim Karkanias" />
  <meta name="citation_author_institution" content="Data Sciences Platform, Chan Zuckerberg Biohub" />
  <meta name="citation_author_orcid" content="0000-0002-8057-6055" />
  <meta name="twitter:creator" content="@jkarkanias" />
  <link rel="canonical" href="https://czbiohub.github.io/kmer-homology-paper/" />
  <meta property="og:url" content="https://czbiohub.github.io/kmer-homology-paper/" />
  <meta property="twitter:url" content="https://czbiohub.github.io/kmer-homology-paper/" />
  <meta name="citation_fulltext_html_url" content="https://czbiohub.github.io/kmer-homology-paper/" />
  <meta name="citation_pdf_url" content="https://czbiohub.github.io/kmer-homology-paper/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://czbiohub.github.io/kmer-homology-paper/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://czbiohub.github.io/kmer-homology-paper/v/629460e0379746b5d38ee3428bcb0e3af169abe3/" />
  <meta name="manubot_html_url_versioned" content="https://czbiohub.github.io/kmer-homology-paper/v/629460e0379746b5d38ee3428bcb0e3af169abe3/" />
  <meta name="manubot_pdf_url_versioned" content="https://czbiohub.github.io/kmer-homology-paper/v/629460e0379746b5d38ee3428bcb0e3af169abe3/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
  
  <!-- pandoc-eqnos: equation style -->
  <style>
    .eqnos { display: inline-block; position: relative; width: 100%; }
    .eqnos br { display: none; }
    .eqnos-number { position: absolute; right: 0em; top: 50%; line-height: 0; }
  </style>
</head>
<body>
<header id="title-block-header">
<h1 class="title"><span class="math inline">\(k\)</span>-mer homology is a transcriptome-first method enabling functional prediction of transcriptomic ‘dark matter’ across species</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://czbiohub.github.io/kmer-homology-paper/v/629460e0379746b5d38ee3428bcb0e3af169abe3/">permalink</a>)
was automatically generated
from <a href="https://github.com/czbiohub/kmer-homology-paper/tree/629460e0379746b5d38ee3428bcb0e3af169abe3">czbiohub/kmer-homology-paper@629460e</a>
on December 7, 2020.
</em></small></p>
<p><span class="banner lightred"><span class="fas fa-info-circle fa-lg"></span> <strong>This is an in progress manuscript.</strong></span></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Olga Borisovna Botvinnik</strong> <sup><a href="#correspondence">✉</a></sup><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4412-7970">0000-0003-4412-7970</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/olgabot">olgabot</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/olgabot">olgabot</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
<li><p><strong>Venkata Naga Pranathi Vemuri</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-5748-9594">0000-0002-5748-9594</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/pranathivemuri">pranathivemuri</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/pranuvemuri">pranuvemuri</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
<li><p><strong>N. Tessa Pierce</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2942-5331">0000-0002-2942-5331</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/bluegenes">bluegenes</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/saltyscientist">saltyscientist</a><br>
<small>
Department of Population Health and Reproduction, University of California, Davis
· Funded by NSF 1711984
</small></p></li>
<li><p><strong>Phoenix Aja Logan</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-4581-0552">0000-0003-4581-0552</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/phoenixAja">phoenixAja</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/phoenixlogan">phoenixlogan</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
<li><p><strong>Saba Nafees</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-3292-7703">0000-0002-3292-7703</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/snafees">snafees</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/sabanafeesTTU">sabanafeesTTU</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub; Department of Biological Sciences, Texas Tech University; Department of Mathematics &amp; Statistics, Texas Tech University
</small></p></li>
<li><p><strong>C. Titus Brown</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6001-2677">0000-0001-6001-2677</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/ctb">ctb</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/ctitusbrown">ctitusbrown</a><br>
<small>
Department of Population Health and Reproduction, University of California, Davis
· Funded by Moore Foundation GBMF4551
</small></p></li>
<li><p><strong>Jim Karkanias</strong> <br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8057-6055">0000-0002-8057-6055</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/jkensai">jkensai</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/jkarkanias">jkarkanias</a><br>
<small>
Data Sciences Platform, Chan Zuckerberg Biohub
</small></p></li>
</ul>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p>Animals spread pathogens to humans in zoonotic diseases, and many of these animals are asymptomatic, and we want to leverage single-cell transcriptomics to identify cell types enabling these organisms to survive without symptoms, and apply them to human therapeutics. However, 99.9% of animal species on the Earth do not have a reference genome, including many pathogen host animals.
Thus, there is an unmet need for transcriptome-first methods to interrogate single cell types that do not require reference genomes or gene annotations.
To this end, we introduce the concept of “<span class="math inline">\(k\)</span>-mer homology,” a method for lightweight detection of orthologous genes directly from transcriptomes without need for a reference genome. Underpinning <span class="math inline">\(k\)</span>-mer homology is the tool <code>leaftea</code>, a novel computational method for extracting high-confidence protein-coding sequences from RNA-seq data into non-coding reads and protein-coding reads which can then be manipulated to reduced amino acid alphabets which are directly comparable across large evolutionary distances.
As the direct assignment of protein-coding sequence skips both traditional alignment and gene orthology assignment it can a) be applied to transcriptomes from organisms with no or poorly annotated genomes, and b) identify putative functions of protein sequences contributing to shared cell types.
We applied these methods to the problem of understanding the unique ability of bats to harbor zoonotic diseases, especially coronaviruses such as SARS, MERS, SARS-CoV-2, an issue of fundamental importance to human health.
However, important immunological genes such as interferon pathway genes have not been adequately identified in the bat genome and thus that aspect of bat biology could not be identified.
With our methods, we were able to detect widespread expression of — in immune tissues of the SARS reservoir species, and the purported SARS-CoV-2 species <em>Rhinolophus sinicus</em> (Chinese horseshoe bat) single-cell RNA-seq transcriptomes, as compared to human tissues.
The <em>R. sinicus</em> interferon genes were slightly different from human interferon genes […]
By enabling analyses across divergent species’ transcriptomes in a genome-, gene annotation-, and orthology-agnostic manner, <code>leaftea</code> illustrates the potential of non-model organisms in building the cell type evolutionary tree of life.</p>
<h2 class="page_break_before" id="introduction">Introduction</h2>
<!-- "---" indicates a number or result that is to be filled out -->
<h3 id="key-sentence-to-identify-cell-types-of-interest-in-host-organisms-single-cell-rna-seq-is-a-powerful-tool-but-it-is-not-readily-accessible-to-organisms-without-well-annotated-reference-genomes.">Key sentence: To identify cell types of interest in host organisms, single-cell RNA-seq is a powerful tool, but it is not readily accessible to organisms without well-annotated reference genomes.</h3>
<p>The current COVID-19 pandemic has exposed inabilities of understanding zoonotic diseases in wild animals such as bats, due to poorly annotated genomes.
However, the asymptomatic nature of infection in zoonotic host organisms and how it could be applied to humans is of immediate concern.
<!-- non-human host organisms such as the bat *Rhinolophus sinicus*. -->
Zoonoses such as SARS-CoV-1 and SARS-CoV-2 are often harbored in wild animals such as bats, that do not have well annotated genomes, but the asymptomatic nature of infection in those organisms and how it could be applied to humans is of immediate concern, due to the recent SARS, MERS, and Ebola outbreaks, and the current COVID-19 pandemic.
Identifying cell types in these host organisms enables studying the broad array of cell types available to living creatures, and can help understand human disease by seeing cell types that are “normal” in one animal, but resemble human disease cell types.
To identify cell types of interest in host organisms, single-cell RNA-seq is a powerful tool, but it is not readily accessible to organisms without well-annotated reference genomes, as 99.999% of the approximate 10 million animal species predicted to exist on Earth have no submitted genome assembly <span class="citation" data-cites="OkFljXKC xJN8WB5G">[<a href="#ref-OkFljXKC" role="doc-biblioref">1</a>,<a href="#ref-xJN8WB5G" role="doc-biblioref">2</a>]</span>.
Current methods for identifying cell types use single-cell RNA-sequencing, however, require three things: (1) high-quality genome assembly, (2) high-quality predicted gene annotations and (3) high-quality orthologous gene mappings across species.
But, we can nonetheless generate single-cell RNA-seq data from these species with poorly annotated genomes, and if we can leverage the annotations from existing organisms, we can better understand why host animals are able to tolerate infection by a pathogen that is deadly in humans.</p>
<h3 id="key-the-most-common-strategies-for-cross-species-transcriptomics-lose-data-due-to-unaligned-sequences-unannotated-genes-and-unmatched-genes-across-species.">key: The most common strategies for cross-species transcriptomics lose data due to unaligned sequences, unannotated genes, and unmatched genes across species.</h3>
<p>There are many strategies for comparative transcriptomics for single cells <span class="citation" data-cites="nI2xGuLe">[<a href="#ref-nI2xGuLe" role="doc-biblioref">3</a>]</span>.
The most commonly used strategies for cross-species gene expression analyses lose data due to unaligned sequences, unannotated genes, and unmatched genes across species.</p>
<p>Unaligned sequences are largely due to errors in genome assembly, or
While high-quality end-to-end genome assemblies are now within reach with the recently published teleomere-to-telomere human genome assembly <span class="citation" data-cites="1C9RVEJCr">[<a href="#ref-1C9RVEJCr" role="doc-biblioref">4</a>]</span>, and high-quality assemblies, including gene annotations, will soon be a possibility thanks to thhe efforts of the Vertebrate Genome Project <span class="citation" data-cites="N285yCi2 1BMDdrblS">[<a href="#ref-N285yCi2" role="doc-biblioref">5</a>,<a href="#ref-1BMDdrblS" role="doc-biblioref">6</a>]</span>.
While the genome assembly provides substantial informationa bout the genomic structure of a species, it does not yet have the matching genes across species.</p>
<h3 id="key-matching-genes-across-species-is-an-open-and-unsolved-problem.">key: Matching genes across species is an open and unsolved problem.</h3>
<p>One of the first fundamental step is creating a common gene expression matrix to compare the cell types.
Matching genes across species is perfolrmed via a process called orthologous gene mapping.
A group that is working on this is the Quest for Orthologs Consortium, which has established benchmarking <span class="citation" data-cites="LkrS4xXB dX45ePUc 19AeqN3SK 1AbwFZgjU">[<a href="#ref-LkrS4xXB" role="doc-biblioref">7</a>,<a href="#ref-dX45ePUc" role="doc-biblioref">8</a>,<a href="#ref-19AeqN3SK" role="doc-biblioref">9</a>,<a href="#ref-1AbwFZgjU" role="doc-biblioref">10</a>]</span> services.
Another option besides using 1:1 orthologs is using orthogroups <span class="citation" data-cites="EvDQzoLf">[<a href="#ref-EvDQzoLf" role="doc-biblioref">11</a>]</span>, aligning genes/reads to a common genome if the species are closely related, such as for primates <span class="citation" data-cites="mtOZyMZw">[<a href="#ref-mtOZyMZw" role="doc-biblioref">12</a>]</span>.</p>
<p>and even fewer have gene predictions and orthologous gene mappings to human, which remains an open problem <span class="citation" data-cites="ogAGO9KH wdzgXUGy">[<a href="#ref-ogAGO9KH" role="doc-biblioref">13</a>,<a href="#ref-wdzgXUGy" role="doc-biblioref">14</a>]</span>.
The most commonly used strategy is subsetting the matrix down to 1:1 orthologous genes, which already have been computed by a database such as ENSEMBL <span class="citation" data-cites="oDiClyTE oDiClyTE">[<a href="#ref-oDiClyTE" role="doc-biblioref">15</a>,<a href="#ref-oDiClyTE" role="doc-biblioref">15</a>]</span>.
Comparative transcriptomics traditionally relies on orthologous gene mapping, which requires a genome.</p>
<p>Another strategy for more distantly related species is to only use transcription factor genes that are part of “core regulatory complex” (CoRC), which again subsets the data to only certain genes.
Not just at alignment, but at aligning to annotated genes, at subsetting to 1:1 orthologs – these are all data lossy steps.
While the first two steps are dependent on the genome assembly and gene annotation quailty, the subsetting to 1:1 orthologs cannot be solved with improved assembly and gene annotation tools.
It can be solved with using other strategies for finding orthologous genes.
Another strategy is aligning reads to the native species’ genome, and using a tool such as LiftOver to remap coordinates into a well annotated genome such as human, and use those genes.</p>
<p>For all of these tools, the problem with them is that they lose data at every step.
These tools are not available to the 99.9% of vertebrates that do not have a reference genome, but transcriptome work, single-cell RNA-seq data is now so easy to obtain, that people already want to do analyses with their single cell data even if their organism doesn’ thave a great reference genome.
And that’s what we are enabling.</p>
<p>Once the genomes of the species are obtained, the work is not over.
For each pariwise species comparison, the orthologous genes must be found.
There are many tools for this, such as OrthoFinder <span class="citation" data-cites="EvDQzoLf">[<a href="#ref-EvDQzoLf" role="doc-biblioref">11</a>]</span>, eggNOG <span class="citation" data-cites="1CZYa2MlR eRqRW5iR">[<a href="#ref-1CZYa2MlR" role="doc-biblioref">16</a>,<a href="#ref-eRqRW5iR" role="doc-biblioref">17</a>]</span>, and highly curated databases such as ENSEMBL [citation].
The time to run these tools is a long time.
But that’s not really the point… what we really care about is using the RNA-seq data directly and not losing data due to mismatched orthology, unannotated genes, or unaligned seuqences.</p>
<h3 id="key-a-method-of-not-losing-data-due-to-unaligned-sequences-or-unannotated-genes-is-using-all-the-data-in-the-form-of-subsampled-k-mers.">key: A method of not losing data due to unaligned sequences or unannotated genes is using all the data in the form of subsampled <span class="math inline">\(k\)</span>-mers.</h3>
<p><span class="math inline">\(k\)</span>-mers from sequencing data for single cells have been used to show that they can identify single cell types as well as gene counts <span class="citation" data-cites="wtk13QGK">[<a href="#ref-wtk13QGK" role="doc-biblioref">18</a>]</span>.
Recently, there has been literature to use subsampled <span class="math inline">\(k\)</span>-mers to identify metagenomes within metagenomic data, but this uses subsampling of the data <span class="citation" data-cites="mH9pzoIn ExvjTLL3 ART4gJKs">[<a href="#ref-mH9pzoIn" role="doc-biblioref">19</a>,<a href="#ref-ExvjTLL3" role="doc-biblioref">20</a>,<a href="#ref-ART4gJKs" role="doc-biblioref">21</a>]</span>.
It turns out that <span class="math inline">\(k\)</span>-mers are quite redundant and it is not necessary to use all of them for genome identification.
Specifically, it is possible to use a fractional subsampling of the original data to represent the data faithfully as if all <span class="math inline">\(k\)</span>-mers were stored, as implemented in the Sourmash software <span class="citation" data-cites="ExvjTLL3 ART4gJKs">[<a href="#ref-ExvjTLL3" role="doc-biblioref">20</a>,<a href="#ref-ART4gJKs" role="doc-biblioref">21</a>]</span>.
<span class="math inline">\(k\)</span>-mers or k-long substrings have a long history in bioinformatics <span class="citation" data-cites="QvbXRDjv">[<a href="#ref-QvbXRDjv" role="doc-biblioref">22</a>]</span> and have been widely used for genome and transcriptome assembly problems. They were recently applied to metagenomics for taxonomic classification.</p>
<p>We show that we use a subsampled set of the <span class="math inline">\(k\)</span>-mers to still identify cell types across and within species.</p>
<p>However, for applying to cell types across species, we add the layer of using reduced amino acid alphabet <span class="math inline">\(k\)</span>-mers.</p>
<p>However, for applying to cell types across species, we add the layer of using protein <span class="math inline">\(k\)</span>-mers.</p>
<h3 id="key-protein-sequence-can-be-inferred-from-rna-seq-due-to-nullomers.">key: Protein sequence can be inferred from RNA-seq due to nullomers.</h3>
<p>Nullomers are sequences that are not present in the genome annotation or protein sequences <span class="citation" data-cites="g2gs9fUL IpiEdhou hMQyspqZ 9kcdMNpl">[<a href="#ref-g2gs9fUL" role="doc-biblioref">23</a>,<a href="#ref-IpiEdhou" role="doc-biblioref">24</a>,<a href="#ref-hMQyspqZ" role="doc-biblioref">25</a>,<a href="#ref-9kcdMNpl" role="doc-biblioref">26</a>]</span> of a species.
While this seems like there are a contradiction between wanting to use all the data, and then only using the data that is present in the database annotation, we show we are able to predict protein coding reads using simulated data.
The signal of a non-protein coding sequence is very obvious – the true coding frames have <span class="math inline">\(k\)</span>-mers in common with the known database, and the non-true coding frames do not have <span class="math inline">\(k\)</span>-mers in common.
There are only so many <span class="math inline">\(k\)</span>-mers that are compatible with eukaryotic, or mammalian life.
There are amino acid combinations that are not seen in the natural world.
Whether nullomers are not present due to natural selection or due to CpG hypermutability <span class="citation" data-cites="17wsdE0b1">[<a href="#ref-17wsdE0b1" role="doc-biblioref">27</a>]</span> remains under debate, there are this phenomena is real and can be exploited to identify putative protein-coding sequences.
By using the “negative space” of unobserved protein sequences, we can identify reading frames of RNA to protein-coding sequences that are unlikely.
Non-productive Reading frames of RNA-seq can be found in the nullomers.
Protein sequences that have been observed are inferred in the <span class="math inline">\(k\)</span>-mers.</p>
<h3 id="key-gene-annotations-can-be-inferred-from-rna-sequencing-data-which-is-easily-obtained-with-commodity-tools">key: Gene annotations can be inferred from RNA-sequencing data, which is easily obtained with commodity tools</h3>
<p>What we can do with <span class="math inline">\(k\)</span>-mers is use a peripherally related genome to align to, and then extract the protein-coding reads and <span class="math inline">\(k\)</span>-mers that are present, then directly compare them.</p>
<p>Transcribed seqeunces in the form of RNA-seq data can be inferred to be transcribed genes for now.</p>
<h3 id="key-reduced-amino-acid-alphabets-provide-a-lightweight-method-of-identifying-matching-genes-across-species.">key: Reduced amino acid alphabets provide a lightweight method of identifying matching genes across species.</h3>
<p>Not only do we use protein <span class="math inline">\(k\)</span>-mers, but we use reduced amino acid alphabet <span class="math inline">\(k\)</span>-mers.</p>
<p>Amino acids can be categorized into their biochemical properites.
Many times,a protein can retain its same phsyical structure and function even with amino acid changes.
These amino acid changes may be retaining the phyiochemical properites of the amino acid.
The space of amino acid substituion was explored back in the 90s to find similarity searches for proteins.
The BLOSUM62, PAM substituion matrices were a method of assigning higher penalties for amino acids that changed their biochemical type.
Reduced amino acid alphabets are a method of reducing blame in changing of amino acid sequences, which is especially compatible with <span class="math inline">\(k\)</span>-mers as if a single letter in a <span class="math inline">\(k\)</span>-mer changes, then the whole <span class="math inline">\(k\)</span>-mer no longer matches.</p>
<p>A method of quickly identifying potential orthologs is using reduced amino acid alphabets <span class="citation" data-cites="Rqlso7kk">[<a href="#ref-Rqlso7kk" role="doc-biblioref">28</a>]</span>.
Reduced amino acid alphabets have been used for database searches of related protein sequences <span class="citation" data-cites="sZoryWIB">[<a href="#ref-sZoryWIB" role="doc-biblioref">29</a>]</span> and for protein structure prediction
<span class="citation" data-cites="Aq7zYbOg 16Kv6RIe3 sZoryWIB f5u3dDtc Aq7zYbOg QMYjnr6l">[<a href="#ref-sZoryWIB" role="doc-biblioref">29</a>,<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>,<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>,<a href="#ref-16Kv6RIe3" role="doc-biblioref">31</a>,<a href="#ref-f5u3dDtc" role="doc-biblioref">32</a>,<a href="#ref-QMYjnr6l" role="doc-biblioref">33</a>]</span></p>
<p>We combine the two concepts of <span class="math inline">\(k\)</span>-mer subsampling and homology identification via reduced amino acid alphabets into what we term “<span class="math inline">\(k\)</span>-mer homology.”</p>
<h3 id="key-there-is-an-unmet-need-to-quantitatively-compare-single-cell-transcriptomes-across-species-without-the-need-for-orthologous-gene-mapping-gene-annotations-or-a-genome-assembly.">key: There is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for orthologous gene mapping, gene annotations, or a genome assembly.</h3>
<p>We translate our protein sequences to protein.
We need to predict protein seuqnece, and turns out most reading frames of RNA don’t need to be predicted – they are already near zero.
They have stop codons.</p>
<p>Thus, there is an unmet need to quantitatively compare single-cell transcriptomes across species, without the need for orthologous gene mapping, gene annotations, or a reference genome.
Short, <span class="math inline">\(k\)</span>-long sequence substrings, or <span class="math inline">\(k\)</span>-mers, have been proposed for clustering single cells <span class="citation" data-cites="wtk13QGK">[<a href="#ref-wtk13QGK" role="doc-biblioref">18</a>]</span> and here we implemented <span class="math inline">\(k\)</span>-mers from putatitvely translated RNA-seq reads with reduced amino acid alphabets <span class="citation" data-cites="Aq7zYbOg 16Kv6RIe3 sZoryWIB f5u3dDtc Aq7zYbOg QMYjnr6l">[<a href="#ref-sZoryWIB" role="doc-biblioref">29</a>,<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>,<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>,<a href="#ref-16Kv6RIe3" role="doc-biblioref">31</a>,<a href="#ref-f5u3dDtc" role="doc-biblioref">32</a>,<a href="#ref-QMYjnr6l" role="doc-biblioref">33</a>]</span>, to find shared cell types across species, and further identify <em>de novo</em> orthologous genes by querying the predicted protein sequences to a reference database.
This method relies solely on divergence time between species, which we show can be estimated from RNA-seq nucleotide <span class="math inline">\(k\)</span>-mers (Supplemental Figure <span class="citation" data-cites="sfig:sfig1">[<span class="citeproc-not-found" data-reference-id="sfig:sfig1"><strong>???</strong></span>]</span>).
As the direct assignment of protein-coding sequence skips both traditional alignment and gene orthology assignment it can a) be applied to transcriptomes from organisms with no or poorly annotated genomes, and b) identify putative functions of protein sequences contributing to shared cell types.</p>
<p>We benchmark the prediction of protein-coding sequence using simulated human RNA-seq data from single-copy genes present across all clades of mammals <span class="citation" data-cites="S53q1T30">[<a href="#ref-S53q1T30" role="doc-biblioref">34</a>]</span>.
We show that one can extract putative protein-coding reads from — mammalian species in BUSCO, and present the best parameters for different divergence times.</p>
<!-- We first apply this method on a bulk comparative transcriptomic dataset consisting of nine amniote species and six tissues [@doi:10.1038/nature10532], showing that we achieve similar clustering results as using only reads mapping to 1:1 orthologs or Hierarchical Orthologous Groups (HOGs) [@doi:10.1371/journal.pone.0053786; @doi:10.1093/bib/bbr034; @doi:10.1093/bioinformatics/btu492] of protein-coding genes, but are able to resolve ... which can only be seen by using the $k$-mer method. -->
<p>We first demonstrate the utility of this method by comparing lung tissue transcriptomes from organisms diverged by approximately 100 million years ago (MYA) <span class="citation" data-cites="1A6CqUMhm">[<a href="#ref-1A6CqUMhm" role="doc-biblioref">35</a>]</span>: mouse from <em>Tabula Muris Senis</em> <span class="citation" data-cites="rSyxV4O8">[<a href="#ref-rSyxV4O8" role="doc-biblioref">36</a>]</span>,
Across this evolutionary distance, only — 1:1 orthologous genes exist as found by …
<!-- and XX HOGs via orthologous matrix (OMA) [@doi:10.7717/peerj.6231; @doi:10.1093/bioinformatics/btx229] -->
We then add single-cell lung data from <em>Rhinolophus sinicus</em>, Chinese greater horseshoe bat, whose single-cell transcriptome was recently published <span class="citation" data-cites="1AkeHrvKP">[<a href="#ref-1AkeHrvKP" role="doc-biblioref">37</a>]</span>..
Using the <span class="math inline">\(k\)</span>-mer homology method, we are able to distinguish the subtypes of T cells in the <em>R. sinicus</em> which were hidden using read counts alone, due to <span class="math inline">\(k\)</span>-mers from genes not present in the genome assembly.
We then show the ability to find cell type enriched sequences that are absent from the reference genome, in particular interferon pathway genes and natural killer receptor genes.</p>
<h3 id="key-k-mer-homology-is-a-transcriptome-first-method-for-cross-species-analyses-implementing-via-k-mers-and-reduced-amino-acid-alphabets">key: <span class="math inline">\(k\)</span>-mer homology is a transcriptome-first method for cross-species analyses implementing via <span class="math inline">\(k\)</span>-mers and reduced amino acid alphabets</h3>
<p>Thus, we have shown the reference-free method using the <span class="math inline">\(k\)</span>-mers from single cells is a novel, annotation-agnostic method for comparing cells across species that is capable of identifying cell states unique to a particular organism, helping to build the cell type evolutionary tree of life.</p>
<p>Similar to <span class="math inline">\(k\)</span>-mer based approaches for transcript quantification <span class="citation" data-cites="RIPzCufe 12KZMHMQl vrqQcFyx">[<a href="#ref-RIPzCufe" role="doc-biblioref">38</a>,<a href="#ref-12KZMHMQl" role="doc-biblioref">39</a>,<a href="#ref-vrqQcFyx" role="doc-biblioref">40</a>]</span>, we implemented <span class="math inline">\(k\)</span>-mer based gene expression analyses across species, but instead of using DNA <span class="math inline">\(k\)</span>-mers, our critical innovation was using translated protein <span class="math inline">\(k\)</span>-mers.
We utilized sequence bloom trees (SBTs) <span class="citation" data-cites="DfZvKOpG">[<a href="#ref-DfZvKOpG" role="doc-biblioref">41</a>]</span> using a bottom-up approach to build them similar to previous work <span class="citation" data-cites="1HpdDzfQX AfFgGSHX">[<a href="#ref-1HpdDzfQX" role="doc-biblioref">42</a>,<a href="#ref-AfFgGSHX" role="doc-biblioref">43</a>]</span> to ensure localization of new datasets, meaning, if two leaves share a parent, they are guaranteed to be more similar than two leaves that do not share a parent.</p>
<p>For widespread accessibility and usage, we implemented <code>leaftea</code> into two distinct Nextflow pipelines following software best practices such as testing and continuous integration: (1) <code>nf-core/kmermaid</code> to compare translated transcriptomes across divergent species, and (2) <code>czbiohub/nf-predictorthologs</code> to infer functions of translated sequences.</p>
<div id="fig:fig1" class="fignos">
<figure>
<img src="images/SVG/figure1.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 1:</span> <strong>A.</strong> Overview of single-cell RNA-seq lung tissue datasets from three species: <em>Homo sapiens</em> (human), <em>Rhinolophus sinicus</em> (Chinese horseshoe bat), and <em>Mus musculus</em>.
<strong>B.</strong> Overview of <span class="math inline">\(k\)</span>-mer homology method. First, RNA-seq reads are translated to protein via the leaftea tool, then a cell type database is built from <span class="math inline">\(k\)</span>-mers from the putative protein-coding sequence, given a query cell of a new species, the best matching cell from the model organism database is found, and the cell label is returned.
<!-- **B.** Overview of computational method. First, in the `kmermaid` pipeline, RNA-seq reads are translated to protein via the leaftea tool, then converted to a reduced amino acid alphabet, decomposed into $k$-mers, and then those $k$-mers are subsampled. Next, the `nf-predictorthologs` pipeline performs differential $k$-mer expression, much like differential gene expression, to find $k$-mers that are enriched in a population, and queries for those differential $k$-mers in databases to identify them. --></figcaption>
</figure>
</div>
<h2 id="results">Results</h2>
<h3 id="reliable-identification-of-putative-protein-coding-sequence-with-leatea-translate-across-150-million-years-of-evolution">Reliable identification of putative protein-coding sequence with <code>leatea translate</code> across ~150 million years of evolution</h3>
<div id="tbl:dayhoff-alphabet" class="tablenos">
<table>
<caption><span>Table 1:</span> Dayhoff encoding of amino acids by categories of biochemical properties, into a six-letter alphabet. Adapted from <span class="citation" data-cites="UCkAlFcx">[<a href="#ref-UCkAlFcx" role="doc-biblioref">44</a>]</span>. For example, the protein sequence <code>LIVING</code> would become <code>eeeecb</code>. </caption>
<thead>
<tr class="header">
<th style="text-align: left;">Amino acid</th>
<th style="text-align: left;">Property</th>
<th style="text-align: left;">Dayhoff</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>C</code></td>
<td style="text-align: left;">Sulfur polymerization</td>
<td style="text-align: left;"><code>a</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>A</code>, <code>G</code>, <code>P</code>, <code>S</code>, <code>T</code></td>
<td style="text-align: left;">Small</td>
<td style="text-align: left;"><code>b</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>D</code>, <code>E</code>, <code>N</code>, <code>Q</code></td>
<td style="text-align: left;">Acid and amide</td>
<td style="text-align: left;"><code>c</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>H</code>, <code>K</code>, <code>R</code></td>
<td style="text-align: left;">Basic</td>
<td style="text-align: left;"><code>d</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>I</code>, <code>L</code>, <code>M</code>, <code>V</code></td>
<td style="text-align: left;">Hydrophobic</td>
<td style="text-align: left;"><code>e</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>F</code>, <code>W</code>, <code>Y</code></td>
<td style="text-align: left;">Aromatic</td>
<td style="text-align: left;"><code>f</code></td>
</tr>
</tbody>
</table>
</div>
<div id="fig:fig2" class="fignos">
<figure>
<img src="images/SVG/figure2.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 2:</span> Application of <span class="math inline">\(k\)</span>-mer homology to mammalian lung cell type identification recapitulates known biology and identifies enriched genes not present in reference genome assembly.
<strong>A.</strong> Overview of <code>leaftea translate</code> method. First, each read is translated into all six possible protein-coding translation frames. Next, reading frames with stop codons are eliminated. Each protein-coding frame is <span class="math inline">\(k\)</span>-merized, then the fraction of <span class="math inline">\(k\)</span>-mers which appear in the known protein-coding database is computed. Reading frames containing 5% or more <span class="math inline">\(k\)</span>-mers matching the reference proteome are inferred to be putatively protein-coding.
<strong>B.</strong> Experimental set up to identify best <span class="math inline">\(k\)</span>-mer sizes for translating protein-coding sequence, given a reference proteome of divergent organism. First, human RNA-seq reads were simulated from genes present in all extant mammals. Next, we used highly curated reference proteomes from BUSCO Mammalia, representing almost 200 million years of evolution, across a wide range of mammalian species. Using <code>leaftea</code>, we partition the reads into protein-coding and noncoding using a variety of amino acid alphabets and <span class="math inline">\(k\)</span>-mer sizes, and then compute classification metrics.
<strong>C.</strong> Predictive power of detecting protein-coding sequence across ~200 million years of evolution, using simulated human RNA-seq reads and mammalian reference proteomes. x-axis, <span class="math inline">\(k\)</span>-mer size used to translate the RNA sequence; y-axis, receiver operating characteristic area under the curve (ROC AUC) of protein-coding prediction at that <span class="math inline">\(k\)</span>-mer size. Dotted vertical lines indicate <span class="math inline">\(k\)</span>-mer size producing maximum ROC AUC for each divergence time.
<strong>D.</strong> Classification metrics of protein-coding prediction with both protein and Dayhoff alphabets, using best-performing <span class="math inline">\(k\)</span>-mer sizes for each alphabet from (<strong>C</strong>). Left, precision; middle, recall; right, <span class="math inline">\(F_{1/2}\)</span> score. x-axis, Estimated divergence time of reference proteome relative to human, from timetree.org; y-axis, value of classification metric.</figcaption>
</figure>
</div>
<p>To extract putative protein coding sequences from RNA-seq reads, we first established whether we could predict protein-coding sequences from simulated human RNA-seq data.
We reliably identified putative protein-coding reads using our tool, <code>leaftea</code>.
<code>leaftea</code> works by performing six frame translation of RNAs, discarding reading frames with stop codons, and decomposing each frame into amino acid <span class="math inline">\(k\)</span>-mers, and checking for the <span class="math inline">\(k\)</span>-mer present in the reference proteome (Fig. <a href="#fig:fig1">1</a>A).
In detecting putative protein coding sequence, there many be small amino acid changes that do not change the overall function of the protein, and to accommodate this, we offer the option to translate using either the original 20-letter amino acid alphabet, or a six-letter reduced amino acid alphabet, the Dayhoff encoding [<a href="#tbl:dayhoff-alphabet">1</a>].
As many organisms of interest are not annotated, but a different organism with a related ancestor is, we wanted to understand the limits of protein-coding detection when the best-annotated organism in the database is divergent with up to approximately 200 million years.
To test this, we simulated human RNA-seq data from genes present in all extant mammals as determined by BUSCO <span class="citation" data-cites="1VJMkE5Z">[<a href="#ref-1VJMkE5Z" role="doc-biblioref">45</a>]</span>, and used reference proteomes from eighteen mammals, ranging in divergence from 0 to almost 200 MYA <span class="citation" data-cites="W97RNFXd">[<a href="#ref-W97RNFXd" role="doc-biblioref">46</a>]</span> (Fig. <a href="#fig:fig1">1</a>B).
We found that we were able maximize classification power, measured via receiver operating characteristic area under the curve (ROC AUC), for most divergence times for a protein alphabet with a <span class="math inline">\(k\)</span>-mer size of 8, and a <span class="math inline">\(k\)</span>-mer size of 17 for a Dayhoff encoding (Fig. <a href="#fig:fig1">1</a>C, left and middle).
Across all mammals tested, it was clear that some were better annotated than others based on the ROC AUC curves (Supplemental Fig. ??).
For example, hedgehog (<em>Erinaceus europaeus</em>) had more successful protein-prediction than species with similar divergence times from human such as rhino (<em>Ceratotherium simum simum</em>), even though both are members of Boroeutheria, a mangnorder of mammals with a most recent common ancestor at an estimated 96 MYA of one another <span class="citation" data-cites="W97RNFXd">[<a href="#ref-W97RNFXd" role="doc-biblioref">46</a>]</span>, containing the superorders Euarchontoglires (includes rodents and primates) and Laurasiatheria (includes bat, hedgehog, hooved animals, cats, and dogs).
We found that a protein alphabet with a <span class="math inline">\(k\)</span>-mer size of 8 was most productive in translating the sequences for closely related species, within Boroeutheria, with a mean ROC AUC of 0.873898 among well-annotated reference proteomes.
For more distantly related species such as platypus (<em>Ornithorhynchus anatinus</em>), a the Dayhoff alphabet had a ROC AUC of 0.547203.
<!-- , however, the precision of rejecting non-coding reads remained quite high across distantly related species (Fig. {@fig:fig1}D, left). -->
Interested in understanding whether <code>leaftea</code>’s strength lies in detecting true protein-coding reads, or eliminating non-coding reads, we computed the precision and recall of protein-coding classification (Fig. ??, left and middle).
Our precision to remove non-coding reads is quite high, especially with the Dayhoff alphabet, which retains a high precision even for the platypus proteome, which is approximately 177 MYA diverged from human.
As this metric performs better in precision, we used an <span class="math inline">\(F_{\beta}\)</span> score, where <span class="math inline">\(\beta = 1/2\)</span>, to indicate a 2x preference for precision over recall (Fig. <a href="#fig:fig1">1</a>D, right).
<!-- Interested in understanding whether `leaftea`'s strength lies in detecting true protein-coding reads, or eliminating non-coding reads, we computed the precision and recall of protein-coding classification (Fig. {@fig:fig1D}).
We found that
Using the reference proteomes from different animals represents extracting protein-coding reads from an organisms whose closely related species in the reference database is 10, 50, etc million years diverged.
Using the reference proteomes, we found that a small fraction of overlapping $k$-mers in the putative protein-coding read were required to identify the read as putatively protein-coding. Our sensitivity to remove non-coding reads is quite high.
We chose a minimum containment threshold of 0.05 as this maximized the number of putative protein coding reads while successfully discarding non-coding reads. -->
This shows that while it is possible to detect putative protein-coding sequence, reliably rejecting non-coding sequences, even when the best annotated organism in the database almost 200 million years diverged from the query organism.</p>
<!-- Putative protein-coding reads from simulated Human RNA-seq data can be identified across the mammalian lineage -->
<h3 id="degenerate-amino-acid-alphabet-k-mers-identify-common-cell-types-across-species-and-finds-genes-absent-from-even-well-annotated-genomes">Degenerate amino acid alphabet <span class="math inline">\(k\)</span>-mers identify common cell types across species and finds genes absent from even well-annotated genomes</h3>
<div id="fig:fig3" class="fignos">
<figure>
<img src="images/SVG/figure3.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 3:</span> <strong>A.</strong> Experimental set up to validate effectiveness of building a cell type database from one species, to label cells from another species. First, we build a cell type database using sequence bloom trees (SBTs) from mouse cells, then search using a human cell for the best matching mouse cell.
<strong>B.</strong> Mean of classification metrics across mouse:human cell type lookup experiments, highlighting the <span class="math inline">\(k\)</span>-mer size that maximizes the mean for each metric and alphabet. Vertical lines, standard deviation of the mean. x-axis, <span class="math inline">\(k\)</span>-mer size in DNA space, i.e. a <span class="math inline">\(k\)</span>-mer size of 21 corresponds to a an amino acid length of 7. y-axis, value of classification metric. Left, adjusted rand score; right, F1 score.
<strong>C.</strong> Molecules and <span class="math inline">\(k\)</span>-mer sizes maximizing classification metrics from (<strong>E</strong>).
<strong>D.</strong> Confusion matrices of classification using Dayhoff encoding and <span class="math inline">\(k\)</span>-mer size of —, the best performing metrics from (<strong>E</strong>). x-axis, predicted cell label; y-axis, ground truth cell label.</figcaption>
</figure>
</div>
<h4 id="key-dayhoff-encoded-k-mers-reliably-predict-cell-types-within-species">key: Dayhoff-encoded <span class="math inline">\(k\)</span>-mers reliably predict cell types within species</h4>
<p>To test identification of cell types using <span class="math inline">\(k\)</span>-mers, we first tested the method within species, with three main questions: Do the parameters change when the cell type lookup is within the same individual, across individual of the same sex, and across individuals of different sex?
<!-- Previous work has used $k$-mers from protein-coding genes of single-cell RNA-seq data within one individual of a species to distinguish cell types [@https://doi.org/10.1101/723833], and we are interested in first comparing effectiveness of $k$-mers to identify cell types across individuals of different species. -->
To test this, we built cell type databases from <span class="math inline">\(k\)</span>-mer signatures using sequence bloom trees (SBTs) <span class="citation" data-cites="DfZvKOpG">[<a href="#ref-DfZvKOpG" role="doc-biblioref">41</a>]</span> using the <code>sourmash</code> software tool <span class="citation" data-cites="ART4gJKs ExvjTLL3">[<a href="#ref-ExvjTLL3" role="doc-biblioref">20</a>,<a href="#ref-ART4gJKs" role="doc-biblioref">21</a>]</span>.
Using the Tabula Muris Senis data, we built a cell type database from one mouse, and queried its own cells (Supplementary Fig. ??A).
We first manually unified cell ontology class names into broader groups to encompass larger cell type labels.
We then created <span class="math inline">\(k\)</span>-mer signatures at many <span class="math inline">\(k\)</span>-mer sizes, from 21 to 90, using the raw nucleotide sequence from the RNA-seq reads, the <code>leaftea</code>-translated protein sequence, or a Dayhoff-encoded version of the <code>leaftea</code> translated sequences.
To reduce batch effects, we removed <span class="math inline">\(k\)</span>-mers from technical artifact genes, such as ribosomal and mitochondrial genes, and genes associated with single-cell dissociation <span class="citation" data-cites="9DeI91qq IVhllwbX 5nByhjEz qtN0W5fQ s4hwSsAX">[<a href="#ref-9DeI91qq" role="doc-biblioref">47</a>,<a href="#ref-IVhllwbX" role="doc-biblioref">48</a>,<a href="#ref-5nByhjEz" role="doc-biblioref">49</a>,<a href="#ref-qtN0W5fQ" role="doc-biblioref">50</a>,<a href="#ref-s4hwSsAX" role="doc-biblioref">51</a>]</span>.
Similar to previous work <span class="citation" data-cites="wtk13QGK">[<a href="#ref-wtk13QGK" role="doc-biblioref">18</a>]</span>, we find that using DNA from coding nucleotides successfully distinguishes cell types within individuals of the same species.
However, we see that even within an individual using the <code>leaftea</code>-translated protein sequences re-encoded into the Dayhoff alphabet outperforms using the DNA alphabet (Supplementary Fig. ??D,E), suggesting that there are enough differences in amplification or sequencing errors to warrant a degenerate protein alphabet, even within individuals of the same species.
<!-- We found that a $k$-mer size of --- with DNA had the best [F1 score / Adjusted Rand Index (ARI)] of --- cell type classification within one male mouse. -->
Interested in how the genomic differences across individuals affect DNA-based cell type classification, we built a cell type database from one male mouse, and queried a cell and returned its most similar matching cell (Supplementary Fig. ??B), and again the protein and Dayhoff alphabets outperformed the DNA alphabets (Supplementary Fig. ??D-F).<!-- We found that across mice, the DNA was still able to identify cell types, but the protein $k$-mers had a higher F1 score of ---. -->
Finally, we also tested a male-female comparison (Supplementary Fig. ??C), where the cell database was built from male cells, and the query was female cells, and found again that the Dayhoff alphabet outperformed all other alphabets (Supplementary Fig. ??D,G).
To compare across the three experiments, we averaged the adjusted rand and F1 scores (Supplementary Fig. ??H) and found that a Dayhoff encoding with a <span class="math inline">\(k\)</span>-mer size of 39, corresponding to a peptide length of 13, maximized the F1 scores to 0.799 (Supplementary Fig. ??I).<!-- We were able to identify cell types across mice by using DNA $k$-mers of size --- and protein $k$-mers of size --- and Dayhoff $k$-mers of size ---. -->
Curious which cell types were most commonly mis-categorized, we created confusion matrices (Supplementary Fig. ??J-L) and found that the most often misclassified cell type was Natural Killer T cells, which were classified to T cells by our method.
Within cell compartments of epithelial, endothelial, myeloid and lymphoid, we had the fewest misclassifications with a — alphabet and k-mer size of —.
As these cell types are often difficult to distinguish, we felt this was sufficient to show the utility of Dayhoff-encoded amino acid <span class="math inline">\(k\)</span>-mers to identify cell types within a species.
Thus, we find that Dayhoff-encoded <code>leaftea</code>-translated <span class="math inline">\(k\)</span>-mers with a DNA <span class="math inline">\(k\)</span>-mer size of 39 works well to “soften the rough edges” introduced by individual genomic variation for cell types and can be used to identify similar cell types within and across individuals of the same species.</p>
<h4 id="key-human-cell-types-can-be-predicted-from-mouse-cells-using-k-mers-from-reduced-amino-acid-alphabets">key: Human cell types can be predicted from mouse cells using <span class="math inline">\(k\)</span>-mers from reduced amino acid alphabets</h4>
<!-- Given putative protein-coding sequences from RNA-seq reads, we now use reduced amino acid alphabets as a lightweight substitute for homology across species. -->
<p>To recapitulate known biology in the mammalian lung, we compared lung cell atlases from species related within the mammalian suborder Euarchontoglires, whose most recent common ancestor is between 85 to 95 million years ago, we applied these methods to a mouse single-cell RNA-seq atlas, <em>Tabula Muris Senis</em> <span class="citation" data-cites="rSyxV4O8">[<a href="#ref-rSyxV4O8" role="doc-biblioref">36</a>]</span>, to a human
lung cell atlas <span class="citation" data-cites="oWw3q9FU">[<a href="#ref-oWw3q9FU" role="doc-biblioref">52</a>]</span>.
Next, we show that we can identify cell types in human RNA-seq data from mouse cell type databases.
We built a cell type database from the 18-month old mouse Lung tissue in Tabula Muris Senis, and queried with human lung cells (Fig. <a href="#fig:fig2">2</a>A).
<!-- Maybe this is for the supplemental text? -->
<!-- Across human-to-mouse, we found that DNA $k$-mers which vary quite greatly across species, were too variable to use to identify cell types. -->
<!-- Instead, we found that even protein $k$-mers cannot reliably identify cell types across species.
We turned to reduced amino acid alphabets such as Dayhoff (Fig. {@fig:fig2}A). -->
We found that a Dayhoff encoding with a minimum <span class="math inline">\(k\)</span>-mer size of 36 maximized the Adjusted Rand index (0.300) and the F1-scores (0.565) for classification of cell types across human and mouse (Fig. <a href="#fig:fig2">2</a>B-C).
Curious what the common misclassifications were, we plotted confusion matrices of the ground truth of human cells, compared to what they were predicted by the mouse cell type database (Fig. <a href="#fig:fig2">2</a>D).</p>
<!-- Comparison to other methods?? -->
<p>To compare the performance of <span class="math inline">\(k\)</span>-mer homology to traditional cross-species methods, we used — tools to compare label propagation across species using 1:1 orthologous gene datasets.
We found that using 1:1 orthologs was not good and resulted in a classification adjusted rand index score of — which was not good.
This enables finding expression of genes that may not be defined in the genome, and certainly genes that do not appear, are excluded by, subsetting to 1:1 orthologs, which is a common strategy for cross-species transcriptome comparisons.</p>
<h4 id="key-reduced-amino-acid-alphabet-k-mers-enable-identifying-cell-type-enriched-sequences-that-may-not-be-present-in-gene-annotation-genome-assembly-or-orthologous-gene-mappings">key: Reduced amino acid alphabet <span class="math inline">\(k\)</span>-mers enable identifying cell type enriched sequences that may not be present in gene annotation, genome assembly, or orthologous gene mappings</h4>
<p>Once we identified common cell types across species, we then dug into genes that are enriched in the cell types that may not be present even in the annotated genome, or present in the orthologous gene mapping (Fig. <a href="#fig:fig2">2</a>E).
To find expression of genes that many not be in the genome, we took two approaches.
First, we used differential <span class="math inline">\(k\)</span>-mer expression to enrich for particular sequences that were highly expressed in a cell type.
We found that the overlap between differential gene expression 1:1 orthologs, and the differential <span class="math inline">\(k\)</span>-mer expression and orthogroup expression, we found — % overlap between al three, which had a — p-value of significance using a chi-squared or hypergeometric test of enrichment.
Additionally, we identified — genes that were enriched in molecular cell types that are not present in the genomes of human and mouse.
Specifically, we found an unannotated <em>pulmonary surfactant-associated protein A</em> gene in the Alveolar Type 2 cells, which makes sense since they express many surfactants.
We found that compared to gene expression, we had a —% overlap with the genes that are known to be enriched by gene expression of 1:1 orthologs, and we found — genes that were enriched in the cell types but were not 1:1 orthologs, and thus would have been missed by traditional methods.
Second, we used <span class="math inline">\(k\)</span>-mers from orthogroups, orthologous groups of genes, defined at the mammalian clade, of genes with common function across species [cite orthogroups paper], computed signatures for the last 100aa of each amino acid sequence of orthogroups to get a relative gene expression (Fig. <a href="#fig:fig2">2</a>F) using containment search <span class="citation" data-cites="5KVVBsT6">[<a href="#ref-5KVVBsT6" role="doc-biblioref">53</a>]</span>.
However, with the <span class="math inline">\(k\)</span>-mers, we were able to find genes such as —, — and — which are enriched in the cell type —, but are not a 1:1 ortholog, and are instead a — ortholog.
Some of these were even not present in the genome assembly, for example — is present in the human genome assembly but not in the mouse assembly.
Thus, using <span class="math inline">\(k\)</span>-mers, we are able to identify genes that are enriched in cell types but are not present in the genome assembly, in unannotated genes, or not 1:1 orthologs.</p>
<h3 id="reduced-amino-acid-k-mers-from-transcriptomes-enabled-identification-of-unannotated-immune-genes-in-chinese-horseshoe-bat-suggesting-immune-adaptations">Reduced amino acid <span class="math inline">\(k\)</span>-mers from transcriptomes enabled identification of unannotated immune genes in Chinese horseshoe bat suggesting immune adaptations</h3>
<p>We refined the lung cell type annotation from the Chinese horseshoe bat using our <span class="math inline">\(k\)</span>-mer based cell type label propagation method. Using these refined labels, we found that — in the bat cell types were more similar to — classical monocytes — in mouse and human than they tended to be to their original cell types (Fig. <a href="#fig:fig3">3</a>A).
This led us to then perform differential expression to find the genes that may be enriched in the bat cell types overall.
Using both logistic regression on the <span class="math inline">\(k\)</span>-mer abundances, and expression of orthogroups containment in the cells, we found — genes were broadly expressed in the bat, as compared to mouse and human (Fig. <a href="#fig:fig3">3</a>B).
We then dug into the genes that are expressed, but not found in the aligned genome or the gene annotation, and found that — gene was a key contributor to — natural killer cell — type identity.</p>
<ul>
<li>Mouse:bat
<ul>
<li>celltype mapping</li>
</ul></li>
<li>mouse:bat:human
<ul>
<li>Identify gene expression for genes missing in genome</li>
</ul></li>
</ul>
<p>To identify molecular cell types unique to the Chinese horseshoe bat within mammals, we added a multi-organ, whole organism single-cell RNA-seq atlas of <em>R. sinicus</em>. We found expression of interferon genes, which have not been identified in the bat genome, but we were able to detect them in the transcriptome. In the bat scrnaseq paper, they performed qPCR to show expression of interferon genes, even though they weren’t identified in the genome, to show those genes do truly exist.</p>
<p>To apply the method to a novel biological problem and create insights into previously unknown biology, we applied the <span class="math inline">\(k\)</span>-mer homology method to a recently published cell atlas of the Chinese horseshoe bat, <em>R. sinicus</em> <span class="citation" data-cites="1AkeHrvKP">[<a href="#ref-1AkeHrvKP" role="doc-biblioref">37</a>]</span>.</p>
<h2 id="discussion">Discussion</h2>
<h3 id="key-single-cell-atlases-need-to-be-made-accessible-to-the-99-of-non-model-organisms">key: Single-cell atlases need to be made accessible to the 99% of non-model organisms</h3>
<!-- The increasing accessibility of single-cell RNA-seq data  provides a  -->
<p>The availability of single-cell whole organism atlases [cite: tabula muris, tabula muris senis, MCA, HCA papers] enables rapid cell type identification in new datasets.
However, this cell type identification is not readily accessible to the 99.99% of animal species without a reference genome [cite vertebrate genome paper].
Indeed, 99.99% of the planet’s rapidly decreasing number of animal species [cite climate change, extinction papers] are unable to directly benefit from the cell typing efforts due to a combination of both lack of a reference genome, and understudied organisms that do not readily have gene mappings across species annotated.
Thus, there is an urgent need for methods that leverage model organism cell atlases for understudied animals.
It’s not just about the genome. The annotation and the orthology of gene annotations mapping across species is the really critical thing for cross-species analyses.
We developed a method to address the gap between model organisms cell atlases and the understudied organisms that may provide with novel cures for disease.</p>
<h3 id="key-putative-protein-coding-sequence-can-be-detected-across-150-million-years-of-evolution">key: Putative protein-coding sequence can be detected across ~150 million years of evolution</h3>
<p>First, we extracted putative protein-coding sequence from an understudied animal, the Chinese horseshoe bat, <em>R. sinicus</em>.
First, we demonstrated the ability to discard unlikely protein-coding frames from RNA-seq data using our tool, <code>leaftea</code>.
Our precision is high, meaning we are able to discard true non-coding frames very easily.
Even for species that are distantly related, ~150 million years between human and the distantly related platypus species, we were able to reject —% of the known non-coding frames from the simulated human data.
The most distantly related organism, platypus, with an estimated divergence time from human ranging from 167-192 MYA (<span class="citation" data-cites="W97RNFXd"><a href="#ref-W97RNFXd" role="doc-biblioref">46</a></span>), was able to identify true protein-coding sequence in human —% of the time.
<!-- This indicates that even if the organism is approximately 200 MYA from the most closely related organism in the database, it is still possible to recover putative protein-coding. -->
Thus, we are able to recover the putative protein coding sequences, even when the most recent ancestor of the closest related organism in the database is within ~150 million years, meaning this method can open up the possibilities for annotating cell types in single-cell RNA-seq atlases for any mammal.
<!-- Thus, this opens up the possibility of studying nearly ---% of organisms on the planet, as most are related within 150 million years of annotated organisms. [Olga: Ask Erich Schwarz for help on how to compute this] --></p>
<!-- With the increasing accessibility of single-cell RNA-seq, it is more and more challenging to organize existing data.
 but the analysis of said data is more and more difficult if the organism is not well anotated in the databses. -->
<!-- [What iss the average distance between unknown/unannotated organisms and annotated organisms? Can this be quantified?]
We used RNA-seq data directly due to the fact that we can extract putative protein-coding sequences from the RNA-seq data. -->
<h3 id="key-cell-type-databases-built-on-protein-k-mers-can-assign-cell-types-across-individuals-within-species">key: Cell type databases built on protein <span class="math inline">\(k\)</span>-mers can assign cell types across individuals within species</h3>
<p>We showed that we can identify cell types across individuals, within species.
Interestingly, we found that even within species and within an individual, re-encoding the predicted protein sequenced found by <code>leaftea</code> into the Dayhoff degenerate amino acid alphabet with an amino acid <span class="math inline">\(k\)</span>-mer length of 13 was most successful in detecting cell types.
Importantly, we also needed to remove genes associated with single-cell dissociation to enable cell:cell type labeling.
This suggests that due to differences in single-cell capture, molecular amplification, and sequencing errors, the nucleotide sequences are different enough that using nucleotide <span class="math inline">\(k\)</span>-mers alone misses some cell types.</p>
<h3 id="key-reduced-amino-acid-alphabet-k-mers-enable-lighweight-orthology-assignment">key: Reduced amino acid alphabet <span class="math inline">\(k\)</span>-mers enable lighweight orthology assignment</h3>
<p>We showed that we can identify cell types across species using <span class="math inline">\(k\)</span>-mers from a degenerate amino acid alphabet.
Across species, the Dayhoff alphabet was again the most successful in identifying cell types across species, specifically at an amino acid <span class="math inline">\(k\)</span>-mer size of 12.
We then looked into the genes that were enriched in cells across species.</p>
<h3 id="key-cell-type-databases-built-on-reduced-amino-acid-alphabet-protein-k-mers-can-assign-cell-types-across-species">key: Cell type databases built on reduced amino acid alphabet protein <span class="math inline">\(k\)</span>-mers can assign cell types across species</h3>
<h3 id="key-lightweight-orthology-assignment-using-k-mers-enabled-reference-independent-cross-species-rna-seq-analyses-and-achieved-insight-into-unannotated-genes-and-genes-absent-from-the-genome-assembly.">key: Lightweight orthology assignment using <span class="math inline">\(k\)</span>-mers enabled reference-independent cross-species RNA-seq analyses and achieved insight into unannotated genes and genes absent from the genome assembly.</h3>
<p>We implemented the concept of lightweight orthology assignment using <span class="math inline">\(k\)</span>-mers to the problem of cross-species RNA-seq analyses and achieved insight into unannotated genes and genes absent from the genome assembly.
By removing the orthology inference step, <code>kmermaid</code> opens up the possibility of finding shared and divergent tissue and cell types across a broad range of species, paving the way for evolutionary analyses of cell types across species.
<code>kmermaid</code> can be used in <em>de novo</em> setting for non-model organisms, finding similar cell types within an organism, or finding similar cell types relative to a reference organism, without the need for a reference genome or transcriptome.
The memory usage of <code>kmermaid</code> is quite low, using only — MB for extracting coding sequences and 50MB for assigning protein <span class="math inline">\(k\)</span>-mer signatures.
As the number of RNA-seq datasets, especially single-cell RNA-seq datasets continues to grow, we expect <code>kmermaid</code> to be widely used for identifying cell types in non-model organisms.</p>
<h3 id="key-reduced-amino-acid-alphabet-k-mer-enrichment-allowed-for-inserting-identifying-genes-that-were-not-present-in-the-genome-assembly-or-gene-annotation-or-gene-orthology">key: Reduced amino acid alphabet <span class="math inline">\(k\)</span>-mer enrichment allowed for inserting /identifying genes that were not present in the genome assembly or gene annotation, or gene orthology</h3>
<p>We were able to identify putative homologs across species by finding genes that were enriched in common cell types</p>
<h3 id="key-we-built-an-orthologous-gene-tree-and-were-able-to-identify-genes-in-cells-across-all-species">key: We built an orthologous gene tree and were able to identify genes in cells across all species</h3>
<!-- Is this a separate paper? -->
<p>Using orthogroups, we built a signature tree database of orthologous genes.
We then were able to quantify gene expression
We were able to detect genes that were not present in the genome, but were present in the orthologous genes databases.
We were thus able to directly compare cell types across species with a common set of orthologous genes.
We used sourmash gather to identify genes that were expressed across multiple cell types</p>
<h3 id="in-summary-we-developed-a-method-to-identify-cell-types-in-a-non-model-organism-using-a-reference-atlas-from-another-organism-without-the-need-for-a-genome-or-gene-annotation-from-the-non-model-organism.">In summary, we developed a method to identify cell types in a non-model organism using a reference atlas from another organism, without the need for a genome or gene annotation from the non-model organism.</h3>
<p>In summary, we developed a method to identify both known cell types in a non-model organism using a reference atlas from another organism, without the need for a genome or gene annotation from the non-model organism.
This method can be used to combine single-cell cell atlases from well-annotated, model organisms, with sequencing data from poorly annotated non-model organisms, to directly find homologous cell types and orthologous genes.
By eliminating read alignment and orthologous gene mapping, <code>kmermaid</code> enables comparison of transcriptomes of the remaining 99.999% Eukaryotic species on Earth without submitted genome assemblies, with the cell atlases of a handful of model organisms to identify shared and novel cell types, and <em>de novo</em> identify orthologous genes.
By identifying homologous cell types across a broad variety of species, we come closer to an understanding of the evolution of genes, cells, and thus life itself.</p>
<p><code>kmermaid</code> is free and open-source software and is available as Supplementary Data and at http://github.com/czbiohub/kmermaid and as a scalable Nextflow workflow at http://github.com/nf-core/nf-kmermaid.</p>
<p>We have provided pre-built cell type databases of <em>Tabula Muris Senis</em> at —.</p>
<h2 id="online-methods">Online Methods</h2>
<!-- This is just a placeholder to separate the supplemental methods from the rest of the paper.
All remaining supplemental sections should fall under this, meaning they need to be formatted
with heading 3 with three hashes: `###` -->
<h3 id="implementation">Implementation</h3>
<h4 id="reduced-alphabets">Reduced alphabets</h4>
<p>At the core of <code>sencha</code> is the ability to cheaply compare sequences using <span class="math inline">\(k\)</span>-mers.
As <span class="math inline">\(k\)</span>-mers are very brittle to substitutions and thus to compare across species, one must allow for minor base substitutions that still maintain similar chemical or structural properties.
A reduced alphabet can encode useful information into a smaller alphabet space, and enable sequence comparisons across a broader variety of species than the original alphabet alone.</p>
<h5 id="reduced-amino-acid-alphabets">Reduced amino acid alphabets</h5>
<p>Reduced amino acid alphabets have been useful for over 50 years <span class="citation" data-cites="66KDHtpG">[<a href="#ref-66KDHtpG" role="doc-biblioref">54</a>]</span> in finding related protein sequences <span class="citation" data-cites="Aq7zYbOg 16Kv6RIe3 sZoryWIB f5u3dDtc QMYjnr6l">[<a href="#ref-sZoryWIB" role="doc-biblioref">29</a>,<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>,<a href="#ref-16Kv6RIe3" role="doc-biblioref">31</a>,<a href="#ref-f5u3dDtc" role="doc-biblioref">32</a>,<a href="#ref-QMYjnr6l" role="doc-biblioref">33</a>]</span>.
Recently, a reduced amino acid alphabet (specifically, <code>aa9</code> below) combined with <span class="math inline">\(k\)</span>-mers were used to find homologous protein-coding sequences <span class="citation" data-cites="Rqlso7kk">[<a href="#ref-Rqlso7kk" role="doc-biblioref">28</a>]</span>.
We build on this concept by enabling prediction of protein-coding sequences from RNA-seq reads, and by enabling users to perform a parameter sweep in an all-by-all comparison to identify putative homologs using a variety of alphabet metrics.</p>
<h6 id="dayhoff-and-hp-alphabets">Dayhoff and <code>HP</code> alphabets</h6>
<div id="tbl:sequence-encodings" class="tablenos">
<table>
<caption><span>Table 2:</span> Dayhoff and hydrophobic-polar encodings are a reduced amino acid
alphabet allowing for permissive cross-species sequence comparisons. For
example, the amino acid sequence <code>SASHAFIERCE</code> would be Dayhoff-encoded
to <code>bbbdbfecdac</code>, and HP-encoded to <code>phpphhhpppp</code>, as below. </caption>
<thead>
<tr class="header">
<th style="text-align: left;">Amino acid</th>
<th style="text-align: left;">Property</th>
<th style="text-align: left;">Dayhoff</th>
<th style="text-align: left;">Hydrophobic-polar (HP)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>C</code></td>
<td style="text-align: left;">Sulfur polymerization</td>
<td style="text-align: left;"><code>a</code></td>
<td style="text-align: left;"><code>p</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>A</code>, <code>G</code>, <code>P</code>, <code>S</code>, <code>T</code></td>
<td style="text-align: left;">Small</td>
<td style="text-align: left;"><code>b</code></td>
<td style="text-align: left;"><code>AGP</code>: <code>h</code><br><code>ST</code>: <code>p</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>D</code>, <code>E</code>, <code>N</code>, <code>Q</code></td>
<td style="text-align: left;">Acid and amide</td>
<td style="text-align: left;"><code>c</code></td>
<td style="text-align: left;"><code>p</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>H</code>, <code>K</code>, <code>R</code></td>
<td style="text-align: left;">Basic</td>
<td style="text-align: left;"><code>d</code></td>
<td style="text-align: left;"><code>p</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>I</code>, <code>L</code>, <code>M</code>, <code>V</code></td>
<td style="text-align: left;">Hydrophobic</td>
<td style="text-align: left;"><code>e</code></td>
<td style="text-align: left;"><code>h</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>F</code>, <code>W</code>, <code>Y</code></td>
<td style="text-align: left;">Aromatic</td>
<td style="text-align: left;"><code>f</code></td>
<td style="text-align: left;"><code>h</code></td>
</tr>
</tbody>
</table>
</div>
<pre><code>protein20: SASHAFIERCE
dayhoff6:  bbbdbfecdac
hp2:       phpphhhpppp</code></pre>
<!--
###### All implemented alphabets (with citations, not as nicely organized)
<!-- Copied this google spreadsheet https://docs.google.com/spreadsheets/d/1RDuQD0aRyv-FnQJbjRosHEJV85_9BNrzmmvgFRQSgN8/edit#gid=0 into https://thisdavej.com/copy-table-in-excel-and-paste-as-a-markdown-table/, reformatted with https://atom.io/packages/markdown-table-formatter -->
<p>[NOTE: maybe this should go into the supplementary? The main alphabets that have been successful for me are dayhoff and HP]</p>
<table>
<colgroup>
<col style="width: 32%" />
<col style="width: 12%" />
<col style="width: 54%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Citation</th>
<th style="text-align: left;">Alphabet</th>
<th style="text-align: left;">Amino acid groups</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Phillips, R., <em>et al</em>. (2012). <span class="citation" data-cites="1GiMcbPLs">[<a href="#ref-1GiMcbPLs" role="doc-biblioref">55</a>]</span></td>
<td style="text-align: left;">hp2</td>
<td style="text-align: left;"><code>AFGILMPVWY</code> <code>CDEHKNQRST</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Peterson, E. L., <em>et al</em>. (2009) <span class="citation" data-cites="Aq7zYbOg">[<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>]</span></td>
<td style="text-align: left;">gbmr4</td>
<td style="text-align: left;"><code>G</code> <code>ADKERNTSQ</code> <code>YFLIVMCWH</code> <code>P</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Dayhoff, M. O., &amp; Eck, R. V. (1968). <span class="citation" data-cites="66KDHtpG">[<a href="#ref-66KDHtpG" role="doc-biblioref">54</a>]</span></td>
<td style="text-align: left;">dayhoff6</td>
<td style="text-align: left;"><code>AGPST</code> <code>HRK</code> <code>DENQ</code> <code>FWY</code> <code>ILMV</code> <code>C</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">This paper</td>
<td style="text-align: left;">botvinnik8</td>
<td style="text-align: left;"><code>AG</code> <code>DE</code> <code>RK</code> <code>NQ</code> <code>ST</code> <code>FY</code> <code>LIV</code> <code>CMWHP</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Hu, X., &amp; Friedberg, I. (2019). <span class="citation" data-cites="Rqlso7kk">[<a href="#ref-Rqlso7kk" role="doc-biblioref">28</a>]</span></td>
<td style="text-align: left;">aa9</td>
<td style="text-align: left;"><code>G</code> <code>AST</code> <code>KR</code> <code>EQ</code> <code>DN</code> <code>CFILMVY</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Peterson, E. L., <em>et al</em>. (2009) <span class="citation" data-cites="Aq7zYbOg">[<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>]</span></td>
<td style="text-align: left;">sdm12</td>
<td style="text-align: left;"><code>G</code> <code>A</code> <code>D</code> <code>KER</code> <code>N</code> <code>TSQ</code> <code>YF</code> <code>LIVM</code> <code>C</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Peterson, E. L., <em>et al</em>. (2009) <span class="citation" data-cites="Aq7zYbOg">[<a href="#ref-Aq7zYbOg" role="doc-biblioref">30</a>]</span></td>
<td style="text-align: left;">hsdm17</td>
<td style="text-align: left;"><code>G</code> <code>A</code> <code>D</code> <code>KE</code> <code>R</code> <code>N</code> <code>T</code> <code>S</code> <code>Q</code> <code>Y</code> <code>F</code> <code>LIV</code> <code>M</code> <code>C</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Dayhoff, M. O., &amp; Eck, R. V. (1968). <span class="citation" data-cites="66KDHtpG">[<a href="#ref-66KDHtpG" role="doc-biblioref">54</a>]</span></td>
<td style="text-align: left;">protein20</td>
<td style="text-align: left;"><code>G</code> <code>A</code> <code>D</code> <code>E</code> <code>K</code> <code>R</code> <code>N</code> <code>T</code> <code>S</code> <code>Q</code> <code>Y</code> <code>F</code> <code>L</code> <code>I</code> <code>V</code> <code>M</code> <code>C</code> <code>W</code> <code>H</code> <code>P</code></td>
</tr>
</tbody>
</table>
<!--
##### Reduced nucleotide alphabets

The IUPAC degenerate nucleotide code [@url:https://www.qmul.ac.uk/sbcs/iubmb/misc/naseq.html] includes several two-letter codes for the original 4-letter nucleobase alphabet.
The first, Weak/Strong, indicates the strength of the hydrogen bond across the double strand.
The bond of adenine to thymine has two hydrogen bonds, making it weak; and the bond of guanine to cytosine has three hydrogen bonds, making it 50% stronger.
The second, Purine/Pyrimidine, encodes the ring size of the nucleobase, where Adenine and Guanine both have larger Purine double rings, while Cytosine and Thymine/Uracil have smaller Pyrimidine rings.
The third, Amino/Keto, designates the main functional group of the ring, where Adenine and Cytosine both have an Amino group, while Guanine and Thymine/Uracil both have a Keto group. -->
<!-- This spreadsheet:https://docs.google.com/spreadsheets/d/1RDuQD0aRyv-FnQJbjRosHEJV85_9BNrzmmvgFRQSgN8/edit#gid=104831627 was copied into this formatter: https://thisdavej.com/copy-table-in-excel-and-paste-as-a-markdown-table/ -->
<!-- | Nucleotide | Hydrogen Bonding | Ring type      | Ring functional group | Nucleobase chemical structure                                                                                                                                   |
|:-----------|:-----------------|:---------------|:----------------------|:------------------------------------------------|
| A          | Weak (W)         | Purine (R)     | Amino (M)             | ![Adenine chemical structure](https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/Adenine.svg/120px-Adenine.svg.png)                                      |
| C          | Strong (S)       | Pyrimidine (Y) | Amino (M)             | ![Cytosine chemical structure](https://upload.wikimedia.org/wikipedia/commons/thumb/1/10/Cytosine_chemical_structure.png/120px-Cytosine_chemical_structure.png) |
| G          | Strong (S)       | Purine (R)     | Keto (K)              | ![Guanine chemical structure](https://upload.wikimedia.org/wikipedia/commons/thumb/3/3f/Guanin.svg/150px-Guanin.svg.png)                                        |
| T          | Weak (W)         | Pyrimidine (Y) | Keto (K)              | ![Thymine chemical structure](https://upload.wikimedia.org/wikipedia/commons/thumb/a/ab/Thymine_skeletal.svg/120px-Thymine_skeletal.svg.png)                    |
| U          | Weak (W)         | Pyrimidine (Y) | Keto (K)              | ![Uracil chemical structure](https://upload.wikimedia.org/wikipedia/commons/thumb/b/b4/Uracil.svg/120px-Uracil.svg.png)                                         |
Thus, the nucleotide string `GATTACA` would be re-encoded into the following:

```
Nucleotide:        GATTACA
Weak/Strong:       SWWWWSW
Purine/Pyrimidine: RRYYRYR
Functional group:  KMKKMMM
```
-->
<h4 id="sencha-translate"><code>sencha translate</code></h4>
<figure>
<img src="images/SVG/supplemental_figure1.svg" id="sfig1:supplemental_figure1" data-tag="supplemental_figure1" style="width:100.0%" alt="" /><figcaption>Supplemental Figure 1: Receiver operating characteristic curves for predicting protein-coding sequence of human RNA-seq data, using mammalian species’ reference proteomes.
<strong>A.</strong> Prediction of protein-coding sequence using protein alphabet.
<strong>B.</strong> Prediction of protein-coding sequence using Dayhoff alphabet.</figcaption>
</figure>
<h5 id="set-jaccard-threshold-of-translate-by-controlling-false-positive-rate-of-protein-coding-prediction">Set Jaccard threshold of <code>translate</code> by controlling false positive rate of protein-coding prediction</h5>
<p>To set a threshold of the minimum Jaccard overlap between a translated read’s frame and the reference proteome, the most statistically principled way is to control the false positive rate of predicing a protein-coding read.</p>
<h6 id="probability-of-random-k-mers-from-a-read">Probability of random <span class="math inline">\(k\)</span>-mers from a read</h6>
<p>If <span class="math inline">\(k\)</span>-mers from reads were independent, identically distributed (i.i.d.) variables, then a translated read of length <span class="math inline">\(L_{\mathrm{translated}}\)</span> drawing letters from the alphabet <span class="math inline">\(\Sigma\)</span>, whose size is <span class="math inline">\(|\Sigma|\)</span>, would contain</p>
<p><span id="eq:n_kmers_per_read_iid" class="eqnos"><span class="math display">\[\left( \frac{1}{\left| \Sigma \right|^k} \right)^{L_{\mathrm{translated}} - k + 1}\]</span><span class="eqnos-number">(1)</span></span> </p>
<p>However, <span class="math inline">\(k\)</span>-mers drawn from reads are not i.i.d.
Let’s take a simple example.
If we have a two-letter alphabet, <span class="math inline">\(\Sigma = \\{a, b\\},\)</span>, thus <span class="math inline">\(|\Sigma| = 2\)</span>.
Let us use an example sequence <span class="math inline">\(S = abbabba\)</span>.
If <span class="math inline">\(k = 4\)</span>, then the first <span class="math inline">\(k\)</span>-mer is <span class="math inline">\(abba\)</span>.
The second <span class="math inline">\(k\)</span>-mer is thus either <span class="math inline">\(bbaa\)</span> or <span class="math inline">\(bbab\)</span>, with equal probability.
We can generalize this: Given the first <span class="math inline">\(k\)</span>-mer, the first <span class="math inline">\(k-1\)</span> letters from the second <span class="math inline">\(k\)</span>-mer are known, and thus the probability of guessing the next <span class="math inline">\(k\)</span>-mer is <span class="math inline">\(\frac{1}{\left|\Sigma\right|}\)</span>.</p>
<figure>
<img src="images/abba_sequence_prob_future_kmers.png" id="sfig:prob_kmers_in_read" data-tag="prob_kmers_in_read" style="width:50.0%" alt="" /><figcaption>Probability of future <span class="math inline">\(k\)</span>-mers is influenced by the existence of previous <span class="math inline">\(k\)</span>-mers.</figcaption>
</figure>
<p>Thus, the probability of a random <span class="math inline">\(k\)</span>-mer from a sequencing read is completely dependent on the alphabet size <span class="math inline">\(|\Sigma|\)</span> and its translated sequence length, <span class="math inline">\(L_{\mathrm{translated}}\)</span>:</p>
<p><span id="eq:prob_kmers_in_read" class="eqnos"><span class="math display">\[
\begin{align}
\mathrm{Pr}(\mathrm{FPR}) &amp;= \left(\frac{1}{|\Sigma|}\right)^k \times \left( \frac{1}{|\Sigma|} \right)^{L_{\mathrm{translated}} - k }\\
&amp;=\frac{1}{| \Sigma |^k  \times |\Sigma|^{L_{\mathrm{translated} - k} }}\\
&amp;= \frac{1}{| \Sigma|^{L_{\mathrm{translated}}}}
\end{align}
\]</span><span class="eqnos-number">(2)</span></span> </p>
<h6 id="bloom-filter-collision-probability">Bloom filter collision probability</h6>
<p>The probability of error of the <code>khmer</code> bloom filter implementation <span class="citation" data-cites="1DpHWPZC9">[<a href="#ref-1DpHWPZC9" role="doc-biblioref">56</a>]</span> used in <code>sencha</code>, given <span class="math inline">\(N\)</span> distinct <span class="math inline">\(k\)</span>-mers counted, a hash table size of <span class="math inline">\(H\)</span>, and <span class="math inline">\(Z\)</span> total number of hash tables, is</p>
<p><span id="eq:bloom_filter_fpr" class="eqnos"><span class="math display">\[\mathrm{Pr}(\mathrm{FPR}_{\mathrm{bloom}}) = \left( 1 - \exp^{N/H}\right)^Z.\]</span><span class="eqnos-number">(3)</span></span> </p>
<p>Theoretically, the total number of <span class="math inline">\(k\)</span>-mers is limited by the alphabet size and choice of <span class="math inline">\(k\)</span>.
Empirically, the number of possible <span class="math inline">\(k\)</span>-mers is limited by the <span class="math inline">\(k\)</span>-mers which are compatible with life, and by <span class="math inline">\(k=5\)</span>, the number of theoretical protein <span class="math inline">\(k\)</span>-mers exceeds the number of observed protein <span class="math inline">\(k\)</span>-mers.
Additionally, the mass of all possible <span class="math inline">\(k\)</span>-mers of a certain size, exceeds the mass of the planet earth by <span class="math inline">\(k = X\)</span> [get the data for this].
The UniProtKB Opisthokonta manually reviewed dataset contains <span class="math inline">\(4.8 \times 10^7\)</span> <span class="math inline">\(7\)</span>-mers in the protein alphabet.
Thus, we can give an upper bound to the number of theoretical <span class="math inline">\(k\)</span>-mers to be <span class="math inline">\(10^8\)</span>.
Therefore, the total number of <span class="math inline">\(k\)</span>-mers in the bloom filter is,</p>
<p><span id="eq:n_proteome_kmers" class="eqnos"><span class="math display">\[N = \min\left( 10^8, |\Sigma|^k \right).\]</span><span class="eqnos-number">(4)</span></span> </p>
<figure>
<img src="images/ksize_vs_n_kmers_theoretical_observed.png" alt="" /><figcaption>Number of theoretical <span class="math inline">\(k\)</span>-mers given alphabet size, compared to observed <span class="math inline">\(k\)</span>-mers in ENSEMBL human translated proteome. The number of observed <span class="math inline">\(k\)</span>-mers is distinct from the number of theoretical <span class="math inline">\(k\)</span>-mers, as the total number of observed <span class="math inline">\(k\)</span>-mers is limted by <span class="math inline">\(k\)</span>-mers compatible with life. Rerun this with uniprot uniref data.</figcaption>
</figure>
<h6 id="false-positive-rate-of-protein-coding-prediction">False positive rate of protein-coding prediction</h6>
<p>Combining Equations <a href="#eq:prob_kmers_in_read">2</a>, <a href="#eq:n_proteome_kmers">4</a>, and <a href="#eq:bloom_filter_fpr">3</a>, for an RNA-seq read of length <span class="math inline">\(L\)</span> where its translated length <span class="math inline">\(L_{\mathrm{translated}} = \lfloor \frac{L}{3} \rfloor\)</span>, containing a possible six frames of translation, then the false-positive rate (FPR) protein-coding read is,</p>
<p><span id="eq:fpr_protein_coding" class="eqnos"><span class="math display">\[
\begin{align}
\mathrm{Pr}\left(\mathrm{FPR}\right) &amp;=
6 \times \left( 1 - \exp^{\min\left( 10^8, |\Sigma|^k \right)} \right)^Z \times \frac{1}{\left| \Sigma \right|^{L_{\mathrm{translated}}}}.
\end{align}
\]</span><span class="eqnos-number">(5)</span></span> </p>
<p>Probability of a random <span class="math inline">\(k\)</span>-mer being present in a perfect protein index
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\frac{N}{\left| \Sigma \right| ^k}
\]</span><span class="eqnos-number">(6)</span></span> </p>
<p>Probability of <em>all</em> <span class="math inline">\(L - k + 1\)</span> <span class="math inline">\(k\)</span>-mers of a read being present in a perfect protein index
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\left( \frac{N}{\left| \Sigma \right| ^k} \right) ^ { L - k + 1 }
\]</span><span class="eqnos-number">(7)</span></span> </p>
<p>Probability of a random <span class="math inline">\(k\)</span>-mer being present in an <em>imperfect</em> protein bloom filter with false positive rate or “collision rate” <span class="math inline">\(p_{\mathrm{collision}}\)</span>
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\frac{N}{p_{\mathrm{collision}} * \left| \Sigma \right| ^k}
\]</span><span class="eqnos-number">(8)</span></span> </p>
<p>Probability of <em>all</em> <span class="math inline">\(L - k + 1\)</span> <span class="math inline">\(k\)</span>-mers of a read being present in an <em>imperfect</em> protein bloom filter with false positive rate or “collision rate” <span class="math inline">\(p_{\mathrm{collision}}\)</span>
<span id="eq:prob_kmer_in_index" class="eqnos"><span class="math display">\[
\left( \frac{N}{p_{\mathrm{collision}} * \left| \Sigma \right| ^k} \right) ^ { L - k + 1 }
\]</span><span class="eqnos-number">(9)</span></span> </p>
<h5 id="similarity-thresholds-for-percentage-of-matching-k-mers">Similarity thresholds for percentage of matching <span class="math inline">\(k\)</span>-mers</h5>
<p>A single SNP in a read affects <span class="math inline">\(k\)</span> <span class="math inline">\(k\)</span>-mers.</p>
<h4 id="sencha-compare-kmer-content-performs-all-by-all-or-pairwise-k-mer-similarity-of-protein-or-nucleotide-sequences-using-reduced-alphabets"><code>sencha compare-kmer-content</code> performs all-by-all or pairwise <span class="math inline">\(k\)</span>-mer similarity of protein or nucleotide sequences using reduced alphabets</h4>
<figure>
<img src="images/SVG/figure3.svg" id="sfig:figure3" data-tag="figure3" style="width:100.0%" alt="" /><figcaption>Overview of <code>sencha compare-kmer-content</code> <strong>A.</strong> Protein sequences are <span class="math inline">\(k\)</span>-merized by converting into a bag of words using a sliding window of size <span class="math inline">\(k\)</span>, potentially re-encoded to a lossy alphabet, and then their fraction of overlapping <span class="math inline">\(k\)</span>-mers is computed into a Jaccard similarity. <strong>B.</strong> One option for <code>sencha compare-kmer-content</code> is to specify a pair of sequence files, and compute a background of <span class="math inline">\(k\)</span>-mer similarty using randomly shuffled pairs. <strong>C.</strong> Another option for <code>sencha compare-kmer-content</code> is to do an all-by-all <span class="math inline">\(k\)</span>-mer similarity comparison.</figcaption>
</figure>
<h3 id="cell-type-predictions">Cell type predictions</h3>
<figure>
<img src="images/SVG/supplemental_figure2.svg" id="sfig2:sfig2" data-tag="supplemental_figure2" style="width:100.0%" alt="" /><figcaption>Supplemental Figure 2: Cell databases built from degenerate protein alphabet <span class="math inline">\(k\)</span>-mers can reliably identify cell types within species, across individuals.
<strong>A.</strong> Experimental set up of testing self:self lookup of cell types within the same organism. Build a cell type sequence bloom tree (SBT) database from a male mouse, then query with cells from the same male mouse, and return the top non-identical hit cell.
<strong>B.</strong> Experimental set up of testing male:male lookup of cell types within the same species, different individuals and same sex. Build a cell type (SBT) database from a male mouse, then query with cells from a different male mouse, and return the most similar cell.
<strong>C.</strong> Experimental set up of testing male:female lookup of cell types within the same species, different individuals and different sex. Build a cell type (SBT) database from a male mouse, then query with cells from a female mouse, and return the most similar cell.
<strong>D.</strong> Mean of dlassification metrics for self:self (left column), male:male (middle column), male:female (right column) cell type lookup, using different molecule types and <span class="math inline">\(k\)</span>-mer sizes, highlighting the <span class="math inline">\(k\)</span>-mer size that maximizes the mean for each metric and alphabet. x-axis, <span class="math inline">\(k\)</span>-mer size in DNA space, i.e. a <span class="math inline">\(k\)</span>-mer size of 21 corresponds to a an amino acid length of 7. y-axis, value of classification metric. Top row, adjusted rand score; bottom row, F1 score.
<strong>E-G.</strong> Heatmap of classification metrics for alphabets and <span class="math inline">\(k\)</span>-mer sizes maximizing classification ability from (<strong>D</strong>), for each mouse:mouse experiment.
<strong>H.</strong> Mean of classification metrics across all three mouse:mouse experiments, highlighting the <span class="math inline">\(k\)</span>-mer size that maximizes the mean for each metric and alphabet. Vertical lines, standard deviation of the mean. x-axis, <span class="math inline">\(k\)</span>-mer size in DNA space. y-axis, value of classification metric. Left, adjusted rand score; right, F1 score.
<strong>I.</strong> Heatmap of classification metrics for alphabets and <span class="math inline">\(k\)</span>-mer sizes maximizing classification ability from (<strong>H</strong>).
<strong>J-L.</strong> Confusion matrices of classification with best performing parameters from (<strong>I</strong>). x-axis, predicted cell label; y-axis, ground truth cell label.
<strong>J.</strong> DNA alphabet with <span class="math inline">\(k\)</span>-mer size of 42.
<strong>K.</strong> Protein 20-letter alphabet with <span class="math inline">\(k\)</span>-mer size of 27.
<strong>L.</strong> Dayhoff 6-letter amino acid encoding with <span class="math inline">\(k\)</span>-mer size of 39.</figcaption>
</figure>
<figure>
<img src="images/SVG/supplemental_figure3.svg" id="sfig3:sfig3" data-tag="supplemental_figure3" style="width:100.0%" alt="" /><figcaption>Supplemental Figure 3: Cell databases built from degenerate protein alphabet <span class="math inline">\(k\)</span>-mers can reliably identify cell types within species, across individuals.
<strong>A.</strong> Experimental set up of testing human:human lookup of cell types within species, across individuals. Build a cell type (SBT) database from one patient mouse, then query with cells from a different patient, and return the most similar cell.
<strong>B.</strong> Mean of classification metrics across human:human experiments, highlighting the <span class="math inline">\(k\)</span>-mer size that maximizes the mean for each metric and alphabet. Vertical lines, standard deviation of the mean. x-axis, <span class="math inline">\(k\)</span>-mer size in DNA space, i.e. a <span class="math inline">\(k\)</span>-mer size of 21 corresponds to a an amino acid length of 7. y-axis, value of classification metric. Left, adjusted rand score; right, F1 score.
<strong>C.</strong> Molecules and <span class="math inline">\(k\)</span>-mer sizes maximizing classification metrics from (<strong>K</strong>).
<strong>D-F.</strong> Confusion matrices of classification with best performing parameters from <strong>E</strong>. x-axis, predicted cell label; y-axis, ground truth cell label.
<strong>D.</strong> DNA alphabet with <span class="math inline">\(k\)</span>-mer size of —.
<strong>E.</strong> Protein 20-letter alphabet with <span class="math inline">\(k\)</span>-mer size of —.
<strong>F.</strong> Dayhoff 6-letter amino acid encoding with <span class="math inline">\(k\)</span>-mer size of —.</figcaption>
</figure>
<h3 id="benchmarking">Benchmarking</h3>
<p>Methods go here.</p>
<h4 id="computational">Computational</h4>
<h5 id="k-mer-comparison-of-orthologous-genes"><span class="math inline">\(k\)</span>-mer comparison of orthologous genes</h5>
<p>We used ENSEMBL version 97.
We did things.
One sentence per line.
Prefer DOI for references, but for Biorxiv use the URL.
DOI example: <span class="citation" data-cites="LkrS4xXB">[<a href="#ref-LkrS4xXB" role="doc-biblioref">7</a>]</span>.
Biorxiv example: <span class="citation" data-cites="46R7aU4N">[<a href="#ref-46R7aU4N" role="doc-biblioref">57</a>]</span>.
Multiple citations per line example: <span class="citation" data-cites="LkrS4xXB 46R7aU4N">[<a href="#ref-LkrS4xXB" role="doc-biblioref">7</a>,<a href="#ref-46R7aU4N" role="doc-biblioref">57</a>]</span>.</p>
<h5 id="extraction-of-putative-coding-reads-from-rna-seq">Extraction of putative coding reads from RNA-seq</h5>
<p>We did things.
One sentence per line.
Prefer DOI for references, but for Biorxiv use the URL.
DOI example: <span class="citation" data-cites="LkrS4xXB">[<a href="#ref-LkrS4xXB" role="doc-biblioref">7</a>]</span>.
Biorxiv example: <span class="citation" data-cites="46R7aU4N">[<a href="#ref-46R7aU4N" role="doc-biblioref">57</a>]</span>.
Multiple citations per line example: <span class="citation" data-cites="LkrS4xXB 46R7aU4N">[<a href="#ref-LkrS4xXB" role="doc-biblioref">7</a>,<a href="#ref-46R7aU4N" role="doc-biblioref">57</a>]</span>.</p>
<h4 id="prediction-of-protein-coding-sequences-across-a-variety-of-species">Prediction of protein-coding sequences across a variety of species</h4>
<p>We used <code>sencha translate</code> to obtain putative protein-coding sequences from a comaparative transcriptomic dataset spanning nine species and six tissues <span class="citation" data-cites="fBCZ87Cf">[<a href="#ref-fBCZ87Cf" role="doc-biblioref">58</a>]</span>.</p>
<h4 id="read-preprocessing">Read preprocessing</h4>
<p>As the protein-coding score is assessed on the entire read, we recommend RNA-seq reads be removed of library artifacts to the best of the user’s ability.
This means, the adapters should be trimmed, and if there was a negative insert size such that the R1 and R2 reads overlap, then the read pairs should be merged.</p>
<h3 id="pipelines">Pipelines</h3>
<!-- ## Figure 3 -- Overview of `kmermaid` and `predictorthologs` pipelines -->
<div id="fig:sfig3" class="fignos">
<figure>
<img src="images/SVG/supplementary_figure3.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 4:</span> <strong>A.</strong> Overview of the <code>kmermaid</code> pipeline. (<strong>a</strong>, <strong>b</strong>, <strong>c</strong>) <code>kmermaid</code> consists of a protein-coding prediction phase (<strong>a</strong>) that is invoked by the command <code>sencha translate</code>, a <span class="math inline">\(k\)</span>-mer sketch computation phase (<strong>b</strong>) invoked by the command <code>sourmash sketch</code>, a signature similarity comparison phase (<strong>c</strong>) invoked by the command <code>sourmash compare</code>, and an optional database-creation phase (<strong>d</strong>) invoked by the command <code>sourmash index</code>. The coding prediction phase has three components: (1) six-frame translation, removal of stop-codon frames, and subsequent <span class="math inline">\(k\)</span>-merization of RNA-sequencing reads; (2) a degenerate protein alphabet which allows for protein-coding detection from a wide variety of species; (3) a bloom filter containing known protein-coding sequences from a well annotated organism; and (4) computation of the Jaccard index of translated RNA-seq reading frames. The sketch computation phase involves randomly subsetting the degenerate peptide <span class="math inline">\(k\)</span>-mers using a MinHash algorithm. The sketch comparison phase consists of computing the Jaccard intersection of MinHashed degenerate peptide <span class="math inline">\(k\)</span>-mers between all pairs of samples.</figcaption>
</figure>
</div>
<div id="fig:kmermaid" class="fignos">
<figure>
<img src="images/SVG/kmermaid_workflow.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 5:</span> Supplemental Figure X.
<strong>A.</strong> Overview of <code>nf-core/kmermaid</code> pipeline to compare DNA/RNA/protein sequences on <span class="math inline">\(k\)</span>-mer content.
1. If input is bam, extract per-cell sequences using <code>bam2fasta percell</code>.
2. Predict amino acid sequence of each RNA-seq read using <code>sencha translate</code>.
3. Randomly subsample amino acid <span class="math inline">\(k\)</span>-mers via MinHash using <code>sourmash sketch</code>.
4. Compare all <span class="math inline">\(k\)</span>-mer sketches to one another using <code>sourmash compare</code> to compute cell-cell Jaccard similarities.
5. Build sequence bloom tree using <code>sourmash index</code>.
6. Build k-nearest neighbor graph using sequence bloom tree.
7. Build UMAP off of KNN.</figcaption>
</figure>
</div>
<div id="fig:predictorthologs" class="fignos">
<figure>
<img src="images/SVG/predictorthologs_workflow.svg" style="width:100.0%" alt="" /><figcaption><span>Figure 6:</span> Supplemental Figure X.
<strong>B.</strong> Overview of <code>czbiohub/nf-predictorthologs</code> pipeline to query putative function of protein sequences.
1. If input is bam, must also have a convert bam reads to raw fastq files using the <code>samtools fastq</code> subcommand (samtools version 1.9). If input is fastqs, go directly to second step.
2. Trim adapters, poly-A, polyG using the <code>fastp</code> tool.
3. Predict protein-coding sequence using khtools extract_coding, using conservative UniProt/SwissProt manually curated database as examples of known protein-coding sequences, for most stringent definition of protein-coding.
4. Query predicted protein in permissive NCBI RefSeq non-redundant protein database for most complete search query.</figcaption>
</figure>
</div>
<h3 id="installation">Installation</h3>
<p><code>sencha</code> can be installed with the Anaconda package manager, <code>conda</code> (preferred),</p>
<pre><code># Note: not actually on bioconda yet ... this is aspirational
conda install --channel bioconda sencha</code></pre>
<p>or from the Python Package Index (PyPI) with the Python package manager, <code>pip</code>,</p>
<pre><code># Note: not actually on PyPI yet ... this is aspirational
pip install sencha</code></pre>
<h3 id="usage">Usage</h3>
<h4 id="creation-of-amino-acid-k-mer-database-with-sencha-index">Creation of amino acid <span class="math inline">\(k\)</span>-mer database with <code>sencha index</code></h4>
<p>Before predicting protein-coding sequences, <code>sencha</code> must create a database of known amino acid <span class="math inline">\(k\)</span>-mers, which is stored in the form of a probabilistic set membership data structure known as a bloom filter.
<code>sencha</code> uses the bloom filter implementation in <code>khmer</code>/<code>oxli</code> <span class="citation" data-cites="G2CzfHHF 1DpHWPZC9">[<a href="#ref-1DpHWPZC9" role="doc-biblioref">56</a>,<a href="#ref-G2CzfHHF" role="doc-biblioref">59</a>]</span>, called a NodeGraph.
We created a dataset of known amino acid <span class="math inline">\(k\)</span>-mers from the manually annotated UniProtKB/Swiss-Prot databases <span class="citation" data-cites="1ZE22clL 1BBNQZrBs">[<a href="#ref-1ZE22clL" role="doc-biblioref">60</a>,<a href="#ref-1BBNQZrBs" role="doc-biblioref">61</a>]</span>.
We used only protein sequences observed in <em>Opisthokont</em> species <span class="citation" data-cites="Ty6517L8">[<a href="#ref-Ty6517L8" role="doc-biblioref">62</a>]</span>, previously known as a “Fungi/Metazoa” group that encompasseses “Fungus-like” <em>Holomycota</em> and “Animal-like” <em>Holozoa</em>. [NOTE: Does this need a figure/phylogenetic timetree?]</p>
<pre><code>sencha index \
  --tablesize 100000000 \
  --molecule protein \
  --peptide-ksize 7 \
  --save-as uniprot-reviewed_yes+taxonomy_2759__molecule-protein_ksize-7.bloomfilter \
  uniprot-reviewed_yes+taxonomy_2759.fasta.gz</code></pre>
<h4 id="prediction-of-protein-coding-sequences-with-sencha-translate">Prediction of protein-coding sequences with <code>sencha translate</code></h4>
<p>We then predicted protein coding reads using the created bloom filter using <code>sencha translate</code>.</p>
<pre><code>sencha translate \
  --molecule protein \
  --coding-nucleotide-fasta SRR306800_GSM752653_ggo_br_F_1__coding_reads_nucleotides.fasta \
  --csv SRR306800_GSM752653_ggo_br_F_1__coding_scores.csv \
  --json-summary SRR306800_GSM752653_ggo_br_F_1__coding_summary.json \
  --jaccard-threshold 0.8 \
  --peptides-are-index \
  uniprot-reviewed_yes+taxonomy_2759__molecule-protein_ksize-7.bloomfilter \
  SRR306800_GSM752653_ggo_br_F_1_trimmed.fq.gz &gt; SRR306800_GSM752653_ggo_br_F_1__coding_reads_peptides.fasta</code></pre>
<figure>
<img src="images/SVG/supplemental_figure2.svg" id="sfig2:sfig2" data-tag="supplemental_figure2" style="width:100.0%" alt="" /><figcaption>Applications of <code>sencha translate</code>. <strong>A.</strong> We simulated RNA-seq data using Opisthokonta species from the Quest for Orthologs dataset for true positive protein-coding RNAs, reads completely contained within intergenic, intronic, and UTR sequences as true positive noncoding RNAs, and reads partially overlapping a coding and noncoding region as an adversarial test set. We then predicted protein-coding sequences and computed false positive and false negative rates. False Positive coding reads were found to be … False negative noncoding reads were found to be … <strong>B.</strong> Number of putative protein-coding sequences per read. <strong>C.</strong> This method could also be used to extract only reads whose putative protein-coding sequences are transcription factors. <strong>D.</strong> We ran <code>sencha translate</code> on the five tissues and nine species from the Brawand 2011 dataset.</figcaption>
</figure>
<!-- ### `sencha compare-kmer-content` is a simple method to identify homologs

![Applications of `sencha compare-kmer-content`. **A.** We used `sencha compare-kmer-content` on pairs of orthologous protein sequences between humans and the remaining Opisthokonta species in the Quest for Orthologs dataset. x-axis, $k$-mer size, y-axis, mean difference. **B.** False positive calls by `sencha compare-kmer-content` are either paralogs or read-through protein products. **C.** We applied `sencha compare-kmer-content` to ... to find putative orthologs. We found ... the accuracy was ...](images/SVG/figure4.svg){#sfig:figure4 tag="figure4" width="100%"} -->
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references hanging-indent" role="doc-bibliography">
<div id="ref-OkFljXKC">
<p>1. <strong>How Many Species Are There on Earth and in the Ocean?</strong> <br />
Camilo Mora, Derek P. Tittensor, Sina Adl, Alastair G. B. Simpson, Boris Worm<br />
<em>PLoS Biology</em> (2011-08-23) <a href="https://doi.org/fpr4z8">https://doi.org/fpr4z8</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pbio.1001127">10.1371/journal.pbio.1001127</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21886479">21886479</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3160336">PMC3160336</a></p>
</div>
<div id="ref-xJN8WB5G">
<p>2. <strong>Genome List - Genome - NCBI</strong> <a href="https://www.ncbi.nlm.nih.gov/genome/browse">https://www.ncbi.nlm.nih.gov/genome/browse</a></p>
</div>
<div id="ref-nI2xGuLe">
<p>3. <strong>Cross-Species Analysis of Single-Cell Transcriptomic Data</strong> <br />
Maxwell E. R. Shafer<br />
<em>Frontiers in Cell and Developmental Biology</em> (2019-09-02) <a href="https://doi.org/gg2rpw">https://doi.org/gg2rpw</a> <br />
DOI: <a href="https://doi.org/10.3389/fcell.2019.00175">10.3389/fcell.2019.00175</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31552245">31552245</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6743501">PMC6743501</a></p>
</div>
<div id="ref-1C9RVEJCr">
<p>4. <strong>Telomere-to-telomere assembly of a complete human X chromosome</strong> <br />
Karen H. Miga, Sergey Koren, Arang Rhie, Mitchell R. Vollger, Ariel Gershman, Andrey Bzikadze, Shelise Brooks, Edmund Howe, David Porubsky, Glennis A. Logsdon, … Adam M. Phillippy<br />
<em>Nature</em> (2020-07-14) <a href="https://doi.org/gg4vw8">https://doi.org/gg4vw8</a> <br />
DOI: <a href="https://doi.org/10.1038/s41586-020-2547-7">10.1038/s41586-020-2547-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32663838">32663838</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7484160">PMC7484160</a></p>
</div>
<div id="ref-N285yCi2">
<p>5. <strong>Towards complete and error-free genome assemblies of all vertebrate species</strong> <br />
Arang Rhie, Shane A. McCarthy, Olivier Fedrigo, Joana Damas, Giulio Formenti, Sergey Koren, Marcela Uliano-Silva, William Chow, Arkarachai Fungtammasan, Gregory L. Gedman, … Erich D. Jarvis<br />
<em>bioRxiv</em> (2020-05-23) <a href="https://www.biorxiv.org/content/10.1101/2020.05.22.110833v1">https://www.biorxiv.org/content/10.1101/2020.05.22.110833v1</a> <br />
DOI: <a href="https://doi.org/10.1101/2020.05.22.110833">10.1101/2020.05.22.110833</a></p>
</div>
<div id="ref-1BMDdrblS">
<p>6. <strong>Complete vertebrate mitogenomes reveal widespread gene duplications and repeats</strong> <br />
Giulio Formenti, Arang Rhie, Jennifer Balacco, Bettina Haase, Jacquelyn Mountcastle, Olivier Fedrigo, Samara Brown, Marco Capodiferro, Farooq O. Al-Ajli, Roberto Ambrosini, … The Vertebrate Genomes Project Consortium<br />
<em>bioRxiv</em> (2020-07-01) <a href="https://www.biorxiv.org/content/10.1101/2020.06.30.177956v2">https://www.biorxiv.org/content/10.1101/2020.06.30.177956v2</a> <br />
DOI: <a href="https://doi.org/10.1101/2020.06.30.177956">10.1101/2020.06.30.177956</a></p>
</div>
<div id="ref-LkrS4xXB">
<p>7. <strong>Standardized benchmarking in the quest for orthologs</strong> <br />
Adrian M Altenhoff, Brigitte Boeckmann, Salvador Capella-Gutierrez, Daniel A Dalquen, Todd DeLuca, Kristoffer Forslund, Jaime Huerta-Cepas, Benjamin Linard, Cécile Pereira, Leszek P Pryszcz, … Quest for Orthologs consortium<br />
<em>Nature Methods</em> (2016-04-04) <a href="https://doi.org/f3rpzx">https://doi.org/f3rpzx</a> <br />
DOI: <a href="https://doi.org/10.1038/nmeth.3830">10.1038/nmeth.3830</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27043882">27043882</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4827703">PMC4827703</a></p>
</div>
<div id="ref-dX45ePUc">
<p>8. <strong>The Quest for Orthologs benchmark service and consensus calls in 2020</strong> <br />
Adrian M Altenhoff, Javier Garrayo-Ventas, Salvatore Cosentino, David Emms, Natasha M Glover, Ana Hernández-Plaza, Yannis Nevers, Vicky Sundesha, Damian Szklarczyk, José M Fernández, … Christophe Dessimoz<br />
<em>Nucleic Acids Research</em> (2020-07-02) <a href="https://doi.org/ghhwqp">https://doi.org/ghhwqp</a> <br />
DOI: <a href="https://doi.org/10.1093/nar/gkaa308">10.1093/nar/gkaa308</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32374845">32374845</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7319555">PMC7319555</a></p>
</div>
<div id="ref-19AeqN3SK">
<p>9. <strong>Big data and other challenges in the quest for orthologs</strong> <br />
E. L. L. Sonnhammer, T. Gabaldon, A. W. Sousa da Silva, M. Martin, M. Robinson-Rechavi, B. Boeckmann, P. D. Thomas, C. Dessimoz, the Quest for Orthologs consortium<br />
<em>Bioinformatics</em> (2014-07-26) <a href="https://doi.org/f6ntvb">https://doi.org/f6ntvb</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btu492">10.1093/bioinformatics/btu492</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25064571">25064571</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4201156">PMC4201156</a></p>
</div>
<div id="ref-1AbwFZgjU">
<p>10. <strong>New Tools in Orthology Analysis: A Brief Review of Promising Perspectives</strong> <br />
Bruno T. L. Nichio, Jeroniza Nunes Marchaukoski, Roberto Tadeu Raittz<br />
<em>Frontiers in Genetics</em> (2017-10-31) <a href="https://doi.org/gcknjp">https://doi.org/gcknjp</a> <br />
DOI: <a href="https://doi.org/10.3389/fgene.2017.00165">10.3389/fgene.2017.00165</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29163633">29163633</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5674930">PMC5674930</a></p>
</div>
<div id="ref-EvDQzoLf">
<p>11. <strong>OrthoFinder: solving fundamental biases in whole genome comparisons dramatically improves orthogroup inference accuracy</strong> <br />
David M. Emms, Steven Kelly<br />
<em>Genome Biology</em> (2015-08-06) <a href="https://doi.org/gfznm8">https://doi.org/gfznm8</a> <br />
DOI: <a href="https://doi.org/10.1186/s13059-015-0721-2">10.1186/s13059-015-0721-2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26243257">26243257</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4531804">PMC4531804</a></p>
</div>
<div id="ref-mtOZyMZw">
<p>12. <strong>Organoid single-cell genomic atlas uncovers human-specific features of brain development</strong> <br />
Sabina Kanton, Michael James Boyle, Zhisong He, Malgorzata Santel, Anne Weigert, Fátima Sanchís-Calleja, Patricia Guijarro, Leila Sidow, Jonas Simon Fleck, Dingding Han, … J. Gray Camp<br />
<em>Nature</em> (2019-10-16) <a href="https://doi.org/dcws">https://doi.org/dcws</a> <br />
DOI: <a href="https://doi.org/10.1038/s41586-019-1654-9">10.1038/s41586-019-1654-9</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31619793">31619793</a></p>
</div>
<div id="ref-ogAGO9KH">
<p>13. <strong>The origin and evolution of cell types</strong> <br />
Detlev Arendt, Jacob M. Musser, Clare V. H. Baker, Aviv Bergman, Connie Cepko, Douglas H. Erwin, Mihaela Pavlicev, Gerhard Schlosser, Stefanie Widder, Manfred D. Laubichler, Günter P. Wagner<br />
<em>Nature Reviews Genetics</em> (2016-11-07) <a href="https://doi.org/f9b62x">https://doi.org/f9b62x</a> <br />
DOI: <a href="https://doi.org/10.1038/nrg.2016.127">10.1038/nrg.2016.127</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27818507">27818507</a></p>
</div>
<div id="ref-wdzgXUGy">
<p>14. <strong>How Single-Cell Genomics Is Changing Evolutionary and Developmental Biology</strong> <br />
John C. Marioni, Detlev Arendt<br />
<em>Annual Review of Cell and Developmental Biology</em> (2017-10-06) <a href="https://doi.org/ggb632">https://doi.org/ggb632</a> <br />
DOI: <a href="https://doi.org/10.1146/annurev-cellbio-100616-060818">10.1146/annurev-cellbio-100616-060818</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28813177">28813177</a></p>
</div>
<div id="ref-oDiClyTE">
<p>15. <strong>Ensembl 2020</strong> <br />
Andrew D Yates, Premanand Achuthan, Wasiu Akanni, James Allen, Jamie Allen, Jorge Alvarez-Jarreta, M Ridwan Amode, Irina M Armean, Andrey G Azov, Ruth Bennett, … Paul Flicek<br />
<em>Nucleic Acids Research</em> (2019-11-06) <a href="https://doi.org/ggqp72">https://doi.org/ggqp72</a> <br />
DOI: <a href="https://doi.org/10.1093/nar/gkz966">10.1093/nar/gkz966</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31691826">31691826</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7145704">PMC7145704</a></p>
</div>
<div id="ref-1CZYa2MlR">
<p>16. <strong>eggNOG 4.5: a hierarchical orthology framework with improved functional annotations for eukaryotic, prokaryotic and viral sequences</strong> <br />
Jaime Huerta-Cepas, Damian Szklarczyk, Kristoffer Forslund, Helen Cook, Davide Heller, Mathias C. Walter, Thomas Rattei, Daniel R. Mende, Shinichi Sunagawa, Michael Kuhn, … Peer Bork<br />
<em>Nucleic Acids Research</em> (2016-01-04) <a href="https://doi.org/f8cmvq">https://doi.org/f8cmvq</a> <br />
DOI: <a href="https://doi.org/10.1093/nar/gkv1248">10.1093/nar/gkv1248</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26582926">26582926</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4702882">PMC4702882</a></p>
</div>
<div id="ref-eRqRW5iR">
<p>17. <strong>Fast Genome-Wide Functional Annotation through Orthology Assignment by eggNOG-Mapper</strong> <br />
Jaime Huerta-Cepas, Kristoffer Forslund, Luis Pedro Coelho, Damian Szklarczyk, Lars Juhl Jensen, Christian von Mering, Peer Bork<br />
<em>Molecular Biology and Evolution</em> (2017-08) <a href="https://doi.org/ggbwwj">https://doi.org/ggbwwj</a> <br />
DOI: <a href="https://doi.org/10.1093/molbev/msx148">10.1093/molbev/msx148</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28460117">28460117</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5850834">PMC5850834</a></p>
</div>
<div id="ref-wtk13QGK">
<p>18. <strong>K-mer counting with low memory consumption enables fast clustering of single-cell sequencing data without read alignment</strong> <br />
Christina Huan Shi, Kevin Y. Yip<br />
<em>bioRxiv</em> (2019-08-02) <a href="https://www.biorxiv.org/content/10.1101/723833v1">https://www.biorxiv.org/content/10.1101/723833v1</a> <br />
DOI: <a href="https://doi.org/10.1101/723833">10.1101/723833</a></p>
</div>
<div id="ref-mH9pzoIn">
<p>19. <strong>Mash: fast genome and metagenome distance estimation using MinHash</strong> <br />
Brian D. Ondov, Todd J. Treangen, Páll Melsted, Adam B. Mallonee, Nicholas H. Bergman, Sergey Koren, Adam M. Phillippy<br />
<em>Genome Biology</em> (2016-06-20) <a href="https://doi.org/gfx74q">https://doi.org/gfx74q</a> <br />
DOI: <a href="https://doi.org/10.1186/s13059-016-0997-x">10.1186/s13059-016-0997-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27323842">27323842</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4915045">PMC4915045</a></p>
</div>
<div id="ref-ExvjTLL3">
<p>20. <strong>sourmash: a library for MinHash sketching of DNA</strong> <br />
C. Titus Brown, Luiz Irber<br />
<em>The Journal of Open Source Software</em> (2016-09-14) <a href="https://doi.org/ghdrk5">https://doi.org/ghdrk5</a> <br />
DOI: <a href="https://doi.org/10.21105/joss.00027">10.21105/joss.00027</a></p>
</div>
<div id="ref-ART4gJKs">
<p>21. <strong>Large-scale sequence comparisons with sourmash</strong> <br />
N. Tessa Pierce, Luiz Irber, Taylor Reiter, Phillip Brooks, C. Titus Brown<br />
<em>F1000Research</em> (2019-07-04) <a href="https://doi.org/gf9v84">https://doi.org/gf9v84</a> <br />
DOI: <a href="https://doi.org/10.12688/f1000research.19675.1">10.12688/f1000research.19675.1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31508216">31508216</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6720031">PMC6720031</a></p>
</div>
<div id="ref-QvbXRDjv">
<p>22. <strong>Bioinformatics algorithms: an active learning approach</strong> <br />
Phillip Compeau<br />
<em>Active Learning Publishers</em> (2018) <br />
ISBN: <a href="https://worldcat.org/isbn/9780990374633">9780990374633</a></p>
</div>
<div id="ref-g2gs9fUL">
<p>23. <strong>ABSENT SEQUENCES: NULLOMERS AND PRIMES</strong> <br />
GREG HAMPIKIAN, TIM ANDERSEN<br />
<em>World Scientific Pub Co Pte Lt</em> (2006-12) <a href="https://doi.org/cpxn43">https://doi.org/cpxn43</a> <br />
DOI: <a href="https://doi.org/10.1142/9789812772435_0034">10.1142/9789812772435_0034</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17990505">17990505</a></p>
</div>
<div id="ref-IpiEdhou">
<p>24. <strong>Nullomers and High Order Nullomers in Genomic Sequences</strong> <br />
Davide Vergni, Daniele Santoni<br />
<em>PLOS ONE</em> (2016-12-01) <a href="https://doi.org/f9jpwv">https://doi.org/f9jpwv</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0164540">10.1371/journal.pone.0164540</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27906971">27906971</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5132333">PMC5132333</a></p>
</div>
<div id="ref-hMQyspqZ">
<p>25. <strong>Significant non-existence of sequences in genomes and proteomes</strong> <br />
Grigorios Koulouras, Martin C. Frith<br />
<em>bioRxiv</em> (2020-06-26) <a href="https://www.biorxiv.org/content/10.1101/2020.06.25.170431v1">https://www.biorxiv.org/content/10.1101/2020.06.25.170431v1</a> <br />
DOI: <a href="https://doi.org/10.1101/2020.06.25.170431">10.1101/2020.06.25.170431</a></p>
</div>
<div id="ref-9kcdMNpl">
<p>26. <strong>Absent from DNA and protein: genomic characterization of nullomers and nullpeptides across functional categories and evolution</strong> <br />
Ilias Georgakopoulos-Soares, Ofer Yizhar Barnea, Ioannis Mouratidis, Martin Hemberg, Nadav Ahituv<br />
<em>bioRxiv</em> (2020-03-03) <a href="https://www.biorxiv.org/content/10.1101/2020.03.02.972422v1">https://www.biorxiv.org/content/10.1101/2020.03.02.972422v1</a> <br />
DOI: <a href="https://doi.org/10.1101/2020.03.02.972422">10.1101/2020.03.02.972422</a></p>
</div>
<div id="ref-17wsdE0b1">
<p>27. <strong>Nullomers: Really a Matter of Natural Selection?</strong> <br />
Claudia Acquisti, George Poste, David Curtiss, Sudhir Kumar<br />
<em>PLoS ONE</em> (2007-10-10) <a href="https://doi.org/dx9qbn">https://doi.org/dx9qbn</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0001022">10.1371/journal.pone.0001022</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/17925870">17925870</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1995752">PMC1995752</a></p>
</div>
<div id="ref-Rqlso7kk">
<p>28. <strong>SwiftOrtho: A fast, memory-efficient, multiple genome orthology classifier</strong> <br />
Xiao Hu, Iddo Friedberg<br />
<em>GigaScience</em> (2019-10) <a href="https://doi.org/ggcr5x">https://doi.org/ggcr5x</a> <br />
DOI: <a href="https://doi.org/10.1093/gigascience/giz118">10.1093/gigascience/giz118</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31648300">31648300</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6812468">PMC6812468</a></p>
</div>
<div id="ref-sZoryWIB">
<p>29. <strong>RAPSearch: a fast protein similarity search tool for short reads</strong> <br />
Yuzhen Ye, Jeong-Hyeon Choi, Haixu Tang<br />
<em>BMC Bioinformatics</em> (2011-05-15) <a href="https://doi.org/dgt5rt">https://doi.org/dgt5rt</a> <br />
DOI: <a href="https://doi.org/10.1186/1471-2105-12-159">10.1186/1471-2105-12-159</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/21575167">21575167</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3113943">PMC3113943</a></p>
</div>
<div id="ref-Aq7zYbOg">
<p>30. <strong>Reduced amino acid alphabets exhibit an improved sensitivity and selectivity in fold assignment</strong> <br />
Eric L. Peterson, Jané Kondev, Julie A. Theriot, Rob Phillips<br />
<em>Bioinformatics</em> (2009-06-01) <a href="https://doi.org/btqmnp">https://doi.org/btqmnp</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btp164">10.1093/bioinformatics/btp164</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19351620">19351620</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2732308">PMC2732308</a></p>
</div>
<div id="ref-16Kv6RIe3">
<p>31. <strong>Fast databank searching with a reduced amino-acid alphabet</strong> <br />
Claudine Landès, Jean-Loup Risler<br />
<em>Bioinformatics</em> (1994) <a href="https://doi.org/cvrjmw">https://doi.org/cvrjmw</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/10.4.453">10.1093/bioinformatics/10.4.453</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/7804879">7804879</a></p>
</div>
<div id="ref-f5u3dDtc">
<p>32. <strong>Simplified amino acid alphabets for protein fold recognition and implications for folding</strong> <br />
Lynne Reed Murphy, Anders Wallqvist, Ronald M. Levy<br />
<em>Protein Engineering, Design and Selection</em> (2000-03) <a href="https://doi.org/bdtngh">https://doi.org/bdtngh</a> <br />
DOI: <a href="https://doi.org/10.1093/protein/13.3.149">10.1093/protein/13.3.149</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/10775656">10775656</a></p>
</div>
<div id="ref-QMYjnr6l">
<p>33. <strong>Local homology recognition and distance measures in linear time using compressed amino acid alphabets</strong> <br />
R. C. Edgar<br />
<em>Nucleic Acids Research</em> (2004-01-02) <a href="https://doi.org/ckg5d4">https://doi.org/ckg5d4</a> <br />
DOI: <a href="https://doi.org/10.1093/nar/gkh180">10.1093/nar/gkh180</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/14729922">14729922</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC373290">PMC373290</a></p>
</div>
<div id="ref-S53q1T30">
<p>34. <strong>BUSCO: assessing genome assembly and annotation completeness with single-copy orthologs</strong> <br />
Felipe A. Simão, Robert M. Waterhouse, Panagiotis Ioannidis, Evgenia V. Kriventseva, Evgeny M. Zdobnov<br />
<em>Bioinformatics</em> (2015-10-01) <a href="https://doi.org/gfznpw">https://doi.org/gfznpw</a> <br />
DOI: <a href="https://doi.org/10.1093/bioinformatics/btv351">10.1093/bioinformatics/btv351</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26059717">26059717</a></p>
</div>
<div id="ref-1A6CqUMhm">
<p>35. <strong>TimeTree :: The Timescale of Life</strong> <a href="http://timetree.org/">http://timetree.org/</a></p>
</div>
<div id="ref-rSyxV4O8">
<p>36. <strong>A single-cell transcriptomic atlas characterizes ageing tissues in the mouse</strong> <br />
The Tabula Muris Consortium<br />
<em>Nature</em> (2020-07-15) <a href="https://doi.org/gg426c">https://doi.org/gg426c</a> <br />
DOI: <a href="https://doi.org/10.1038/s41586-020-2496-1">10.1038/s41586-020-2496-1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32669714">32669714</a></p>
</div>
<div id="ref-1AkeHrvKP">
<p>37. <strong>Single-cell transcriptional atlas of the Chinese horseshoe bat (Rhinolophus sinicus) provides insight into the cellular mechanisms which enable bats to be viral reservoirs</strong> <br />
Lili Ren, Chao Wu, Li Guo, Jiacheng Yao, Conghui Wang, Yan Xiao, Angela Oliveira Pisco, Zhiqiang Wu, Xiaobo Lei, Yiwei Liu, … Jianwei Wang<br />
<em>bioRxiv</em> (2020-06-30) <a href="https://www.biorxiv.org/content/10.1101/2020.06.30.175778v1">https://www.biorxiv.org/content/10.1101/2020.06.30.175778v1</a> <br />
DOI: <a href="https://doi.org/10.1101/2020.06.30.175778">10.1101/2020.06.30.175778</a></p>
</div>
<div id="ref-RIPzCufe">
<p>38. <strong>Sailfish enables alignment-free isoform quantification from RNA-seq reads using lightweight algorithms</strong> <br />
Rob Patro, Stephen M Mount, Carl Kingsford<br />
<em>Nature Biotechnology</em> (2014-04-20) <a href="https://doi.org/gfghc2">https://doi.org/gfghc2</a> <br />
DOI: <a href="https://doi.org/10.1038/nbt.2862">10.1038/nbt.2862</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24752080">24752080</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4077321">PMC4077321</a></p>
</div>
<div id="ref-12KZMHMQl">
<p>39. <strong>Near-optimal probabilistic RNA-seq quantification</strong> <br />
Nicolas L Bray, Harold Pimentel, Páll Melsted, Lior Pachter<br />
<em>Nature Biotechnology</em> (2016-04-04) <a href="https://doi.org/f8nvsp">https://doi.org/f8nvsp</a> <br />
DOI: <a href="https://doi.org/10.1038/nbt.3519">10.1038/nbt.3519</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27043002">27043002</a></p>
</div>
<div id="ref-vrqQcFyx">
<p>40. <strong>Salmon provides fast and bias-aware quantification of transcript expression</strong> <br />
Rob Patro, Geet Duggal, Michael I Love, Rafael A Irizarry, Carl Kingsford<br />
<em>Nature Methods</em> (2017-03-06) <a href="https://doi.org/gcw9f5">https://doi.org/gcw9f5</a> <br />
DOI: <a href="https://doi.org/10.1038/nmeth.4197">10.1038/nmeth.4197</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28263959">28263959</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5600148">PMC5600148</a></p>
</div>
<div id="ref-DfZvKOpG">
<p>41. <strong>Fast search of thousands of short-read sequencing experiments</strong> <br />
Brad Solomon, Carl Kingsford<br />
<em>Nature Biotechnology</em> (2016-02-08) <a href="https://doi.org/f8ddk3">https://doi.org/f8ddk3</a> <br />
DOI: <a href="https://doi.org/10.1038/nbt.3442">10.1038/nbt.3442</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26854477">26854477</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4804353">PMC4804353</a></p>
</div>
<div id="ref-1HpdDzfQX">
<p>42. <strong>Improved Search of Large Transcriptomic Sequencing Databases Using Split Sequence Bloom Trees</strong> <br />
Brad Solomon, Carl Kingsford<br />
<em>Journal of Computational Biology</em> (2018-07) <a href="https://doi.org/gdxhgz">https://doi.org/gdxhgz</a> <br />
DOI: <a href="https://doi.org/10.1089/cmb.2017.0265">10.1089/cmb.2017.0265</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29641248">29641248</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6067102">PMC6067102</a></p>
</div>
<div id="ref-AfFgGSHX">
<p>43. <strong>AllSome Sequence Bloom Trees</strong> <br />
Chen Sun, Robert S. Harris, Rayan Chikhi, Paul Medvedev<br />
<em>Journal of Computational Biology</em> (2018-05) <a href="https://doi.org/gdkn3v">https://doi.org/gdkn3v</a> <br />
DOI: <a href="https://doi.org/10.1089/cmb.2017.0258">10.1089/cmb.2017.0258</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29620920">29620920</a></p>
</div>
<div id="ref-UCkAlFcx">
<p>44. <strong>IgTM: An algorithm to predict transmembrane domains and topology in proteins</strong> <br />
Piedachu Peris, Damián López, Marcelino Campos<br />
<em>BMC Bioinformatics</em> (2008-09-10) <a href="https://doi.org/dmcb9p">https://doi.org/dmcb9p</a> <br />
DOI: <a href="https://doi.org/10.1186/1471-2105-9-367">10.1186/1471-2105-9-367</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/18783592">18783592</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2566583">PMC2566583</a></p>
</div>
<div id="ref-1VJMkE5Z">
<p>45. <strong>BUSCO: Assessing Genome Assembly and Annotation Completeness</strong> <br />
Mathieu Seppey, Mosè Manni, Evgeny M. Zdobnov<br />
<em>Methods in Molecular Biology</em> (2019) <a href="https://doi.org/gg7zjc">https://doi.org/gg7zjc</a> <br />
DOI: <a href="https://doi.org/10.1007/978-1-4939-9173-0_14">10.1007/978-1-4939-9173-0_14</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31020564">31020564</a></p>
</div>
<div id="ref-W97RNFXd">
<p>46. <strong>TimeTree :: The Timescale of Life</strong> <a href="http://timetree.org/">http://timetree.org/</a></p>
</div>
<div id="ref-9DeI91qq">
<p>47. <strong>The effects of death and post-mortem cold ischemia on human tissue transcriptomes</strong> <br />
Pedro G. Ferreira, Manuel Muñoz-Aguirre, Ferran Reverter, Caio P. Sá Godinho, Abel Sousa, Alicia Amadoz, Reza Sodaei, Marta R. Hidalgo, Dmitri Pervouchine, Jose Carbonell-Caballero, … Roderic Guigó<br />
<em>Nature Communications</em> (2018-02-13) <a href="https://doi.org/gc2xd8">https://doi.org/gc2xd8</a> <br />
DOI: <a href="https://doi.org/10.1038/s41467-017-02772-x">10.1038/s41467-017-02772-x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29440659">29440659</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5811508">PMC5811508</a></p>
</div>
<div id="ref-IVhllwbX">
<p>48. <strong>Transcriptional Profiling of Quiescent Muscle Stem Cells In Vivo</strong> <br />
Cindy T. J. van Velthoven, Antoine de Morree, Ingrid M. Egner, Jamie O. Brett, Thomas A. Rando<br />
<em>Cell Reports</em> (2017-11) <a href="https://doi.org/gdxwmg">https://doi.org/gdxwmg</a> <br />
DOI: <a href="https://doi.org/10.1016/j.celrep.2017.10.037">10.1016/j.celrep.2017.10.037</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/29141228">29141228</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5711481">PMC5711481</a></p>
</div>
<div id="ref-5nByhjEz">
<p>49. <strong>Single-cell sequencing reveals dissociation-induced gene expression in tissue subpopulations</strong> <br />
Susanne C van den Brink, Fanny Sage, Ábel Vértesy, Bastiaan Spanjaard, Josi Peterson-Maduro, Chloé S Baron, Catherine Robin, Alexander van Oudenaarden<br />
<em>Nature Methods</em> (2017-09-29) <a href="https://doi.org/gf3hwq">https://doi.org/gf3hwq</a> <br />
DOI: <a href="https://doi.org/10.1038/nmeth.4437">10.1038/nmeth.4437</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28960196">28960196</a></p>
</div>
<div id="ref-qtN0W5fQ">
<p>50. <strong>Sampling time-dependent artifacts in single-cell genomics studies</strong> <br />
Ramon Massoni-Badosa, Giovanni Iacono, Catia Moutinho, Marta Kulis, Núria Palau, Domenica Marchese, Javier Rodríguez-Ubreva, Esteban Ballestar, Gustavo Rodriguez-Esteban, Sara Marsal, … Holger Heyn<br />
<em>Genome Biology</em> (2020-05-11) <a href="https://doi.org/ggzqfj">https://doi.org/ggzqfj</a> <br />
DOI: <a href="https://doi.org/10.1186/s13059-020-02032-0">10.1186/s13059-020-02032-0</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32393363">32393363</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7212672">PMC7212672</a></p>
</div>
<div id="ref-s4hwSsAX">
<p>51. <strong>Streamlining Data-Intensive Biology With Workflow Systems</strong> <br />
Taylor Reiter, Phillip T. Brooks, Luiz Irber, Shannon E. K. Joslin, Charles M. Reid, Camille Scott, C. Titus Brown, N. Tessa Pierce<br />
<em>bioRxiv</em> (2020-11-16) <a href="https://www.biorxiv.org/content/10.1101/2020.06.30.178673v3">https://www.biorxiv.org/content/10.1101/2020.06.30.178673v3</a> <br />
DOI: <a href="https://doi.org/10.1101/2020.06.30.178673">10.1101/2020.06.30.178673</a></p>
</div>
<div id="ref-oWw3q9FU">
<p>52. <strong>A molecular cell atlas of the human lung from single-cell RNA sequencing</strong> <br />
Kyle J. Travaglini, Ahmad N. Nabhan, Lolita Penland, Rahul Sinha, Astrid Gillich, Rene V. Sit, Stephen Chang, Stephanie D. Conley, Yasuo Mori, Jun Seita, … Mark A. Krasnow<br />
<em>Nature</em> (2020-11-18) <a href="https://doi.org/ghk5bp">https://doi.org/ghk5bp</a> <br />
DOI: <a href="https://doi.org/10.1038/s41586-020-2922-4">10.1038/s41586-020-2922-4</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33208946">33208946</a></p>
</div>
<div id="ref-5KVVBsT6">
<p>53. <strong>dib-lab/2020-paper-sourmash-gather</strong> <br />
The Lab for Data Intensive Biology<br />
(2020-12-02) <a href="https://github.com/dib-lab/2020-paper-sourmash-gather">https://github.com/dib-lab/2020-paper-sourmash-gather</a></p>
</div>
<div id="ref-66KDHtpG">
<p>54. <strong>Atlas of protein sequence and structure</strong> <br />
Margaret O Dayhoff<br />
<em>National Biomedical Research Foundation.</em> (1969)</p>
</div>
<div id="ref-1GiMcbPLs">
<p>55. <strong>Physical biology of the cell</strong> <br />
Rob Phillips, Julie Theriot, Jane Kondev, Hernan Garcia<br />
<em>Garland Science</em> (2012)</p>
</div>
<div id="ref-1DpHWPZC9">
<p>56. <strong>These Are Not the K-mers You Are Looking For: Efficient Online K-mer Counting Using a Probabilistic Data Structure</strong> <br />
Qingpeng Zhang, Jason Pell, Rosangela Canino-Koning, Adina Chuang Howe, C. Titus Brown<br />
<em>PLoS ONE</em> (2014-07-25) <a href="https://doi.org/f6kb9b">https://doi.org/f6kb9b</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pone.0101271">10.1371/journal.pone.0101271</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/25062443">25062443</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4111482">PMC4111482</a></p>
</div>
<div id="ref-46R7aU4N">
<p>57. <strong>OrthoFinder: phylogenetic orthology inference for comparative genomics</strong> <br />
David M. Emms, Steven Kelly<br />
<em>bioRxiv</em> (2019-04-24) <a href="https://www.biorxiv.org/content/10.1101/466201v2">https://www.biorxiv.org/content/10.1101/466201v2</a> <br />
DOI: <a href="https://doi.org/10.1101/466201">10.1101/466201</a></p>
</div>
<div id="ref-fBCZ87Cf">
<p>58. <strong>The evolution of gene expression levels in mammalian organs</strong> <br />
David Brawand, Magali Soumillon, Anamaria Necsulea, Philippe Julien, Gábor Csárdi, Patrick Harrigan, Manuela Weier, Angélica Liechti, Ayinuer Aximu-Petri, Martin Kircher, … Henrik Kaessmann<br />
<em>Nature</em> (2011-10-19) <a href="https://doi.org/fcvk54">https://doi.org/fcvk54</a> <br />
DOI: <a href="https://doi.org/10.1038/nature10532">10.1038/nature10532</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22012392">22012392</a></p>
</div>
<div id="ref-G2CzfHHF">
<p>59. <strong>The khmer software package: enabling efficient nucleotide sequence analysis</strong> <br />
Michael R. Crusoe, Hussien F. Alameldin, Sherine Awad, Elmar Boucher, Adam Caldwell, Reed Cartwright, Amanda Charbonneau, Bede Constantinides, Greg Edvenson, Scott Fay, … C. Titus Brown<br />
<em>F1000Research</em> (2015-09-25) <a href="https://doi.org/9qp">https://doi.org/9qp</a> <br />
DOI: <a href="https://doi.org/10.12688/f1000research.6924.1">10.12688/f1000research.6924.1</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26535114">26535114</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4608353">PMC4608353</a></p>
</div>
<div id="ref-1ZE22clL">
<p>60. <strong>UniProt: a worldwide hub of protein knowledge</strong> <br />
The UniProt Consortium<br />
<em>Nucleic Acids Research</em> (2019-01-08) <a href="https://doi.org/gfwqck">https://doi.org/gfwqck</a> <br />
DOI: <a href="https://doi.org/10.1093/nar/gky1049">10.1093/nar/gky1049</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/30395287">30395287</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6323992">PMC6323992</a></p>
</div>
<div id="ref-1BBNQZrBs">
<p>61. <strong>UniProtKB/Swiss-Prot, the Manually Annotated Section of the UniProt KnowledgeBase: How to Use the Entry View</strong> <br />
Emmanuel Boutet, Damien Lieberherr, Michael Tognolli, Michel Schneider, Parit Bansal, Alan J. Bridge, Sylvain Poux, Lydie Bougueleret, Ioannis Xenarios<br />
<em>Methods in Molecular Biology</em> (2016) <a href="https://doi.org/f79kbb">https://doi.org/f79kbb</a> <br />
DOI: <a href="https://doi.org/10.1007/978-1-4939-3167-5_2">10.1007/978-1-4939-3167-5_2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26519399">26519399</a></p>
</div>
<div id="ref-Ty6517L8">
<p>62. <strong>Opisthokont</strong> <br />
Wikipedia<br />
(2020-11-27) <a href="https://en.wikipedia.org/w/index.php?title=Opisthokont&amp;oldid=990958852">https://en.wikipedia.org/w/index.php?title=Opisthokont&amp;oldid=990958852</a></p>
</div>
</div>
<!--

## Outline

- Kmers can approximate orthologies
  - Jaccard similarity of orthologues is higher than non-orthologues
  - Benchmarking using https://orthology.benchmarkservice.org/cgi-bin/gateway.pl
  - Finding orthologues
    - Gold standard
      - ENSEMBL COMPARA
      - Quest for Orthologs consortium, Altenhoff, A. M., Boeckmann, B., Capella-Gutierrez, S., Dalquen, D. A., DeLuca, T., et al. (2016). Standardized benchmarking in the quest for orthologs. Nature Methods, 13(5), 425–430. http://doi.org/10.1038/nmeth.3830 [@doi:10.1038/nmeth.3830]
    - Orthologous groups/Conserved Domain Database [@url:https://www.ebi.ac.uk/miriam/main/collections/MIR:00000119]


<!-- ## Figure 2 -- $k$-mers from lossily-encoded putative protein-coding reads faithfully pull out reads from protein-coding genes within amniotes -->
<!-- - Overview of kmermaid pipeline
  - Comparison of tissue across species
    - Partition reads to coding/noncoding bins
    - MinHash the Dayhoff-encoded coding sequences
    - Jaccard similarity on the MinHashes
- Which reads are found to have coding features but didn’t map to the genome?
- Do these features map to novel genes or gene fusions?
- Kmers can find correct reading from of RNA-seq reads
  - Human peptides → human, chimp, bonobo, orangutan, gorilla, macaque, mouse, opossum, playtpus, chicken RNAseq from Brawand2011 data
- Comparison to other methods: RNASamba [@doi:10.1101/620880] -->
<!-- ### Figure 3 -- $k$-mers can pull out only reads from transcription factors and Amniotes can be compared on the MinHashes of their protein-coding sequences -->
<!-- ![Figure 3.](images/figure3.svg){#fig:fig3 width="100%"} -->
<!-- ### Some potential references

Gene expression evolution through duplications

- Farre, D., & Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 [@doi:10.1093/molbev/msp242]
- Thornton, J. W., & DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 [@doi:10.1146/annurev.genom.1.1.41]
- Farre, D., & Alba, M. M. (2010). Heterogeneous Patterns of Gene-Expression Diversification in Mammalian Gene Duplicates. Molecular Biology and Evolution, 27(2), 325–335. http://doi.org/10.1093/molbev/msp242 [@doi:10.1093/molbev/msp242]

Taxa-restricted genes

- Human-specific genes in fetal neocortex
Florio, M., Heide, M., Pinson, A., Brandl, H., Albert, M., Winkler, S., et al. (2018). Evolution and cell-type specificity of human-specific genes preferentially expressed in progenitors of fetal neocortex. eLife, 7, D635. http://doi.org/10.7554/eLife.32332 [@doi:10.7554/eLife.32332]
- Insects -- Santos, M. E., Le Bouquin, A., Crumière, A. J. J., & Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 [@doi:10.1126/science.aan2748]


Correlated evolution of celltypes?

- Liang, C., Musser, J. M., Cloutier, A., Prum, R. O., & Wagner, G. P. (2018). Pervasive Correlated Evolution in Gene Expression Shapes Cell and Tissue Type Transcriptomes. Genome Biology and Evolution, 10(2), 538–552. http://doi.org/10.1093/gbe/evy016 [@doi:10.1093/gbe/evy016]

Cell type homology

- Thornton, J. W., & DeSalle, R. (2000). Gene family evolution and homology: genomics meets phylogenetics. Annual Review of Genomics and Human Genetics, 1(1), 41–73. http://doi.org/10.1146/annurev.genom.1.1.41 [@doi:10.1146/annurev.genom.1.1.41]
- Tschopp, P., & Tabin, C. J. (2017). Deep homology in the age of next-generation sequencing. Philosophical Transactions of the Royal Society B: Biological Sciences, 372(1713), 20150475–8. http://doi.org/10.1098/rstb.2015.0475 [@doi:10.1098/rstb.2015.0475]
- Hejnol, A., & Lowe, C. J. (2015). Embracing the comparative approach: how robust phylogenies and broader developmental sampling impacts the understanding of nervous system evolution. Philosophical Transactions of the Royal Society B: Biological Sciences, 370(1684), 20150045–16. http://doi.org/10.1098/rstb.2015.0045 [@doi:10.1098/rstb.2015.0045]
- Santos, M. E., Le Bouquin, A., Crumière, A. J. J., & Khila, A. (2017). Taxon-restricted genes at the origin of a novel trait allowing access to a new environment. Science, 358(6361), 386–390. http://doi.org/10.1126/science.aan2748 [@doi:10.1126/science.aan2748]
- Mammalian decidual cell

Cell type evolution

- Erkenbrack, E. M., Maziarz, J. D., Griffith, O. W., Liang, C., Chavan, A. R., Nnamani, M. C., & Wagner, G. P. (2018). The mammalian decidual cell evolved from a cellular stress response. PLOS Biology, 16(8), e2005594–27. http://doi.org/10.1371/journal.pbio.2005594 [@doi:10.1371/journal.pbio.2005594] -->
<!-- # Discussion -->
<p>–&gt;</p>
<!-- default theme -->

<style>
    /* import google fonts */
    @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
    @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

    /* -------------------------------------------------- */
    /* global */
    /* -------------------------------------------------- */

    /* all elements */
    * {
        /* force sans-serif font unless specified otherwise */
        font-family: "Open Sans", "Helvetica", sans-serif;

        /* prevent text inflation on some mobile browsers */
        -webkit-text-size-adjust: none !important;
        -moz-text-size-adjust: none !important;
        -o-text-size-adjust: none !important;
        text-size-adjust: none !important;
    }

    @media only screen {
        /* "page" element */
        body {
            position: relative;
            box-sizing: border-box;
            font-size: 12pt;
            line-height: 1.5;
            max-width: 8.5in;
            margin: 20px auto;
            padding: 40px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* "page" element */
        body {
            padding: 20px;
            margin: 0;
            border-radius: 0;
            border: none;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
            background: none;
        }
    }

    /* -------------------------------------------------- */
    /* headings */
    /* -------------------------------------------------- */

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
        margin: 20px 0;
        padding: 0;
        font-weight: bold;
    }

    /* biggest heading */
    h1 {
        margin: 40px 0;
        text-align: center;
    }

    /* second biggest heading */
    h2 {
        margin-top: 30px;
        padding-bottom: 5px;
        border-bottom: solid 1px #bdbdbd;
    }

    /* heading font sizes */
    h1 {
        font-size: 2em;
    }
    h2 {
        font-size: 1.5em;
    }
    h3{
        font-size: 1.35em;
    }
    h4 {
        font-size: 1.25em;
    }
    h5 {
        font-size: 1.15em;
    }
    h6 {
        font-size: 1em;
    }

    /* -------------------------------------------------- */
    /* manuscript header */
    /* -------------------------------------------------- */

    /* manuscript title */
    header > h1 {
        margin: 0;
    }

    /* manuscript title caption text (ie "automatically generated on") */
    header + p {
        text-align: center;
        margin-top: 10px;
    }

    /* -------------------------------------------------- */
    /* text elements */
    /* -------------------------------------------------- */

    /* links */
    a {
        color: #2196f3;
        overflow-wrap: break-word;
    }

    /* normal links (not empty, not button link, not syntax highlighting link) */
    a:not(:empty):not(.button):not(.sourceLine) {
        padding-left: 1px;
        padding-right: 1px;
    }

    /* superscripts and subscripts */
    sub,
    sup {
        /* prevent from affecting line height */
        line-height: 0;
    }

    /* unordered and ordered lists*/
    ul,
    ol {
        padding-left: 20px;
    }

    /* class for styling text semibold */
    .semibold {
        font-weight: 600;
    }

    /* class for styling elements horizontally left aligned */
    .left {
        display: block;
        text-align: left;
        margin-left: auto;
        margin-right: 0;
        justify-content: left;
    }

    /* class for styling elements horizontally centered */
    .center {
        display: block;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
        justify-content: center;
    }

    /* class for styling elements horizontally right aligned */
    .right {
        display: block;
        text-align: right;
        margin-left: 0;
        margin-right: auto;
        justify-content: right;
    }

    /* -------------------------------------------------- */
    /* section elements */
    /* -------------------------------------------------- */

    /* horizontal divider line */
    hr {
        border: none;
        height: 1px;
        background: #bdbdbd;
    }

    /* paragraphs, horizontal dividers, figures, tables, code */
    p,
    hr,
    figure,
    table,
    pre {
        /* treat all as "paragraphs", with consistent vertical margins */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* figures */
    /* -------------------------------------------------- */

    /* figure */
    figure {
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure caption */
    figcaption {
        padding: 0;
        padding-top: 10px;
    }

    /* figure image element */
    figure > img,
    figure > svg {
        max-width: 100%;
        display: block;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure auto-number */
    img + figcaption > span:first-of-type,
    svg + figcaption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* tables */
    /* -------------------------------------------------- */

    /* table */
    table {
        border-collapse: collapse;
        border-spacing: 0;
        width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* table cells */
    th,
    td {
        border: solid 1px #bdbdbd;
        padding: 10px;
        /* squash table if too wide for page by forcing line breaks */
        overflow-wrap: break-word;
        word-break: break-word;
    }

    /* header row and even rows */
    th,
    tr:nth-child(2n) {
        background-color: #fafafa;
    }

    /* odd rows */
    tr:nth-child(2n + 1) {
        background-color: #ffffff;
    }

    /* table caption */
    caption {
        text-align: left;
        padding: 0;
        padding-bottom: 10px;
    }

    /* table auto-number */
    table > caption > span:first-of-type,
    div.table_wrapper > table > caption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* code */
    /* -------------------------------------------------- */

    /* multi-line code block */
    pre {
        padding: 10px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
        break-inside: avoid;
        text-align: left;
    }

    /* inline code, ie code within normal text */
    :not(pre) > code {
        padding: 0 4px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
    }

    /* code text */
    /* apply all children, to reach syntax highlighting sub-elements */
    code,
    code * {
        /* force monospace font */
        font-family: "Source Code Pro", "Courier New", monospace;
    }

    /* -------------------------------------------------- */
    /* quotes */
    /* -------------------------------------------------- */

    /* quoted text */
    blockquote {
        margin: 0;
        padding: 0;
        border-left: 4px solid #bdbdbd;
        padding-left: 16px;
        break-inside: avoid;
    }

    /* -------------------------------------------------- */
    /* banners */
    /* -------------------------------------------------- */

    /* info banners */
    .banner {
        box-sizing: border-box;
        display: block;
        position: relative;
        width: 100%;
        margin-top: 20px;
        margin-bottom: 20px;
        padding: 20px;
        text-align: center;
    }

    /* paragraph in banner */
    .banner > p {
        margin: 0;
    }

    /* -------------------------------------------------- */
    /* highlight colors */
    /* -------------------------------------------------- */

    .white {
        background: #ffffff;
    }
    .lightgrey {
        background: #eeeeee;
    }
    .grey {
        background: #757575;
    }
    .darkgrey {
        background: #424242;
    }
    .black {
        background: #000000;
    }
    .lightred {
        background: #ffcdd2;
    }
    .lightyellow {
        background: #ffecb3;
    }
    .lightgreen {
        background: #dcedc8;
    }
    .lightblue {
        background: #e3f2fd;
    }
    .lightpurple {
        background: #f3e5f5;
    }
    .red {
        background: #f44336;
    }
    .orange {
        background: #ff9800;
    }
    .yellow {
        background: #ffeb3b;
    }
    .green {
        background: #4caf50;
    }
    .blue {
        background: #2196f3;
    }
    .purple {
        background: #9c27b0;
    }
    .white,
    .lightgrey,
    .lightred,
    .lightyellow,
    .lightgreen,
    .lightblue,
    .lightpurple,
    .orange,
    .yellow,
    .white a,
    .lightgrey a,
    .lightred a,
    .lightyellow a,
    .lightgreen a,
    .lightblue a,
    .lightpurple a,
    .orange a,
    .yellow a {
        color: #000000;
    }
    .grey,
    .darkgrey,
    .black,
    .red,
    .green,
    .blue,
    .purple,
    .grey a,
    .darkgrey a,
    .black a,
    .red a,
    .green a,
    .blue a,
    .purple a {
        color: #ffffff;
    }

    /* -------------------------------------------------- */
    /* buttons */
    /* -------------------------------------------------- */

    /* class for styling links like buttons */
    .button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        margin: 5px;
        padding: 10px 20px;
        font-size: 0.75em;
        font-weight: 600;
        text-transform: uppercase;
        text-decoration: none;
        letter-spacing: 1px;
        background: none;
        color: #2196f3;
        border: solid 1px #bdbdbd;
        border-radius: 5px;
    }

    /* buttons when hovered */
    .button:hover:not([disabled]),
    .icon_button:hover:not([disabled]) {
        cursor: pointer;
        background: #f5f5f5;
    }

    /* buttons when disabled */
    .button[disabled],
    .icon_button[disabled] {
        opacity: 0.35;
        pointer-events: none;
    }

    /* class for styling buttons containg only single icon */
    .icon_button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        margin: 0;
        padding: 0;
        background: none;
        border-radius: 5px;
        border: none;
        width: 20px;
        height: 20px;
        min-width: 20px;
        min-height: 20px;
    }

    /* icon button inner svg image */
    .icon_button > svg {
        height: 16px;
    }

    /* -------------------------------------------------- */
    /* icons */
    /* -------------------------------------------------- */

    /* class for styling icons inline with text */
    .inline_icon {
        height: 1em;
        position: relative;
        top: 0.125em;
    }

    /* -------------------------------------------------- */
    /* print control */
    /* -------------------------------------------------- */

    @media print {
        @page {
            /* suggested printing margin */
            margin: 0.5in;
        }

        /* document and "page" elements */
        html, body {
            margin: 0;
            padding: 0;
            width: 100%;
            height: 100%;
        }

        /* "page" element */
        body {
            font-size: 11pt !important;
            line-height: 1.35;
        }

        /* all headings */
        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
            margin: 15px 0;
        }

        /* figures and tables */
        figure, table {
            font-size: 0.85em;
        }

        /* table cells */
        th,
        td {
            padding: 5px;
        }

        /* shrink font awesome icons */
        i.fas,
        i.fab,
        i.far,
        i.fal {
            transform: scale(0.85);
        }

        /* decrease banner margins */
        .banner {
            margin-top: 15px;
            margin-bottom: 15px;
            padding: 15px;
        }

        /* class for centering an element vertically on its own page */
        .page_center {
            margin: auto;
            width: 100%;
            height: 100%;
            display: flex;
            align-items: center;
            vertical-align: middle;
            break-before: page;
            break-after: page;
        }

        /* always insert a page break before the element */
        .page_break_before {
            break-before: page;
        }

        /* always insert a page break after the element */
        .page_break_after {
            break-after: page;
        }

        /* avoid page break before the element */
        .page_break_before_avoid {
            break-before: avoid;
        }

        /* avoid page break after the element */
        .page_break_after_avoid {
            break-after: avoid;
        }

        /* avoid page break inside the element */
        .page_break_inside_avoid {
            break-inside: avoid;
        }
    }

    /* -------------------------------------------------- */
    /* override pandoc css quirks */
    /* -------------------------------------------------- */

    .sourceCode {
        /* prevent unsightly overflow in wide code blocks */
        overflow: auto !important;
    }

    div.sourceCode {
        /* prevent background fill on top-most code block  container */
        background: none !important;
    }

    .sourceCode * {
        /* force consistent line spacing */
        line-height: 1.5 !important;
    }

    div.sourceCode {
        /* style code block margins same as <pre> element */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* tablenos */
    /* -------------------------------------------------- */

    /* tablenos wrapper */
    .tablenos {
        /* show scrollbar on tables if necessary to prevent overflow */
        width: 100%;
        margin: 20px 0;
    }

    .tablenos > table {
        /* move margins from table to table_wrapper to allow margin collapsing */
        margin: 0;
    }

    @media only screen {
        /* tablenos wrapper */
        .tablenos {
            /* show scrollbar on tables if necessary to prevent overflow */
            overflow-x: auto !important;
        }

        .tablenos th,
        .tablenos td {
            overflow-wrap: unset !important;
            word-break: unset !important;
        }

        /* table in wrapper */
        .tablenos table,
        .tablenos table * {
            /* don't break table words */
            overflow-wrap: normal !important;
        }
    }

    /* -------------------------------------------------- */
    /* mathjax */
    /* -------------------------------------------------- */

    /* mathjax containers */
    .math.display > span:not(.MathJax_Preview) {
        /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
        display: flex !important;
        overflow-x: auto !important;
        overflow-y: hidden !important;
        justify-content: center;
        align-items: center;
        margin: 0 !important;
    }

    /* right click menu */
    .MathJax_Menu {
        border-radius: 5px !important;
        border: solid 1px #bdbdbd !important;
        box-shadow: none !important;
    }

    /* equation auto-number */
    span[id^="eq:"] > span.math.display + span {
        font-weight: 600;
    }

    /* equation */
    span[id^="eq:"] > span.math.display > span {
        /* nudge to make room for equation auto-number and anchor */
        margin-right: 60px !important;
    }

    /* -------------------------------------------------- */
    /* anchors plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anchor button */
        .anchor {
            opacity: 0;
            margin-left: 5px;
        }

        /* anchor buttons within <h2>'s */
        h2 .anchor {
            margin-left: 10px;
        }

        /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
        *:hover > .anchor,
        .anchor:hover,
        .anchor:focus {
            opacity: 1;
        }

        /* anchor button when hovered */
        .anchor:hover {
            cursor: pointer;
        }
    }

    /* always show anchor button on devices with no mouse/hover ability */
    @media (hover: none) {
        .anchor {
            opacity: 1;
        }
    }

    /* always hide anchor button on print */
    @media only print {
        .anchor {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* accordion plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* accordion arrow button */
        .accordion_arrow {
            margin-right: 10px;
        }

        /* arrow icon when <h2> data-collapsed attribute true */
        h2[data-collapsed="true"] > .accordion_arrow > svg {
            transform: rotate(-90deg);
        }

        /* all elements (except <h2>'s) when data-collapsed attribute true */
        *:not(h2)[data-collapsed="true"] {
            display: none;
        }

        /* accordion arrow button when hovered and <h2>'s when hovered */
        .accordion_arrow:hover,
        h2[data-collapsed="true"]:hover,
        h2[data-collapsed="false"]:hover {
            cursor: pointer;
        }
    }

    /* always hide accordion arrow button on print */
    @media only print {
        .accordion_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* tooltips plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* tooltip container */
        #tooltip {
            position: absolute;
            width: 50%;
            min-width: 240px;
            max-width: 75%;
            z-index: 1;
        }

        /* tooltip content */
        #tooltip_content {
            margin-bottom: 5px;
            padding: 20px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
            overflow-wrap: break-word;
        }

        /* tooltip copy of paragraphs and figures */
        #tooltip_content > p,
        #tooltip_content > figure {
            margin: 0;
            max-height: 320px;
            overflow-y: auto;
        }

        /* tooltip copy of <img> */
        #tooltip_content > figure > img,
        #tooltip_content > figure > svg {
            max-height: 260px;
        }

        /* navigation bar */
        #tooltip_nav_bar {
            margin-top: 10px;
            text-align: center;
        }

        /* navigation bar previous/next buton */
        #tooltip_nav_bar > .icon_button {
            position: relative;
            top: 3px;
        }

        /* navigation bar previous button */
        #tooltip_nav_bar > .icon_button:first-of-type {
            margin-right: 5px;
        }

        /* navigation bar next button */
        #tooltip_nav_bar > .icon_button:last-of-type {
            margin-left: 5px;
        }
    }

    /* always hide tooltip on print */
    @media only print {
        #tooltip {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* jump to first plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* jump button */
        .jump_arrow {
            position: relative;
            top: 0.125em;
            margin-right: 5px;
        }
    }

    /* always hide jump button on print */
    @media only print {
        .jump_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* link highlight plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anything with data-highlighted attribute true */
        [data-highlighted="true"] {
            background: #ffeb3b;
        }

        /* anything with data-selected attribute true */
        [data-selected="true"] {
            background: #ff8a65 !important;
        }

        /* animation definition for glow */
        @keyframes highlight_glow {
            0% {
                background: none;
            }
            10% {
                background: #bbdefb;
            }
            100% {
                background: none;
            }
        }

        /* anything with data-glow attribute true */
        [data-glow="true"] {
            animation: highlight_glow 2s;
        }
    }

    /* -------------------------------------------------- */
    /* table of contents plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* toc panel */
        #toc_panel {
            box-sizing: border-box;
            position: fixed;
            top: 0;
            left: 0;
            background: #ffffff;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            z-index: 2;
        }

        /* toc panel when closed */
        #toc_panel[data-open="false"] {
            min-width: 60px;
            width: 60px;
            height: 60px;
            border-right: solid 1px #bdbdbd;
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc panel when open */
        #toc_panel[data-open="true"] {
            min-width: 260px;
            max-width: 480px;
            /* keep panel edge consistent distance away from "page" edge */
            width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
            bottom: 0;
            border-right: solid 1px #bdbdbd;
        }

        /* toc panel header */
        #toc_header {
            box-sizing: border-box;
            display: flex;
            flex-direction: row;
            align-items: center;
            height: 60px;
            margin: 0;
            padding: 20px;
        }

        /* toc panel header when hovered */
        #toc_header:hover {
            cursor: pointer;
        }

        /* toc panel header when panel open */
        #toc_panel[data-open="true"] > #toc_header {
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc open/close header button */
        #toc_button {
            margin-right: 20px;
        }

        /* hide toc list and header text when closed */
        #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
        #toc_panel[data-open="false"] > #toc_list {
            display: none;
        }

        /* toc list of entries */
        #toc_list {
            box-sizing: border-box;
            width: 100%;
            padding: 20px;
            position: absolute;
            top: calc(60px + 1px);
            bottom: 0;
            overflow: auto;
        }

        /* toc entry, link to section in document */
        .toc_link {
            display: block;
            padding: 5px;
            position: relative;
            font-weight: 600;
            text-decoration: none;
        }

        /* toc entry when hovered or when "viewed" */
        .toc_link:hover,
        .toc_link[data-viewing="true"] {
            background: #f5f5f5;
        }

        /* toc entry, level 1 indentation */
        .toc_link[data-level="1"] {
            margin-left: 0;
        }

        /* toc entry, level 2 indentation */
        .toc_link[data-level="2"] {
            margin-left: 20px;
        }

        /* toc entry, level 3 indentation */
        .toc_link[data-level="3"] {
            margin-left: 40px;
        }

        /* toc entry, level 4 indentation */
        .toc_link[data-level="4"] {
            margin-left: 60px;
        }

        /* toc entry bullets */
        #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
            position: absolute;
            left: -15px;
            top: -1px;
            font-size: 1.5em;
        }

        /* toc entry, level 2 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
            content: "\2022";
        }

        /* toc entry, level 3 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
            content: "\25AB";
        }

        /* toc entry, level 4 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
            content: "-";
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* push <body> ("page") element down to make room for toc icon */
        .toc_body_nudge {
            padding-top: 60px;
        }

        /* toc icon when panel closed and not hovered */
        #toc_panel[data-open="false"]:not(:hover) {
            background: rgba(255, 255, 255, 0.75);
        }
    }

    /* always hide toc panel on print */
    @media only print {
        #toc_panel {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* lightbox plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* regular <img> in document when hovered */
        img.lightbox_document_img:hover {
            cursor: pointer;
        }

        .body_no_scroll {
            overflow: hidden !important;
        }

        /* screen overlay */
        #lightbox_overlay {
            display: flex;
            flex-direction: column;
            position: fixed;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            background: rgba(0, 0, 0, 0.75);
            z-index: 3;
        }

        /* middle area containing lightbox image */
        #lightbox_image_container {
            flex-grow: 1;
            display: flex;
            justify-content: center;
            align-items: center;
            overflow: hidden;
            position: relative;
            padding: 20px;
        }

        /* bottom area containing caption */
        #lightbox_bottom_container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100px;
            min-height: 100px;
            max-height: 100px;
            background: rgba(0, 0, 0, 0.5);
        }

        /* image number info text box */
        #lightbox_number_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            left: 2px;
            top: 0;
            z-index: 4;
        }

        /* zoom info text box */
        #lightbox_zoom_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            right: 2px;
            top: 0;
            z-index: 4;
        }

        /* copy of image caption */
        #lightbox_caption {
            box-sizing: border-box;
            display: inline-block;
            width: 100%;
            max-height: 100%;
            padding: 10px 0;
            text-align: center;
            overflow-y: auto;
            color: #ffffff;
        }

        /* navigation previous/next button */
        .lightbox_button {
            width: 100px;
            height: 100%;
            min-width: 100px;
            min-height: 100%;
            color: #ffffff;
        }

        /* navigation previous/next button when hovered */
        .lightbox_button:hover {
            background: none !important;
        }

        /* navigation button icon */
        .lightbox_button > svg {
            height: 25px;
        }

        /* figure auto-number */
        #lightbox_caption > span:first-of-type {
            font-weight: bold;
            margin-right: 5px;
        }

        /* lightbox image when hovered */
        #lightbox_img:hover {
            cursor: grab;
        }

        /* lightbox image when grabbed */
        #lightbox_img:active {
            cursor: grabbing;
        }
    }

    /* when on screen < 480px wide */
    @media only screen and (max-width: 480px) {
        /* make navigation buttons skinnier on small screens to make more room for caption text */
        .lightbox_button {
            width: 50px;
            min-width: 50px;
        }
    }

    /* always hide lightbox on print */
    @media only print {
        #lightbox_overlay {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* hypothesis (annotations) plugin */
    /* -------------------------------------------------- */

    /* hypothesis activation button */
    #hypothesis_button {
        box-sizing: border-box;
        position: fixed;
        top: 0;
        right: 0;
        width: 60px;
        height: 60px;
        background: #ffffff;
        border-radius: 0;
        border-left: solid 1px #bdbdbd;
        border-bottom: solid 1px #bdbdbd;
        box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
        z-index: 2;
    }

    /* hypothesis button svg */
    #hypothesis_button > svg {
        position: relative;
        top: -4px;
    }

    /* hypothesis annotation count */
    #hypothesis_count {
        position: absolute;
        left: 0;
        right: 0;
        bottom: 5px;
    }

    /* side panel */
    .annotator-frame {
        width: 280px !important;
    }

    /* match highlight color to rest of theme */
    .annotator-highlights-always-on .annotator-hl {
        background-color: #ffeb3b !important;
    }

    /* match focused color to rest of theme */
    .annotator-hl.annotator-hl-focused {
        background-color: #ff8a65 !important;
    }

    /* match bucket bar color to rest of theme */
    .annotator-bucket-bar {
        background: #f5f5f5 !important;
    }

    /* always hide button, toolbar, and tooltip on print */
    @media only print {
        #hypothesis_button {
            display: none;
        }

        .annotator-frame {
            display: none !important;
        }

        hypothesis-adder {
            display: none !important;
        }
    }
</style>
<!-- anchors plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds an anchor next to each of a certain type
        // of element that provides a human-readable url to that specific
        // item/position in the document (eg "manuscript.html#abstract"). It
        // also makes it such that scrolling out of view of a target removes
        // its identifier from the url.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'anchors';

        // default plugin options
        const options = {
            // which types of elements to add anchors next to, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3, [id^="fig:"], [id^="tbl:"], [id^="eq:"]',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // add anchor to each element of specified types
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements)
                addAnchor(element);

            // attach scroll listener to window
            window.addEventListener('scroll', onScroll);
        }

        // when window is scrolled
        function onScroll() {
            // if url has hash and user has scrolled out of view of hash
            // target, remove hash from url
            const tolerance = 100;
            const target = getHashTarget();
            if (target) {
                if (
                    target.getBoundingClientRect().top >
                        window.innerHeight + tolerance ||
                    target.getBoundingClientRect().bottom < 0 - tolerance
                )
                    history.pushState(null, null, ' ');
            }
        }

        // add anchor to element
        function addAnchor(element) {
            let addTo; // element to add anchor button to

            // if figure or table, modify withId and addTo to get expected
            // elements
            if (element.id.indexOf('fig:') === 0) {
                addTo = element.querySelector('figcaption');
            } else if (element.id.indexOf('tbl:') === 0) {
                addTo = element.querySelector('caption');
            } else if (element.id.indexOf('eq:') === 0) {
                addTo = element.querySelector('.eqnos-number');
            }

            addTo = addTo || element;
            const id = element.id || null;

            // do not add anchor if element doesn't have assigned id.
            // id is generated by pandoc and is assumed to be unique and
            // human-readable
            if (!id)
                return;

            // create anchor button
            const anchor = document.createElement('a');
            anchor.innerHTML = document.querySelector('.icon_link').innerHTML;
            anchor.title = 'Link to this part of the document';
            anchor.classList.add('icon_button', 'anchor');
            anchor.dataset.ignore = 'true';
            anchor.href = '#' + id;
            addTo.appendChild(anchor);
        }

        // get element that is target of link or url hash
        function getHashTarget() {
            const hash = window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- link icon -->

<template class="icon_link">
    <!-- modified from: https://fontawesome.com/icons/link -->
    <svg width="16" height="16" viewBox="0 0 512 512">
        <path
            fill="currentColor"
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        ></path>
    </svg>
</template>
<!-- accordion plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows sections of content under <h2> headings
        // to be collapsible.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'accordion';

        // default plugin options
        const options = {
            // whether to always start expanded ('false'), always start
            // collapsed ('true'), or start collapsed when screen small ('auto')
            startCollapsed: 'auto',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <h2> heading
            const headings = document.querySelectorAll('h2');
            for (const heading of headings) {
                addArrow(heading);

                // start expanded/collapsed based on option
                if (
                    options.startCollapsed === 'true' ||
                    (options.startCollapsed === 'auto' && isSmallScreen())
                )
                    collapseHeading(heading);
                else
                    expandHeading(heading);
            }

            // attach hash change listener to window
            window.addEventListener('hashchange', onHashChange);
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                goToElement(target);
        }

        // add arrow to heading
        function addArrow(heading) {
            // add arrow button
            const arrow = document.createElement('button');
            arrow.innerHTML = document.querySelector(
                '.icon_angle_down'
            ).innerHTML;
            arrow.classList.add('icon_button', 'accordion_arrow');
            heading.insertBefore(arrow, heading.firstChild);

            // attach click listener to heading and button
            heading.addEventListener('click', onHeadingClick);
            arrow.addEventListener('click', onArrowClick);
        }

        // determine if on mobile-like device with small screen
        function isSmallScreen() {
            return Math.min(window.innerWidth, window.innerHeight) < 480;
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get element that is target of hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // when <h2> heading is clicked
        function onHeadingClick(event) {
            // only collapse if <h2> itself is target of click (eg, user did
            // not click on anchor within <h2>)
            if (event.target === this)
                toggleCollapse(this);
        }

        // when arrow button is clicked
        function onArrowClick() {
            toggleCollapse(this.parentNode);
        }

        // collapse section if expanded, expand if collapsed
        function toggleCollapse(heading) {
            if (heading.dataset.collapsed === 'false')
                collapseHeading(heading);
            else
                expandHeading(heading);
        }

        // elements to exclude from collapse, such as table of contents panel,
        // hypothesis panel, etc
        const exclude = '#toc_panel, div.annotator-frame, #lightbox_overlay';

        // collapse section
        function collapseHeading(heading) {
            heading.setAttribute('data-collapsed', 'true');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'true');
        }

        // expand section
        function expandHeading(heading) {
            heading.setAttribute('data-collapsed', 'false');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'false');
        }

        // get list of elements between this <h2> and next <h2> or <h1>
        // ("children" of the <h2> section)
        function getChildren(heading) {
            return nextUntil(heading, 'h2, h1', exclude);
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get list of elements after a start element up to element matching
        // query
        function nextUntil(element, query, exclude) {
            const elements = [];
            while (element = element.nextElementSibling, element) {
                if (element.matches(query))
                    break;
                if (!element.matches(exclude))
                    elements.push(element);
            }
            return elements;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
    <!-- modified from: https://fontawesome.com/icons/angle-down -->
    <svg width="16" height="16" viewBox="0 0 448 512">
        <path
            fill="currentColor"
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        ></path>
    </svg>
</template>
<!-- tooltips plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when the user hovers or
        // focuses a link to a citation or figure, a tooltip appears with a
        // preview of the reference content, along with arrows to navigate
        // between instances of the same reference in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tooltips';

        // default plugin options
        const options = {
            // whether user must click off to close tooltip instead of just
            // un-hovering
            clickClose: 'false',
            // delay (in ms) between opening and closing tooltip
            delay: '100',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach hover and focus listeners to link
                link.addEventListener('mouseover', onLinkHover);
                link.addEventListener('mouseleave', onLinkUnhover);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('touchend', onLinkTouch);
            }

            // attach mouse, key, and resize listeners to window
            window.addEventListener('mousedown', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('keyup', onKeyUp);
            window.addEventListener('resize', onResize);
        }

        // when link is hovered
        function onLinkHover() {
            // function to open tooltip
            const delayOpenTooltip = function() {
                openTooltip(this);
            }.bind(this);

            // run open function after delay
            this.openTooltipTimer = window.setTimeout(
                delayOpenTooltip,
                options.delay
            );
        }

        // when mouse leaves link
        function onLinkUnhover() {
            // cancel opening tooltip
            window.clearTimeout(this.openTooltipTimer);

            // don't close on unhover if option specifies
            if (options.clickClose === 'true')
                return;

            // function to close tooltip
            const delayCloseTooltip = function() {
                // if tooltip open and if mouse isn't over tooltip, close
                const tooltip = document.getElementById('tooltip');
                if (tooltip && !tooltip.matches(':hover'))
                    closeTooltip();
            };

            // run close function after delay
            this.closeTooltipTimer = window.setTimeout(
                delayCloseTooltip,
                options.delay
            );
        }

        // when link is focused (tabbed to)
        function onLinkFocus(event) {
            openTooltip(this);
        }

        // when link is touched on touch screen
        function onLinkTouch(event) {
            // attempt to force hover state on first tap always, and trigger
            // regular link click (and navigation) on second tap
            if (event.target === document.activeElement)
                event.target.click();
            else {
                document.activeElement.blur();
                event.target.focus();
            }
            if (event.cancelable)
                event.preventDefault();
            event.stopPropagation();
            return false;
        }

        // when mouse is clicked anywhere in window
        function onClick(event) {
            closeTooltip();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'tooltip_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'tooltip_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeTooltip();
                    break;
            }
        }

        // when window is resized or zoomed
        function onResize() {
            closeTooltip();
        }

        // get all links of types we wish to handle
        function getLinks() {
            const queries = [];
            // exclude buttons, anchor links, toc links, etc
            const exclude =
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            queries.push('a[href^="#ref-"]' + exclude); // citation links
            queries.push('a[href^="#fig:"]' + exclude); // figure links
            const query = queries.join(', ');
            return document.querySelectorAll(query);
        }

        // get links with same target, get index of link in set, get total
        // same links
        function getSameLinks(link) {
            const sameLinks = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    sameLinks.push(otherLink);
            }

            return {
                elements: sameLinks,
                index: sameLinks.indexOf(link),
                total: sameLinks.length
            };
        }

        // open tooltip
        function openTooltip(link) {
            // delete tooltip if it exists, start fresh
            closeTooltip();

            // make tooltip element
            const tooltip = makeTooltip(link);

            // if source couldn't be found and tooltip not made, exit
            if (!tooltip)
                return;

            // make navbar elements
            const navBar = makeNavBar(link);
            if (navBar)
                tooltip.firstElementChild.appendChild(navBar);

            // attach tooltip to page
            document.body.appendChild(tooltip);

            // position tooltip
            const position = function() {
                positionTooltip(link);
            };
            position();

            // if tooltip contains images, position again after they've loaded
            const imgs = tooltip.querySelectorAll('img');
            for (const img of imgs)
                img.addEventListener('load', position);
        }

        // close (delete) tooltip
        function closeTooltip() {
            const tooltip = document.getElementById('tooltip');
            if (tooltip)
                tooltip.remove();
        }

        // make tooltip
        function makeTooltip(link) {
            // get target element that link points to
            const source = getSource(link);

            // if source can't be found, exit
            if (!source)
                return;

            // create new tooltip
            const tooltip = document.createElement('div');
            tooltip.id = 'tooltip';
            const tooltipContent = document.createElement('div');
            tooltipContent.id = 'tooltip_content';
            tooltip.appendChild(tooltipContent);

            // make copy of source node and put in tooltip
            const sourceCopy = makeCopy(source);
            tooltipContent.appendChild(sourceCopy);

            // attach mouse event listeners
            tooltip.addEventListener('click', onTooltipClick);
            tooltip.addEventListener('mousedown', onTooltipClick);
            tooltip.addEventListener('touchstart', onTooltipClick);
            tooltip.addEventListener('mouseleave', onTooltipUnhover);

            // (for interaction with lightbox plugin)
            // transfer click on tooltip copied img to original img
            const sourceImg = source.querySelector('img');
            const sourceCopyImg = sourceCopy.querySelector('img');
            if (sourceImg && sourceCopyImg) {
                const clickImg = function() {
                    sourceImg.click();
                    closeTooltip();
                };
                sourceCopyImg.addEventListener('click', clickImg);
            }

            return tooltip;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // when tooltip is clicked
        function onTooltipClick(event) {
            // when user clicks on tooltip, stop click from transferring
            // outside of tooltip (eg, click off to close tooltip, or eg click
            // off to unhighlight same refs)
            event.stopPropagation();
        }

        // when tooltip is unhovered
        function onTooltipUnhover(event) {
            if (options.clickClose === 'true')
                return;

            // make sure new mouse/touch/focus no longer over tooltip or any
            // element within it
            const tooltip = document.getElementById('tooltip');
            if (!tooltip)
                return;
            if (this.contains(event.relatedTarget))
                return;

            closeTooltip();
        }

        // make nav bar to go betwen prev/next instances of same reference
        function makeNavBar(link) {
            // find other links to the same source
            const sameLinks = getSameLinks(link);

            // don't show nav bar when singular reference
            if (sameLinks.total <= 1)
                return;

            // find prev/next links with same target
            const prevLink = getPrevLink(link, sameLinks);
            const nextLink = getNextLink(link, sameLinks);

            // create nav bar
            const navBar = document.createElement('div');
            navBar.id = 'tooltip_nav_bar';
            const text = sameLinks.index + 1 + ' of ' + sameLinks.total;

            // create nav bar prev/next buttons
            const prevButton = document.createElement('button');
            const nextButton = document.createElement('button');
            prevButton.id = 'tooltip_prev_button';
            nextButton.id = 'tooltip_next_button';
            prevButton.title =
                'Jump to the previous occurence of this item in the document [←]';
            nextButton.title =
                'Jump to the next occurence of this item in the document [→]';
            prevButton.classList.add('icon_button');
            nextButton.classList.add('icon_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;
            navBar.appendChild(prevButton);
            navBar.appendChild(document.createTextNode(text));
            navBar.appendChild(nextButton);

            // attach click listeners to buttons
            prevButton.addEventListener('click', function() {
                onPrevNextClick(link, prevLink);
            });
            nextButton.addEventListener('click', function() {
                onPrevNextClick(link, nextLink);
            });

            return navBar;
        }

        // get previous link with same target
        function getPrevLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if < 1
            let index;
            if (sameLinks.index - 1 >= 0)
                index = sameLinks.index - 1;
            else
                index = sameLinks.total - 1;
            return sameLinks.elements[index];
        }

        // get next link with same target
        function getNextLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if > total
            let index;
            if (sameLinks.index + 1 <= sameLinks.total - 1)
                index = sameLinks.index + 1;
            else
                index = 0;
            return sameLinks.elements[index];
        }

        // get element that is target of link or url hash
        function getSource(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if ref or figure, modify target to get expected element
            if (id.indexOf('ref-') === 0)
                target = target.querySelector('p');
            else if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');

            return target;
        }

        // when prev/next arrow button is clicked
        function onPrevNextClick(link, prevNextLink) {
            if (link && prevNextLink)
                goToElement(prevNextLink, window.innerHeight * 0.5);
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // determine position to place tooltip based on link position in
        // viewport and tooltip size
        function positionTooltip(link, left, top) {
            const tooltipElement = document.getElementById('tooltip');
            if (!tooltipElement)
                return;

            // get convenient vars for position/dimensions of
            // link/tooltip/page/view
            link = getRectInPage(link);
            const tooltip = getRectInPage(tooltipElement);
            const view = getRectInPage();

            // horizontal positioning
            if (left)
                // use explicit value
                left = left;
            else if (link.left + tooltip.width < view.right)
                // fit tooltip to right of link
                left = link.left;
            else if (link.right - tooltip.width > view.left)
                // fit tooltip to left of link
                left = link.right - tooltip.width;
            // center tooltip in view
            else
                left = (view.right - view.left) / 2 - tooltip.width / 2;

            // vertical positioning
            if (top)
                // use explicit value
                top = top;
            else if (link.top - tooltip.height > view.top)
                // fit tooltip above link
                top = link.top - tooltip.height;
            else if (link.bottom + tooltip.height < view.bottom)
                // fit tooltip below link
                top = link.bottom;
            else {
                // center tooltip in view
                top = view.top + view.height / 2 - tooltip.height / 2;
                // nudge off of link to left/right if possible
                if (link.right + tooltip.width < view.right)
                    left = link.right;
                else if (link.left - tooltip.width > view.left)
                    left = link.left - tooltip.width;
            }

            tooltipElement.style.left = left + 'px';
            tooltipElement.style.top = top + 'px';
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get position of element relative to page
        function getRectInPage(element) {
            const rect = getRectInView(element);
            const body = getRectInView(document.body);

            const newRect = {};
            newRect.left = rect.left - body.left;
            newRect.top = rect.top - body.top;
            newRect.right = rect.right - body.left;
            newRect.bottom = rect.bottom - body.top;
            newRect.width = rect.width;
            newRect.height = rect.height;

            return newRect;
        }

        // (for interaction with accordion plugin)
        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // (for interaction with accordion plugin)
        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- jump to first plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds a button next to each reference entry,
        // figure, and table that jumps the page to the first occurrence of a
        // link to that item in the manuscript.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'jumpToFirst';

        // default plugin options
        const options = {
            // whether to add buttons next to reference entries
            references: 'true',
            // whether to add buttons next to figures
            figures: 'true',
            // whether to add buttons next to tables
            tables: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            if (options.references !== 'false')
                makeReferenceButtons();
            if (options.figures !== 'false')
                makeFigureButtons();
            if (options.tables !== 'false')
                makeTableButtons();
        }

        // when jump button clicked
        function onButtonClick() {
            const first = getFirstOccurrence(this.dataset.id);
            if (!first)
                return;

            // update url hash so navigating "back" in history will return
            // user to jump button
            window.location.hash = this.dataset.id;
            // scroll to link
            window.setTimeout(function() {
                goToElement(first, window.innerHeight * 0.5);
            }, 0);
        }

        // get first occurence of link to item in document
        function getFirstOccurrence(id) {
            let query = 'a';
            query += '[href="#' + id + '"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelector(query);
        }

        // add button next to each reference entry
        function makeReferenceButtons() {
            const references = document.querySelectorAll('div[id^="ref-"]');
            for (const reference of references) {
                // get reference id and element to add button to
                const id = reference.id;
                const container = reference.firstElementChild;
                const first = getFirstOccurrence(id);

                // if can't find link to reference, ignore
                if (!first)
                    continue;

                // make jump button
                let button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this reference in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.innerHTML = button.outerHTML + container.innerHTML;
                button = container.firstElementChild;
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeFigureButtons() {
            const figures = document.querySelectorAll('[id^="fig:"]');
            for (const figure of figures) {
                // get figure id and element to add button to
                const id = figure.id;
                const container = figure.querySelector('figcaption') || figure;
                const first = getFirstOccurrence(id);

                // if can't find link to figure, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this figure in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeTableButtons() {
            const tables = document.querySelectorAll('[id^="tbl:"]');
            for (const table of tables) {
                // get ref id and element to add button to
                const id = table.id;
                const container = table.querySelector('caption') || table;
                const first = getFirstOccurrence(id);

                // if can't find link to table, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this table in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
    <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
    <svg width="16" height="16" viewBox="0 0 320 512">
        <path
            fill="currentColor"
            d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
        ></path>
    </svg>
</template>
<!-- link highlight plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user hovers or
        // focuses a link, other links that have the same target will be
        // highlighted. It also makes it such that when clicking a link, the
        // target of the link (eg reference, figure, table) is briefly
        // highlighted.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'linkHighlight';

        // default plugin options
        const options = {
            // whether to also highlight links that go to external urls
            externalLinks: 'false',
            // whether user must click off to unhighlight instead of just
            // un-hovering
            clickUnhighlight: 'false',
            // whether to also highlight links that are unique
            highlightUnique: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach mouse and focus listeners to link
                link.addEventListener('mouseenter', onLinkFocus);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('mouseleave', onLinkUnhover);
            }

            // attach click and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('hashchange', onHashChange);

            // run hash change on window load in case user has navigated
            // directly to hash
            onHashChange();
        }

        // when link is focused (tabbed to) or hovered
        function onLinkFocus() {
            highlight(this);
        }

        // when link is unhovered
        function onLinkUnhover() {
            if (options.clickUnhighlight !== 'true')
                unhighlightAll();
        }

        // when the mouse is clicked anywhere in window
        function onClick(event) {
            unhighlightAll();
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                glowElement(target);
        }

        // get element that is target of link or url hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            return target;
        }

        // start glow sequence on an element
        function glowElement(element) {
            const startGlow = function() {
                onGlowEnd();
                element.dataset.glow = 'true';
                element.addEventListener('animationend', onGlowEnd);
            };
            const onGlowEnd = function() {
                element.removeAttribute('data-glow');
                element.removeEventListener('animationend', onGlowEnd);
            };
            startGlow();
        }

        // highlight link and all others with same target
        function highlight(link) {
            // force unhighlight all to start fresh
            unhighlightAll();

            // get links with same target
            if (!link)
                return;
            const sameLinks = getSameLinks(link);

            // if link unique and option is off, exit and don't highlight
            if (sameLinks.length <= 1 && options.highlightUnique !== 'true')
                return;

            // highlight all same links, and "select" (special highlight) this
            // one
            for (const sameLink of sameLinks) {
                if (sameLink === link)
                    sameLink.setAttribute('data-selected', 'true');
                else
                    sameLink.setAttribute('data-highlighted', 'true');
            }
        }

        // unhighlight all links
        function unhighlightAll() {
            const links = getLinks();
            for (const link of links) {
                link.setAttribute('data-selected', 'false');
                link.setAttribute('data-highlighted', 'false');
            }
        }

        // get links with same target
        function getSameLinks(link) {
            const results = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    results.push(otherLink);
            }
            return results;
        }

        // get all links of types we wish to handle
        function getLinks() {
            let query = 'a';
            if (options.externalLinks !== 'true')
                query += '[href^="#"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelectorAll(query);
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin provides a "table of contents" (toc) panel on
        // the side of the document that allows the user to conveniently
        // navigate between sections of the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableOfContents';

        // default plugin options
        const options = {
            // which types of elements to add links for, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3',
            // whether toc starts open. use 'true' or 'false', or 'auto' to
            // use 'true' behavior when screen wide enough and 'false' when not
            startOpen: 'false',
            // whether toc closes when clicking on toc link. use 'true' or
            // 'false', or 'auto' to use 'false' behavior when screen wide
            // enough and 'true' when not
            clickClose: 'auto',
            // if list item is more than this many characters, text will be
            // truncated
            charLimit: '50',
            // whether or not to show bullets next to each toc item
            bullets: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // make toc panel and populate with entries (links to document
            // sections)
            const panel = makePanel();
            if (!panel)
                return;
            makeEntries(panel);
            // attach panel to document after making entries, so 'toc' heading
            // in panel isn't included in toc
            document.body.insertBefore(panel, document.body.firstChild);

            // initial panel state
            if (
                options.startOpen === 'true' ||
                (options.startOpen === 'auto' && !isSmallScreen())
            )
                openPanel();
            else
                closePanel();

            // attach click, scroll, and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('scroll', onScroll);
            window.addEventListener('hashchange', onScroll);
            window.addEventListener('keyup', onKeyUp);
            onScroll();

            // add class to push document body down out of way of toc button
            document.body.classList.add('toc_body_nudge');
        }

        // determine if screen wide enough to fit toc panel
        function isSmallScreen() {
            // in default theme:
            // 816px = 8.5in = width of "page" (<body>) element
            // 260px = min width of toc panel (*2 for both sides of <body>)
            return window.innerWidth < 816 + 260 * 2;
        }

        // when mouse is clicked anywhere in window
        function onClick() {
            if (isSmallScreen())
                closePanel();
        }

        // when window is scrolled or hash changed
        function onScroll() {
            highlightViewed();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            // close on esc
            if (event.key === 'Escape')
                closePanel();
        }

        // find entry of currently viewed document section in toc and highlight
        function highlightViewed() {
            const firstId = getFirstInView(options.typesQuery);

            // get toc entries (links), unhighlight all, then highlight viewed
            const list = document.getElementById('toc_list');
            if (!firstId || !list)
                return;
            const links = list.querySelectorAll('a');
            for (const link of links)
                link.dataset.viewing = 'false';
            const link = list.querySelector('a[href="#' + firstId + '"]');
            if (!link)
                return;
            link.dataset.viewing = 'true';
        }

        // get first or previous toc listed element in top half of view
        function getFirstInView(query) {
            // get all elements matching query and with id
            const elements = document.querySelectorAll(query);
            const elementsWithIds = [];
            for (const element of elements) {
                if (element.id)
                    elementsWithIds.push(element);
            }


            // get first or previous element in top half of view
            for (let i = 0; i < elementsWithIds.length; i++) {
                const element = elementsWithIds[i];
                const prevElement = elementsWithIds[Math.max(0, i - 1)];
                if (element.getBoundingClientRect().top >= 0) {
                    if (
                        element.getBoundingClientRect().top <
                        window.innerHeight / 2
                    )
                        return element.id;
                    else
                        return prevElement.id;
                }
            }
        }

        // make panel
        function makePanel() {
            // create panel
            const panel = document.createElement('div');
            panel.id = 'toc_panel';
            if (options.bullets === 'true')
                panel.dataset.bullets = 'true';

            // create header
            const header = document.createElement('div');
            header.id = 'toc_header';

            // create toc button
            const button = document.createElement('button');
            button.id = 'toc_button';
            button.innerHTML = document.querySelector('.icon_th_list').innerHTML;
            button.title = 'Table of Contents';
            button.classList.add('icon_button');

            // create header text
            const text = document.createElement('h4');
            text.innerHTML = 'Table of Contents';

            // create container for toc list
            const list = document.createElement('div');
            list.id = 'toc_list';

            // attach click listeners
            panel.addEventListener('click', onPanelClick);
            header.addEventListener('click', onHeaderClick);
            button.addEventListener('click', onButtonClick);

            // attach elements
            header.appendChild(button);
            header.appendChild(text);
            panel.appendChild(header);
            panel.appendChild(list);

            return panel;
        }

        // create toc entries (links) to each element of the specified types
        function makeEntries(panel) {
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements) {
                // do not add link if element doesn't have assigned id
                if (!element.id)
                    continue;

                // create link/list item
                const link = document.createElement('a');
                link.classList.add('toc_link');
                switch (element.tagName.toLowerCase()) {
                    case 'h1':
                        link.dataset.level = '1';
                        break;
                    case 'h2':
                        link.dataset.level = '2';
                        break;
                    case 'h3':
                        link.dataset.level = '3';
                        break;
                    case 'h4':
                        link.dataset.level = '4';
                        break;
                }
                link.title = element.innerText;
                let text = element.innerText;
                if (text.length > options.charLimit)
                    text = text.slice(0, options.charLimit) + '...';
                link.innerHTML = text;
                link.href = '#' + element.id;
                link.addEventListener('click', onLinkClick);

                // attach link
                panel.querySelector('#toc_list').appendChild(link);
            }
        }

        // when panel is clicked
        function onPanelClick(event) {
            // stop click from propagating to window/document and closing panel
            event.stopPropagation();
        }

        // when header itself is clicked
        function onHeaderClick(event) {
            togglePanel();
        }

        // when button is clicked
        function onButtonClick(event) {
            togglePanel();
            // stop header underneath button from also being clicked
            event.stopPropagation();
        }

        // when link is clicked
        function onLinkClick(event) {
            if (
                options.clickClose === 'true' ||
                (options.clickClose === 'auto' && isSmallScreen())
            )
                closePanel();
            else
                openPanel();
        }

        // open panel if closed, close if opened
        function togglePanel() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                closePanel();
            else
                openPanel();
        }

        // open panel
        function openPanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'true';
        }

        // close panel
        function closePanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'false';
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- th list icon -->

<template class="icon_th_list">
    <!-- modified from: https://fontawesome.com/icons/th-list -->
    <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
        <path
            fill="currentColor"
            d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- lightbox plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user clicks on an
        // image, the image fills the screen and the user can pan/drag/zoom
        // the image and navigate between other images in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'lightbox';

        // default plugin options
        const options = {
            // list of possible zoom/scale factors
            zoomSteps:
                '0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1,' +
                '1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8',
            // whether to fit image to view ('fit'), display at 100% and shrink
            // if necessary ('shrink'), or always display at 100% ('100')
            defaultZoom: 'fit',
            // whether to zoom in/out toward center of view ('true') or mouse
            // ('false')
            centerZoom: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <img> element
            const imgs = document.querySelectorAll('figure > img');
            let count = 1;
            for (const img of imgs) {
                img.classList.add('lightbox_document_img');
                img.dataset.number = count;
                img.dataset.total = imgs.length;
                img.addEventListener('click', openLightbox);
                count++;
            }

            // attach mouse and key listeners to window
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('keyup', onKeyUp);
        }

        // when mouse is moved anywhere in window
        function onWindowMouseMove(event) {
            window.mouseX = event.clientX;
            window.mouseY = event.clientY;
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'lightbox_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'lightbox_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeLightbox();
                    break;
            }
        }

        // open lightbox
        function openLightbox() {
            const lightbox = makeLightbox(this);
            if (!lightbox)
                return;

            blurBody(lightbox);
            document.body.appendChild(lightbox);
        }

        // make lightbox
        function makeLightbox(img) {
            // delete lightbox if it exists, start fresh
            closeLightbox();

            // create screen overlay containing lightbox
            const overlay = document.createElement('div');
            overlay.id = 'lightbox_overlay';

            // create image info boxes
            const numberInfo = document.createElement('div');
            const zoomInfo = document.createElement('div');
            numberInfo.id = 'lightbox_number_info';
            zoomInfo.id = 'lightbox_zoom_info';

            // create container for image
            const imageContainer = document.createElement('div');
            imageContainer.id = 'lightbox_image_container';
            const lightboxImg = makeLightboxImg(
                img,
                imageContainer,
                numberInfo,
                zoomInfo
            );
            imageContainer.appendChild(lightboxImg);

            // create bottom container for caption and navigation buttons
            const bottomContainer = document.createElement('div');
            bottomContainer.id = 'lightbox_bottom_container';
            const caption = makeCaption(img);
            const prevButton = makePrevButton(img);
            const nextButton = makeNextButton(img);
            bottomContainer.appendChild(prevButton);
            bottomContainer.appendChild(caption);
            bottomContainer.appendChild(nextButton);

            // attach top middle and bottom to overlay
            overlay.appendChild(numberInfo);
            overlay.appendChild(zoomInfo);
            overlay.appendChild(imageContainer);
            overlay.appendChild(bottomContainer);

            return overlay;
        }

        // make <img> object that is intuitively draggable and zoomable
        function makeLightboxImg(
            sourceImg,
            container,
            numberInfoBox,
            zoomInfoBox
        ) {
            // create copy of source <img>
            const img = sourceImg.cloneNode(true);
            img.classList.remove('lightbox_document_img');
            img.removeAttribute('id');
            img.removeAttribute('width');
            img.removeAttribute('height');
            img.style.position = 'unset';
            img.style.margin = '0';
            img.style.padding = '0';
            img.style.width = '';
            img.style.height = '';
            img.style.minWidth = '';
            img.style.minHeight = '';
            img.style.maxWidth = '';
            img.style.maxHeight = '';
            img.id = 'lightbox_img';

            // build sorted list of unique zoomSteps, always including a 100%
            let zoomSteps = [];
            const optionsZooms = options.zoomSteps.split(/[^0-9.]/);
            for (const optionZoom of optionsZooms) {
                const newZoom = parseFloat(optionZoom);
                if (newZoom && !zoomSteps.includes(newZoom))
                    zoomSteps.push(newZoom);
            }
            if (!zoomSteps.includes(1))
                zoomSteps.push(1);
            zoomSteps = zoomSteps.sort(function sortNumber(a, b) {
                return a - b;
            });

            // <img> object property variables
            let zoom = 1;
            let translateX = 0;
            let translateY = 0;
            let clickMouseX = undefined;
            let clickMouseY = undefined;
            let clickTranslateX = undefined;
            let clickTranslateY = undefined;

            updateNumberInfo();

            // update image numbers displayed in info box
            function updateNumberInfo() {
                numberInfoBox.innerHTML =
                    sourceImg.dataset.number + ' of ' + sourceImg.dataset.total;
            }

            // update zoom displayed in info box
            function updateZoomInfo() {
                let zoomInfo = zoom * 100;
                if (!Number.isInteger(zoomInfo))
                    zoomInfo = zoomInfo.toFixed(2);
                zoomInfoBox.innerHTML = zoomInfo + '%';
            }

            // move to closest zoom step above current zoom
            const zoomIn = function() {
                for (const zoomStep of zoomSteps) {
                    if (zoomStep > zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                updateTransform();
            };

            // move to closest zoom step above current zoom
            const zoomOut = function() {
                zoomSteps.reverse();
                for (const zoomStep of zoomSteps) {
                    if (zoomStep < zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                zoomSteps.reverse();

                updateTransform();
            };

            // update display of <img> based on scale/translate properties
            const updateTransform = function() {
                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                // get new width/height after scale
                const rect = img.getBoundingClientRect();
                // limit translate
                translateX = Math.max(translateX, -rect.width / 2);
                translateX = Math.min(translateX, rect.width / 2);
                translateY = Math.max(translateY, -rect.height / 2);
                translateY = Math.min(translateY, rect.height / 2);

                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                updateZoomInfo();
            };

            // fit <img> to container
            const fit = function() {
                // no x/y offset, 100% zoom by default
                translateX = 0;
                translateY = 0;
                zoom = 1;

                // widths of <img> and container
                const imgWidth = img.naturalWidth;
                const imgHeight = img.naturalHeight;
                const containerWidth = parseFloat(
                    window.getComputedStyle(container).width
                );
                const containerHeight = parseFloat(
                    window.getComputedStyle(container).height
                );

                // how much zooming is needed to fit <img> to container
                const xRatio = imgWidth / containerWidth;
                const yRatio = imgHeight / containerHeight;
                const maxRatio = Math.max(xRatio, yRatio);
                const newZoom = 1 / maxRatio;

                // fit <img> to container according to option
                if (options.defaultZoom === 'shrink') {
                    if (maxRatio > 1)
                        zoom = newZoom;
                } else if (options.defaultZoom === 'fit')
                    zoom = newZoom;

                updateTransform();
            };

            // when mouse wheel is rolled anywhere in container
            const onContainerWheel = function(event) {
                if (!event)
                    return;

                // let ctrl + mouse wheel to zoom behave as normal
                if (event.ctrlKey)
                    return;

                // prevent normal scroll behavior
                event.preventDefault();
                event.stopPropagation();

                // point around which to scale img
                const viewRect = container.getBoundingClientRect();
                const viewX = (viewRect.left + viewRect.right) / 2;
                const viewY = (viewRect.top + viewRect.bottom) / 2;
                const originX = options.centerZoom === 'true' ? viewX : mouseX;
                const originY = options.centerZoom === 'true' ? viewY : mouseY;

                // get point on image under origin
                const oldRect = img.getBoundingClientRect();
                const oldPercentX = (originX - oldRect.left) / oldRect.width;
                const oldPercentY = (originY - oldRect.top) / oldRect.height;

                // increment/decrement zoom
                if (event.deltaY < 0)
                    zoomIn();
                if (event.deltaY > 0)
                    zoomOut();

                // get offset between previous image point and origin
                const newRect = img.getBoundingClientRect();
                const offsetX =
                    originX - (newRect.left + newRect.width * oldPercentX);
                const offsetY =
                    originY - (newRect.top + newRect.height * oldPercentY);

                // translate image to keep image point under origin
                translateX += offsetX;
                translateY += offsetY;

                // perform translate
                updateTransform();
            };

            // when container is clicked
            function onContainerClick(event) {
                // if container itself is target of click, and not other
                // element above it
                if (event.target === this)
                    closeLightbox();
            }

            // when mouse button is pressed on image
            const onImageMouseDown = function(event) {
                // store original mouse position relative to image
                clickMouseX = window.mouseX;
                clickMouseY = window.mouseY;
                clickTranslateX = translateX;
                clickTranslateY = translateY;
                event.stopPropagation();
                event.preventDefault();
            };

            // when mouse button is released anywhere in window
            const onWindowMouseUp = function(event) {
                // reset original mouse position
                clickMouseX = undefined;
                clickMouseY = undefined;
                clickTranslateX = undefined;
                clickTranslateY = undefined;

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mouseup', onWindowMouseUp);
            };

            // when mouse is moved anywhere in window
            const onWindowMouseMove = function(event) {
                if (
                    clickMouseX === undefined ||
                    clickMouseY === undefined ||
                    clickTranslateX === undefined ||
                    clickTranslateY === undefined
                )
                    return;

                // offset image based on original and current mouse position
                translateX = clickTranslateX + window.mouseX - clickMouseX;
                translateY = clickTranslateY + window.mouseY - clickMouseY;
                updateTransform();
                event.preventDefault();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mousemove', onWindowMouseMove);
            };

            // when window is resized
            const onWindowResize = function(event) {
                fit();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('resize', onWindowResize);
            };

            // attach the necessary event listeners
            img.addEventListener('dblclick', fit);
            img.addEventListener('mousedown', onImageMouseDown);
            container.addEventListener('wheel', onContainerWheel);
            container.addEventListener('mousedown', onContainerClick);
            container.addEventListener('touchstart', onContainerClick);
            window.addEventListener('mouseup', onWindowMouseUp);
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('resize', onWindowResize);

            // run fit() after lightbox atttached to document and <img> Loaded
            // so needed container and img dimensions available
            img.addEventListener('load', fit);

            return img;
        }

        // make caption
        function makeCaption(img) {
            const caption = document.createElement('div');
            caption.id = 'lightbox_caption';
            const captionSource = img.nextElementSibling;
            if (captionSource.tagName.toLowerCase() === 'figcaption') {
                const captionCopy = makeCopy(captionSource);
                caption.innerHTML = captionCopy.innerHTML;
            }

            caption.addEventListener('touchstart', function(event) {
                event.stopPropagation();
            });

            return caption;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // make button to jump to previous image in document
        function makePrevButton(img) {
            const prevButton = document.createElement('button');
            prevButton.id = 'lightbox_prev_button';
            prevButton.title = 'Jump to the previous image in the document [←]';
            prevButton.classList.add('icon_button', 'lightbox_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;

            // attach click listeners to button
            prevButton.addEventListener('click', function() {
                getPrevImg(img).click();
            });

            return prevButton;
        }

        // make button to jump to next image in document
        function makeNextButton(img) {
            const nextButton = document.createElement('button');
            nextButton.id = 'lightbox_next_button';
            nextButton.title = 'Jump to the next image in the document [→]';
            nextButton.classList.add('icon_button', 'lightbox_button');
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;

            // attach click listeners to button
            nextButton.addEventListener('click', function() {
                getNextImg(img).click();
            });

            return nextButton;
        }

        // get previous image in document
        function getPrevImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if < 1
            if (index - 1 >= 0)
                index--;
            else
                index = imgs.length - 1;
            return imgs[index];
        }

        // get next image in document
        function getNextImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if > total
            if (index + 1 <= imgs.length - 1)
                index++;
            else
                index = 0;
            return imgs[index];
        }

        // close lightbox
        function closeLightbox() {
            focusBody();

            const lightbox = document.getElementById('lightbox_overlay');
            if (lightbox)
                lightbox.remove();
        }

        // make all elements behind lightbox non-focusable
        function blurBody(overlay) {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.tabIndex = -1;
            document.body.classList.add('body_no_scroll');
        }

        // make all elements focusable again
        function focusBody() {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.removeAttribute('tabIndex');
            document.body.classList.remove('body_no_scroll');
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- attributes plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows arbitrary HTML attributes to be attached
        // to (almost) any element. Place an HTML comment inside or next to the
        // desired element in the format <!-- $attribute="value" -->

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'attributes';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // get list of comments in document
            const comments = findComments();

            for(const comment of comments)
                if (comment.parentElement)
                    addAttributes(
                        comment.parentElement,
                        comment.nodeValue.trim()
                    );
        }

        // add html attributes to specified element based on string of 
        // html attributes and values
        function addAttributes(element, text) {
            // regex's for finding attribute/value pairs in the format of
            // attribute="value" or attribute='value
            const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
            const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

            // loop through attribute/value pairs
            let match;
            while(match = text.match(regex2) || text.match(regex1)) {
                // get attribute and value from regex capture groups
                let attribute = match[1];
                let value = match[2];

                // remove from string
                text = text.substring(match.index + match[0].length);

                if (!attribute || !value)
                    break;

                // set attribute of parent element
                try {
                    element.setAttribute(attribute, value);
                } catch(error) {
                    console.log(error);
                }

                // special case for colspan
                if (attribute === 'colspan')
                    removeTableCells(element, value);
            }
        }

        // get list of comment elements in document
        function findComments() {
            const comments = [];

            // iterate over comment nodes in document
            function acceptNode(node) {
                return NodeFilter.FILTER_ACCEPT;
            }
            const iterator = document.createNodeIterator(
                document.body,
                NodeFilter.SHOW_COMMENT,
                acceptNode
            );
            let node;
            while(node = iterator.nextNode())
                comments.push(node);

            return comments;
        }

        // remove certain number of cells after specified cell
        function removeTableCells(cell, number) {
            number = parseInt(number);
            if (!number)
                return;

            // remove elements
            for(; number > 1; number--) {
                if (cell.nextElementSibling)
                    cell.nextElementSibling.remove();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- mathjax plugin configuration -->

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        "CommonHTML": { linebreaks: { automatic: true } },
        "HTML-CSS": { linebreaks: { automatic: true } },
        "SVG": { linebreaks: { automatic: true } },
        "fast-preview": { disabled: true }
  });
</script>

<!-- mathjax plugin -->

<script
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
    integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
    crossorigin="anonymous"
>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'MathJax' allows the proper rendering of
    // math/equations written in LaTeX.

    // https://www.mathjax.org/
</script>
<!-- annotations plugin -->

<script>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'Hypothesis' allows public annotation of the
    // manuscript.

    // https://web.hypothes.is/

    // plugin configuration
    window.hypothesisConfig = function() {
        return {
            branding: {
                accentColor: '#2196f3',
                appBackgroundColor: '#f8f8f8',
                ctaBackgroundColor: '#f8f8f8',
                ctaTextColor: '#000000',
                selectionFontFamily: 'Open Sans, Helvetica, sans serif',
                annotationFontFamily: 'Open Sans, Helvetica, sans serif'
            }
        };
    };

    // hypothesis client script
    const embed = 'https://hypothes.is/embed.js';
    // hypothesis annotation count query url
    const query = 'https://api.hypothes.is/api/search?limit=0&url='

    
    // start script
    function start() {
        const button = makeButton();
        document.body.insertBefore(button, document.body.firstChild);
        insertCount(button);
    }

    // make button
    function makeButton() {
        // create button
        const button = document.createElement('button');
        button.id = 'hypothesis_button';
        button.innerHTML = document.querySelector('.icon_hypothesis').innerHTML;
        button.title = 'Hypothesis annotations';
        button.classList.add('icon_button');

        function onClick(event) {
            onButtonClick(event, button);
        }

        // attach click listeners
        button.addEventListener('click', onClick);

        return button;
    }

    // insert annotations count
    async function insertCount(button) {
        // get annotation count from Hypothesis based on url
        let count = '-';
        try {
            const canonical = document.querySelector('link[rel="canonical"]');
            const location = window.location;
            const url = encodeURIComponent((canonical || location).href);
            const response = await fetch(query + url);
            const json = await response.json();
            count = json.total || '-';
        } catch(error) {
            console.log(error);
        }
        
        // put count into button
        const counter = document.createElement('span');
        counter.id = 'hypothesis_count';
        counter.innerHTML = count;
        button.title = 'View ' + count + ' Hypothesis annotations';
        button.append(counter);
    }

    // when button is clicked
    function onButtonClick(event, button) {
        const script = document.createElement('script');
        script.src = embed;
        document.body.append(script);
        button.remove();
    }

    window.addEventListener('load', start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
    <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
    <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
        <path
            fill="currentColor"
            d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- analytics plugin -->

<!-- copy and paste code from Google Analytics or similar service here -->
</body>
</html>
